---
title: "CRISPR analysis of Rin3 and Ezr enahcner region"
author: "Khyati Dalal"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    depth: 3
    theme: sandstone
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
  word_document:
    toc: yes
header-includes:
- \usepackage{fancyhdr}
- \usepackage{lipsum}
- \pagestyle{fancy}
- \fancyfoot[CO,CE]{1028-02098-001-001}
- \fancyfoot[LE,RO]{\thepage}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options:
  chunk_output_type: console
---

# Introduction

Performed two CRISPR experiment

1) At Ezr enhacner region where Tead4 double motif on chr17:6,827,802-6,827,811 (mm10) was mutated from ACATTCCAGA (wild type) to GCATTCCAGGAATTCCA (mutant). 

![Ezr enhancer regions](/n/projects/kd2200/publication/bpnet/analysis/figures/14_crispr_analysis/ezr_snapgene.png)

2) At Rin3 enhacner region where Tead4 motif on chr12:102,262,024-102,262,033 (mm10) was mutated from CACATTCCTA (wild type) to CACCGTCCTA (mutant) and re-inserted at position (CACATTCCTA) at 60bp downstream from Tfap2c motif (GGGCCCCAGGGCC) in a sequential manner where re-insertion followed by dinucleotide mutation

![Rin3 enhacner regions](/n/projects/kd2200/publication/bpnet/analysis/figures/14_crispr_analysis/rin3_snapgene.png)

# Computational Setup

```{r, warning=F, message=F}
#Standard packages
library(rtracklayer) ; library(GenomicRanges); library(magrittr) ; library(Biostrings);library(parallel)
library(ggplot2) ; library(reshape2); library(plyranges); library(Rsamtools);library(BSgenome.Mmusculus.UCSC.mm10)
library(dplyr); library(data.table); library(patchwork); library(readr);library(TxDb.Mmusculus.UCSC.mm10.knownGene)

#KNITR Options
setwd("/n/projects/kd2200/publication/bpnet/analysis/")
options(knitr.figure_dir="figures/14_crispr_analysis/", java.parameters = "- Xmx6g")

#Lab sources
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/granges_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/knitr_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/caching.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_functions.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/multiplot.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/motif_functions.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_new.r")
#Special sources
library(ggpubr)
library(ggrastr)

# region of interest
## ezr and rin3 genomic enhancer region
gr <- GRanges(seqnames = c("chr17","chr12"), strand = c("+","+"),
              ranges = IRanges(start = c(6827588,102261800), end=c(6828029,102262303)))

# to get mapped motifs
motifs.gr <- read_bed("bed/curated_motif_regoion_tsc_model_motifs_fold5_0based_noerv_nopromoter.bed")

```

## Ezr CRISPR analysis

```{r, ezr crispr,fig.height = 6, fig.width = 6, fig.align = "center"}

##ezronly
ezronly_crispr_chip_set_rpm.bw.list<-list(ezrmut_tead4_c = list(
    pos = "../../data_preparation/crispr/bw/mm10_mtsc_h32_ezr_td_singletodouble_mut/nexus/combined/mtsc_ezrmut_tead4_nexus_combined_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10_mtsc_h32_ezr_td_singletodouble_mut/nexus/combined/mtsc_ezrmut_tead4_nexus_combined_normalized_negative.bw"),
wt_tead4_c = list(
    pos = "../../data_preparation/crispr/bw/mm10/nexus/combined/mtsc_tead4_nexus_combined_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10/nexus/combined/mtsc_tead4_nexus_combined_normalized_negative.bw"),
ezrmut_tead4_rep1 = list(
    pos = "../../data_preparation/crispr/bw/mm10_mtsc_h32_ezr_td_singletodouble_mut/nexus/individual/mtsc_ezrmut_tead4_nexus_1_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10_mtsc_h32_ezr_td_singletodouble_mut/nexus/individual/mtsc_ezrmut_tead4_nexus_1_normalized_negative.bw"),
wt_tead4_rep1 = list(
    pos = "../../data_preparation/crispr/bw/mm10/nexus/individual/mtsc_tead4_nexus_4_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10/nexus/individual/mtsc_tead4_nexus_4_normalized_negative.bw"),
ezrmut_tead4_rep2 = list(
    pos = "../../data_preparation/crispr/bw/mm10_mtsc_h32_ezr_td_singletodouble_mut/nexus/individual/mtsc_ezrmut_tead4_nexus_2_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10_mtsc_h32_ezr_td_singletodouble_mut/nexus/individual/mtsc_ezrmut_tead4_nexus_2_normalized_negative.bw"),
wt_tead4_rep2 = list(
    pos = "../../data_preparation/crispr/bw/mm10/nexus/individual/mtsc_tead4_nexus_5_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10/nexus/individual/mtsc_tead4_nexus_5_normalized_negative.bw"))

gr_ezr <- gr[1]
test <- findOverlaps(gr_ezr, motifs.gr, ignore.strand=TRUE)
gr_ezr_motif <- motifs.gr[test@to]
gr_ezr_motif <- gr_ezr_motif[2]
gr_ezr_region <- resize(gr_ezr_motif, 2000, "center")

#combined reps
gr_ezr_region$signal_wt_tead4 <- nexus_regionSums(gr_ezr_region, ezronly_crispr_chip_set_rpm.bw.list$wt_tead4_c)
gr_ezr_region$signal_mut_tead4 <- nexus_regionSums(gr_ezr_region, ezronly_crispr_chip_set_rpm.bw.list$ezrmut_tead4_c)

#rep1
gr_ezr_region$signal_wt_tead4_1 <- nexus_regionSums(gr_ezr_region, ezronly_crispr_chip_set_rpm.bw.list$wt_tead4_rep1)
gr_ezr_region$signal_mut_tead4_1 <- nexus_regionSums(gr_ezr_region, ezronly_crispr_chip_set_rpm.bw.list$ezrmut_tead4_rep1)

#rep2
gr_ezr_region$signal_wt_tead4_2 <- nexus_regionSums(gr_ezr_region, ezronly_crispr_chip_set_rpm.bw.list$wt_tead4_rep2)
gr_ezr_region$signal_mut_tead4_2 <- nexus_regionSums(gr_ezr_region, ezronly_crispr_chip_set_rpm.bw.list$ezrmut_tead4_rep2)
# 
# # gr_ezr_region
# # GRanges object with 1 range and 8 metadata columns:
# #       seqnames          ranges strand |          name     score signal_wt_tead4 signal_mut_tead4 signal_wt_tead4_1 signal_mut_tead4_1 signal_wt_tead4_2
# #          <Rle>       <IRanges>  <Rle> |   <character> <numeric>       <numeric>        <numeric>         <numeric>          <numeric>         <numeric>
# #   [1]    chr17 6826806-6828805      + | yap1_0_100333         0         17.6619          77.9322           17.7828            74.5628           17.1651
# #       signal_mut_tead4_2
# #                <numeric>
# #   [1]            83.8178
# 


# tead4 observed metapeak
gr_ezr <- gr[1]
test <- findOverlaps(gr_ezr, motifs.gr, ignore.strand=TRUE)
gr_ezr_motif <- motifs.gr[test@to]
gr_ezr_motif <- gr_ezr_motif[2]

tead4_ezrcrispr_chip_set_rpm.bw.list<-list(
  b_ezrmut_tead4 = list(
    pos = "../../data_preparation/crispr/bw/mm10_mtsc_h32_ezr_td_singletodouble_mut/nexus/combined/mtsc_ezrmut_tead4_nexus_combined_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10_mtsc_h32_ezr_td_singletodouble_mut/nexus/combined/mtsc_ezrmut_tead4_nexus_combined_normalized_negative.bw"),
  a_wt_tead4 = list(
    pos = "../../data_preparation/crispr/bw/mm10/nexus/combined/mtsc_tead4_nexus_combined_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10/nexus/combined/mtsc_tead4_nexus_combined_normalized_negative.bw"))


tead4_metapeak.df <- get_multi_bw_exo_metapeak(resize(gr_ezr_motif,1,"start"), bw.list = tead4_ezrcrispr_chip_set_rpm.bw.list, upstream = 100, downstream = 100, smooth = 1)
g <- ggplot(tead4_metapeak.df,aes(tss_distance, reads, color = sample_name))+
  geom_line(position="identity")+
  ggtitle("Tead4 binding at Ezr region")+
  scale_x_continuous(name = "Position (bp)")+
  scale_y_continuous(name = "Normalized Reads")+facet_wrap(~sample_name, scales = "fixed", ncol =2)+
  theme_classic()+
  theme(text=element_text(size=10))

ggsave("figures/14_crispr_analysis/tead4_nexus_combined_wt_ezr_mut.pdf",g,height= 4, width=8)
```


## Rin3 CRISPR analysis

```{r, rin3 crispr,fig.height = 6, fig.width = 6, fig.align = "center"}

rin3only_crispr_chip_set_rpm.bw.list<-list(
  rin3mut_polii_c = list(
    pos = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/nexus/combined/mtsc_rin3mut_polii_nexus_combined_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/nexus/combined/mtsc_rin3mut_polii_nexus_combined_normalized_negative.bw"), 
  rin3mut_h3k27ac_c = list(
    pos = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/seq/combined/mtsc_rin3mut_h3k27ac_seq_combined_rpkm_normalized.bw",
    neg = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/seq/combined/mtsc_rin3mut_h3k27ac_seq_combined_rpkm_normalized.bw"),
  rin3mut_tead4_c = list(
    pos = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/nexus/combined/mtsc_rin3mut_tead4_nexus_combined_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/nexus/combined/mtsc_rin3mut_tead4_nexus_combined_normalized_negative.bw"),
  wt_polii_c = list(
    pos = "../../data_preparation/crispr/bw/mm10/nexus/combined/mtsc_polii_nexus_combined_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10/nexus/combined/mtsc_polii_nexus_combined_normalized_negative.bw"),
  wt_h3k27ac_c = list(
    pos = "../../data_preparation/crispr/bw/mm10/seq/combined/mtsc_h3k27ac_seq_combined_ratio_rpkm.bw",
    neg = "../../data_preparation/crispr/bw/mm10/seq/combined/mtsc_h3k27ac_seq_combined_ratio_rpkm.bw"),
  wt_tead4_c = list(
    pos = "../../data_preparation/crispr/bw/mm10/nexus/combined/mtsc_tead4_nexus_combined_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10/nexus/combined/mtsc_tead4_nexus_combined_normalized_negative.bw"),
  rin3mut_polii_rep1 = list(
    pos = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/nexus/individual/mtsc_rin3mut_polii_nexus_1_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/nexus/individual/mtsc_rin3mut_polii_nexus_1_normalized_negative.bw"), 
  rin3mut_h3k27ac_rep1 = list(
    pos = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/seq/individual/mtsc_rin3mut_h3k27ac_seq_1_rpkm_normalized.bw",
    neg = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/seq/individual/mtsc_rin3mut_h3k27ac_seq_1_rpkm_normalized.bw"),
  rin3mut_tead4_rep1 = list(
    pos = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/nexus/individual/mtsc_rin3mut_tead4_nexus_1_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/nexus/individual/mtsc_rin3mut_tead4_nexus_1_normalized_negative.bw"),
  wt_polii_rep1 = list(
    pos = "../../data_preparation/crispr/bw/mm10/nexus/individual/mtsc_polii_nexus_5_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10/nexus/individual/mtsc_polii_nexus_5_normalized_negative.bw"),
  wt_h3k27ac_rep1 = list(
    pos = "../../data_preparation/crispr/bw/mm10/seq/individual/mtsc_h3k27ac_seq_4_rpkm_normalized.bw",
    neg = "../../data_preparation/crispr/bw/mm10/seq/individual/mtsc_h3k27ac_seq_4_rpkm_normalized.bw"),
  wt_tead4_rep1 = list(
    pos = "../../data_preparation/crispr/bw/mm10/nexus/individual/mtsc_tead4_nexus_4_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10/nexus/individual/mtsc_tead4_nexus_4_normalized_negative.bw"),
  rin3mut_polii_rep2 = list(
    pos = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/nexus/individual/mtsc_rin3mut_polii_nexus_2_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/nexus/individual/mtsc_rin3mut_polii_nexus_2_normalized_negative.bw"), 
  rin3mut_h3k27ac_rep2 = list(
    pos = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/seq/individual/mtsc_rin3mut_h3k27ac_seq_2_rpkm_normalized.bw",
    neg = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/seq/individual/mtsc_rin3mut_h3k27ac_seq_2_rpkm_normalized.bw"),
  rin3mut_tead4_rep2 = list(
    pos = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/nexus/individual/mtsc_rin3mut_tead4_nexus_2_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/nexus/individual/mtsc_rin3mut_tead4_nexus_2_normalized_negative.bw"),
  wt_polii_rep2 = list(
    pos = "../../data_preparation/crispr/bw/mm10/nexus/individual/mtsc_polii_nexus_6_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10/nexus/individual/mtsc_polii_nexus_6_normalized_negative.bw"),
  wt_h3k27ac_rep2 = list(
    pos = "../../data_preparation/crispr/bw/mm10/seq/individual/mtsc_h3k27ac_seq_5_rpkm_normalized.bw",
    neg = "../../data_preparation/crispr/bw/mm10/seq/individual/mtsc_h3k27ac_seq_5_rpkm_normalized.bw"),
  wt_tead4_rep2 = list(
    pos = "../../data_preparation/crispr/bw/mm10/nexus/individual/mtsc_tead4_nexus_5_normalized_positive.bw",
    neg = "../../data_preparation/crispr/bw/mm10/nexus/individual/mtsc_tead4_nexus_5_normalized_negative.bw"))

gr_rin3 <- gr[2]
test <- findOverlaps(gr_rin3, motifs.gr, ignore.strand=TRUE)
gr_rin3_motif <- motifs.gr[test@to]
gr_rin3_motif1 <- gr_rin3_motif[4]
gr_rin3_region <- resize(gr_rin3_motif1, 2000, "center")

#combined reps
gr_rin3_region$signal_wt_h3 <- regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$wt_h3k27ac_c$pos)
gr_rin3_region$signal_mut_h3 <- regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$rin3mut_h3k27ac_c$pos)
gr_rin3_region$signal_wt_polii <- nexus_regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$wt_polii_c)
gr_rin3_region$signal_mut_polii <- nexus_regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$rin3mut_polii_c)
gr_rin3_region$signal_wt_tead4 <- nexus_regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$wt_tead4_c)
gr_rin3_region$signal_mut_tead4 <- nexus_regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$rin3mut_tead4_c)

#rep1
gr_rin3_region$signal_wt_h3_1 <- regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$wt_h3k27ac_rep1$pos)
gr_rin3_region$signal_mut_h3_1 <- regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$rin3mut_h3k27ac_rep1$pos)
gr_rin3_region$signal_wt_polii_1 <- nexus_regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$wt_polii_rep1)
gr_rin3_region$signal_mut_polii_1 <- nexus_regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$rin3mut_polii_rep1)
gr_rin3_region$signal_wt_tead4_1 <- nexus_regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$wt_tead4_rep1)
gr_rin3_region$signal_mut_tead4_1 <- nexus_regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$rin3mut_tead4_rep1)

#rep2
gr_rin3_region$signal_wt_h3_2 <- regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$wt_h3k27ac_rep2$pos)
gr_rin3_region$signal_mut_h3_2 <- regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$rin3mut_h3k27ac_rep2$pos)
gr_rin3_region$signal_wt_polii_2 <- nexus_regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$wt_polii_rep2)
gr_rin3_region$signal_mut_polii_2 <- nexus_regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$rin3mut_polii_rep2)
gr_rin3_region$signal_wt_tead4_2 <- nexus_regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$wt_tead4_rep2)
gr_rin3_region$signal_mut_tead4_2 <- nexus_regionSums(gr_rin3_region, rin3only_crispr_chip_set_rpm.bw.list$rin3mut_tead4_rep2)

#  gr_rin3_region
# GRanges object with 1 range and 20 metadata columns:
#       seqnames              ranges strand |         name     score signal_wt_h3 signal_mut_h3 signal_wt_polii signal_mut_polii signal_wt_tead4 signal_mut_tead4
#          <Rle>           <IRanges>  <Rle> |  <character> <numeric>    <numeric>     <numeric>       <numeric>        <numeric>       <numeric>        <numeric>
#   [1]    chr12 102261029-102263028      + | yap1_0_80735         0      19350.9       8155.88         8.56459         0.935513         26.8195          9.02066
#       signal_wt_h3_1 signal_mut_h3_1 signal_wt_polii_1 signal_mut_polii_1 signal_wt_tead4_1 signal_mut_tead4_1 signal_wt_h3_2 signal_mut_h3_2 signal_wt_polii_2
#            <numeric>       <numeric>         <numeric>          <numeric>         <numeric>          <numeric>      <numeric>       <numeric>         <numeric>
#   [1]        21842.5         8076.95            9.7032            1.18104           22.5693            7.62644        19841.7         8322.11           5.52636
#       signal_mut_polii_2 signal_wt_tead4_2 signal_mut_tead4_2
#                <numeric>         <numeric>          <numeric>
#   [1]           0.588868           41.3071            9.87788
#   -------
#   seqinfo: 22 sequences from an unspecified genome; no seqlengths

activity=rep(c("Tead4","H3K27ac","PolII"),each=2)
motif_name=rep(c("a_WT_tead4", "b_WT_h3k27c","c_WT_polii","d_mut_tead4", "e_mut_h3k27c","f_mut_polii"),each=2)
average=c(a=c(26.8195,0),b=c(19350.9,0),c=c(8.56459,0), d=c(9.02066,0),e=c(8155.88,0),f=c(0.935513,0))
rep=c(a=c(22.5693,41.3071),b=c(21842.5, 19841.7), c=c(9.7032,5.52636), d=c(7.62644,9.87788), e=c(8076.95,8322.11),f=c(1.18104,0.588868))

df <- data.frame(motif_name, average, rep, activity)

p <- ggplot(data=df, aes(x=motif_name, y=average)) +facet_wrap(~activity, scales = "free", ncol =3)+
  geom_bar(stat="identity",fill="#1ba14a",width=0.7)+geom_point(stat = "identity", aes(y=rep))+theme_classic2()+ylab("Normalized Reads")+xlab("")
p
ggsave("figures/14_crispr_analysis/rin3_enhanceractivity_barplot.pdf",p,height= 4, width=8)

# barplot of only tead4 signal at rin3
gr_rin3_motif$signal_wt_tead4_c <- nexus_regionSums(resize(gr_rin3_motif,500,"center"), rin3only_crispr_chip_set_rpm.bw.list$wt_tead4_c)
gr_rin3_motif$signal_mut_tead4_c <- nexus_regionSums(resize(gr_rin3_motif,500,"center"), rin3only_crispr_chip_set_rpm.bw.list$rin3mut_tead4_c)

gr_rin3_motif$signal_wt_tead4_1 <- nexus_regionSums(resize(gr_rin3_motif,500,"center"), rin3only_crispr_chip_set_rpm.bw.list$wt_tead4_rep1)
gr_rin3_motif$signal_mut_tead4_1 <- nexus_regionSums(resize(gr_rin3_motif,500,"center"), rin3only_crispr_chip_set_rpm.bw.list$rin3mut_tead4_rep1)

gr_rin3_motif$signal_wt_tead4_2 <- nexus_regionSums(resize(gr_rin3_motif,500,"center"), rin3only_crispr_chip_set_rpm.bw.list$wt_tead4_rep2)
gr_rin3_motif$signal_mut_tead4_2 <- nexus_regionSums(resize(gr_rin3_motif,500,"center"), rin3only_crispr_chip_set_rpm.bw.list$rin3mut_tead4_rep2)

activity=rep(c("Tead4"),each=2)
motif_name=rep(c("a_WT_tead4","d_mut_tead4"),each=2)
average=c(a=c(22.5987,0), d=c(6.79203,0))
rep=c(a=c(18.6428,36.1022), d=c(5.91325,7.33097))

df <- data.frame(motif_name, average, rep, activity)

q <- ggplot(data=df, aes(x=motif_name, y=average)) +facet_wrap(~activity, scales = "free", ncol =3)+
  geom_bar(stat="identity",fill="blue",width=0.7)+geom_point(stat = "identity", aes(y=rep))+theme_classic2()+ylab("Normalized Reads")+xlab("")

ggsave("figures/14_crispr_analysis/rin3_tead4binding_barplot.pdf",q,height= 4, width=4)




##perform H3K27ac bar plot at rin3 with log2_normalized track

rin3_h3k27ac_crispr_chip_set_rpm.bw.list<-list(
    rin3mut_h3k27ac_c = list(
        pos = "../../data_preperation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/seq/combined/mtsc_rin3mut_h3k27ac_mm10_seq_combined_log2_normalized.bw",
        neg = "../../data_preperation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/seq/combined/mtsc_rin3mut_h3k27ac_mm10_seq_combined_log2_normalized.bw"),
    rin3mut_h3k27ac_rep1 = list(
        pos = "../../data_preperation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/seq/individual/mtsc_rin3mut_h3k27ac_1_seq_log2_normalized.bw",
        neg = "../../data_preperation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/seq/individual/mtsc_rin3mut_h3k27ac_1_seq_log2_normalized.bw"), 
    rin3mut_h3k27ac_rep2 = list(
        pos = "../../data_preperation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/seq/individual/mtsc_rin3mut_h3k27ac_2_seq_log2_normalized.bw",
        neg = "../../data_preperation/crispr/bw/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/seq/individual/mtsc_rin3mut_h3k27ac_2_seq_log2_normalized.bw"), 
    wt_h3k27ac_c = list(
        pos = "../../data_preperation/crispr/bw/mm10/seq/combined/mtsc_h3k27ac_mm10_seq_combined_log2_normalized.bw",
        neg = "../../data_preperation/crispr/bw/mm10/seq/combined/mtsc_h3k27ac_mm10_seq_combined_log2_normalized.bw"),
    wt_h3k27ac_rep1 = list(
        pos = "../../data_preperation/crispr/bw/mm10/seq/individual/mtsc_h3k27ac_4_seq_log2_normalized.bw",
        neg = "../../data_preperation/crispr/bw/mm10/seq/individual/mtsc_h3k27ac_4_seq_log2_normalized.bw"),
    wt_h3k27ac_rep2 = list(
        pos ="../../data_preperation/crispr/bw/mm10/seq/individual/mtsc_h3k27ac_5_seq_log2_normalized.bw",
        neg = "../../data_preperation/crispr/bw/mm10/seq/individual/mtsc_h3k27ac_5_seq_log2_normalized.bw"))


gr_rin3 <- gr[2]
test <- findOverlaps(gr_rin3, motifs.gr, ignore.strand=TRUE)
gr_rin3_motif <- motifs.gr[test@to]
gr_rin3_motif1 <- gr_rin3_motif[4]
gr_rin3_region <- resize(gr_rin3_motif1, 2000, "center")

#combined reps
gr_rin3_region$signal_wt_h3 <- regionMeans(gr_rin3_region, rin3_h3k27ac_crispr_chip_set_rpm.bw.list$wt_h3k27ac_c$pos)
gr_rin3_region$signal_mut_h3 <- regionMeans(gr_rin3_region, rin3_h3k27ac_crispr_chip_set_rpm.bw.list$rin3mut_h3k27ac_c$pos)
#rep1
gr_rin3_region$signal_wt_h3_1 <- regionMeans(gr_rin3_region, rin3_h3k27ac_crispr_chip_set_rpm.bw.list$wt_h3k27ac_rep1$pos)
gr_rin3_region$signal_mut_h3_1 <- regionMeans(gr_rin3_region, rin3_h3k27ac_crispr_chip_set_rpm.bw.list$rin3mut_h3k27ac_rep1$pos)
#rep2
gr_rin3_region$signal_wt_h3_2 <- regionMeans(gr_rin3_region, rin3_h3k27ac_crispr_chip_set_rpm.bw.list$wt_h3k27ac_rep2$pos)
gr_rin3_region$signal_mut_h3_2 <- regionMeans(gr_rin3_region, rin3_h3k27ac_crispr_chip_set_rpm.bw.list$rin3mut_h3k27ac_rep2$pos)

gr_rin3_region
# GRanges object with 1 range and 8 metadata columns:
#GRanges object with 1 range and 8 metadata columns:
  #    seqnames              ranges strand |         name     score signal_wt_h3 signal_mut_h3 signal_wt_h3_1
  #       <Rle>           <IRanges>  <Rle> |  <character> <numeric>    <numeric>     <numeric>      <numeric>
 # [1]    chr12 102261029-102263028      + | yap1_0_80735         0      3.44307        2.2304        3.60705
#      signal_mut_h3_1 signal_wt_h3_2 signal_mut_h3_2
#            <numeric>      <numeric>       <numeric>
#  [1]         2.21515        3.32602         2.23995

activity=rep(c("H3K27ac"),each=2)
motif_name=rep(c("b_WT_h3k27c", "e_mut_h3k27c"),each=2)
average=c(b=c(3.5,0),e=c(2.3,0))
rep=c(b=c(3.6,3.3), e=c(2.2, 2.4))

df <- data.frame(motif_name, average, rep, activity)

p <- ggplot(data=df, aes(x=motif_name, y=average)) +facet_wrap(~activity, scales = "free", ncol =3)+
    geom_bar(stat="identity",fill="brown",width=0.7)+geom_point(stat = "identity", aes(y=rep))+theme_classic2()+ylab("log2 fold-change(H3K27ac/WCE)")+xlab("")
p
ggsave("figures/14_crispr_analysis/rin3_h3k27ac_log2_enhanceractivity_barplot.pdf",p,height= 4, width=6)
```

#predicted signal at rin3 and ezr
```{r,predicted signal,fig.height = 6, fig.width = 6, fig.align = "center"}

## ezr
profile_preds.df <- readr::read_csv('csv/for_preds/predictions/can_ezr_predict.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_ezr_coordinates.csv')
profile_preds_pos.df <- profile_preds.df %>% dplyr::filter(position>=250 & position<=450)
tead4_profile_preds_pos.df <- profile_preds_pos.df %>% dplyr::filter(task=="tead4")
tead4_profile_preds_pos.df$region_idx %>% table()
wt_preds.df <- dplyr::filter(tead4_profile_preds_pos.df,region_idx==0)
mut_preds1.df <- dplyr::filter(tead4_profile_preds_pos.df,region_idx==1)


g <- ggplot()
for(row in 1:nrow(motifs.df)){
  g<-g + annotate("rect", xmin = motifs.df$start[row], xmax = motifs.df$end[row],
                  ymin = -Inf, ymax = Inf, alpha = .2)
  g<-g + annotate("text", x = motifs.df$start[row], label = motifs.df$name[row], y = Inf, vjust = 1.5)
}
g<- g+ geom_area(data = wt_preds.df, mapping = aes(position, prediction, group = strand, fill = task), alpha = .7)+
  geom_line(data = mut_preds1.df, mapping = aes(position, prediction, group = strand, color = task))+
  facet_grid(task ~ ., scales = "fixed")+
  scale_x_continuous(limits = c(250,450),
                     name = "Position (bp)")+
  scale_y_continuous(name = "BPNet predictions")+
  scale_fill_manual(values = c( '#d896ff'), name = "WT")+
  scale_color_manual(values = c('#800080'), name = "Mut")+
  theme_classic()+ggtitle("Ezr: WT & mut")
g

ggsave("figures/14_crispr_analysis/tead4_predicted_profile_at_ezr.pdf",g,height=4, width=8)
ggsave("figures/14_crispr_analysis/tead4_predicted_profile_at_ezr.png",g,height=4, width=8)


## rin3
profile_preds.df <- readr::read_csv('csv/for_preds/predictions/can_rin3_predict.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_rin3_coordinates.csv')
profile_preds_pos.df <- profile_preds.df %>% dplyr::filter(position>=350 & position<=550)
tead4_profile_preds_pos.df <- profile_preds_pos.df %>% dplyr::filter(task=="tead4")
tead4_profile_preds_pos.df$region_idx %>% table()
wt_preds.df <- dplyr::filter(tead4_profile_preds_pos.df,region_idx==0)
mut_preds1.df <- dplyr::filter(tead4_profile_preds_pos.df,region_idx==1)
mut_preds2.df <- dplyr::filter(tead4_profile_preds_pos.df,region_idx==2)
mut_preds3.df <- dplyr::filter(tead4_profile_preds_pos.df,region_idx==3)

g <- ggplot()
for(row in 1:nrow(motifs.df)){
  g<-g + annotate("rect", xmin = motifs.df$start[row], xmax = motifs.df$end[row],
                  ymin = -Inf, ymax = Inf, alpha = .2)
  g<-g + annotate("text", x = motifs.df$start[row], label = motifs.df$name[row], y = Inf, vjust = 1.5)
}
g<- g+ geom_area(data = wt_preds.df, mapping = aes(position, prediction, group = strand, fill = task), alpha = .7)+
  geom_line(data = mut_preds2.df, mapping = aes(position, prediction, group = strand, color = task))+
  facet_grid(task ~ ., scales = "fixed")+
  scale_x_continuous(limits = c(350,550),
                     name = "Position (bp)")+
  scale_y_continuous(name = "BPNet predictions")+
  scale_fill_manual(values = c( '#66b2b2'), name = "WT")+
  scale_color_manual(values = c('#006666'), name = "Mut")+
  theme_classic()+ggtitle("Rin3: WT & mut")
g

ggsave("figures/14_crispr_analysis/tead4_predicted_profile_at_rin3.pdf",g,height=4, width=8)
ggsave("figures/14_crispr_analysis/tead4_predicted_profile_at_rin3.png",g,height=4, width=8)
```


# best fit line at tead4 ear and rin3
```{r,ezr best fit,fig.height = 6, fig.width = 6, fig.align = "center"}

## regions coordinates of luciferase assay
gr <- GRanges(seqnames = c("chr17","chr12","chr15","chr9","chr7","chr1","chr2","chr19","chr10"), strand = c("+","+","+","+","+","+","+","+","+"),
ranges = IRanges(start = c(6827588,102261800,102016018,102726228,65430364,33980924,172759732,55713879,17579465), end=c(6828029,102262303,102016539,102726721,65430848,33981418,172760998,55714382,17579914)))

## ezr
#tead4 peaks
wt_td_peaks.gr <- read_narrowpeaks('../../data_preparation/crispr/peaks/mm10/nexus/combined/mtsc_tead4_nexus_combined_peaks.narrowPeak') %>% keepStandardChromosomes(., pruning.mode = "coarse")
ezr_td_peaks.gr <- read_narrowpeaks('../../data_preparation/crispr/peaks/mm10_mtsc_h32_ezr_td_singletodouble_mut/nexus/combined/mtsc_ezrmut_tead4_nexus_combined_peaks.narrowPeak')%>% keepStandardChromosomes(., pruning.mode = "coarse")


# find overlaps
peaks_common_ezr.gr <- findOverlapPairs(ezr_td_peaks.gr, wt_td_peaks.gr, ignore.strand=T)

# ezr overlap with motifs
peaks_sel <- peaks_common_ezr.gr@second
region_of_interest <- findOverlaps(peaks_sel, resize(gr_ezr_motif,500,"center"), ignore.strand=T)
peak_with_ezr_region_motifs <- peaks_sel[region_of_interest@from]
peaks_resize.gr <- resize(peaks_sel, 500,"center")

# calculate signal
peaks_resize.gr$signal_wt <- nexus_regionSums(peaks_resize.gr, ezronly_crispr_chip_set_rpm.bw.list$wt_tead4_c)
peaks_resize.gr$signal_mut <- nexus_regionSums(peaks_resize.gr, ezronly_crispr_chip_set_rpm.bw.list$ezrmut_tead4_c)
peaks_resize.df <- as.data.frame(peaks_resize.gr)
peak_with_ezr_region_motifs$name
x <- unique(peak_with_ezr_region_motifs)
a <- peaks_resize.df[peaks_resize.df$name %in% x$name,]

#plot
g <- peaks_resize.df %>%
    ggplot(aes(x=log2(signal_wt),y=log2(signal_mut))) +
              ggrastr::geom_point_rast(alpha=0.3, color="grey") +
              ggrastr::geom_point_rast(data=peaks_resize.df  %>% dplyr::filter(name =="mtsc_tead4_nexus_combined_peak_3447" | name == "mtsc_tead4_nexus_combined_peak_30616"| name =="mtsc_tead4_nexus_combined_peak_86495"| name =="mtsc_tead4_nexus_combined_peak_138870"| name =="mtsc_tead4_nexus_combined_peak_154095"| name =="mtsc_tead4_nexus_combined_peak_219378"| name == "mtsc_tead4_nexus_combined_peak_315279" | name =="mtsc_tead4_nexus_combined_peak_190953"| name=="mtsc_tead4_nexus_combined_peak_364433"),
              aes(x=log2(signal_wt),y=log2(signal_mut)),color='#038948',size=1.5)+
              ggrastr::geom_point_rast(data=a,aes(x=log2(signal_wt),y=log2(signal_mut)),color='red',size=1.5)+geom_abline(intercept = 0, 
              linetype =  "dashed", colour="black")+xlab("log2(WT)")+ylab("log2(Ezr CRISPR)")+ggtitle("Tead4 ChIP-nexus")+
              stat_cor(data =  peaks_resize.df, aes(x = log2(signal_wt), y = log2(signal_mut)),method = "pearson")+theme_classic()

m_theme = theme(
  axis.title.x = element_text(size = 16),
  axis.text.x = element_text(size = 14),
  axis.title.y = element_text(size = 16),
  axis.text.y = element_text(size = 14))
z <- g + m_theme
z
ggsave("figures/14_crispr_analysis/tead4_ezr_correlation_wtmut_bestfit.pdf",z,height= 6, width=6)

```

# rin3 best fit

```{r, rin3 tead4 best fit,fig.height = 6, fig.width = 6, fig.align = "center"}


wt_td_peaks.gr <- read_narrowpeaks('../../data_preparation/crispr/peaks/mm10/nexus/combined/mtsc_tead4_nexus_combined_peaks.narrowPeak') %>% keepStandardChromosomes(., pruning.mode = "coarse")
rin3_td_peaks.gr <- read_narrowpeaks('../../data_preparation/crispr/peaks/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/nexus/combined/mtsc_rin3mut_tead4_nexus_combined_peaks.narrowPeak')%>% keepStandardChromosomes(., pruning.mode = "coarse")

# find overlaps
peaks_common_rin3.gr <- findOverlapPairs(rin3_td_peaks.gr, wt_td_peaks.gr, ignore.strand=T)

# rin3 overlap with motifs
peaks_sel <- peaks_common_rin3.gr@second
region_of_interest <- findOverlaps(peaks_sel, resize(gr,500,"center"), ignore.strand=T)
peak_with_rin3_region_motifs <- peaks_sel[region_of_interest@from]
peaks_resize.gr <- resize(peaks_sel, 500,"center")

# calculate signal
peaks_resize.gr$signal_wt <- nexus_regionSums(peaks_resize.gr, rin3only_crispr_chip_set_rpm.bw.list$wt_tead4_c)
peaks_resize.gr$signal_mut <- nexus_regionSums(peaks_resize.gr, rin3only_crispr_chip_set_rpm.bw.list$rin3mut_tead4_c)
peaks_resize.df <- as.data.frame(peaks_resize.gr)
peak_with_rin3_region_motifs$name
x <- unique(peak_with_rin3_region_motifs)
a <- peaks_resize.df[peaks_resize.df$name %in% x$name,]
a<- a[3,]

# plot
r <- peaks_resize.df %>%
    ggplot(aes(x=log2(signal_wt),y=log2(signal_mut))) +
    ggrastr::geom_point_rast(alpha=0.3, color="grey") +
    ggrastr::geom_point_rast(data=peaks_resize.df %>% dplyr::filter(name =="mtsc_tead4_nexus_combined_peak_3447" | name == "mtsc_tead4_nexus_combined_peak_30616"| name =="mtsc_tead4_nexus_combined_peak_86495"| name =="mtsc_tead4_nexus_combined_peak_138870"| name =="mtsc_tead4_nexus_combined_peak_154095"| name =="mtsc_tead4_nexus_combined_peak_219378"| name == "mtsc_tead4_nexus_combined_peak_315279" | name=="mtsc_tead4_nexus_combined_peak_364433"),
              aes(x=log2(signal_wt),y=log2(signal_mut)),color='#038948',size=1.5)+
              ggrastr::geom_point_rast(data=a,aes(x=log2(signal_wt),y=log2(signal_mut)),color='red',size=1.5)+
              geom_abline(intercept = 0, linetype = "dashed", colour="black")+xlab("log2(WT)")+ylab("log2(Rin3 CRISPR)")+
              ggtitle("Tead4 ChIP-nexus")+stat_cor(data =  peaks_resize.df, aes(x = log2(signal_wt), y = log2(signal_mut)),
                                                   method = "pearson")+theme_classic()

m_theme = theme(
  axis.title.x = element_text(size = 16),
  axis.text.x = element_text(size = 14),
  axis.title.y = element_text(size = 16),
  axis.text.y = element_text(size = 14))
rx <- r + m_theme

# gr_1 <- GRanges(seqnames = c("chr17","chr12"), strand = c("+","+"),
#                 ranges = IRanges(start = c(6827588,102261800), end=c(6828029,102262303)))
# gr_rin3 <- gr_1[2]
# e<- findOverlaps(peaks_sel, resize(gr_rin3,500,"center"), ignore.strand=T)
# ee <- peaks_sel[e@from]

ggsave("figures/14_crispr_analysis/tead4_rin3_correlation_wtmut_bestfit.pdf",rx,height= 6, width=4)
```


```{r, rin3 polii best fit,fig.height = 6, fig.width = 6, fig.align = "center"}

wt_polii_peaks.gr <- read_narrowpeaks('../../data_preparation/crispr/peaks/mm10/nexus/combined/mtsc_polii_nexus_combined_peaks.narrowPeak') %>% keepStandardChromosomes(., pruning.mode = "coarse")
rin3_polii_peaks.gr <- read_narrowpeaks('../../data_preparation/crispr/peaks/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/nexus/combined/mtsc_rin3mut_polii_nexus_combined_peaks.narrowPeak')%>% keepStandardChromosomes(., pruning.mode = "coarse")
peaks_common_rin3.gr <- findOverlapPairs(rin3_polii_peaks.gr, wt_polii_peaks.gr, ignore.strand=T)

peaks_sel <- peaks_common_rin3.gr@second
region_of_interest <- findOverlaps(peaks_sel, resize(gr,500,"center"), ignore.strand=T)
peak_with_rin3_region_motifs <- peaks_sel[region_of_interest@from]
peaks_resize.gr <- resize(peaks_sel, 500,"center")

peaks_resize.gr$signal_polii_wt <- nexus_regionSums(peaks_resize.gr, rin3only_crispr_chip_set_rpm.bw.list$wt_polii_c)
peaks_resize.gr$signal_polii_mut <- nexus_regionSums(peaks_resize.gr, rin3only_crispr_chip_set_rpm.bw.list$rin3mut_polii_c)
peaks_resize.df <- as.data.frame(peaks_resize.gr)
peak_with_rin3_region_motifs$name
x <- unique(peak_with_rin3_region_motifs)
a <- peaks_resize.df[peaks_resize.df$name %in% x$name,]
a <- unique(a)
a<- a[3,]

rp <- peaks_resize.df %>%
    ggplot(aes(x=log2(signal_polii_wt),y=log2(signal_polii_mut))) +
    ggrastr::geom_point_rast(alpha=0.3, color="grey") +
    ggrastr::geom_point_rast(data=peaks_resize.df %>% dplyr::filter(name =="mtsc_polii_nexus_combined_peak_1352" | name == "mtsc_polii_nexus_combined_peak_15732" | name == "mtsc_polii_nexus_combined_peak_50617" | name == "mtsc_polii_nexus_combined_peak_79731"  | name == "mtsc_polii_nexus_combined_peak_88871" | name ==  "mtsc_polii_nexus_combined_peak_112011"| name ==  "mtsc_polii_nexus_combined_peak_130491" | name == "mtsc_polii_nexus_combined_peak_188711" | name == "mtsc_polii_nexus_combined_peak_218716"),aes(x=log2(signal_polii_wt),y=log2(signal_polii_mut)),color='#038948',size=1.5)+
    ggrastr::geom_point_rast(data=a %>% dplyr::filter(name =="mtsc_polii_nexus_combined_peak_50617"),           aes(x=log2(signal_polii_wt),y=log2(signal_polii_mut)),color='red',size=1.5)+geom_abline(intercept = 0, linetype = "dashed",colour="black")+ xlab("log2(WT)")+ ylab("log2(Rin3 CRISPR)")+ggtitle("PolII ChIP-nexus")+ stat_cor(data =  peaks_resize.df, aes(x = log2(signal_polii_wt), y = log2(signal_polii_mut)),method = "pearson")+theme_classic()

m_theme = theme(
    axis.title.x = element_text(size = 16),
    axis.text.x = element_text(size = 14),
    axis.title.y = element_text(size = 16),
    axis.text.y = element_text(size = 14))
rx_p <- rp + m_theme

# e<- findOverlaps(peaks_sel, resize(gr_rin3,500,"center"), ignore.strand=T)
# ee <- peaks_sel[e@from]

ggsave("figures/14_crispr_analysis/polii_rin3_correlation_wtmut_bestfit.pdf",g,height= 6, width=6)
```

```{r, rin3 h3k27ac best fit,fig.height = 6, fig.width = 6, fig.align = "center"}

wt_h3k_peaks.gr <- read_narrowpeaks('../../data_preparation/crispr/peaks/mm10/seq/combined/mtsc_h3k27ac_seq_combined_peaks.narrowPeak') %>% keepStandardChromosomes(., pruning.mode = "coarse")
rin3_h3k_peaks.gr <- read_narrowpeaks('../../data_preparation/crispr/peaks/mm10_mtsc_h32_rin3_ttdis_mut_2bpdel/seq/combined/mtsc_rin3mut_h3k27ac_seq_combined_peaks.narrowPeak')%>% keepStandardChromosomes(., pruning.mode = "coarse")
peaks_common_rin3.gr <- findOverlapPairs(rin3_h3k_peaks.gr, wt_h3k_peaks.gr, ignore.strand=T)

peaks_sel <- peaks_common_rin3.gr@second
region_of_interest <- findOverlaps(peaks_sel, resize(gr,500,"center"), ignore.strand=T)
peak_with_rin3_region_motifs <- peaks_sel[region_of_interest@from]
peaks_resize.gr <- resize(peaks_sel, 500,"center")

peaks_resize.gr$signal_h3k_wt <- regionSums(resize(peaks_resize.gr, 1000, "center"), rin3only_crispr_chip_set_rpm.bw.list$wt_h3k27ac_c$pos)
peaks_resize.gr$signal_h3k_mut <- regionSums(resize(peaks_resize.gr, 1000, "center"), rin3only_crispr_chip_set_rpm.bw.list$rin3mut_h3k27ac_c$pos)
peaks_resize.df <- as.data.frame(peaks_resize.gr)
peak_with_rin3_region_motifs$name
x <- unique(peak_with_rin3_region_motifs)
a <- peaks_resize.df[peaks_resize.df$name %in% x$name,]
a<- unique(a)
a<-a[3,]

hr <- peaks_resize.df %>%
        ggplot(aes(x=log2(signal_h3k_wt),y=log2(signal_h3k_mut))) +
        ggrastr::geom_point_rast(alpha=0.3, color="grey") +
        ggrastr::geom_point_rast(data=peaks_resize.df %>% dplyr::filter(name ==  "mtsc_h3k27ac_seq_combined_peak_680"  | name ==   "mtsc_h3k27ac_seq_combined_peak_10247" | name ==  "mtsc_h3k27ac_seq_combined_peak_37457" | name == "mtsc_h3k27ac_seq_combined_peak_59044" | name ==  "mtsc_h3k27ac_seq_combined_peak_65322" | name ==  "mtsc_h3k27ac_seq_combined_peak_83799" | name == "mtsc_h3k27ac_seq_combined_peak_96976" | name ==  "mtsc_h3k27ac_seq_combined_peak_138813" | name ==  "mtsc_h3k27ac_seq_combined_peak_138814"| name ==  "mtsc_h3k27ac_seq_combined_peak_161977"),
      aes(x=log2(signal_h3k_wt),y=log2(signal_h3k_mut)),color='#038948',size=1.5)+
      ggrastr::geom_point_rast(data=a %>% dplyr::filter(name =="mtsc_h3k27ac_seq_combined_peak_37457"),aes(x=log2(signal_h3k_wt),y=log2(signal_h3k_mut)),
     color='red',size=1.5)+geom_abline(intercept = 0, linetype = "dashed", colour="black")+xlab("log2(WT)")+ylab("log2(Rin3 CRISPR)")+ggtitle("H3K27ac ChIP-seq")+
        stat_cor(data =  peaks_resize.df, aes(x = log2(signal_h3k_wt), y = log2(signal_h3k_mut)),method = "pearson")+theme_classic()
    
m_theme = theme(
        axis.title.x = element_text(size = 16),
        axis.text.x = element_text(size = 14),
        axis.title.y = element_text(size = 16),
        axis.text.y = element_text(size = 14))
rx_h <- hr + m_theme
rx_h

ggsave("figures/14_crispr_analysis/h3k27ac_rin3_correlation_wtmut_bestfit.pdf",g,height= 6, width=6)
    
g<-rx + rx_h +rx_p + patchwork::plot_layout(nrow = 1, widths = c(.5, .5, .5))

```  


 # ezr TPM plot 
```{r, ezr_tpm,fig.height = 6, fig.width = 6, fig.align = "center"}
##TPM values : WT and Ezr 
#(values for TPM came from file : /n/projects/kd2200/publication/data_preparation/rsem/mm10/rna/tables) for Ezr (ENSMUSG00000052397) and Dynlt1f (ENSMUSG00000095677)

motif_name=rep(c("a_Ezr_wt", "b_Ezr_mut","c_Dynlt1f_wt","d_Dynlt1f_mut"),each=3)
average=c(a=c(512,0,0),b=c(678,0,0),c=c(307,0,0),d=c(282,0,0))
rep=c(a=c(459.22,496.08,582.65),b=c(668.76, 650.3,713.1),c=c(331.63,281.76,310.04),d=c(284.87,274.68,285.88))

df <- data.frame(motif_name, average, rep)

p <- ggplot(data=df, aes(x=motif_name, y=average)) +
    geom_bar(stat="identity",fill="lightblue")+geom_point(stat = "identity", aes(y=rep))+theme_classic2()+ylab(paste("TPM"))+xlab("")
p
ggsave("figures/14_crispr_analysis/ezr_tpm_barplot.pdf",p,height= 6, width=6)

library(rstatix)
library(ggpubr)
stat.test <- df %>%
    pairwise_t_test(
        rep ~ motif_name, paired = FALSE,
        p.adjust.method = "bonferroni"
    )
stat.test

stat.test <- stat.test %>%
    mutate(y.position = c(800))
z <- p+stat_pvalue_manual(
    stat.test,  label = "p.adj.signif", tip.length = 0,y.position = "y.position")

ggsave("figures/14_crispr_analysis/ezr_tpm_barplot_stats.pdf",z,height= 6, width=6)



```


# SessionInfo

For the purposes of reproducibility, the R/Bioconductor session information is printed below:

```{r sessioninfo}
sessionInfo()
```






























