---
title: "plot predictions and contribution for luciferase region"
author: "Khyati Dalal"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    depth: 3
    theme: sandstone
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
  word_document:
    toc: yes
header-includes:
- \usepackage{fancyhdr}
- \usepackage{lipsum}
- \pagestyle{fancy}
- \fancyfoot[CO,CE]{1028-02098-001-001}
- \fancyfoot[LE,RO]{\thepage}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options:
  chunk_output_type: console
---
# Introduction

The analysis is about plotting the predictions and contribution at the regions that I used for lucifearse assay.
   
# Computational Setup

```{r, warning=F, message=F}
#Standard packages
library(rtracklayer) ; library(GenomicRanges); library(magrittr) ; library(Biostrings);library(parallel);library(data.table)
library(ggplot2) ; library(reshape2); library(plyranges); library(Rsamtools);library(BSgenome.Mmusculus.UCSC.mm10);library(xtable)
library(dplyr); library(data.table); library(patchwork); library(readr);library(TxDb.Mmusculus.UCSC.mm10.knownGene)

#KNITR Options
setwd("/n/projects/kd2200/publication/bpnet/analysis/")
options(knitr.figure_dir="figures/15_luci_regions_predict_plots/", java.parameters = "- Xmx6g")

#Lab sources
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/granges_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/knitr_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/caching.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_functions.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/multiplot.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/motif_functions.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_new.r")

#Specific packages
library(scales)
library(ggseqlogo)

```


```{r, functions}

#Functions
myceiling<-function(x, base = 10){
  return( ceiling(x / base) * base)
}
myfloor<-function(x, base = 10){
  return( floor(x / base) * base)
}
```

## Read in predictions generated from the `15_.ipynb`

## cited2 contribution
```{r,cited2 contribution}

profile_contrib.df <- readr::read_csv('csv/for_preds/predictions/can_cited2_contrib.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_cited2_coordinates.csv')
wt_contrib.df <- dplyr::filter(profile_contrib.df,region_idx==0)
mut_contrib1.df <- dplyr::filter(profile_contrib.df,region_idx==1)
mut_contrib2.df <- dplyr::filter(profile_contrib.df,region_idx==2)
mut_contrib3.df <- dplyr::filter(profile_contrib.df,region_idx==3)
mut_contrib4.df <- dplyr::filter(profile_contrib.df,region_idx==4)
mut_contrib5.df <- dplyr::filter(profile_contrib.df,region_idx==5)
mut_contrib6.df <- dplyr::filter(profile_contrib.df,region_idx==6)

xlim_genomic<-c(myfloor(motifs.df$start, 50)-50, myceiling(motifs.df$end, 50)+50)
xlim_window<-xlim_genomic- motifs.df$start

plot_contribution_scores<-function(contrib.df, motifs_across_window.df, xlim, subset_by = "task"){
    
    #Format to ggseqlogo
    mat.list<-lapply(unique(contrib.df[[subset_by]]), function(x){
        c.df<-contrib.df[contrib.df[[subset_by]]==x,]
        c.df %>% dplyr::filter(task == x) %>% .[,1:4] %>% t %>% as.matrix
    })
    names(mat.list)<-unique(contrib.df$task)
    
    #Plot using ggseqlogo
    g<-ggplot()
    
    #Add in motif boxes
    for(row in 1:nrow(motifs_across_window.df)){
        g<-g + annotate("rect", xmin = motifs_across_window.df$start[row], xmax = motifs_across_window.df$end[row],
                        ymin = -Inf, ymax = Inf, alpha = .2)
        g<-g + annotate("text", x = motifs_across_window.df$start[row], label = motifs_across_window.df$name[row], y = Inf, vjust = 1.5)
    }
    
    g<-g + geom_logo(data = mat.list, method='custom', seq_type='dna') + 
        scale_x_continuous(limits = c(200,400))+
        scale_y_continuous(name = 'contribution')+
        facet_wrap(~seq_group, ncol=1, scales='free')+
        theme_classic()+ggtitle("cited2")+
        theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())
    return(g)
    
}


a <- plot_contribution_scores(contrib.df = wt_contrib.df  %>% dplyr::filter(region_idx==0), 
                            motifs_across_window.df = motifs.df)
b <- plot_contribution_scores(contrib.df = mut_contrib1.df %>% dplyr::filter(region_idx==1), 
                            motifs_across_window.df = motifs.df)
c <- plot_contribution_scores(contrib.df = mut_contrib2.df  %>% dplyr::filter(region_idx==2), 
                          motifs_across_window.df = motifs.df)
d <- plot_contribution_scores(contrib.df = mut_contrib3.df  %>% dplyr::filter(region_idx==3), 
                          motifs_across_window.df = motifs.df)
e <- plot_contribution_scores(contrib.df = mut_contrib4.df  %>% dplyr::filter(region_idx==4), 
                          motifs_across_window.df = motifs.df)
f <- plot_contribution_scores(contrib.df = mut_contrib5.df  %>% dplyr::filter(region_idx==5), 
                          motifs_across_window.df = motifs.df)
g <- plot_contribution_scores(contrib.df = mut_contrib6.df  %>% dplyr::filter(region_idx==6), 
                          motifs_across_window.df = motifs.df)



 #Save plots
   # ggsave("predictions_contributions/can_cited2_wt.pdf",a,height = 12, width = 24)
   # ggsave("predictions_contributions/can_cited2_good_center.pdf",b,height =12, width = 24)
   # ggsave("predictions_contributions/can_cited2_good_start.pdf",c,height = 12, width = 24)
   # ggsave("predictions_contributions/can_cited2_good_end.pdf",d,height = 12, width = 24)
   # ggsave("predictions_contributions/can_cited2_bad_center.pdf",e,height = 12, width = 24)
   # ggsave("predictions_contributions/can_cited2_bad_start.pdf",f,height = 12, width = 24)
   # ggsave("predictions_contributions/can_cited2_bad_end.pdf",g,height = 12, width = 24)
```

# cited2 prediction
```{r,cited2 prediction}
profile_preds.df <- readr::read_csv('csv/for_preds/predictions/can_cited2_predict.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_cited2_coordinates.csv')
wt_preds.df <- dplyr::filter(profile_preds.df,region_idx==0)
mut_preds1.df <- dplyr::filter(profile_preds.df,region_idx==1)
mut_preds2.df <- dplyr::filter(profile_preds.df,region_idx==2)
mut_preds3.df <- dplyr::filter(profile_preds.df,region_idx==3)
mut_preds4.df <- dplyr::filter(profile_preds.df,region_idx==4)
mut_preds5.df <- dplyr::filter(profile_preds.df,region_idx==5)
mut_preds6.df <- dplyr::filter(profile_preds.df,region_idx==6)

g <- ggplot()

for(row in 1:nrow(motifs.df)){
    g<-g + annotate("rect", xmin = motifs.df$start[row], xmax = motifs.df$end[row],
                    ymin = -Inf, ymax = Inf, alpha = .2)
    g<-g + annotate("text", x = motifs.df$start[row], label = motifs.df$name[row], y = Inf, vjust = 1.5)
}

g<- g+ geom_area(data = wt_preds.df, mapping = aes(position, prediction, group = strand, fill = task), alpha = .7)+
    geom_line(data = mut_preds1.df, mapping = aes(position, prediction, group = strand, color = task))+
    facet_grid(task ~ ., scales = "free")+
    scale_x_continuous(limits = c(200,400),
                       name = "Position (bp)")+
    scale_y_continuous(name = "BPNet predictions")+
    scale_fill_manual(values = c('#6e828e', '#7f778c', '#b47382','#788e6e', '#968b71', '#96718f'), name = "WT")+
    scale_color_manual(values = c('#314d5e', '#493d5c','#94384d','#3e8f1a', '#a87c14','#8c1f78'), name = "Mut")+
    theme_classic()+ggtitle("can_cited2")
  
    # ggsave("predictions_contributions/can_cited2_wt_mut_good_center.pdf",g,height = 20, width = 8)
    # ggsave("predictions_contributions/can_cited2_wt_mut_good_start.pdf",g,height = 20, width = 8)
    # ggsave("predictions_contributions/can_cited2_wt_mut_good_end.pdf",g,height = 20, width = 8)
    # ggsave("predictions_contributions/can_cited2_wt_mut_bad_center.pdf",g,height = 20, width = 8)
    # ggsave("predictions_contributions/can_cited2_wt_mut_bad_start.pdf",g,height = 20, width = 8)
    # ggsave("predictions_contributions/can_cited2_wt_mut_bad_end.pdf",g,height = 20, width = 8)
```

## Tjp1 
# tjp1 contributions
```{r,tjp1 contributions}
profile_contrib.df <- readr::read_csv('csv/for_preds/predictions/can_tjp1_contrib_negstrand.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_tjp1_coordinates_negstrand.csv')
wt_contrib.df <- dplyr::filter(profile_contrib.df,region_idx==0)
mut_contrib1.df <- dplyr::filter(profile_contrib.df,region_idx==1)
mut_contrib2.df <- dplyr::filter(profile_contrib.df,region_idx==2)
mut_contrib3.df <- dplyr::filter(profile_contrib.df,region_idx==3)


xlim_genomic<-c(myfloor(motifs.df$start, 50)-50, myceiling(motifs.df$end, 50)+50)
xlim_window<-xlim_genomic- motifs.df$start

plot_contribution_scores<-function(contrib.df, motifs_across_window.df, xlim, subset_by = "task"){
    
    #Format to ggseqlogo
    mat.list<-lapply(unique(contrib.df[[subset_by]]), function(x){
        c.df<-contrib.df[contrib.df[[subset_by]]==x,]
        c.df %>% dplyr::filter(task == x) %>% .[,1:4] %>% t %>% as.matrix
    })
    names(mat.list)<-unique(contrib.df$task)
    
    #Plot using ggseqlogo
    g<-ggplot()
    
    #Add in motif boxes
    for(row in 1:nrow(motifs_across_window.df)){
        g<-g + annotate("rect", xmin = motifs_across_window.df$start[row], xmax = motifs_across_window.df$end[row],
                        ymin = -Inf, ymax = Inf, alpha = .2)
        g<-g + annotate("text", x = motifs_across_window.df$start[row], label = motifs_across_window.df$name[row], y = Inf, vjust = 1.5)
    }
    
    g<-g + geom_logo(data = mat.list, method='custom', seq_type='dna') + 
        scale_x_continuous(limits = c(400,600))+
        scale_y_continuous(name = 'contribution')+
        facet_wrap(~seq_group, ncol=1, scales='free')+
        theme_classic()+ggtitle("can_tjp1")+
        theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())
    return(g)
    
}


q <- plot_contribution_scores(contrib.df = wt_contrib.df  %>% dplyr::filter(region_idx==0), 
                            motifs_across_window.df = motifs.df)
 

w <- plot_contribution_scores(contrib.df = mut_contrib1.df %>% dplyr::filter(region_idx==1), 
                            motifs_across_window.df = motifs.df)

e <- plot_contribution_scores(contrib.df = mut_contrib2.df  %>% dplyr::filter(region_idx==2), 
                          motifs_across_window.df = motifs.df)

r <- plot_contribution_scores(contrib.df = mut_contrib3.df  %>% dplyr::filter(region_idx==3), 
                          motifs_across_window.df = motifs.df)


#Save plots
   # ggsave("predictions_contributions/can_tjp1_negstrand_wt.pdf",q,height = 12, width = 24)
   # ggsave("predictions_contributions/can_tjp1_negstrand_mut.pdf",w,height = 12, width = 24)
   # ggsave("predictions_contributions/can_tjp1_negstrand_firsthalfmut.pdf",e,height = 12, width = 24)
   # ggsave("predictions_contributions/can_tjp1_negstrand_secondhalfmut.pdf",r,height = 12, width = 24)

```

# tjp1 predictions
```{r,tjp1 predictions}
profile_preds.df <- readr::read_csv('csv/for_preds/predictions/can_tjp1_predict_negstrand.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_tjp1_coordinates_negstrand.csv')
wt_preds.df <- dplyr::filter(profile_preds.df,region_idx==0)
mut_preds1.df <- dplyr::filter(profile_preds.df,region_idx==1)
mut_preds2.df <- dplyr::filter(profile_preds.df,region_idx==2)
mut_preds3.df <- dplyr::filter(profile_preds.df,region_idx==3)

g <- ggplot()

for(row in 1:nrow(motifs.df)){
    g<-g + annotate("rect", xmin = motifs.df$start[row], xmax = motifs.df$end[row],
                    ymin = -Inf, ymax = Inf, alpha = .2)
    g<-g + annotate("text", x = motifs.df$start[row], label = motifs.df$name[row], y = Inf, vjust = 1.5)
}

g<- g+ geom_area(data = wt_preds.df, mapping = aes(position, prediction, group = strand, fill = task), alpha = .7)+
    geom_line(data = mut_preds1.df, mapping = aes(position, prediction, group = strand, color = task))+
    facet_grid(task ~ ., scales = "free")+
    scale_x_continuous(limits = c(400,600),
                       name = "Position (bp)")+
    scale_y_continuous(name = "BPNet predictions")+
    scale_fill_manual(values = c('#6e828e', '#7f778c', '#b47382','#788e6e', '#968b71', '#96718f'), name = "WT")+
    scale_color_manual(values = c('#314d5e', '#493d5c','#94384d','#3e8f1a', '#a87c14','#8c1f78'), name = "Mut")+
    theme_classic()+ggtitle("can_tjp1")
g


  ##save plots
    # ggsave("predictions_contributions/can_tjp1_negstrand_wt_mut_pred.pdf",g,height = 20, width = 8)
    # ggsave("predictions_contributions/can_tjp1_negstrand_wt_mut1_pred.pdf",g,height = 20, width = 8)
    # ggsave("predictions_contributions/can_tjp1_negstrand_wt_mut2_pred.pdf",g,height = 20, width = 8)
```


## ezr
# ezr predictions
```{r,ezr predictions}

profile_preds.df <- readr::read_csv('csv/for_preds/predictions/can_ezr_predict.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_ezr_coordinates.csv')
wt_preds.df <- dplyr::filter(profile_preds.df,region_idx==0)
mut_preds1.df <- dplyr::filter(profile_preds.df,region_idx==1)

g <- ggplot()

for(row in 1:nrow(motifs.df)){
    g<-g + annotate("rect", xmin = motifs.df$start[row], xmax = motifs.df$end[row],
                    ymin = -Inf, ymax = Inf, alpha = .2)
    g<-g + annotate("text", x = motifs.df$start[row], label = motifs.df$name[row], y = Inf, vjust = 1.5)
}

g<- g+ geom_area(data = wt_preds.df, mapping = aes(position, prediction, group = strand, fill = task), alpha = .7)+
    geom_line(data = mut_preds1.df, mapping = aes(position, prediction, group = strand, color = task))+
    facet_grid(task ~ ., scales = "free")+
    scale_x_continuous(limits = c(250,450),
                       name = "Position (bp)")+
    scale_y_continuous(name = "BPNet predictions")+
    scale_fill_manual(values = c('#6e828e', '#7f778c', '#b47382','#788e6e', '#968b71', '#96718f'), name = "WT")+
    scale_color_manual(values = c('#314d5e', '#493d5c','#94384d','#3e8f1a', '#a87c14','#8c1f78'), name = "Variant")+
    theme_classic()+ggtitle("can_ezr")
g
# ggsave("predictions_contributions/can_ezr_wt_tdbl_pred.pdf",g,height = 20, width = 8)
# ggsave("predictions_contributions/can_ezr_wt_only_pred.pdf",g,height = 20, width = 8)
# ggsave("predictions_contributions/can_ezr_tdbl_only_pred.pdf",g,height = 20, width = 8)

```

# ezr contributions
```{r,ezr contributions}
profile_contrib.df <- readr::read_csv('csv/for_preds/predictions/can_ezr_contrib.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_ezr_coordinates.csv')
wt_contrib.df <- dplyr::filter(profile_contrib.df,region_idx==0)
mut_contrib1.df <- dplyr::filter(profile_contrib.df,region_idx==1)
mut_contrib2.df <- dplyr::filter(profile_contrib.df,region_idx==2)


xlim_genomic<-c(myfloor(motifs.df$start, 50)-50, myceiling(motifs.df$end, 50)+50)
xlim_window<-xlim_genomic- motifs.df$start

plot_contribution_scores<-function(contrib.df, motifs_across_window.df, xlim, subset_by = "task"){
    
    #Format to ggseqlogo
    mat.list<-lapply(unique(contrib.df[[subset_by]]), function(x){
        c.df<-contrib.df[contrib.df[[subset_by]]==x,]
        c.df %>% dplyr::filter(task == x) %>% .[,1:4] %>% t %>% as.matrix
    })
    names(mat.list)<-unique(contrib.df$task)
    
    #Plot using ggseqlogo
    g<-ggplot()
    
    #Add in motif boxes
    for(row in 1:nrow(motifs_across_window.df)){
        g<-g + annotate("rect", xmin = motifs_across_window.df$start[row], xmax = motifs_across_window.df$end[row],
                        ymin = -Inf, ymax = Inf, alpha = .2)
        g<-g + annotate("text", x = motifs_across_window.df$start[row], label = motifs_across_window.df$name[row], y = Inf, vjust = 1.5)
    }
    
    g<-g + geom_logo(data = mat.list, method='custom', seq_type='dna') + 
        scale_x_continuous(limits = c(250,450))+
        scale_y_continuous(name = 'contribution')+
        facet_wrap(~seq_group, ncol=1, scales='free')+
        theme_classic()+ggtitle("can_ezr")+
        theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())
    return(g)
    
}

q <- plot_contribution_scores(contrib.df = wt_contrib.df  %>% dplyr::filter(region_idx==0), 
                              motifs_across_window.df = motifs.df)
w <- plot_contribution_scores(contrib.df = mut_contrib1.df %>% dplyr::filter(region_idx==1),
                               motifs_across_window.df = motifs.df)

# ggsave("predictions_contributions/can_ezr_wt.pdf",q,height = 12, width = 24)
# ggsave("predictions_contributions/can_ezr_tdbl.pdf",w,height =12, width = 24)

```


## rin3
# rin3 predictions
```{r, rin3 predictions}

profile_preds.df <- readr::read_csv('csv/for_preds/predictions/can_rin3_predict.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_rin3_coordinates.csv')
wt_preds.df <- dplyr::filter(profile_preds.df,region_idx==0)
mut_preds1.df <- dplyr::filter(profile_preds.df,region_idx==1)
mut_preds2.df <- dplyr::filter(profile_preds.df,region_idx==2)
mut_preds3.df <- dplyr::filter(profile_preds.df,region_idx==3)

g <- ggplot()

for(row in 1:nrow(motifs.df)){
    g<-g + annotate("rect", xmin = motifs.df$start[row], xmax = motifs.df$end[row],
                    ymin = -Inf, ymax = Inf, alpha = .2)
    g<-g + annotate("text", x = motifs.df$start[row], label = motifs.df$name[row], y = Inf, vjust = 1.5)
}

g<- g+ geom_area(data = wt_preds.df, mapping = aes(position, prediction, group = strand, fill = task), alpha = .7)+
    geom_line(data = mut_preds1.df, mapping = aes(position, prediction, group = strand, color = task))+
    facet_grid(task ~ ., scales = "free")+
    scale_x_continuous(limits = c(350,550),
                       name = "Position (bp)")+
    scale_y_continuous(name = "BPNet predictions")+
    scale_fill_manual(values = c('#6e828e', '#7f778c', '#b47382','#788e6e', '#968b71', '#96718f'), name = "WT")+
    scale_color_manual(values = c('#314d5e', '#493d5c','#94384d','#3e8f1a', '#a87c14','#8c1f78'), name = "Variant")+
    theme_classic()+ggtitle("can_rin3")
g
# ggsave("predictions_contributions/can_rin3_wt_mut_pred.pdf",g,height = 20, width = 8)
# ggsave("predictions_contributions/can_rin3_wt_2bpdelmut_pred.pdf",g,height = 20, width = 8)
# ggsave("predictions_contributions/can_rin3_wt_2bpdelmut_onesnp_pred.pdf",g,height = 20, width = 8)

```

# rin3 contributions
```{r, rin3 contributions}
profile_contrib.df <- readr::read_csv('csv/for_preds/predictions/can_rin3_contrib.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_rin3_coordinates.csv')
wt_contrib.df <- dplyr::filter(profile_contrib.df,region_idx==0)
mut_contrib1.df <- dplyr::filter(profile_contrib.df,region_idx==1)
mut_contrib2.df <- dplyr::filter(profile_contrib.df,region_idx==2)
mut_contrib3.df <- dplyr::filter(profile_contrib.df,region_idx==3)

xlim_genomic<-c(myfloor(motifs.df$start, 50)-50, myceiling(motifs.df$end, 50)+50)
xlim_window<-xlim_genomic- motifs.df$start

plot_contribution_scores<-function(contrib.df, motifs_across_window.df, xlim, subset_by = "task"){
    
    #Format to ggseqlogo
    mat.list<-lapply(unique(contrib.df[[subset_by]]), function(x){
        c.df<-contrib.df[contrib.df[[subset_by]]==x,]
        c.df %>% dplyr::filter(task == x) %>% .[,1:4] %>% t %>% as.matrix
    })
    names(mat.list)<-unique(contrib.df$task)
    
    #Plot using ggseqlogo
    g<-ggplot()
    
    #Add in motif boxes
    for(row in 1:nrow(motifs_across_window.df)){
        g<-g + annotate("rect", xmin = motifs_across_window.df$start[row], xmax = motifs_across_window.df$end[row],
                        ymin = -Inf, ymax = Inf, alpha = .2)
        g<-g + annotate("text", x = motifs_across_window.df$start[row], label = motifs_across_window.df$name[row], y = Inf, vjust = 1.5)
    }
    
    g<-g + geom_logo(data = mat.list, method='custom', seq_type='dna') + 
        scale_x_continuous(limits = c(350,550))+
        scale_y_continuous(name = 'contribution')+
        facet_wrap(~seq_group, ncol=1, scales='free')+
        theme_classic()+ggtitle("can_ezr")+
        theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())
    return(g)
    
}

q <- plot_contribution_scores(contrib.df = wt_contrib.df  %>% dplyr::filter(region_idx==0), 
                              motifs_across_window.df = motifs.df)
w <- plot_contribution_scores(contrib.df = mut_contrib1.df %>% dplyr::filter(region_idx==1),
                               motifs_across_window.df = motifs.df)
e <- plot_contribution_scores(contrib.df = mut_contrib2.df %>% dplyr::filter(region_idx==2),
                               motifs_across_window.df = motifs.df)
r <- plot_contribution_scores(contrib.df = mut_contrib3.df %>% dplyr::filter(region_idx==3),
                               motifs_across_window.df = motifs.df)

# ggsave("predictions_contributions/can_rin3_wt.pdf",q,height = 12, width = 24)
# ggsave("predictions_contributions/can_rin3_mut.pdf",w,height =12, width = 24)
# ggsave("predictions_contributions/can_rin3_2bpdelmut.pdf",e,height = 12, width = 24)
# ggsave("predictions_contributions/can_rin3_2bpdelmut_onesnp.pdf",r,height =12, width = 24)

```

## can bmp7 (yap1 task prediction)
```{r, bmp7 predictions}

profile_preds.df <- readr::read_csv('csv/for_preds/predictions/can_bamp7_tty_predict_fold5.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_bmp7_coordinates.csv')

wt_profile.df <- dplyr::filter(profile_preds.df,region_idx==0) %>% dplyr::mutate(.,name="a_wildtype")
mut_profile1.df <- dplyr::filter(profile_preds.df,region_idx==1)%>% dplyr::mutate(.,name="b_complete_mut")
mut_profile2.df <- dplyr::filter(profile_preds.df,region_idx==2)%>% dplyr::mutate(.,name="c_tfap2c_mut")
mut_profile3.df <- dplyr::filter(profile_preds.df,region_idx==3)%>% dplyr::mutate(.,name="d_tead4_mut")
mut_profile4.df <- dplyr::filter(profile_preds.df,region_idx==4)%>% dplyr::mutate(.,name="e_yap1_mut")

profile_preds.df <- rbind(wt_profile.df ,mut_profile1.df ,mut_profile2.df ,mut_profile3.df ,mut_profile4.df )
  
sub_wt_preds.df <- profile_preds.df %>% dplyr::filter(position %in% (425:575))
sums_sub_preds.df <- sub_wt_preds.df %>% dplyr::group_by(task,name) %>% dplyr::mutate(avg=mean(abs(prediction)))
td_single <- sums_sub_preds.df  %>% dplyr::filter(task=="yap1")
g <- ggplot(td_single , aes(name, avg))+geom_bar(stat="identity")+facet_wrap(~task)+theme_classic()

#ggsave("predictions_contributions/can_bmp7_wt_mut_all_yaptaskonly.pdf",g,height = 4, width =6)

```


# can tjp1 sum reads upon mut in barplots.
```{r}

profile_preds.df <- readr::read_csv('csv/for_preds/predictions/can_tjp1_predict.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_tjp1_coordinates.csv')
wt_preds.df <- dplyr::filter(profile_preds.df,region_idx==0)
mut_preds1.df <- dplyr::filter(profile_preds.df,region_idx==1)
mut_preds2.df <- dplyr::filter(profile_preds.df,region_idx==2) #right side mut based on 5>3 orientation
mut_preds3.df <- dplyr::filter(profile_preds.df,region_idx==3) #left side mut based on 5>3 orientation

tead4_tjp1 <- profile_preds.df %>% dplyr::filter(position>=450 & position<=550)
t <- tead4_tjp1  %>% dplyr::group_by(task,region_idx) %>% summarize(sum_preds= sum(abs(prediction)))
t <- t %>% dplyr::filter(task=="tead4" |task== "yap1")
t$region_idx_2 <- paste(t$region_idx)
w <- ggplot(t, aes(x = region_idx_2, y = sum_preds))+
    geom_bar(stat = "identity", fill="#848716")+
    facet_wrap(.~task, scales = "free", ncol = 1)+ xlab("")+ylab("Predicted signal")+
    theme(axis.text.x = element_text(angle = 45, hjust = 1))+theme_classic()
# ggsave("predictions_contributions/tjp1_pred_signal_barplots.pdf",w,height = 8, width = 4)
```


## tjp1 and amotl2 muts

```{r,Tjp1}
profile_preds.df <- readr::read_csv('csv/for_preds/predictions/can_tjp1_predict.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_tjp1_coordinates.csv')
wt_preds.df <- dplyr::filter(profile_preds.df,region_idx==0)
mut_preds1.df <- dplyr::filter(profile_preds.df,region_idx==1)
mut_preds2.df <- dplyr::filter(profile_preds.df,region_idx==2)
mut_preds3.df <- dplyr::filter(profile_preds.df,region_idx==3)


g <- ggplot()

for(row in 1:nrow(motifs.df)){
    g<-g + annotate("rect", xmin = motifs.df$start[row], xmax = motifs.df$end[row],
                    ymin = -Inf, ymax = Inf, alpha = .2)
    g<-g + annotate("text", x = motifs.df$start[row], label = motifs.df$name[row], y = Inf, vjust = 1.5)
}

g<- g+ geom_area(data = wt_preds.df, mapping = aes(position, prediction, group = strand, fill = task), alpha = .7)+
    geom_line(data = mut_preds2.df, mapping = aes(position, prediction, group = strand, color = task))+
    facet_grid(task ~ ., scales = "free")+
    scale_x_continuous(limits = c(400,600),
                       name = "Position (bp)")+
    scale_y_continuous(name = "BPNet predictions")+
    scale_fill_manual(values = c('#6e828e', '#7f778c', '#b47382','#788e6e', '#968b71', '#96718f'), name = "WT")+
    scale_color_manual(values = c('#314d5e', '#493d5c','#94384d','#3e8f1a', '#a87c14','#8c1f78'), name = "Mut")+
    theme_classic()+ggtitle("can_4_WT & mut_2_both")
g

# ggsave("predictions_contributions/can_4_td_double_wt.png",g,height = 18, width = 12)
# ggsave("predictions_contributions/can_4_td_double_wt.pdf",g,height = 18, width = 12)
# ggsave("predictions_contributions/can_4_td_double_first_start.pdf",g,height = 18, width = 12)
# ggsave("predictions_contributions/can_4_td_double_first_start.png",g,height = 18, width = 12)
# ggsave("predictions_contributions/can_4_td_wt_&_double_first_start.png",g,height = 18, width = 12)
# ggsave("predictions_contributions/can_4_td_wt_&_double_first_start.pdf",g,height = 18, width = 12)

```
## amotl2 muts
```{r,amotl2}

#contribution
profile_contrib.df <- readr::read_csv('csv/for_preds/predictions/can_amotl2_contrib.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_amotl2_coordinates.csv')
wt_contrib.df <- dplyr::filter(profile_contrib.df,region_idx==0)
mut_contrib2.df <- dplyr::filter(profile_contrib.df,region_idx==2) #this is the >> right side mut (so left motif contrib only)
mut_contrib3.df <- dplyr::filter(profile_contrib.df,region_idx==3) #this is the >> left side mut (so right motif contrib only)

tead4_amotl2 <- profile_contrib.df %>% dplyr::filter(position>=475 & position<=525)
t <- tead4_amotl2  %>% dplyr::group_by(task,region_idx) %>% summarize(sum_contrib=sum(sum_a=sum(A),sum_c=sum(C),sum_g=sum(G),sum_t=sum(T)))
t <- t %>% dplyr::filter(task=="tead4")
w <- ggplot(t, aes(x = region_idx, y = sum_contrib))+
    geom_bar(stat = "identity", fill="#848716")+
    facet_wrap(.~task, scales = "free", ncol = 1)+ xlab("")+ylab("Contribution signal")+
    theme(axis.text.x = element_text(angle = 45, hjust = 1))+theme_classic()
# ggsave("predictions_contributions/amotl2_contrib_signal_barplots.pdf",w,height = 8, width = 4)


#prediction
profile_preds.df <- readr::read_csv('csv/for_preds/predictions/can_amotl2_predict.csv')
motifs.df <- readr::read_csv('csv/for_preds/can_amotl2_coordinates.csv')
wt_preds.df <- dplyr::filter(profile_preds.df,region_idx==0)
mut_preds2.df <- dplyr::filter(profile_preds.df,region_idx==2) #this is the >> right side mut
mut_preds3.df <- dplyr::filter(profile_preds.df,region_idx==3)  #this is the >> left side mut 

g <- ggplot()

for(row in 1:nrow(motifs.df)){
    g<-g + annotate("rect", xmin = motifs.df$start[row], xmax = motifs.df$end[row],
                    ymin = -Inf, ymax = Inf, alpha = .2)
    g<-g + annotate("text", x = motifs.df$start[row], label = motifs.df$name[row], y = Inf, vjust = 1.5)
}

g<- g+ geom_area(data = wt_preds.df, mapping = aes(position, prediction, group = strand, fill = task), alpha = .7)+
    geom_line(data = mut_preds3.df, mapping = aes(position, prediction, group = strand, color = task))+
    facet_grid(task ~ ., scales = "free")+
    scale_x_continuous(limits = c(400,600),
                       name = "Position (bp)")+
    scale_y_continuous(name = "BPNet predictions")+
    scale_fill_manual(values = c('#6e828e', '#7f778c', '#b47382','#788e6e', '#968b71', '#96718f'), name = "WT")+
    scale_color_manual(values = c('#314d5e', '#493d5c','#94384d','#3e8f1a', '#a87c14','#8c1f78'), name = "Mut")+
    theme_classic()+ggtitle("can_4_WT & mut_2_both")

# ggsave("predictions_contributions/can_amotl2_wt_td_mut_left.png",g,height = 20, width = 8)
# ggsave("predictions_contributions/can_amotl2_wt_td_mut_left.pdf",g,height = 20, width = 8)
# ggsave("predictions_contributions/can_amotl2_wt_td_mut_right.pdf",g,height = 20, width = 8)
# ggsave("predictions_contributions/can_amotl2_wt_td_mut_right.png",g,height = 20, width = 8)
```


# Session information

For reproducibility, this analysis was performed with the following R/Bioconductor session:

``` {r session_info, echo=FALSE, comment=NA}
sessionInfo()
```
