{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "The purpose of this `.ipynb` is to use the `TSC TFs tranied BPNet model` to predict binding on genomic regions where Tfap2c and Tead4 have distance replationship and plot change in predicted binding upon their distance manipulation.\n",
    "\n",
    "# Computational setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from tensorflow.python.util import deprecation\n",
    "deprecation._PRINT_DEPRECATION_WARNINGS = False\n",
    "\n",
    "#Packages\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pybedtools import BedTool\n",
    "from genomelake.extractors import FastaExtractor\n",
    "from bpnet.extractors import Interval\n",
    "from bpnet.cli.contrib import bpnet_contrib\n",
    "from bpnet.cli.modisco import cwm_scan\n",
    "from bpnet.cli.contrib import ContribFile\n",
    "from concise.preprocessing import encodeDNA\n",
    "from bpnet.cli.contrib import ContribFile\n",
    "from bpnet.simulate import random_seq, insert_motif\n",
    "from bpnet.plot.tracks import plot_tracks, to_neg\n",
    "from plotnine import *\n",
    "from tqdm import tqdm\n",
    "from bpnet.BPNet import BPNetSeqModel\n",
    "from bpnet.utils import create_tf_session\n",
    "\n",
    "#create_tf_session('0', .4)\n",
    "\n",
    "# Settings\n",
    "os.chdir(f'/n/projects/kd2200/publication/bpnet/analysis/test/')\n",
    "pd.set_option('display.max_columns', 100)\n",
    "\n",
    "\n",
    "# Custom commands\n",
    "sys.path.insert(0, f'/n/projects/kd2200/publication/bpnet/analysis/scripts/bpnet/scripts')\n",
    "from data_format_functions import myround, myfloor, myceiling, df_to_intervals,tidy_bpnet_predictions_nexus, one_hot_decode\n",
    "from motif_functions import remove_palindromic_motif_duplicates, filter_overlapping_motifs_by_priority\n",
    "\n",
    "\n",
    "# function to return key for any value \n",
    "def get_key(val, my_dict): \n",
    "    for key, value in my_dict.items(): \n",
    "        if val == value: \n",
    "            return key \n",
    "    return \"key doesn't exist\"\n",
    "\n",
    "#Function to save contribution information with 4 columns and a task (does not use hypoth. contrib)\n",
    "def tidy_bpnet_contributions(seq, contrib, tasks, contrib_type = 'profile'):\n",
    "    contrib_df = pd.DataFrame()\n",
    "    for task in tasks:\n",
    "        c = contrib[f'{task}/{contrib_type}'] * seq\n",
    "        df = pd.DataFrame(c, columns = ['A','C','G','T'])\n",
    "        df['task'] = task\n",
    "        contrib_df=contrib_df.append(df)\n",
    "    return(contrib_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Pre-existing variables\n",
    "fasta_file = f'/n/projects/kd2200/publication/bpnet/fasta/mm10.fa'\n",
    "model_dir = f'/n/projects/kd2200/publication/bpnet/model/dataspec.yaml_default_fold_5/'\n",
    "contrib_file = f\"/n/projects/kd2200/publication/bpnet/contrib/dataspec.yaml_default_fold_5/contrib.h5\"\n",
    "modisco_dir = f'/n/projects/kd2200/publication/bpnet/modisco/profile/dataspec.yaml_default_fold_5/'\n",
    "modisco_all_region_cwm_dir = f'/n/projects/kd2200/publication/bpnet/cwm_all_regions/'\n",
    "tasks = ['tead4','tfap2c','yap1','gata3','cdx2']\n",
    "\n",
    "# Independent variables\n",
    "motifs_of_interest_dict = {'tead4': 'tead4/m0_p0',\n",
    "                           'tfap2c': 'tfap2c/m0_p0',\n",
    "                            'cdx2': 'cdx2/m0_p0',\n",
    "                            'gata3': 'gata3/m0_p4'}\n",
    "# Dependent variables\n",
    "enhancers_path = f'enhancers_genes_tead4_tfap2c_dis_0based.bed'\n",
    "enhancers_bed3 = f'enhancers_genes_tead4_tfap2c_dis_0based.bed3'\n",
    "tfap2c_path = f'genes_tfap2c_dis_0based.bed'\n",
    "tead4_path = f'genes_tead4_dis_0based.bed'\n",
    "prefix = 'tt_enhancers'\n",
    "enhancer_bed_file = f'{prefix}.bed'\n",
    "cwm_scan_prefix = f'cwm_scan/{prefix}'\n",
    "enhancer_contrib_file = f\"contrib/{prefix}_contrib.h5\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script false --no-raise-error\n",
    "!mkdir -p {cwm_scan_prefix}\n",
    "!mkdir -p bed\n",
    "!mkdir -p contrib\n",
    "!mkdir -p figures/1_{prefix}_tt_distance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate contributions based on enhancer coordinates\n",
    "\n",
    "Generate contributions, then run CWM-scanning to map all pertinent motifs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cut -f 1-3 {enhancers_path} > {enhancers_bed3}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script false --no-raise-error\n",
    "!bpnet contrib --regions {enhancers_bed3} --fasta-file {fasta_file} --method deeplift --overwrite  {model_dir} {enhancer_contrib_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-12-28 12:11:03,898 [INFO] Using tasks: ['tead4']\n",
      "2023-12-28 12:11:03,900 [INFO] Loading the contribution scores from: contrib/tt_enhancers_contrib.h5\n",
      "2023-12-28 12:11:03,907 [INFO] Centroid matches already exist.\n",
      "2023-12-28 12:11:03,909 [INFO] Loading centroid matches from /n/projects/kd2200/publication/bpnet/modisco/profile/dataspec.yaml_default_fold_5/tead4/cwm-scan-seqlets.trim-frac=0.08.csv.gz\n",
      "2023-12-28 12:11:04,285 [INFO] Scanning for patterns\n",
      "100%|██████████| 11/11 [00:21<00:00,  1.94s/it]\n",
      "2023-12-28 12:11:25,672 [INFO] Merging\n",
      "2023-12-28 12:11:25,721 [INFO] Append ranges\n",
      "2023-12-28 12:11:25,732 [INFO] Table info\n",
      "2023-12-28 12:11:25,745 [INFO] Writing the resuling pd.DataFrame of shape (838, 30) to cwm_scan/tt_enhancers/enhancer_cwm_scan_instances.tead4.tsv.gz\n",
      "2023-12-28 12:11:25,747 [INFO] Writing a tsv file\n",
      "2023-12-28 12:11:25,801 [INFO] Done!\n",
      "2023-12-28 12:11:25,826 [INFO] Using tasks: ['tfap2c']\n",
      "2023-12-28 12:11:25,826 [INFO] Loading the contribution scores from: contrib/tt_enhancers_contrib.h5\n",
      "2023-12-28 12:11:25,830 [INFO] Centroid matches already exist.\n",
      "2023-12-28 12:11:25,831 [INFO] Loading centroid matches from /n/projects/kd2200/publication/bpnet/modisco/profile/dataspec.yaml_default_fold_5/tfap2c/cwm-scan-seqlets.trim-frac=0.08.csv.gz\n",
      "2023-12-28 12:11:26,195 [INFO] Scanning for patterns\n",
      "100%|██████████| 5/5 [00:03<00:00,  1.38it/s]\n",
      "2023-12-28 12:11:29,835 [INFO] Merging\n",
      "2023-12-28 12:11:29,874 [INFO] Append ranges\n",
      "2023-12-28 12:11:29,882 [INFO] Table info\n",
      "2023-12-28 12:11:29,893 [INFO] Writing the resuling pd.DataFrame of shape (628, 30) to cwm_scan/tt_enhancers/enhancer_cwm_scan_instances.tfap2c.tsv.gz\n",
      "2023-12-28 12:11:29,895 [INFO] Writing a tsv file\n",
      "2023-12-28 12:11:29,926 [INFO] Done!\n",
      "2023-12-28 12:11:29,948 [INFO] Using tasks: ['cdx2']\n",
      "2023-12-28 12:11:29,950 [INFO] Loading the contribution scores from: contrib/tt_enhancers_contrib.h5\n",
      "2023-12-28 12:11:29,952 [INFO] Centroid matches already exist.\n",
      "2023-12-28 12:11:29,954 [INFO] Loading centroid matches from /n/projects/kd2200/publication/bpnet/modisco/profile/dataspec.yaml_default_fold_5/cdx2/cwm-scan-seqlets.trim-frac=0.08.csv.gz\n",
      "2023-12-28 12:11:30,163 [INFO] Scanning for patterns\n",
      "100%|██████████| 21/21 [00:15<00:00,  1.35it/s]\n",
      "2023-12-28 12:11:45,688 [INFO] Merging\n",
      "2023-12-28 12:11:45,764 [INFO] Append ranges\n",
      "2023-12-28 12:11:45,774 [INFO] Table info\n",
      "2023-12-28 12:11:45,787 [INFO] Writing the resuling pd.DataFrame of shape (1147, 30) to cwm_scan/tt_enhancers/enhancer_cwm_scan_instances.cdx2.tsv.gz\n",
      "2023-12-28 12:11:45,789 [INFO] Writing a tsv file\n",
      "2023-12-28 12:11:45,847 [INFO] Done!\n",
      "2023-12-28 12:11:45,867 [INFO] Using tasks: ['gata3']\n",
      "2023-12-28 12:11:45,868 [INFO] Loading the contribution scores from: contrib/tt_enhancers_contrib.h5\n",
      "2023-12-28 12:11:45,870 [INFO] Centroid matches already exist.\n",
      "2023-12-28 12:11:45,871 [INFO] Loading centroid matches from /n/projects/kd2200/publication/bpnet/modisco/profile/dataspec.yaml_default_fold_5/gata3/cwm-scan-seqlets.trim-frac=0.08.csv.gz\n",
      "2023-12-28 12:11:46,067 [INFO] Scanning for patterns\n",
      "100%|██████████| 14/14 [00:09<00:00,  1.54it/s]\n",
      "2023-12-28 12:11:55,175 [INFO] Merging\n",
      "2023-12-28 12:11:55,218 [INFO] Append ranges\n",
      "2023-12-28 12:11:55,225 [INFO] Table info\n",
      "2023-12-28 12:11:55,236 [INFO] Writing the resuling pd.DataFrame of shape (486, 30) to cwm_scan/tt_enhancers/enhancer_cwm_scan_instances.gata3.tsv.gz\n",
      "2023-12-28 12:11:55,237 [INFO] Writing a tsv file\n",
      "2023-12-28 12:11:55,262 [INFO] Done!\n",
      "2023-12-28 12:11:55,281 [INFO] Using tasks: ['yap1']\n",
      "2023-12-28 12:11:55,282 [INFO] Loading the contribution scores from: contrib/tt_enhancers_contrib.h5\n",
      "2023-12-28 12:11:55,283 [INFO] Centroid matches already exist.\n",
      "2023-12-28 12:11:55,285 [INFO] Loading centroid matches from /n/projects/kd2200/publication/bpnet/modisco/profile/dataspec.yaml_default_fold_5/yap1/cwm-scan-seqlets.trim-frac=0.08.csv.gz\n",
      "2023-12-28 12:11:55,486 [INFO] Scanning for patterns\n",
      "100%|██████████| 26/26 [00:21<00:00,  1.24it/s]\n",
      "2023-12-28 12:12:16,535 [INFO] Merging\n",
      "2023-12-28 12:12:16,643 [INFO] Append ranges\n",
      "2023-12-28 12:12:16,655 [INFO] Table info\n",
      "2023-12-28 12:12:16,673 [INFO] Writing the resuling pd.DataFrame of shape (5881, 30) to cwm_scan/tt_enhancers/enhancer_cwm_scan_instances.yap1.tsv.gz\n",
      "2023-12-28 12:12:16,675 [INFO] Writing a tsv file\n",
      "2023-12-28 12:12:17,026 [INFO] Done!\n"
     ]
    }
   ],
   "source": [
    "%%script false --no-raise-error\n",
    "# For each task, CWM-scan the enhancers. The `example_idx` will be the same across tasks\n",
    "for task in ['tead4','tfap2c','cdx2','gata3','yap1']:\n",
    "    print(task)\n",
    "    cwm_scan(modisco_dir = f'{modisco_dir}/{task}', \n",
    "             output_file = f'{cwm_scan_prefix}/enhancer_cwm_scan_instances.{task}.tsv.gz', \n",
    "             contrib_file = enhancer_contrib_file)     "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Collect annotated genomic loci\n",
    "\n",
    "Import the regions for downstream analysis. Because the enhancers are already a bed file, no conversion is needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import regions and the genomic coordinates.\n",
    "\n",
    "enhancers_df = BedTool(enhancers_path).to_dataframe().sort_values(['name']).reset_index(drop = True)\n",
    "tfap2c_df = BedTool(tfap2c_path).to_dataframe().sort_values(['name']).reset_index(drop = True)\n",
    "tead4_df = BedTool(tead4_path).to_dataframe().sort_values(['name']).reset_index(drop = True)\n",
    "enhancers_bed = BedTool.from_dataframe(enhancers_df)\n",
    "tfap2c_bed = BedTool.from_dataframe(tfap2c_df)\n",
    "tead4_bed = BedTool.from_dataframe(tead4_df)\n",
    "\n",
    "#Ensure everything is in the same order and unique\n",
    "assert(all(enhancers_df.name.values==tfap2c_df.name.values)==all(enhancers_df.name.values==tead4_df.name.values), \n",
    "       'Either motifs occur more than once in region OR the motifs and enhancers are not in the same order')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "enhancers_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tfap2c_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import enhancers, motifs, and collect coordinates for sequence mutation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Assign in-window coordinates to these motifs (watch enhancer id!)\n",
    "tfap2c_df['window_enhancer_start'] = enhancers_df.start\n",
    "tead4_df['window_enhancer_start'] = enhancers_df.start\n",
    "tfap2c_df['window_enhancer_end'] = enhancers_df.end\n",
    "tead4_df['window_enhancer_end'] = enhancers_df.end\n",
    "\n",
    "tfap2c_df['window_motif_start'] = [m.start - m.window_enhancer_start for i,m in tfap2c_df.iterrows()]\n",
    "tead4_df['window_motif_start'] = [m.start - m.window_enhancer_start for i,m in tead4_df.iterrows()]\n",
    "\n",
    "tfap2c_df['window_motif_end'] = [m.end - m.window_enhancer_start for i,m in tfap2c_df.iterrows()]\n",
    "tead4_df['window_motif_end'] = [m.end - m.window_enhancer_start for i,m in tead4_df.iterrows()]\n",
    "\n",
    "tfap2c_df['window_motif_center'] = np.floor((tfap2c_df.window_motif_end - tfap2c_df.window_motif_start)/2) + tfap2c_df.window_motif_start\n",
    "tead4_df['window_motif_center'] = np.floor((tead4_df.window_motif_end - tead4_df.window_motif_start)/2) + tead4_df.window_motif_start\n",
    "\n",
    "# Mark where the motif boundaries are and store in enhancer_df\n",
    "enhancers_df['tfap2c_start_genomic'] = tfap2c_df.start\n",
    "enhancers_df['tfap2c_end_genomic'] = tfap2c_df.end\n",
    "enhancers_df['tead4_start_genomic'] = tead4_df.start\n",
    "enhancers_df['tead4_end_genomic'] = tead4_df.end\n",
    "\n",
    "enhancers_df['tfap2c_center_genomic'] = np.floor((tfap2c_df.end - tfap2c_df.start)/2) + tfap2c_df.start\n",
    "enhancers_df['tead4_center_genomic'] = np.floor((tead4_df.end - tead4_df.start)/2) + tead4_df.start\n",
    "\n",
    "#Mark distance between the two motifs, with tfap2c as the anchor.\n",
    "enhancers_df['tead4_minus_tfap2c_distance'] = enhancers_df['tfap2c_center_genomic'] - enhancers_df['tead4_center_genomic']\n",
    "\n",
    "#enhancers_df[enhancers_df.name==119]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/n/projects/kd2200/publication/bpnet/fasta/mm10.fa'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tfap2c_df.head(n=15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tfap2c_df.head(n=15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Confirm that the sequences found by BPNet match the sequences based off of the genomic coordinates of the .fasta file. To do this, we need to collect sequences and hyper-contributions across each enhancer in order to mutate them at the targeted sites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get sequence and contribution \n",
    "c = ContribFile(enhancer_contrib_file)\n",
    "seqs = c.get_seq()\n",
    "coords = c.get_ranges()\n",
    "hyp_contrib = c.get_hyp_contrib()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>window_enhancer_start</th>\n",
       "      <th>window_enhancer_end</th>\n",
       "      <th>window_motif_start</th>\n",
       "      <th>window_motif_end</th>\n",
       "      <th>window_motif_center</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>33981094</td>\n",
       "      <td>33981106</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>33980621</td>\n",
       "      <td>33981621</td>\n",
       "      <td>473</td>\n",
       "      <td>485</td>\n",
       "      <td>479.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>34716429</td>\n",
       "      <td>34716441</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>146482883</td>\n",
       "      <td>146483883</td>\n",
       "      <td>-111766454</td>\n",
       "      <td>-111766442</td>\n",
       "      <td>-111766448.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>38259714</td>\n",
       "      <td>38259726</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>3206597</td>\n",
       "      <td>3207597</td>\n",
       "      <td>35053117</td>\n",
       "      <td>35053129</td>\n",
       "      <td>35053123.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>53939591</td>\n",
       "      <td>53939603</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>4716425</td>\n",
       "      <td>4717425</td>\n",
       "      <td>49223166</td>\n",
       "      <td>49223178</td>\n",
       "      <td>49223172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>60852606</td>\n",
       "      <td>60852618</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>18749498</td>\n",
       "      <td>18750498</td>\n",
       "      <td>42103108</td>\n",
       "      <td>42103120</td>\n",
       "      <td>42103114.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>chr19</td>\n",
       "      <td>54982668</td>\n",
       "      <td>54982680</td>\n",
       "      <td>150</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>79261639</td>\n",
       "      <td>79262639</td>\n",
       "      <td>-24278971</td>\n",
       "      <td>-24278959</td>\n",
       "      <td>-24278965.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>chr19</td>\n",
       "      <td>55714436</td>\n",
       "      <td>55714448</td>\n",
       "      <td>151</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>84128795</td>\n",
       "      <td>84129795</td>\n",
       "      <td>-28414359</td>\n",
       "      <td>-28414347</td>\n",
       "      <td>-28414353.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>chrX</td>\n",
       "      <td>93690800</td>\n",
       "      <td>93690812</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>87089787</td>\n",
       "      <td>87090787</td>\n",
       "      <td>6601013</td>\n",
       "      <td>6601025</td>\n",
       "      <td>6601019.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>chrX</td>\n",
       "      <td>167414659</td>\n",
       "      <td>167414671</td>\n",
       "      <td>153</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>32632323</td>\n",
       "      <td>32633323</td>\n",
       "      <td>134782336</td>\n",
       "      <td>134782348</td>\n",
       "      <td>134782342.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>chr12</td>\n",
       "      <td>102261989</td>\n",
       "      <td>102262003</td>\n",
       "      <td>155</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>102261651</td>\n",
       "      <td>102262651</td>\n",
       "      <td>338</td>\n",
       "      <td>352</td>\n",
       "      <td>345.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>155 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end  name  score strand  window_enhancer_start  \\\n",
       "0     chr1   33981094   33981106     0      0      +               33980621   \n",
       "1     chr1   34716429   34716441     1      0      -              146482883   \n",
       "2     chr1   38259714   38259726     2      0      +                3206597   \n",
       "3     chr1   53939591   53939603     3      0      -                4716425   \n",
       "4     chr1   60852606   60852618     4      0      +               18749498   \n",
       "..     ...        ...        ...   ...    ...    ...                    ...   \n",
       "150  chr19   54982668   54982680   150      0      +               79261639   \n",
       "151  chr19   55714436   55714448   151      0      +               84128795   \n",
       "152   chrX   93690800   93690812   152      0      +               87089787   \n",
       "153   chrX  167414659  167414671   153      0      +               32632323   \n",
       "154  chr12  102261989  102262003   155      0      +              102261651   \n",
       "\n",
       "     window_enhancer_end  window_motif_start  window_motif_end  \\\n",
       "0               33981621                 473               485   \n",
       "1              146483883          -111766454        -111766442   \n",
       "2                3207597            35053117          35053129   \n",
       "3                4717425            49223166          49223178   \n",
       "4               18750498            42103108          42103120   \n",
       "..                   ...                 ...               ...   \n",
       "150             79262639           -24278971         -24278959   \n",
       "151             84129795           -28414359         -28414347   \n",
       "152             87090787             6601013           6601025   \n",
       "153             32633323           134782336         134782348   \n",
       "154            102262651                 338               352   \n",
       "\n",
       "     window_motif_center  \n",
       "0                  479.0  \n",
       "1           -111766448.0  \n",
       "2             35053123.0  \n",
       "3             49223172.0  \n",
       "4             42103114.0  \n",
       "..                   ...  \n",
       "150          -24278965.0  \n",
       "151          -28414353.0  \n",
       "152            6601019.0  \n",
       "153          134782342.0  \n",
       "154                345.0  \n",
       "\n",
       "[155 rows x 11 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x =tead4_bed[1]\n",
    "x\n",
    "#[one_hot_decode(i) for i in FastaExtractor(fasta_file)([Interval.from_pybedtools(i) for i in x])]\n",
    "tfap2c_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a check that looks at every sequence to make sure it is correct."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "encoded array not the same length as given seq",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-49-e11679d10e24>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#Extract BPNet-contained sequences\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mtfap2c_bpnet_seqs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mone_hot_decode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseqs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwindow_motif_start\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwindow_motif_end\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mrow\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtfap2c_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miterrows\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mtfap2c_fasta_seqs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mone_hot_decode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mFastaExtractor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfasta_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mInterval\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_pybedtools\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtfap2c_bed\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;31m#assert tfap2c_bpnet_seqs==tfap2c_fasta_seqs, 'BPNet sequences and genomic sequences do not match. Correct this before proceeding.'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/bpnet/lib/python3.6/site-packages/genomelake/extractors.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, intervals, out, **kwargs)\u001b[0m\n\u001b[1;32m     24\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__call__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mintervals\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m         \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_check_or_create_output_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mintervals\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 26\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_extract\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mintervals\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     27\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/bpnet/lib/python3.6/site-packages/genomelake/extractors.py\u001b[0m in \u001b[0;36m_extract\u001b[0;34m(self, intervals, out, **kwargs)\u001b[0m\n\u001b[1;32m     92\u001b[0m             seq = self.fasta.fetch(str(interval.chrom), interval.start,\n\u001b[1;32m     93\u001b[0m                                        interval.stop)\n\u001b[0;32m---> 94\u001b[0;31m             \u001b[0mone_hot_encode_sequence\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     95\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     96\u001b[0m             \u001b[0;31m# reverse-complement seq the negative strand\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mgenomelake/util.pyx\u001b[0m in \u001b[0;36mgenomelake.util.one_hot_encode_sequence\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mgenomelake/util.pyx\u001b[0m in \u001b[0;36mgenomelake.util.one_hot_encode_sequence\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: encoded array not the same length as given seq"
     ]
    }
   ],
   "source": [
    "#Extract BPNet-contained sequences\n",
    "tfap2c_bpnet_seqs = [one_hot_decode(seqs[row.name][row.window_motif_start:row.window_motif_end]) for i,row in tfap2c_df.iterrows()]\n",
    "tfap2c_fasta_seqs = [one_hot_decode(i) for i in FastaExtractor(fasta_file)([Interval.from_pybedtools(i) for i in tfap2c_bed])]\n",
    "#assert tfap2c_bpnet_seqs==tfap2c_fasta_seqs, 'BPNet sequences and genomic sequences do not match. Correct this before proceeding.'\n",
    "\n",
    "#tead4_bpnet_seqs = [one_hot_decode(seqs[row.name][row.window_motif_start:row.window_motif_end]) for i,row in tead4_df.iterrows()]\n",
    "#tead4_fasta_seqs = [one_hot_decode(i) for i in FastaExtractor(fasta_file)([Interval.from_pybedtools(i) for i in tead4_bed])]\n",
    "#assert tead4_bpnet_seqs==tead4_fasta_seqs, 'BPNet sequences and genomic sequences do not match. Correct this before proceeding.'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Begin generating perturbations based on 2-mutation dynamic approach\n",
    "\n",
    "TODO: Explain approach in normal words."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define distances that we want to inject motifs across."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "distances = range(-200, 200, 1)\n",
    "mutation_count = 2 #how many bases we will mutate in a targeted sense"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate all sequences \n",
    "\n",
    "Across every enhancer, motif, and distance generate the WT, mutant, and mutant-distance injection. Combine into array while storing index information in a separate `pd.df`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "155it [39:34, 15.32s/it]\n"
     ]
    }
   ],
   "source": [
    "#Preallocate objects for writing.\n",
    "wt_seqs = []\n",
    "mut_seqs = []\n",
    "mut_inj_seqs = []\n",
    "wt_info_all_df = pd.DataFrame()\n",
    "mut_inj_info_all_df = pd.DataFrame()\n",
    "\n",
    "#test = enhancers_df.head(n=118)\n",
    "for enhancer_id,enhancer_row in tqdm(enhancers_df.iterrows()):\n",
    "    \n",
    "    for anchor_motif in ['tfap2c', 'tead4']:\n",
    "        #Define motif coordinates\n",
    "        if anchor_motif=='tfap2c':\n",
    "            anchor_motif_coords = tfap2c_df.loc[enhancer_id] \n",
    "            moved_motif = 'tead4'\n",
    "            moved_motif_coords = tead4_df.loc[enhancer_id]\n",
    "        else:\n",
    "            anchor_motif_coords = tead4_df.loc[enhancer_id]\n",
    "            moved_motif = 'tfap2c'\n",
    "            moved_motif_coords = tfap2c_df.loc[enhancer_id]\n",
    "        #Define input sequence across enhancer\n",
    "        wt_seq = seqs[enhancer_id] \n",
    "        wt_seqs.append(wt_seq)\n",
    "        \n",
    "        #Find the minimum hyp contribution across all positions\n",
    "        hyp_contrib_across_motif_arr = hyp_contrib[moved_motif][enhancer_id][moved_motif_coords.window_motif_start:moved_motif_coords.window_motif_end]\n",
    "        min_contrib_across_motif_arr = np.min(hyp_contrib_across_motif_arr, axis = 1)\n",
    "        min_indices_across_motif_arr = np.stack([np.where(pos==min_contrib_across_motif_arr[i])[0].flatten() for i,pos in enumerate(hyp_contrib_across_motif_arr)], axis = 0).flatten()\n",
    "        seq_across_motif_arr = wt_seq[moved_motif_coords.window_motif_start:moved_motif_coords.window_motif_end]\n",
    "\n",
    "        #Find the top contributing indices across the motif\n",
    "        contrib_across_motif_arr = np.sum(hyp_contrib_across_motif_arr * seq_across_motif_arr, axis = 1)\n",
    "        top_contributing_indices = (-contrib_across_motif_arr).argsort()[:mutation_count]\n",
    "\n",
    "        #Reassign the worst contributing nt for each mutation position.\n",
    "        mut_motif_seq = []\n",
    "        for i in range(seq_across_motif_arr.shape[0]):\n",
    "            if i in top_contributing_indices:\n",
    "                row = np.array([0,0,0,0], dtype = 'float32')\n",
    "                row[min_indices_across_motif_arr[i]] = 1\n",
    "            else: \n",
    "                row = seq_across_motif_arr[i]\n",
    "            mut_motif_seq.append(row)\n",
    "        mut_motif_seq = np.stack(mut_motif_seq)\n",
    "\n",
    "        #Reinject the motif sequence back into the entire window\n",
    "        mut_seq = wt_seq.copy()\n",
    "        mut_seq[moved_motif_coords.window_motif_start:moved_motif_coords.window_motif_end] = mut_motif_seq\n",
    "        mut_seqs.append(mut_seq)\n",
    "  \n",
    "        #Record all perturbation information in a single pd.df (defined by enhancer_id)\n",
    "        wt_info_df = pd.DataFrame([enhancer_id, anchor_motif, anchor_motif_coords.window_motif_start, anchor_motif_coords.window_motif_end, anchor_motif_coords.window_motif_center,\n",
    "                                moved_motif, moved_motif_coords.window_motif_start, moved_motif_coords.window_motif_end, moved_motif_coords.window_motif_center]).transpose()\n",
    "        wt_info_df.columns = ['enhancer_id',\n",
    "                           'anchor_motif','anchor_motif_window_start_0based','anchor_motif_window_end_0based','anchor_motif_window_center_0based',\n",
    "                           'moved_motif','moved_motif_window_orig_start_0based','moved_motif_window_orig_end_0based','moved_motif_window_orig_center_0based']\n",
    "        wt_info_df['rank_positions'] = ','.join([str(x) for x in top_contributing_indices])\n",
    "        wt_info_df['injected_motif_seq'] = one_hot_decode(seq_across_motif_arr)\n",
    "        wt_info_df['mutated_motif_seq'] = one_hot_decode(mut_motif_seq)\n",
    "        wt_info_df['anchor_motif_seq'] = one_hot_decode(wt_seq[anchor_motif_coords.window_motif_start:anchor_motif_coords.window_motif_end])\n",
    "        wt_info_all_df = wt_info_all_df.append(wt_info_df)\n",
    "        \n",
    "        for d in distances:\n",
    "            \n",
    "            #Add check to make sure distance injection doesn't move off window\n",
    "            check1 = (d + anchor_motif_coords.window_motif_start) <= 0\n",
    "            check2 = (d + anchor_motif_coords.window_motif_end) >= (anchor_motif_coords.window_enhancer_end - anchor_motif_coords.window_enhancer_start - 1) \n",
    "            if not (check1 or check2):\n",
    "\n",
    "                #Inject original motif into the designated distance \n",
    "                mut_inj_seq = mut_seq.copy()\n",
    "\n",
    "                #Define distance boundaries based on anchor center\n",
    "                injected_moved_motif_center = int(anchor_motif_coords.window_motif_center + d)\n",
    "                upstream_to_center_moved_motif_distance = moved_motif_coords.window_motif_center - moved_motif_coords.window_motif_start\n",
    "                downstream_to_center_moved_motif_distance = moved_motif_coords.window_motif_end - moved_motif_coords.window_motif_center \n",
    "                injected_moved_motif_boundaries = (int(injected_moved_motif_center - upstream_to_center_moved_motif_distance), \n",
    "                                                   int(injected_moved_motif_center + downstream_to_center_moved_motif_distance))\n",
    "                \n",
    "                #Add check to make sure distance doesn't inject over another motif.\n",
    "                inj_motif_int = pd.Interval(injected_moved_motif_boundaries[0], injected_moved_motif_boundaries[1])\n",
    "                anchor_motif_int = pd.Interval(anchor_motif_coords.window_motif_start, anchor_motif_coords.window_motif_end)\n",
    "                if not inj_motif_int.overlaps(anchor_motif_int):\n",
    "                    #Add check to make sure your mutation and injected sequence cover the same coordinates\n",
    "                    test_mut = mut_inj_seq[injected_moved_motif_boundaries[0]:injected_moved_motif_boundaries[1]].shape == seq_across_motif_arr.shape\n",
    "                    assert(test_mut, 'You are injecting a misproportioned sequence, check coordinates.')\n",
    "                    mut_inj_seq[injected_moved_motif_boundaries[0]:injected_moved_motif_boundaries[1]] = seq_across_motif_arr\n",
    "                    mut_inj_seqs.append(mut_inj_seq)\n",
    "\n",
    "                    #Record all perturbation information in a single pd.df (distance_id)\n",
    "                    mut_inj_info_df = wt_info_df\n",
    "                    mut_inj_info_df['distance_between_moved_center_and_anchor_center'] = d\n",
    "                    mut_inj_info_df['distance_id'] = str(mut_inj_info_df.enhancer_id.values[0]) + '_' + str(d)\n",
    "                    mut_inj_info_all_df = mut_inj_info_all_df.append(mut_inj_info_df)\n",
    "                ##below commented out is the alternative approach to get rank, high contributing nucleotides and mutated nucleotides:\n",
    "                # df = pd.DataFrame([top_contributing_indices, list(range(mutation_count)), \n",
    "                #              list(one_hot_decode(seq_across_motif_arr[top_contributing_indices])),\n",
    "                #              list(one_hot_decode(mut_motif_seq[top_contributing_indices]))]).transpose()\n",
    "                # df.columns = ['mutation_index_across_motif', 'mutation_rank', 'wt_nt','mut_nt']\n",
    "                # df.pivot(columns='mutation_rank', values=['mutation_index_across_motif', 'wt_nt','mut_nt'])\n",
    "wt_seqs = np.array(wt_seqs)\n",
    "mut_seqs = np.array(mut_seqs)\n",
    "mut_inj_seqs = np.array(mut_inj_seqs)\n",
    "\n",
    "#Add checks to make sure correct dimensions match enhancer ids\n",
    "assert wt_seqs.shape[0]==mut_seqs.shape[0]==wt_info_all_df.shape[0], 'WT and mut arrays dont match their index'\n",
    "assert mut_inj_seqs.shape[0]==mut_inj_info_all_df.shape[0], 'WT and mut arrays dont match their index'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method NDFrame.head of    enhancer_id anchor_motif anchor_motif_window_start_0based  \\\n",
       "0            0       tfap2c                              473   \n",
       "0            0       tfap2c                              473   \n",
       "0            0       tfap2c                              473   \n",
       "0            0       tfap2c                              473   \n",
       "0            0       tfap2c                              473   \n",
       "..         ...          ...                              ...   \n",
       "0          154        tead4                              372   \n",
       "0          154        tead4                              372   \n",
       "0          154        tead4                              372   \n",
       "0          154        tead4                              372   \n",
       "0          154        tead4                              372   \n",
       "\n",
       "   anchor_motif_window_end_0based anchor_motif_window_center_0based  \\\n",
       "0                             485                               479   \n",
       "0                             485                               479   \n",
       "0                             485                               479   \n",
       "0                             485                               479   \n",
       "0                             485                               479   \n",
       "..                            ...                               ...   \n",
       "0                             382                               377   \n",
       "0                             382                               377   \n",
       "0                             382                               377   \n",
       "0                             382                               377   \n",
       "0                             382                               377   \n",
       "\n",
       "   moved_motif moved_motif_window_orig_start_0based  \\\n",
       "0        tead4                                  516   \n",
       "0        tead4                                  516   \n",
       "0        tead4                                  516   \n",
       "0        tead4                                  516   \n",
       "0        tead4                                  516   \n",
       "..         ...                                  ...   \n",
       "0       tfap2c                                  338   \n",
       "0       tfap2c                                  338   \n",
       "0       tfap2c                                  338   \n",
       "0       tfap2c                                  338   \n",
       "0       tfap2c                                  338   \n",
       "\n",
       "   moved_motif_window_orig_end_0based moved_motif_window_orig_center_0based  \\\n",
       "0                                 526                                   521   \n",
       "0                                 526                                   521   \n",
       "0                                 526                                   521   \n",
       "0                                 526                                   521   \n",
       "0                                 526                                   521   \n",
       "..                                ...                                   ...   \n",
       "0                                 352                                   345   \n",
       "0                                 352                                   345   \n",
       "0                                 352                                   345   \n",
       "0                                 352                                   345   \n",
       "0                                 352                                   345   \n",
       "\n",
       "   rank_positions injected_motif_seq mutated_motif_seq anchor_motif_seq  \\\n",
       "0             6,7         CCTGGAATGT        CCTGGAGAGT     CCCCCTGAGGCA   \n",
       "0             6,7         CCTGGAATGT        CCTGGAGAGT     CCCCCTGAGGCA   \n",
       "0             6,7         CCTGGAATGT        CCTGGAGAGT     CCCCCTGAGGCA   \n",
       "0             6,7         CCTGGAATGT        CCTGGAGAGT     CCCCCTGAGGCA   \n",
       "0             6,7         CCTGGAATGT        CCTGGAGAGT     CCCCCTGAGGCA   \n",
       "..            ...                ...               ...              ...   \n",
       "0            10,4     AGGGCCCCAGGGCC    AGGGACCCAGCGCC       CACATTCCTA   \n",
       "0            10,4     AGGGCCCCAGGGCC    AGGGACCCAGCGCC       CACATTCCTA   \n",
       "0            10,4     AGGGCCCCAGGGCC    AGGGACCCAGCGCC       CACATTCCTA   \n",
       "0            10,4     AGGGCCCCAGGGCC    AGGGACCCAGCGCC       CACATTCCTA   \n",
       "0            10,4     AGGGCCCCAGGGCC    AGGGACCCAGCGCC       CACATTCCTA   \n",
       "\n",
       "    distance_between_moved_center_and_anchor_center distance_id  \n",
       "0                                              -200      0_-200  \n",
       "0                                              -199      0_-199  \n",
       "0                                              -198      0_-198  \n",
       "0                                              -197      0_-197  \n",
       "0                                              -196      0_-196  \n",
       "..                                              ...         ...  \n",
       "0                                               195     154_195  \n",
       "0                                               196     154_196  \n",
       "0                                               197     154_197  \n",
       "0                                               198     154_198  \n",
       "0                                               199     154_199  \n",
       "\n",
       "[117486 rows x 15 columns]>"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_inj_info_all_df.head"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Export full seqs to csv.gz and make sure full seqs match."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "wt_test_df = pd.DataFrame([list(one_hot_decode(wt_seqs[i])) for i in range(wt_seqs.shape[0])]).transpose()\n",
    "mut_test_df = pd.DataFrame([list(one_hot_decode(mut_seqs[i])) for i in range(mut_seqs.shape[0])]).transpose()\n",
    "mut_inj_test_df = pd.DataFrame([list(one_hot_decode(mut_inj_seqs[i])) for i in range(mut_inj_seqs.shape[0])]).transpose()\n",
    "wt_test_df.to_csv('tmp/wt_seqs.tsv', sep = '\\t')\n",
    "mut_test_df.to_csv('tmp/mut_seqs.tsv', sep = '\\t')\n",
    "mut_inj_test_df.to_csv('tmp/mut_inj_seqs.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Predict each sequence set\n",
    "\n",
    "Predict the WT, mutated, and mutated-injected sequences that were generated above. First, import the model of interest to generate predictions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-12-28 13:06:10,788 [WARNING] From /home/kd2200/anaconda3/envs/bpnet/lib/python3.6/site-packages/Keras-2.2.4-py3.6.egg/keras/backend/tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.\n",
      "\n",
      "2023-12-28 13:06:10,820 [WARNING] From /home/kd2200/anaconda3/envs/bpnet/lib/python3.6/site-packages/Keras-2.2.4-py3.6.egg/keras/backend/tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instead.\n",
      "\n",
      "2023-12-28 13:06:11,453 [WARNING] From /home/kd2200/anaconda3/envs/bpnet/lib/python3.6/site-packages/Keras-2.2.4-py3.6.egg/keras/backend/tensorflow_backend.py:174: The name tf.get_default_session is deprecated. Please use tf.compat.v1.get_default_session instead.\n",
      "\n",
      "2023-12-28 13:06:11,455 [WARNING] From /home/kd2200/anaconda3/envs/bpnet/lib/python3.6/site-packages/Keras-2.2.4-py3.6.egg/keras/backend/tensorflow_backend.py:181: The name tf.ConfigProto is deprecated. Please use tf.compat.v1.ConfigProto instead.\n",
      "\n",
      "2023-12-28 13:06:11,456 [WARNING] From /home/kd2200/anaconda3/envs/bpnet/lib/python3.6/site-packages/Keras-2.2.4-py3.6.egg/keras/backend/tensorflow_backend.py:186: The name tf.Session is deprecated. Please use tf.compat.v1.Session instead.\n",
      "\n",
      "2023-12-28 13:06:11,484 [WARNING] From /home/kd2200/anaconda3/envs/bpnet/lib/python3.6/site-packages/Keras-2.2.4-py3.6.egg/keras/backend/tensorflow_backend.py:190: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n",
      "2023-12-28 13:06:11,486 [WARNING] From /home/kd2200/anaconda3/envs/bpnet/lib/python3.6/site-packages/Keras-2.2.4-py3.6.egg/keras/backend/tensorflow_backend.py:199: The name tf.is_variable_initialized is deprecated. Please use tf.compat.v1.is_variable_initialized instead.\n",
      "\n",
      "2023-12-28 13:06:11,631 [WARNING] From /home/kd2200/anaconda3/envs/bpnet/lib/python3.6/site-packages/Keras-2.2.4-py3.6.egg/keras/backend/tensorflow_backend.py:206: The name tf.variables_initializer is deprecated. Please use tf.compat.v1.variables_initializer instead.\n",
      "\n",
      "2023-12-28 13:06:11,813 [WARNING] \n",
      "The TensorFlow contrib module will not be included in TensorFlow 2.0.\n",
      "For more information, please see:\n",
      "  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md\n",
      "  * https://github.com/tensorflow/addons\n",
      "  * https://github.com/tensorflow/io (for I/O related ops)\n",
      "If you depend on functionality not listed there, please file an issue.\n",
      "\n",
      "2023-12-28 13:06:12,580 [WARNING] From /home/kd2200/anaconda3/envs/bpnet/lib/python3.6/site-packages/Keras-2.2.4-py3.6.egg/keras/backend/tensorflow_backend.py:986: The name tf.assign_add is deprecated. Please use tf.compat.v1.assign_add instead.\n",
      "\n",
      "2023-12-28 13:06:12,874 [WARNING] From /home/kd2200/anaconda3/envs/bpnet/lib/python3.6/site-packages/Keras-2.2.4-py3.6.egg/keras/backend/tensorflow_backend.py:973: The name tf.assign is deprecated. Please use tf.compat.v1.assign instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "model = BPNetSeqModel.from_mdir(model_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get predictions from each set of sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get predictions from sequences\n",
    "wt_preds = model.predict(wt_seqs)\n",
    "mut_preds = model.predict(mut_seqs)\n",
    "mut_inj_preds = model.predict(mut_inj_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Measure the whole-window counts summary of these perturbations, and link them into a single data frame. Add motif-range measurements in addition to whole-window measurements. Make this dynamic based on the injection site of distance.\n",
    "\n",
    "First, define the window range for the motif-range measurements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_range_width = 50"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, collect measurements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Measure WT and mutant task predictions across the whole window\n",
    "wt_preds_df = pd.DataFrame.from_dict({k: np.sum(v, axis = (1,2)) for k,v in wt_preds.items()})\n",
    "wt_preds_df.columns = [i + '_wt_whole-range' for i in wt_preds_df.columns]\n",
    "mut_preds_df = pd.DataFrame.from_dict({k:np.sum(v, axis = (1,2)) for k,v in mut_preds.items()})\n",
    "mut_preds_df.columns = [i + '_mut_whole-range' for i in mut_preds_df.columns]\n",
    "wt_info_w_preds_df = pd.concat([wt_info_all_df.reset_index(), wt_preds_df.reset_index(), mut_preds_df.reset_index()], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wt_info_w_preds_df.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Measure mut_inj predictions across the whole window for every distance\n",
    "mut_inj_preds_df = pd.DataFrame.from_dict({k: np.sum(v, axis = (1,2)) for k,v in mut_inj_preds.items()})\n",
    "mut_inj_preds_df.columns = [i + '_mut-inj_whole-range' for i in mut_inj_preds_df.columns]\n",
    "mut_inj_info_w_preds_df = pd.concat([mut_inj_info_all_df.reset_index(), mut_inj_preds_df.reset_index()], axis=1)\n",
    "mut_inj_info_w_preds_df['genomic_moved_minus_anchor_motif_distance'] = mut_inj_info_w_preds_df.moved_motif_window_orig_center_0based - mut_inj_info_w_preds_df.anchor_motif_window_center_0based\n",
    "whole_window_summary_df = mut_inj_info_w_preds_df.merge(wt_info_w_preds_df[['enhancer_id', 'anchor_motif', 'moved_motif'] + list(wt_preds_df.columns) +  list(mut_preds_df.columns)], \n",
    "                              on=['enhancer_id', 'anchor_motif', 'moved_motif'], how = 'left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "117486"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_inj_info_w_preds_df.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each set of enhancers, anchor/move motifs, and distances, extract the motif-window measurements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "117486it [27:06, 72.21it/s]\n"
     ]
    }
   ],
   "source": [
    "motif_window_predictions = pd.DataFrame()\n",
    "for i,row in tqdm(whole_window_summary_df.iterrows()):\n",
    "    injected_center = row.anchor_motif_window_center_0based - row.distance_between_moved_center_and_anchor_center\n",
    "    injected_coords = (injected_center - (row.moved_motif_window_orig_center_0based - row.moved_motif_window_orig_start_0based),\n",
    "                       injected_center - (row.moved_motif_window_orig_end_0based - row.moved_motif_window_orig_center_0based))\n",
    "    injected_motif_window = (int(injected_center - np.floor(motif_range_width/2)), \n",
    "                             int(injected_center + np.floor(motif_range_width/2)))\n",
    "    \n",
    "    #Define enhancer_id and anchor_motif to index WT_preds and MUT_preds\n",
    "    enhancer_id = row.enhancer_id\n",
    "    anchor_motif = row.anchor_motif\n",
    "    wt_and_mut_index = int(wt_info_w_preds_df[(wt_info_w_preds_df.enhancer_id==enhancer_id) & (wt_info_w_preds_df.anchor_motif==anchor_motif)].index.values[0])\n",
    "    print(wt_and_mut_index)                   \n",
    "    #Measure motif range across 3 types for injected window\n",
    "    i_wt_preds_df = pd.DataFrame({k: np.sum(v[wt_and_mut_index, injected_motif_window[0]:injected_motif_window[1], :], axis = (0,1)) \n",
    "                                          for k,v in wt_preds.items()}, index=[0])\n",
    "    i_wt_preds_df.columns = [i + '_wt_motif-range_inj-window' for i in i_wt_preds_df.columns]\n",
    "    i_mut_preds_df = pd.DataFrame({k:np.sum(v[wt_and_mut_index, injected_motif_window[0]:injected_motif_window[1], :], axis = (0,1)) \n",
    "                                           for k,v in mut_preds.items()}, index=[0])\n",
    "    i_mut_preds_df.columns = [i + '_mut_motif-range_inj-window' for i in i_mut_preds_df.columns]    \n",
    "    i_mut_inj_preds_df = pd.DataFrame({k: np.sum(v[i, injected_motif_window[0]:injected_motif_window[1], :], axis = (0,1)) \n",
    "                                               for k,v in mut_inj_preds.items()}, index=[0])\n",
    "    i_mut_inj_preds_df.columns = [i + '_mut-inj_motif-range_inj-window' for i in i_mut_inj_preds_df.columns]\n",
    "                      \n",
    "    #Measure motif range across 3 types for anchor window\n",
    "    anchor_motif_window = (int(row.anchor_motif_window_center_0based - np.floor(motif_range_width/2)), \n",
    "                           int(row.anchor_motif_window_center_0based + np.floor(motif_range_width/2)))                            \n",
    "    a_wt_preds_df = pd.DataFrame({k: np.sum(v[wt_and_mut_index, anchor_motif_window[0]:anchor_motif_window[1], :], axis = (0,1)) \n",
    "                                          for k,v in wt_preds.items()}, index=[0])\n",
    "    a_wt_preds_df.columns = [i + '_wt_motif-range_anchor-window' for i in a_wt_preds_df.columns]\n",
    "    a_mut_preds_df = pd.DataFrame({k:np.sum(v[wt_and_mut_index, anchor_motif_window[0]:anchor_motif_window[1], :], axis = (0,1)) \n",
    "                                           for k,v in mut_preds.items()}, index=[0])\n",
    "    a_mut_preds_df.columns = [i + '_mut_motif-range_anchor-window' for i in a_mut_preds_df.columns]   \n",
    "    a_mut_inj_preds_df = pd.DataFrame({k: np.sum(v[i, anchor_motif_window[0]:anchor_motif_window[1], :], axis = (0,1)) \n",
    "                                               for k,v in mut_inj_preds.items()}, index=[0])\n",
    "    a_mut_inj_preds_df.columns = [i + '_mut-inj_motif-range_anchor-window' for i in a_mut_inj_preds_df.columns]\n",
    "    \n",
    "    #Measure motif range across 3 types for mutation window\n",
    "    moved_orig_motif_window = (int(row.moved_motif_window_orig_center_0based - np.floor(motif_range_width/2)), \n",
    "                               int(row.moved_motif_window_orig_center_0based + np.floor(motif_range_width/2)))                            \n",
    "    m_wt_preds_df = pd.DataFrame({k: np.sum(v[wt_and_mut_index, moved_orig_motif_window[0]:moved_orig_motif_window[1], :], axis = (0,1)) \n",
    "                                          for k,v in wt_preds.items()}, index=[0])\n",
    "    m_wt_preds_df.columns = [i + '_wt_motif-range_mut-window' for i in m_wt_preds_df.columns]\n",
    "    m_mut_preds_df = pd.DataFrame({k:np.sum(v[wt_and_mut_index, moved_orig_motif_window[0]:moved_orig_motif_window[1], :], axis = (0,1)) \n",
    "                                           for k,v in mut_preds.items()}, index=[0])\n",
    "    m_mut_preds_df.columns = [i + '_mut_motif-range_mut-window' for i in m_mut_preds_df.columns]    \n",
    "    m_mut_inj_preds_df = pd.DataFrame({k: np.sum(v[i, moved_orig_motif_window[0]:moved_orig_motif_window[1], :], axis = (0,1)) \n",
    "                                               for k,v in mut_inj_preds.items()}, index=[0])\n",
    "    m_mut_inj_preds_df.columns = [i + '_mut-inj_motif-range_mut-window' for i in m_mut_inj_preds_df.columns]\n",
    "                               \n",
    "    motif_window_prediction = pd.concat([a_wt_preds_df, a_mut_preds_df, a_mut_inj_preds_df,\n",
    "                                         m_wt_preds_df, m_mut_preds_df, m_mut_inj_preds_df,\n",
    "                                         i_wt_preds_df, i_mut_preds_df, i_mut_inj_preds_df],axis = 1)  \n",
    "\n",
    "    motif_window_predictions = motif_window_predictions.append(motif_window_prediction)\n",
    "                               \n",
    "assert motif_window_predictions.shape[0]==whole_window_summary_df.shape[0]\n",
    "final_pred_summary_df = pd.concat([whole_window_summary_df,motif_window_predictions.reset_index()], axis = 1)                              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>enhancer_id</th>\n",
       "      <th>anchor_motif</th>\n",
       "      <th>anchor_motif_window_start_0based</th>\n",
       "      <th>anchor_motif_window_end_0based</th>\n",
       "      <th>anchor_motif_window_center_0based</th>\n",
       "      <th>moved_motif</th>\n",
       "      <th>moved_motif_window_orig_start_0based</th>\n",
       "      <th>moved_motif_window_orig_end_0based</th>\n",
       "      <th>moved_motif_window_orig_center_0based</th>\n",
       "      <th>rank_positions</th>\n",
       "      <th>injected_motif_seq</th>\n",
       "      <th>mutated_motif_seq</th>\n",
       "      <th>anchor_motif_seq</th>\n",
       "      <th>distance_between_moved_center_and_anchor_center</th>\n",
       "      <th>distance_id</th>\n",
       "      <th>index</th>\n",
       "      <th>cdx2_mut-inj_whole-range</th>\n",
       "      <th>tfap2c_mut-inj_whole-range</th>\n",
       "      <th>tead4_mut-inj_whole-range</th>\n",
       "      <th>yap1_mut-inj_whole-range</th>\n",
       "      <th>gata3_mut-inj_whole-range</th>\n",
       "      <th>genomic_moved_minus_anchor_motif_distance</th>\n",
       "      <th>cdx2_wt_whole-range</th>\n",
       "      <th>tfap2c_wt_whole-range</th>\n",
       "      <th>tead4_wt_whole-range</th>\n",
       "      <th>yap1_wt_whole-range</th>\n",
       "      <th>gata3_wt_whole-range</th>\n",
       "      <th>cdx2_mut_whole-range</th>\n",
       "      <th>tfap2c_mut_whole-range</th>\n",
       "      <th>tead4_mut_whole-range</th>\n",
       "      <th>yap1_mut_whole-range</th>\n",
       "      <th>gata3_mut_whole-range</th>\n",
       "      <th>index</th>\n",
       "      <th>cdx2_wt_motif-range_anchor-window</th>\n",
       "      <th>tfap2c_wt_motif-range_anchor-window</th>\n",
       "      <th>tead4_wt_motif-range_anchor-window</th>\n",
       "      <th>yap1_wt_motif-range_anchor-window</th>\n",
       "      <th>gata3_wt_motif-range_anchor-window</th>\n",
       "      <th>cdx2_mut_motif-range_anchor-window</th>\n",
       "      <th>tfap2c_mut_motif-range_anchor-window</th>\n",
       "      <th>tead4_mut_motif-range_anchor-window</th>\n",
       "      <th>yap1_mut_motif-range_anchor-window</th>\n",
       "      <th>gata3_mut_motif-range_anchor-window</th>\n",
       "      <th>cdx2_mut-inj_motif-range_anchor-window</th>\n",
       "      <th>tfap2c_mut-inj_motif-range_anchor-window</th>\n",
       "      <th>tead4_mut-inj_motif-range_anchor-window</th>\n",
       "      <th>yap1_mut-inj_motif-range_anchor-window</th>\n",
       "      <th>gata3_mut-inj_motif-range_anchor-window</th>\n",
       "      <th>cdx2_wt_motif-range_mut-window</th>\n",
       "      <th>tfap2c_wt_motif-range_mut-window</th>\n",
       "      <th>tead4_wt_motif-range_mut-window</th>\n",
       "      <th>yap1_wt_motif-range_mut-window</th>\n",
       "      <th>gata3_wt_motif-range_mut-window</th>\n",
       "      <th>cdx2_mut_motif-range_mut-window</th>\n",
       "      <th>tfap2c_mut_motif-range_mut-window</th>\n",
       "      <th>tead4_mut_motif-range_mut-window</th>\n",
       "      <th>yap1_mut_motif-range_mut-window</th>\n",
       "      <th>gata3_mut_motif-range_mut-window</th>\n",
       "      <th>cdx2_mut-inj_motif-range_mut-window</th>\n",
       "      <th>tfap2c_mut-inj_motif-range_mut-window</th>\n",
       "      <th>tead4_mut-inj_motif-range_mut-window</th>\n",
       "      <th>yap1_mut-inj_motif-range_mut-window</th>\n",
       "      <th>gata3_mut-inj_motif-range_mut-window</th>\n",
       "      <th>cdx2_wt_motif-range_inj-window</th>\n",
       "      <th>tfap2c_wt_motif-range_inj-window</th>\n",
       "      <th>tead4_wt_motif-range_inj-window</th>\n",
       "      <th>yap1_wt_motif-range_inj-window</th>\n",
       "      <th>gata3_wt_motif-range_inj-window</th>\n",
       "      <th>cdx2_mut_motif-range_inj-window</th>\n",
       "      <th>tfap2c_mut_motif-range_inj-window</th>\n",
       "      <th>tead4_mut_motif-range_inj-window</th>\n",
       "      <th>yap1_mut_motif-range_inj-window</th>\n",
       "      <th>gata3_mut_motif-range_inj-window</th>\n",
       "      <th>cdx2_mut-inj_motif-range_inj-window</th>\n",
       "      <th>tfap2c_mut-inj_motif-range_inj-window</th>\n",
       "      <th>tead4_mut-inj_motif-range_inj-window</th>\n",
       "      <th>yap1_mut-inj_motif-range_inj-window</th>\n",
       "      <th>gata3_mut-inj_motif-range_inj-window</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>tfap2c</td>\n",
       "      <td>473</td>\n",
       "      <td>485</td>\n",
       "      <td>479</td>\n",
       "      <td>tead4</td>\n",
       "      <td>516</td>\n",
       "      <td>526</td>\n",
       "      <td>521</td>\n",
       "      <td>6,7</td>\n",
       "      <td>CCTGGAATGT</td>\n",
       "      <td>CCTGGAGAGT</td>\n",
       "      <td>CCCCCTGAGGCA</td>\n",
       "      <td>-200</td>\n",
       "      <td>0_-200</td>\n",
       "      <td>0</td>\n",
       "      <td>32.586403</td>\n",
       "      <td>17.439480</td>\n",
       "      <td>41.252026</td>\n",
       "      <td>17.924112</td>\n",
       "      <td>11.688992</td>\n",
       "      <td>42</td>\n",
       "      <td>40.289711</td>\n",
       "      <td>22.881042</td>\n",
       "      <td>82.098846</td>\n",
       "      <td>25.748079</td>\n",
       "      <td>12.507404</td>\n",
       "      <td>32.175621</td>\n",
       "      <td>18.674313</td>\n",
       "      <td>21.664646</td>\n",
       "      <td>13.90106</td>\n",
       "      <td>12.43005</td>\n",
       "      <td>0</td>\n",
       "      <td>3.78507</td>\n",
       "      <td>5.57162</td>\n",
       "      <td>9.336608</td>\n",
       "      <td>3.556051</td>\n",
       "      <td>1.061003</td>\n",
       "      <td>1.98358</td>\n",
       "      <td>4.03511</td>\n",
       "      <td>1.170352</td>\n",
       "      <td>0.901606</td>\n",
       "      <td>0.900368</td>\n",
       "      <td>1.864665</td>\n",
       "      <td>3.676927</td>\n",
       "      <td>1.489836</td>\n",
       "      <td>1.030099</td>\n",
       "      <td>0.835613</td>\n",
       "      <td>5.880242</td>\n",
       "      <td>5.489297</td>\n",
       "      <td>44.698868</td>\n",
       "      <td>7.368331</td>\n",
       "      <td>1.461488</td>\n",
       "      <td>2.872119</td>\n",
       "      <td>3.414649</td>\n",
       "      <td>2.980212</td>\n",
       "      <td>1.528275</td>\n",
       "      <td>1.171574</td>\n",
       "      <td>2.771478</td>\n",
       "      <td>3.135449</td>\n",
       "      <td>3.741745</td>\n",
       "      <td>1.714167</td>\n",
       "      <td>1.067181</td>\n",
       "      <td>2.067455</td>\n",
       "      <td>0.840098</td>\n",
       "      <td>1.286956</td>\n",
       "      <td>0.963832</td>\n",
       "      <td>0.465170</td>\n",
       "      <td>1.758967</td>\n",
       "      <td>0.807522</td>\n",
       "      <td>0.957595</td>\n",
       "      <td>0.819860</td>\n",
       "      <td>0.502480</td>\n",
       "      <td>1.610182</td>\n",
       "      <td>0.697082</td>\n",
       "      <td>1.257510</td>\n",
       "      <td>0.873194</td>\n",
       "      <td>0.457523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>tfap2c</td>\n",
       "      <td>473</td>\n",
       "      <td>485</td>\n",
       "      <td>479</td>\n",
       "      <td>tead4</td>\n",
       "      <td>516</td>\n",
       "      <td>526</td>\n",
       "      <td>521</td>\n",
       "      <td>6,7</td>\n",
       "      <td>CCTGGAATGT</td>\n",
       "      <td>CCTGGAGAGT</td>\n",
       "      <td>CCCCCTGAGGCA</td>\n",
       "      <td>-199</td>\n",
       "      <td>0_-199</td>\n",
       "      <td>1</td>\n",
       "      <td>33.794708</td>\n",
       "      <td>17.476225</td>\n",
       "      <td>42.365692</td>\n",
       "      <td>17.766081</td>\n",
       "      <td>12.389852</td>\n",
       "      <td>42</td>\n",
       "      <td>40.289711</td>\n",
       "      <td>22.881042</td>\n",
       "      <td>82.098846</td>\n",
       "      <td>25.748079</td>\n",
       "      <td>12.507404</td>\n",
       "      <td>32.175621</td>\n",
       "      <td>18.674313</td>\n",
       "      <td>21.664646</td>\n",
       "      <td>13.90106</td>\n",
       "      <td>12.43005</td>\n",
       "      <td>0</td>\n",
       "      <td>3.78507</td>\n",
       "      <td>5.57162</td>\n",
       "      <td>9.336608</td>\n",
       "      <td>3.556051</td>\n",
       "      <td>1.061003</td>\n",
       "      <td>1.98358</td>\n",
       "      <td>4.03511</td>\n",
       "      <td>1.170352</td>\n",
       "      <td>0.901606</td>\n",
       "      <td>0.900368</td>\n",
       "      <td>1.830293</td>\n",
       "      <td>3.551563</td>\n",
       "      <td>1.324419</td>\n",
       "      <td>1.040776</td>\n",
       "      <td>0.900902</td>\n",
       "      <td>5.880242</td>\n",
       "      <td>5.489297</td>\n",
       "      <td>44.698868</td>\n",
       "      <td>7.368331</td>\n",
       "      <td>1.461488</td>\n",
       "      <td>2.872119</td>\n",
       "      <td>3.414649</td>\n",
       "      <td>2.980212</td>\n",
       "      <td>1.528275</td>\n",
       "      <td>1.171574</td>\n",
       "      <td>2.722175</td>\n",
       "      <td>3.169663</td>\n",
       "      <td>3.403313</td>\n",
       "      <td>1.693549</td>\n",
       "      <td>1.144201</td>\n",
       "      <td>2.089602</td>\n",
       "      <td>0.852262</td>\n",
       "      <td>1.302774</td>\n",
       "      <td>0.977092</td>\n",
       "      <td>0.464553</td>\n",
       "      <td>1.776769</td>\n",
       "      <td>0.816880</td>\n",
       "      <td>0.968045</td>\n",
       "      <td>0.827349</td>\n",
       "      <td>0.502215</td>\n",
       "      <td>1.674533</td>\n",
       "      <td>0.727360</td>\n",
       "      <td>1.219269</td>\n",
       "      <td>0.910385</td>\n",
       "      <td>0.484783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>tfap2c</td>\n",
       "      <td>473</td>\n",
       "      <td>485</td>\n",
       "      <td>479</td>\n",
       "      <td>tead4</td>\n",
       "      <td>516</td>\n",
       "      <td>526</td>\n",
       "      <td>521</td>\n",
       "      <td>6,7</td>\n",
       "      <td>CCTGGAATGT</td>\n",
       "      <td>CCTGGAGAGT</td>\n",
       "      <td>CCCCCTGAGGCA</td>\n",
       "      <td>-198</td>\n",
       "      <td>0_-198</td>\n",
       "      <td>2</td>\n",
       "      <td>33.564041</td>\n",
       "      <td>17.344700</td>\n",
       "      <td>39.337379</td>\n",
       "      <td>17.659176</td>\n",
       "      <td>11.963973</td>\n",
       "      <td>42</td>\n",
       "      <td>40.289711</td>\n",
       "      <td>22.881042</td>\n",
       "      <td>82.098846</td>\n",
       "      <td>25.748079</td>\n",
       "      <td>12.507404</td>\n",
       "      <td>32.175621</td>\n",
       "      <td>18.674313</td>\n",
       "      <td>21.664646</td>\n",
       "      <td>13.90106</td>\n",
       "      <td>12.43005</td>\n",
       "      <td>0</td>\n",
       "      <td>3.78507</td>\n",
       "      <td>5.57162</td>\n",
       "      <td>9.336608</td>\n",
       "      <td>3.556051</td>\n",
       "      <td>1.061003</td>\n",
       "      <td>1.98358</td>\n",
       "      <td>4.03511</td>\n",
       "      <td>1.170352</td>\n",
       "      <td>0.901606</td>\n",
       "      <td>0.900368</td>\n",
       "      <td>1.922206</td>\n",
       "      <td>3.684247</td>\n",
       "      <td>1.677499</td>\n",
       "      <td>1.035957</td>\n",
       "      <td>0.869245</td>\n",
       "      <td>5.880242</td>\n",
       "      <td>5.489297</td>\n",
       "      <td>44.698868</td>\n",
       "      <td>7.368331</td>\n",
       "      <td>1.461488</td>\n",
       "      <td>2.872119</td>\n",
       "      <td>3.414649</td>\n",
       "      <td>2.980212</td>\n",
       "      <td>1.528275</td>\n",
       "      <td>1.171574</td>\n",
       "      <td>2.793713</td>\n",
       "      <td>3.068845</td>\n",
       "      <td>4.105606</td>\n",
       "      <td>1.703379</td>\n",
       "      <td>1.111650</td>\n",
       "      <td>2.117435</td>\n",
       "      <td>0.865505</td>\n",
       "      <td>1.338787</td>\n",
       "      <td>0.993072</td>\n",
       "      <td>0.462888</td>\n",
       "      <td>1.799637</td>\n",
       "      <td>0.826136</td>\n",
       "      <td>0.990214</td>\n",
       "      <td>0.835365</td>\n",
       "      <td>0.501156</td>\n",
       "      <td>1.666581</td>\n",
       "      <td>0.716631</td>\n",
       "      <td>1.310965</td>\n",
       "      <td>0.905573</td>\n",
       "      <td>0.463220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>tfap2c</td>\n",
       "      <td>473</td>\n",
       "      <td>485</td>\n",
       "      <td>479</td>\n",
       "      <td>tead4</td>\n",
       "      <td>516</td>\n",
       "      <td>526</td>\n",
       "      <td>521</td>\n",
       "      <td>6,7</td>\n",
       "      <td>CCTGGAATGT</td>\n",
       "      <td>CCTGGAGAGT</td>\n",
       "      <td>CCCCCTGAGGCA</td>\n",
       "      <td>-197</td>\n",
       "      <td>0_-197</td>\n",
       "      <td>3</td>\n",
       "      <td>34.376530</td>\n",
       "      <td>18.410126</td>\n",
       "      <td>42.624908</td>\n",
       "      <td>18.048157</td>\n",
       "      <td>12.253555</td>\n",
       "      <td>42</td>\n",
       "      <td>40.289711</td>\n",
       "      <td>22.881042</td>\n",
       "      <td>82.098846</td>\n",
       "      <td>25.748079</td>\n",
       "      <td>12.507404</td>\n",
       "      <td>32.175621</td>\n",
       "      <td>18.674313</td>\n",
       "      <td>21.664646</td>\n",
       "      <td>13.90106</td>\n",
       "      <td>12.43005</td>\n",
       "      <td>0</td>\n",
       "      <td>3.78507</td>\n",
       "      <td>5.57162</td>\n",
       "      <td>9.336608</td>\n",
       "      <td>3.556051</td>\n",
       "      <td>1.061003</td>\n",
       "      <td>1.98358</td>\n",
       "      <td>4.03511</td>\n",
       "      <td>1.170352</td>\n",
       "      <td>0.901606</td>\n",
       "      <td>0.900368</td>\n",
       "      <td>1.855693</td>\n",
       "      <td>3.767128</td>\n",
       "      <td>1.407524</td>\n",
       "      <td>1.061882</td>\n",
       "      <td>0.875847</td>\n",
       "      <td>5.880242</td>\n",
       "      <td>5.489297</td>\n",
       "      <td>44.698868</td>\n",
       "      <td>7.368331</td>\n",
       "      <td>1.461488</td>\n",
       "      <td>2.872119</td>\n",
       "      <td>3.414649</td>\n",
       "      <td>2.980212</td>\n",
       "      <td>1.528275</td>\n",
       "      <td>1.171574</td>\n",
       "      <td>2.795211</td>\n",
       "      <td>3.304552</td>\n",
       "      <td>3.568894</td>\n",
       "      <td>1.751415</td>\n",
       "      <td>1.134392</td>\n",
       "      <td>2.145015</td>\n",
       "      <td>0.879383</td>\n",
       "      <td>1.375235</td>\n",
       "      <td>1.010723</td>\n",
       "      <td>0.463652</td>\n",
       "      <td>1.820538</td>\n",
       "      <td>0.836888</td>\n",
       "      <td>1.016771</td>\n",
       "      <td>0.847623</td>\n",
       "      <td>0.502921</td>\n",
       "      <td>1.760733</td>\n",
       "      <td>0.773272</td>\n",
       "      <td>1.338766</td>\n",
       "      <td>0.963139</td>\n",
       "      <td>0.479209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>tfap2c</td>\n",
       "      <td>473</td>\n",
       "      <td>485</td>\n",
       "      <td>479</td>\n",
       "      <td>tead4</td>\n",
       "      <td>516</td>\n",
       "      <td>526</td>\n",
       "      <td>521</td>\n",
       "      <td>6,7</td>\n",
       "      <td>CCTGGAATGT</td>\n",
       "      <td>CCTGGAGAGT</td>\n",
       "      <td>CCCCCTGAGGCA</td>\n",
       "      <td>-196</td>\n",
       "      <td>0_-196</td>\n",
       "      <td>4</td>\n",
       "      <td>32.639538</td>\n",
       "      <td>17.831081</td>\n",
       "      <td>33.506481</td>\n",
       "      <td>16.410358</td>\n",
       "      <td>12.066856</td>\n",
       "      <td>42</td>\n",
       "      <td>40.289711</td>\n",
       "      <td>22.881042</td>\n",
       "      <td>82.098846</td>\n",
       "      <td>25.748079</td>\n",
       "      <td>12.507404</td>\n",
       "      <td>32.175621</td>\n",
       "      <td>18.674313</td>\n",
       "      <td>21.664646</td>\n",
       "      <td>13.90106</td>\n",
       "      <td>12.43005</td>\n",
       "      <td>0</td>\n",
       "      <td>3.78507</td>\n",
       "      <td>5.57162</td>\n",
       "      <td>9.336608</td>\n",
       "      <td>3.556051</td>\n",
       "      <td>1.061003</td>\n",
       "      <td>1.98358</td>\n",
       "      <td>4.03511</td>\n",
       "      <td>1.170352</td>\n",
       "      <td>0.901606</td>\n",
       "      <td>0.900368</td>\n",
       "      <td>1.918224</td>\n",
       "      <td>3.825709</td>\n",
       "      <td>1.486437</td>\n",
       "      <td>0.996670</td>\n",
       "      <td>0.870556</td>\n",
       "      <td>5.880242</td>\n",
       "      <td>5.489297</td>\n",
       "      <td>44.698868</td>\n",
       "      <td>7.368331</td>\n",
       "      <td>1.461488</td>\n",
       "      <td>2.872119</td>\n",
       "      <td>3.414649</td>\n",
       "      <td>2.980212</td>\n",
       "      <td>1.528275</td>\n",
       "      <td>1.171574</td>\n",
       "      <td>2.772931</td>\n",
       "      <td>3.166693</td>\n",
       "      <td>3.726338</td>\n",
       "      <td>1.651039</td>\n",
       "      <td>1.100751</td>\n",
       "      <td>2.176472</td>\n",
       "      <td>0.893857</td>\n",
       "      <td>1.404888</td>\n",
       "      <td>1.029042</td>\n",
       "      <td>0.466761</td>\n",
       "      <td>1.844383</td>\n",
       "      <td>0.848398</td>\n",
       "      <td>1.036326</td>\n",
       "      <td>0.859559</td>\n",
       "      <td>0.506482</td>\n",
       "      <td>1.747933</td>\n",
       "      <td>0.770797</td>\n",
       "      <td>1.290250</td>\n",
       "      <td>0.908727</td>\n",
       "      <td>0.484424</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index enhancer_id anchor_motif anchor_motif_window_start_0based  \\\n",
       "0      0           0       tfap2c                              473   \n",
       "1      0           0       tfap2c                              473   \n",
       "2      0           0       tfap2c                              473   \n",
       "3      0           0       tfap2c                              473   \n",
       "4      0           0       tfap2c                              473   \n",
       "\n",
       "  anchor_motif_window_end_0based anchor_motif_window_center_0based  \\\n",
       "0                            485                               479   \n",
       "1                            485                               479   \n",
       "2                            485                               479   \n",
       "3                            485                               479   \n",
       "4                            485                               479   \n",
       "\n",
       "  moved_motif moved_motif_window_orig_start_0based  \\\n",
       "0       tead4                                  516   \n",
       "1       tead4                                  516   \n",
       "2       tead4                                  516   \n",
       "3       tead4                                  516   \n",
       "4       tead4                                  516   \n",
       "\n",
       "  moved_motif_window_orig_end_0based moved_motif_window_orig_center_0based  \\\n",
       "0                                526                                   521   \n",
       "1                                526                                   521   \n",
       "2                                526                                   521   \n",
       "3                                526                                   521   \n",
       "4                                526                                   521   \n",
       "\n",
       "  rank_positions injected_motif_seq mutated_motif_seq anchor_motif_seq  \\\n",
       "0            6,7         CCTGGAATGT        CCTGGAGAGT     CCCCCTGAGGCA   \n",
       "1            6,7         CCTGGAATGT        CCTGGAGAGT     CCCCCTGAGGCA   \n",
       "2            6,7         CCTGGAATGT        CCTGGAGAGT     CCCCCTGAGGCA   \n",
       "3            6,7         CCTGGAATGT        CCTGGAGAGT     CCCCCTGAGGCA   \n",
       "4            6,7         CCTGGAATGT        CCTGGAGAGT     CCCCCTGAGGCA   \n",
       "\n",
       "   distance_between_moved_center_and_anchor_center distance_id  index  \\\n",
       "0                                             -200      0_-200      0   \n",
       "1                                             -199      0_-199      1   \n",
       "2                                             -198      0_-198      2   \n",
       "3                                             -197      0_-197      3   \n",
       "4                                             -196      0_-196      4   \n",
       "\n",
       "   cdx2_mut-inj_whole-range  tfap2c_mut-inj_whole-range  \\\n",
       "0                 32.586403                   17.439480   \n",
       "1                 33.794708                   17.476225   \n",
       "2                 33.564041                   17.344700   \n",
       "3                 34.376530                   18.410126   \n",
       "4                 32.639538                   17.831081   \n",
       "\n",
       "   tead4_mut-inj_whole-range  yap1_mut-inj_whole-range  \\\n",
       "0                  41.252026                 17.924112   \n",
       "1                  42.365692                 17.766081   \n",
       "2                  39.337379                 17.659176   \n",
       "3                  42.624908                 18.048157   \n",
       "4                  33.506481                 16.410358   \n",
       "\n",
       "   gata3_mut-inj_whole-range genomic_moved_minus_anchor_motif_distance  \\\n",
       "0                  11.688992                                        42   \n",
       "1                  12.389852                                        42   \n",
       "2                  11.963973                                        42   \n",
       "3                  12.253555                                        42   \n",
       "4                  12.066856                                        42   \n",
       "\n",
       "   cdx2_wt_whole-range  tfap2c_wt_whole-range  tead4_wt_whole-range  \\\n",
       "0            40.289711              22.881042             82.098846   \n",
       "1            40.289711              22.881042             82.098846   \n",
       "2            40.289711              22.881042             82.098846   \n",
       "3            40.289711              22.881042             82.098846   \n",
       "4            40.289711              22.881042             82.098846   \n",
       "\n",
       "   yap1_wt_whole-range  gata3_wt_whole-range  cdx2_mut_whole-range  \\\n",
       "0            25.748079             12.507404             32.175621   \n",
       "1            25.748079             12.507404             32.175621   \n",
       "2            25.748079             12.507404             32.175621   \n",
       "3            25.748079             12.507404             32.175621   \n",
       "4            25.748079             12.507404             32.175621   \n",
       "\n",
       "   tfap2c_mut_whole-range  tead4_mut_whole-range  yap1_mut_whole-range  \\\n",
       "0               18.674313              21.664646              13.90106   \n",
       "1               18.674313              21.664646              13.90106   \n",
       "2               18.674313              21.664646              13.90106   \n",
       "3               18.674313              21.664646              13.90106   \n",
       "4               18.674313              21.664646              13.90106   \n",
       "\n",
       "   gata3_mut_whole-range  index  cdx2_wt_motif-range_anchor-window  \\\n",
       "0               12.43005      0                            3.78507   \n",
       "1               12.43005      0                            3.78507   \n",
       "2               12.43005      0                            3.78507   \n",
       "3               12.43005      0                            3.78507   \n",
       "4               12.43005      0                            3.78507   \n",
       "\n",
       "   tfap2c_wt_motif-range_anchor-window  tead4_wt_motif-range_anchor-window  \\\n",
       "0                              5.57162                            9.336608   \n",
       "1                              5.57162                            9.336608   \n",
       "2                              5.57162                            9.336608   \n",
       "3                              5.57162                            9.336608   \n",
       "4                              5.57162                            9.336608   \n",
       "\n",
       "   yap1_wt_motif-range_anchor-window  gata3_wt_motif-range_anchor-window  \\\n",
       "0                           3.556051                            1.061003   \n",
       "1                           3.556051                            1.061003   \n",
       "2                           3.556051                            1.061003   \n",
       "3                           3.556051                            1.061003   \n",
       "4                           3.556051                            1.061003   \n",
       "\n",
       "   cdx2_mut_motif-range_anchor-window  tfap2c_mut_motif-range_anchor-window  \\\n",
       "0                             1.98358                               4.03511   \n",
       "1                             1.98358                               4.03511   \n",
       "2                             1.98358                               4.03511   \n",
       "3                             1.98358                               4.03511   \n",
       "4                             1.98358                               4.03511   \n",
       "\n",
       "   tead4_mut_motif-range_anchor-window  yap1_mut_motif-range_anchor-window  \\\n",
       "0                             1.170352                            0.901606   \n",
       "1                             1.170352                            0.901606   \n",
       "2                             1.170352                            0.901606   \n",
       "3                             1.170352                            0.901606   \n",
       "4                             1.170352                            0.901606   \n",
       "\n",
       "   gata3_mut_motif-range_anchor-window  \\\n",
       "0                             0.900368   \n",
       "1                             0.900368   \n",
       "2                             0.900368   \n",
       "3                             0.900368   \n",
       "4                             0.900368   \n",
       "\n",
       "   cdx2_mut-inj_motif-range_anchor-window  \\\n",
       "0                                1.864665   \n",
       "1                                1.830293   \n",
       "2                                1.922206   \n",
       "3                                1.855693   \n",
       "4                                1.918224   \n",
       "\n",
       "   tfap2c_mut-inj_motif-range_anchor-window  \\\n",
       "0                                  3.676927   \n",
       "1                                  3.551563   \n",
       "2                                  3.684247   \n",
       "3                                  3.767128   \n",
       "4                                  3.825709   \n",
       "\n",
       "   tead4_mut-inj_motif-range_anchor-window  \\\n",
       "0                                 1.489836   \n",
       "1                                 1.324419   \n",
       "2                                 1.677499   \n",
       "3                                 1.407524   \n",
       "4                                 1.486437   \n",
       "\n",
       "   yap1_mut-inj_motif-range_anchor-window  \\\n",
       "0                                1.030099   \n",
       "1                                1.040776   \n",
       "2                                1.035957   \n",
       "3                                1.061882   \n",
       "4                                0.996670   \n",
       "\n",
       "   gata3_mut-inj_motif-range_anchor-window  cdx2_wt_motif-range_mut-window  \\\n",
       "0                                 0.835613                        5.880242   \n",
       "1                                 0.900902                        5.880242   \n",
       "2                                 0.869245                        5.880242   \n",
       "3                                 0.875847                        5.880242   \n",
       "4                                 0.870556                        5.880242   \n",
       "\n",
       "   tfap2c_wt_motif-range_mut-window  tead4_wt_motif-range_mut-window  \\\n",
       "0                          5.489297                        44.698868   \n",
       "1                          5.489297                        44.698868   \n",
       "2                          5.489297                        44.698868   \n",
       "3                          5.489297                        44.698868   \n",
       "4                          5.489297                        44.698868   \n",
       "\n",
       "   yap1_wt_motif-range_mut-window  gata3_wt_motif-range_mut-window  \\\n",
       "0                        7.368331                         1.461488   \n",
       "1                        7.368331                         1.461488   \n",
       "2                        7.368331                         1.461488   \n",
       "3                        7.368331                         1.461488   \n",
       "4                        7.368331                         1.461488   \n",
       "\n",
       "   cdx2_mut_motif-range_mut-window  tfap2c_mut_motif-range_mut-window  \\\n",
       "0                         2.872119                           3.414649   \n",
       "1                         2.872119                           3.414649   \n",
       "2                         2.872119                           3.414649   \n",
       "3                         2.872119                           3.414649   \n",
       "4                         2.872119                           3.414649   \n",
       "\n",
       "   tead4_mut_motif-range_mut-window  yap1_mut_motif-range_mut-window  \\\n",
       "0                          2.980212                         1.528275   \n",
       "1                          2.980212                         1.528275   \n",
       "2                          2.980212                         1.528275   \n",
       "3                          2.980212                         1.528275   \n",
       "4                          2.980212                         1.528275   \n",
       "\n",
       "   gata3_mut_motif-range_mut-window  cdx2_mut-inj_motif-range_mut-window  \\\n",
       "0                          1.171574                             2.771478   \n",
       "1                          1.171574                             2.722175   \n",
       "2                          1.171574                             2.793713   \n",
       "3                          1.171574                             2.795211   \n",
       "4                          1.171574                             2.772931   \n",
       "\n",
       "   tfap2c_mut-inj_motif-range_mut-window  \\\n",
       "0                               3.135449   \n",
       "1                               3.169663   \n",
       "2                               3.068845   \n",
       "3                               3.304552   \n",
       "4                               3.166693   \n",
       "\n",
       "   tead4_mut-inj_motif-range_mut-window  yap1_mut-inj_motif-range_mut-window  \\\n",
       "0                              3.741745                             1.714167   \n",
       "1                              3.403313                             1.693549   \n",
       "2                              4.105606                             1.703379   \n",
       "3                              3.568894                             1.751415   \n",
       "4                              3.726338                             1.651039   \n",
       "\n",
       "   gata3_mut-inj_motif-range_mut-window  cdx2_wt_motif-range_inj-window  \\\n",
       "0                              1.067181                        2.067455   \n",
       "1                              1.144201                        2.089602   \n",
       "2                              1.111650                        2.117435   \n",
       "3                              1.134392                        2.145015   \n",
       "4                              1.100751                        2.176472   \n",
       "\n",
       "   tfap2c_wt_motif-range_inj-window  tead4_wt_motif-range_inj-window  \\\n",
       "0                          0.840098                         1.286956   \n",
       "1                          0.852262                         1.302774   \n",
       "2                          0.865505                         1.338787   \n",
       "3                          0.879383                         1.375235   \n",
       "4                          0.893857                         1.404888   \n",
       "\n",
       "   yap1_wt_motif-range_inj-window  gata3_wt_motif-range_inj-window  \\\n",
       "0                        0.963832                         0.465170   \n",
       "1                        0.977092                         0.464553   \n",
       "2                        0.993072                         0.462888   \n",
       "3                        1.010723                         0.463652   \n",
       "4                        1.029042                         0.466761   \n",
       "\n",
       "   cdx2_mut_motif-range_inj-window  tfap2c_mut_motif-range_inj-window  \\\n",
       "0                         1.758967                           0.807522   \n",
       "1                         1.776769                           0.816880   \n",
       "2                         1.799637                           0.826136   \n",
       "3                         1.820538                           0.836888   \n",
       "4                         1.844383                           0.848398   \n",
       "\n",
       "   tead4_mut_motif-range_inj-window  yap1_mut_motif-range_inj-window  \\\n",
       "0                          0.957595                         0.819860   \n",
       "1                          0.968045                         0.827349   \n",
       "2                          0.990214                         0.835365   \n",
       "3                          1.016771                         0.847623   \n",
       "4                          1.036326                         0.859559   \n",
       "\n",
       "   gata3_mut_motif-range_inj-window  cdx2_mut-inj_motif-range_inj-window  \\\n",
       "0                          0.502480                             1.610182   \n",
       "1                          0.502215                             1.674533   \n",
       "2                          0.501156                             1.666581   \n",
       "3                          0.502921                             1.760733   \n",
       "4                          0.506482                             1.747933   \n",
       "\n",
       "   tfap2c_mut-inj_motif-range_inj-window  \\\n",
       "0                               0.697082   \n",
       "1                               0.727360   \n",
       "2                               0.716631   \n",
       "3                               0.773272   \n",
       "4                               0.770797   \n",
       "\n",
       "   tead4_mut-inj_motif-range_inj-window  yap1_mut-inj_motif-range_inj-window  \\\n",
       "0                              1.257510                             0.873194   \n",
       "1                              1.219269                             0.910385   \n",
       "2                              1.310965                             0.905573   \n",
       "3                              1.338766                             0.963139   \n",
       "4                              1.290250                             0.908727   \n",
       "\n",
       "   gata3_mut-inj_motif-range_inj-window  \n",
       "0                              0.457523  \n",
       "1                              0.484783  \n",
       "2                              0.463220  \n",
       "3                              0.479209  \n",
       "4                              0.484424  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_pred_summary_df.head(n=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "## save each enhancer_id as seperate tsv\n",
    "for id in final_pred_summary_df['enhancer_id'].unique():\n",
    "    subset_df = final_pred_summary_df[final_pred_summary_df['enhancer_id'] == id]\n",
    "    filename = f'two_mutation/tsv/individual_pred_summary_by_enhancer_id/' + str(id) + '_pred_summary_df' + '.tsv.gz'\n",
    "    subset_df.to_csv(filename, index=False, sep='\\t')  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(117486, 79)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#save `final_pred_summary_df` as .tsv.gz\n",
    "final_pred_summary_df.to_csv(f'two_mutation/tsv/combined/{prefix}_final_pred_summary_df.tsv.gz', sep = '\\t', index = False)\n",
    "final_pred_summary_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(117486, 79)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%script false --no-raise-error\n",
    "#read csv\n",
    "final_pred_summary_df= pd.read_csv(f'two_mutation/tsv/combined/{prefix}_final_pred_summary_df.tsv.gz', sep = '\\t')\n",
    "final_pred_summary_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Measure entire profile signals across desired distances of mut/inj, mut, and wt predictions. Store in a tidy `pd.df` and export to a `tsv.gz` for plotting using `plotnine` or `ggplot` in R. First, specify distances that you want to preserve based on injection distance from the anchor motif."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0,\n",
       " 15,\n",
       " 30,\n",
       " 45,\n",
       " 60,\n",
       " 75,\n",
       " 90,\n",
       " 105,\n",
       " 120,\n",
       " 135,\n",
       " 150,\n",
       " 165,\n",
       " 180,\n",
       " 195,\n",
       " 0,\n",
       " -15,\n",
       " -30,\n",
       " -45,\n",
       " -60,\n",
       " -75,\n",
       " -90,\n",
       " -105,\n",
       " -120,\n",
       " -135,\n",
       " -150,\n",
       " -165,\n",
       " -180,\n",
       " -195]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "distances_to_store_profile = list(range(0, 200, 15)) + [i * -1 for i in list(range(0, 200, 15))]\n",
    "distances_to_store_profile\n",
    "\n",
    "#distances_to_store_profile = list(range(0, 200, 10)) + [i * -1 for i in list(range(0, 200, 10))]\n",
    "#distances_to_store_profile"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, extract and tidy profiles, eventually saving them as a `.tsv.gz`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_profiles_df = pd.DataFrame()\n",
    "\n",
    "for task in model.tasks:\n",
    "    ## Generate WT profiles\n",
    "    #Allocate all positive reads as 2D matrix and then a data frame\n",
    "    wt_pos_profile_df = pd.DataFrame(data=wt_preds[task][:,:,0], columns=list(range(0, model.input_seqlen())))\n",
    "    wt_pos_profile_df['enhancer_id'] = wt_info_w_preds_df.enhancer_id\n",
    "    wt_pos_profile_df['anchor_motif'] = wt_info_w_preds_df.anchor_motif\n",
    "    wt_pos_profile_df['moved_motif'] = wt_info_w_preds_df.moved_motif\n",
    "    wt_pos_profile_df = wt_pos_profile_df.melt(id_vars = ['enhancer_id', 'anchor_motif', 'moved_motif'], \n",
    "                                               var_name = 'position', value_name = 'signal')\n",
    "    wt_pos_profile_df['type'] = 'wt'\n",
    "    wt_pos_profile_df['strand'] = 'pos'\n",
    "    wt_pos_profile_df['task'] = task\n",
    "\n",
    "    #Allocate all negative reads as 2D matrix and then a data frame\n",
    "    wt_neg_profile_df = pd.DataFrame(data=wt_preds[task][:,:,1], columns=list(range(0, model.input_seqlen())))\n",
    "    wt_neg_profile_df['enhancer_id'] = wt_info_w_preds_df.enhancer_id\n",
    "    wt_neg_profile_df['anchor_motif'] = wt_info_w_preds_df.anchor_motif\n",
    "    wt_neg_profile_df['moved_motif'] = wt_info_w_preds_df.moved_motif\n",
    "    wt_neg_profile_df = wt_neg_profile_df.melt(id_vars = ['enhancer_id', 'anchor_motif', 'moved_motif'], \n",
    "                                               var_name = 'position', value_name = 'signal')\n",
    "    wt_neg_profile_df['type'] = 'wt'\n",
    "    wt_neg_profile_df['strand'] = 'neg'\n",
    "    wt_neg_profile_df['signal'] = -1 * wt_neg_profile_df['signal']\n",
    "    wt_neg_profile_df['task'] = task\n",
    "\n",
    "    ## Generate mut profiles\n",
    "    #Allocate all positive reads as 2D matrix and then a data frame\n",
    "    mut_pos_profile_df = pd.DataFrame(data=mut_preds[task][:,:,0], columns=list(range(0, model.input_seqlen())))\n",
    "    mut_pos_profile_df['enhancer_id'] = wt_info_w_preds_df.enhancer_id\n",
    "    mut_pos_profile_df['anchor_motif'] = wt_info_w_preds_df.anchor_motif\n",
    "    mut_pos_profile_df['moved_motif'] = wt_info_w_preds_df.moved_motif\n",
    "    mut_pos_profile_df = mut_pos_profile_df.melt(id_vars = ['enhancer_id', 'anchor_motif', 'moved_motif'], \n",
    "                                                 var_name = 'position', value_name = 'signal')\n",
    "    mut_pos_profile_df['type'] = 'mut'\n",
    "    mut_pos_profile_df['strand'] = 'pos'\n",
    "    mut_pos_profile_df['task'] = task\n",
    "\n",
    "    #Allocate all negative reads as 2D matrix and then a data frame\n",
    "    mut_neg_profile_df = pd.DataFrame(data=mut_preds[task][:,:,1], columns=list(range(0, model.input_seqlen())))\n",
    "    mut_neg_profile_df['enhancer_id'] = wt_info_w_preds_df.enhancer_id\n",
    "    mut_neg_profile_df['anchor_motif'] = wt_info_w_preds_df.anchor_motif\n",
    "    mut_neg_profile_df['moved_motif'] = wt_info_w_preds_df.moved_motif\n",
    "    mut_neg_profile_df = mut_neg_profile_df.melt(id_vars = ['enhancer_id', 'anchor_motif', 'moved_motif'], \n",
    "                                                 var_name = 'position', value_name = 'signal')\n",
    "    mut_neg_profile_df['type'] = 'mut'\n",
    "    mut_neg_profile_df['strand'] = 'neg'\n",
    "    mut_neg_profile_df['signal'] = -1 * mut_neg_profile_df['signal']\n",
    "    mut_neg_profile_df['task'] = task\n",
    "\n",
    "    ## Generate mut and injectedprofiles\n",
    "    #Allocate all positive reads as 2D matrix and then a data frame\n",
    "    mut_inj_pos_profile_df = pd.DataFrame(data=mut_inj_preds[task][:,:,0], columns=list(range(0, model.input_seqlen())))\n",
    "    mut_inj_pos_profile_df['enhancer_id'] = mut_inj_info_w_preds_df.enhancer_id\n",
    "    mut_inj_pos_profile_df['anchor_motif'] = mut_inj_info_w_preds_df.anchor_motif\n",
    "    mut_inj_pos_profile_df['moved_motif'] = mut_inj_info_w_preds_df.moved_motif\n",
    "    mut_inj_pos_profile_df['type'] = ['mut_inj/' +  str(i) for i in mut_inj_info_w_preds_df.distance_between_moved_center_and_anchor_center.values]\n",
    "    mut_inj_pos_profile_df = mut_inj_pos_profile_df.melt(id_vars = ['enhancer_id', 'anchor_motif', 'moved_motif', 'type'], \n",
    "                                                         var_name = 'position', value_name = 'signal')\n",
    "    mut_inj_pos_profile_df['strand'] = 'pos'\n",
    "    mut_inj_pos_profile_df['task'] = task\n",
    "    \n",
    "    #Label distances desired to store\n",
    "    distances_to_store_profile_label = ['mut_inj/' +  str(i) for i in distances_to_store_profile]\n",
    "\n",
    "    #Filter out distances not desired to store        \n",
    "    mut_inj_pos_profile_df = mut_inj_pos_profile_df[mut_inj_pos_profile_df['type'].isin(list(distances_to_store_profile_label))]\n",
    "    \n",
    "    #Allocate all negative reads as 2D matrix and then a data frame\n",
    "    mut_inj_neg_profile_df = pd.DataFrame(data=mut_inj_preds[task][:,:,1], columns=list(range(0, model.input_seqlen())))\n",
    "    mut_inj_neg_profile_df['enhancer_id'] = mut_inj_info_w_preds_df.enhancer_id\n",
    "    mut_inj_neg_profile_df['anchor_motif'] = mut_inj_info_w_preds_df.anchor_motif\n",
    "    mut_inj_neg_profile_df['moved_motif'] = mut_inj_info_w_preds_df.moved_motif\n",
    "    mut_inj_neg_profile_df['type'] = ['mut_inj/' +  str(i) for i in mut_inj_info_w_preds_df.distance_between_moved_center_and_anchor_center.values]\n",
    "    mut_inj_neg_profile_df = mut_inj_neg_profile_df.melt(id_vars = ['enhancer_id', 'anchor_motif', 'moved_motif', 'type'], \n",
    "                                                         var_name = 'position', value_name = 'signal')\n",
    "    mut_inj_neg_profile_df['strand'] = 'neg'\n",
    "    mut_inj_neg_profile_df['signal'] = -1 * mut_inj_neg_profile_df['signal']\n",
    "    mut_inj_neg_profile_df['task'] = task\n",
    "    mut_inj_neg_profile_df = mut_inj_neg_profile_df[mut_inj_neg_profile_df['type'].isin(list(distances_to_store_profile_label))]\n",
    "    \n",
    "    all_profile_df = pd.concat([wt_pos_profile_df, wt_neg_profile_df, mut_pos_profile_df, mut_neg_profile_df, \n",
    "                                mut_inj_pos_profile_df, mut_inj_neg_profile_df], axis = 0).reset_index()\n",
    "    all_profiles_df = all_profiles_df.append(all_profile_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(86800000, 9)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Save as tsv.gz\n",
    "all_profiles_df.to_csv(f'two_mutation/tsv/combined/{prefix}_all_profiles_multiples_15_df.tsv.gz', sep = '\\t', index = False)\n",
    "all_profiles_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf= pd.read_csv(f'two_mutation/tsv/combined/{prefix}_all_profiles_multiples_15_df.tsv.gz',sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "## save each enhancer_id as seperate tsv\n",
    "for id in pdf['enhancer_id'].unique():\n",
    "    subset_df = pdf[pdf['enhancer_id'] == id]\n",
    "    filename = f'two_mutation/tsv/individual_profile_by_enhancer_id_dis_15_multiples/' + str(id) + '_profiles_df' + '.tsv.gz'\n",
    "    subset_df.to_csv(filename, index=False, sep='\\t')  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bpnet-gpu",
   "language": "python",
   "name": "bpnet-gpu"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
