---
title: 'Summary of TSC model motifs'
author: "Khyati Dalal"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    depth: 3
    theme: sandstone
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
  word_document:
    toc: yes
header-includes:
- \usepackage{fancyhdr}
- \usepackage{lipsum}
- \pagestyle{fancy}
- \fancyfoot[CO,CE]{1028-02098-001-001}
- \fancyfoot[LE,RO]{\thepage}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options:
  chunk_output_type: console
---

# Introduction

The goal of this analysis is to summarize the PWM, CWM, ChIP-nexus binding, and contribution of the instances mapped by CWM scanning from the motifs found by the tsc TF-MoDISco. We will collect these values and plot them in a metaplot. Note that we will be collecting the binding and contribution information from the mapped instances, but the PWMs and CWMs were derived from the original TF-MoDISco run.

# Computational Setup

```{r, warning=F, message=F}
#Standard packages
library(rtracklayer) ; library(GenomicRanges); library(magrittr) ; library(Biostrings);library(BSgenome.Mmusculus.UCSC.mm10)
library(ggplot2) ; library(reshape2); library(plyranges); library(Rsamtools);library(parallel)
library(dplyr); library(data.table); library(patchwork); library(readr); library(testit)

#KNITR Options
setwd("/n/projects/kd2200/publication/bpnet/analysis/")
options(knitr.figure_dir="figures/1b_motif_summary/", java.parameters = "- Xmx6g")

#Lab sources
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/granges_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/knitr_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/caching.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_functions.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/multiplot.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/motif_functions.R")

#Specific sources
library(ggseqlogo)
source("scripts/1.motif_summary/tsc_variables.R")

#Pre-existing variables
modisco_dir <- '/n/projects/kd2200/publication/bpnet/modisco/profile/dataspec.yaml_default_fold_5/'
#modisco_dir <- '/n/projects/kd2200/publication/bpnet/modisco/profile/dataspec.yaml_default_fold_4/'
#modisco_dir <- '/n/projects/kd2200/publication/bpnet/modisco/profile/dataspec.yaml_default_fold_3/'
tasks <- c('cdx2','tead4','yap1','tfap2c','gata3') 

```

# Import motifs of interest

```{r, import motifs}

# fold 5
motifs.gr<-readr::read_tsv('tsv/curated_motif_regoion_tsc_model_motifs_fold5_1based.tsv.gz') %>% 
    GenomicRanges::makeGRangesFromDataFrame(keep.extra.columns = T, starts.in.df.are.0based = F)

# fold 4
#motifs.gr<-readr::read_tsv('tsv/curated_motif_regoion_tsc_model_motifs_fold4_1based.tsv.gz') %>% 
#    GenomicRanges::makeGRangesFromDataFrame(keep.extra.columns = T, starts.in.df.are.0based = F)

# fold 3
#motifs.gr<-readr::read_tsv('../motif_curation/tsv/curated_motif_regoion_tsc_model_motifs_fold3_1based.tsv.gz') %>% 
#    GenomicRanges::makeGRangesFromDataFrame(keep.extra.columns = T, starts.in.df.are.0based = F)

```

# Extract information for metaplot

## Function for extraction

+ Inputs:
  + motif_name: string naming motif to filter `motifs.gr` from
  + motifs.gr: GRanges object with `motif_name` pattern as a `pattern_name` column
  + motif_logo_h5_path: path to .h5 object that is compatible with code to extract logos at `/n/projects/mw2098/shared_code/bpnet/bpnet_extract_modisco_logos.py`
  + footprint_window: window width to extract metapeaks around [default = 400]
+ Output: list of following elements:
  + motif_name: motif
  + motif_count: number of motifs mapped using CWM scanning
  + actual_metapeaks.df: df of metapeak of averaged motifs
  + contrib_averages.df: df of metapeak of averaged contribution across motifs
  + ppm.mat: motif PPM
  + cwm.mat: motif CWM

# generate modisco_logos.h5 files for each modisco tasks

```{r,eval = F}
#bash scripts/1.motif_summary/tsc_modisco_logos.sh
```

```{r}
extract_metaplot_information<-function(motif_name, motifs.gr, motif_logo_h5_path, footprint_window = 500, threads = 6){
  library(rhdf5)
  
  motif.gr<-motifs.gr %>% plyranges::filter(pattern_name==motif_name)
  
  #Extract average task contribution
  contrib_averages.df<-lapply(tsc.contrib.profile.bws, function(x){
    regionMeans(regions = motif.gr, cvg = x) %>% mean(., na.rm = T)
  }) %>% as.data.frame() %>% data.table() %>% 
    melt.data.table(variable.name = 'task_name', value.name = 'contrib') %>% 
    dplyr::mutate(motif = motif_name)
  
  #Extract average task profiles
  actual_metapeaks.df<-mclapply(names(tsc.actual.norm.bws), function(x){
    mp.df<-exo_metapeak(gr = motif.gr %>% resize(1, "center"), sample = tsc.actual.norm.bws[[x]], 
                        upstream = floor(footprint_window/2), downstream = ceiling(footprint_window/2)) %>%
      dplyr::mutate(task_name = x)
  }, mc.cores = threads) %>% rbindlist() %>% dplyr::mutate(motif = motif_name)
  
  #Extract PPMs and CWMs
  testit::assert('This motif set has varying pattern identifiers.', (motif.gr$pattern %>% unique %>% length)==1)
  testit::assert('This motif set has varying modisco tasks.', (motif.gr$modisco_task %>% unique %>% length)==1)

  ppm.mat<-rhdf5::h5read(motif_logo_h5_path, paste0('/', motif.gr$pattern[1], '/PPM/trimmed'))
  cwm.mat<-rhdf5::h5read(motif_logo_h5_path, paste0('/', motif.gr$pattern[1], '/CWM/trimmed/', motif.gr$modisco_task[1]))

  return(list(motif_name = motif_name, motif_count = length(motif.gr),
              contrib_averages.df = contrib_averages.df, 
              actual_metapeaks.df = actual_metapeaks.df, 
              ppm.mat = ppm.mat, cwm.mat = cwm.mat))
  
}
```

## Extract information from relevant .bw files and .h5 files.

```{r, eval=FALSE}
metaplot.list<-lapply(motifs.gr$pattern_name %>% unique, function(x){
  message(x)
  motif.gr<-motifs.gr %>% plyranges::filter(pattern_name==x)
  info.list<-extract_metaplot_information(motif_name = x, motifs.gr = motifs.gr,
                                          motif_logo_h5_path = paste(modisco_dir, motif.gr$modisco_task[1], 'modisco_logos.h5', sep = '/'),
                                          footprint_window =250)

})
names(metaplot.list)<-motifs.gr$pattern_name %>% unique

saveRDS(metaplot.list, 'rds/1.motif_summary/tsc_curated_motif_metaplot_information_fold5.list.rds')
#saveRDS(metaplot.list, 'rds/1.motif_summary/tsc_curated_motif_metaplot_information_fold4.list.rds')
#saveRDS(metaplot.list, 'rds/1.motif_summary/tsc_curated_motif_metaplot_information_fold3.list.rds')

```

## Organize across different categories.

```{r}

#all motifs with all profile
#metaplot.list<-readRDS('rds/1.motif_summary/tsc_curated_motif_metaplot_information_fold5.list.rds')
#metaplot.list <- metaplot.list[c('tfap2c_0','tead4_0','tead4_double_1','cdx2_0','gata3_1','gata3_4','jun_fos','elf5','ctcf')]

#less motifs with all profile
metaplot.list<-readRDS('rds/1.motif_summary/tsc_curated_motif_metaplot_information_fold5.list.rds')
metaplot.list <- metaplot.list[c('tfap2c_0','tead4_0','tead4_double_1','cdx2_0','gata3_1','gata3_4')]


#all motifs with all profile for fold 3
# metaplot.list<-readRDS('tsc_curated_motif_metaplot_information_fold3.list.rds')
# metaplot.list <- metaplot.list[c('tfap2c_0','tead4_0','tead4_double_1','cdx2_0','gata3_2','gata3_1','gata3_4','jun_fos','elf5','ctcf')]

#all motifs with all profile for fold 4
# metaplot.list<-readRDS('tsc_curated_motif_metaplot_information_fold4.list.rds')
# metaplot.list <- metaplot.list[c('tfap2c_0','tead4_0','tead4_double_1','cdx2_0','gata3_2','gata3_1','jun_fos','elf5','ctcf')]
```


Here, in order to properly categorize the PWMs in terms of entropy, we will extract count frequency information directly.


```{r}

seq.list<-lapply(names(metaplot.list), function(x){ 
  motif.gr<-motifs.gr %>% plyranges::filter(pattern_name==x)
  seqlet.gr<-rtracklayer::import(paste0(modisco_dir, '/', motif.gr$modisco_task[1], '/seqlets/', motif.gr$pattern[1], '.bed.gz'))
  if(x=='tead4_double_1'| x=='tead4_0'){
    seqlet.gr<-seqlet.gr %>% plyranges::mutate(strand = ifelse(strand == '+', '-', '+'))
  }
  seqlet.seq<-getSeq(BSgenome.Mmusculus.UCSC.mm10, seqlet.gr, as.character = T)
})
names(seq.list)<-names(metaplot.list)

#Collect cwm other information
cwm.list<-lapply(metaplot.list, function(x) {
  mat<-x$cwm.mat
  if(x=='tead4_double_1' | x=='tead4_0'){
    mat_rev<-mat[,ncol(mat):1]
    rownames(mat_rev)<-c('T','G','C','A')
    mat.df<-mat_rev %>% as.data.frame
    mat<-mat.df[order(row.names(mat.df)),] %>% as.matrix(.)
    colnames(mat)<-NULL
  }else{
    rownames(mat)<-c('A','C','G','T')
  }
  return(mat)
})

#Collect pwm other information
pwm.list<-lapply(metaplot.list, function(x) {
  mat<-x$ppm.mat
  if(x=='tead4_double_1'| x=='tead4_0'){
    mat_rev<-mat[,ncol(mat):1]
    rownames(mat_rev)<-c('T','G','C','A')
    mat.df<-mat_rev %>% as.data.frame
    mat<-mat.df[order(row.names(mat.df)),] %>% as.matrix(.)
    colnames(mat)<-NULL
  }else{
    rownames(mat)<-c('A','C','G','T')
  }
  return(mat)
})

metapeaks.df<-lapply(metaplot.list, function(x) x$actual_metapeaks.df) %>% rbindlist
contribs.df<-lapply(metaplot.list, function(x) x$contrib_averages.df) %>% rbindlist
labels.df<-lapply(metaplot.list, function(x) data.frame(motif_name = x$motif_name, motif_count = x$motif_count)) %>% rbindlist

```

# Plot metaplot

```{r, fig.height = 8, fig.width = 30}
# for fold 5
#motif_order<-c('tfap2c_0','tead4_0','tead4_double_1','cdx2_0','gata3_1','gata3_4','jun_fos','elf5','ctcf')
motif_order<-c('tead4_0','tead4_double_1','tfap2c_0','cdx2_0','gata3_4','gata3_1')

# for fold 3&4
#motif_order<-c('tead4_0','tead4_double_1','tfap2c_0','cdx2_0','gata3_2','gata3_1','elf5','jun_fos','ctcf')


tf_order<-c('tead4','yap1','tfap2c','cdx2','gata3')


labels.df$motif_name<-factor(labels.df$motif_name, levels = motif_order)

label.plot<-ggplot(labels.df)+
  geom_text(aes(x= .5, y = .5, label = paste0(motif_name, '\n(', motif_count, ')')))+
  facet_grid(motif_name~.)+
  ggtitle('Motif name')+
  theme_classic()+theme(axis.line = element_blank(), axis.text = element_blank(), axis.ticks = element_blank(), axis.title = element_blank(), 
                        strip.background = element_blank(), strip.text = element_blank())

seq.list<-lapply(motif_order, function(x) seq.list[[x]])
pwm.plot<-ggplot() + geom_logo(seq.list) + 
  facet_grid(seq_group~., scales = 'fixed') +
  theme_classic()+
  ggtitle('PWM')+
  theme(axis.line = element_blank(), axis.text = element_blank(), axis.ticks = element_blank(), axis.title = element_blank(), 
        strip.background = element_blank(), strip.text = element_blank())

cwm.list<-lapply(motif_order, function(x) cwm.list[[x]])
cwm.plot<-ggplot() + geom_logo(cwm.list, method='custom', seq_type='dna') +
  facet_grid(seq_group~., scales = 'free_y')+
  theme_classic()+
  ggtitle('CWM')+
  theme(axis.line = element_blank(), axis.text = element_blank(), axis.ticks = element_blank(), axis.title = element_blank(), 
        strip.background = element_blank(), strip.text = element_blank())

metapeaks.df$task_name<-factor(metapeaks.df$task_name, levels = tf_order)

metapeaks.df$motif<-factor(metapeaks.df$motif, levels = motif_order)

mp.plot<-ggplot(metapeaks.df, aes(x = tss_distance, y = reads, alpha = strand,color=metapeaks.df$task_name))+
    geom_line(size=1.3)+scale_alpha_manual(values = c(.9, 1.2))+
    scale_color_manual(values=c('#00B3B4','#F47F20','#C54B9A','#B1D03E','#373493'))+
    scale_x_continuous(name = 'Distance from motif center (bp)', breaks = seq(-500, 500, 250))+
    scale_y_continuous(name = 'Actual ChIP-nexus RPM')+
    facet_grid(motif ~ metapeaks.df$task_name, scales = 'free')+ #fixed
    ggtitle('Actual ChIP-nexus metapeaks (RPM)')+
    theme_classic()+
    theme(legend.position = 'none', strip.background = element_blank())


metapeaks.df$task_name<-factor(metapeaks.df$task_name, levels = tf_order)
metapeaks.df$motif<-factor(metapeaks.df$motif, levels = motif_order)
mp.plot<-ggplot(metapeaks.df, aes(x = tss_distance, y = reads, color = task_name, alpha = strand))+
  geom_line(aes(group=strand))+
  scale_color_manual(values = tsc.color.vec[tf_order], name = 'task')+
  scale_alpha_manual(values = c(.7, 1))+
  scale_x_continuous(name = 'Distance from motif center (bp)', breaks = seq(-500, 500, 250))+
  scale_y_continuous(name = 'Actual ChIP-nexus RPM')+
  facet_grid(motif ~ task_name, scales = 'free_y')+
  ggtitle('Actual ChIP-nexus metapeaks (RPM)')+
  theme_classic()+
  theme(legend.position = 'none', strip.background = element_blank())

contribs.df$task_name<-factor(contribs.df$task_name, levels = tf_order)
contribs.df$motif<-factor(contribs.df$motif, levels = rev(motif_order))
contrib.plot<-ggplot(contribs.df, aes(x = task_name, y = motif, fill = contrib))+
  geom_tile()+
  scale_fill_gradient(low = 'white', high = '#1c5b78', name = 'Profile\ncontrib')+
  scale_x_discrete(name = 'TF')+
  scale_y_discrete(name = 'Motif')+
  ggtitle('DeepLIFT profile contribution')+
  theme_classic()+
  theme()

g<-label.plot + pwm.plot + cwm.plot + mp.plot + contrib.plot + patchwork::plot_layout(nrow = 1, widths = c(.1, .4, .4, 1.2, .4))


#For plotting pwm and cwm in a plot
#g<-label.plot + pwm.plot + cwm.plot + patchwork::plot_layout(nrow = 1, widths = c(.1, .5, .5))
#ggsave('figures/1_motif_summary/all_motifs_curation_fold5_logos.png', g, height = 8, width = 10)
#ggsave('figures/1_motif_summary/all_motifs_curation_fold5_logos.pdf', g, height = 8, width = 10)

#ggsave('figures/1_motif_summary/final_curated_motif_metaplot.png', g, height = 8, width = 30)
#ggsave('figures/1_motif_summary/final_curated_motif_metaplot.pdf', g, height = 8, width = 30)


# for fold 3
#ggsave('figures/curated_motif_metaplot_fold3_all.png', g, height = 8, width = 30)
#ggsave('figures/curated_motif_metaplot_fold3_all.pdf', g, height = 8, width = 30)

# for fold 4
#ggsave('figures/curated_motif_metaplot_fold4.png', g, height = 8, width = 30)
#ggsave('figures/curated_motif_metaplot_fold4.pdf', g, height = 8, width = 30)

# for Yap1 profile chip-nexus metapeak on a sperate axis
metapeaks.df$task_name<-factor(metapeaks.df$task_name, levels = tf_order)
metapeaks.df$motif<-factor(metapeaks.df$motif, levels = motif_order)
yap1_profile.plot<-ggplot(dplyr::filter(metapeaks.df,task_name=="yap1"), aes(x = tss_distance, y = reads, color = task_name, alpha = strand))+
    geom_line(aes(group=strand))+
    scale_color_manual(values = tsc.color.vec[tf_order], name = 'task')+
    scale_alpha_manual(values = c(.7, 1))+
    scale_x_continuous(name = 'Distance from motif center (bp)', breaks = seq(-500, 500, 250))+
    scale_y_continuous(name = 'Actual ChIP-nexus RPM')+
    facet_grid(motif ~ task_name, scales = 'fixed')+
    ggtitle('Actual ChIP-nexus metapeaks (RPM)')+
    theme_classic()+
    theme(legend.position = 'none', strip.background = element_blank())

ggsave('figures/1_motif_summary/curated_motif_metaplot_yap1only.pdf', yap1_profile.plot, height = 8, width = 4)


# Tead4 and Yap1 profile contributions

metaplot.list<-readRDS('rds/1.motif_summary/tsc_curated_motif_metaplot_information_fold5.list.rds')
metaplot.list <- metaplot.list[c('tfap2c_0','tead4_0','tead4_double_1','cdx2_0','gata3_4','gata3_1','elf5','jun_fos','ctcf')]
contribs.df<-lapply(metaplot.list, function(x) x$contrib_averages.df) %>% rbindlist

contribs_yap1.df <- dplyr::filter(contribs.df,task_name=="yap1") ##repeat with tead4 to get seperate scale


motif_order<-c('tead4_0','tead4_double_1','jun_fos','tfap2c_0','ctcf','elf5','gata3_4','cdx2_0','gata3_1')

tf_order<-c('tead4','yap1','tfap2c','cdx2','gata3')

#yap1
contribs_yap1.df$task_name<-factor(contribs_yap1.df$task_name, levels = tf_order)
contribs_yap1.df$motif<-factor(contribs_yap1.df$motif, levels = rev(motif_order))
contrib.plot<-ggplot(contribs_yap1.df, aes(x = task_name, y = motif, fill = contrib))+
    geom_tile()+
    scale_fill_gradient(low = 'white', high = '#1c5b78', name = 'Profile\ncontrib')+
    scale_x_discrete(name = 'TF')+
    scale_y_discrete(name = 'Motif')+
    ggtitle('DeepLIFT Yap1 profile contribution')+
    theme_classic()+
    theme()

ggsave('figures/1_motif_summary/yap1_profile_contrib_only.pdf', contrib.plot, height =6, width = 4)


# tead4
contribs_tead4.df <- dplyr::filter(contribs.df,task_name=="tead4")
contribs_tead4.df$task_name<-factor(contribs_tead4.df$task_name, levels = tf_order)
contribs_tead4.df$motif<-factor(contribs_tead4.df$motif, levels = rev(motif_order))
contrib_tead4.plot<-ggplot(contribs_tead4.df, aes(x = task_name, y = motif, fill = contrib))+
    geom_tile()+
    scale_fill_gradient(low = 'white', high = '#1c5b78', name = 'Profile\ncontrib')+
    scale_x_discrete(name = 'TF')+
    scale_y_discrete(name = 'Motif')+
    ggtitle('DeepLIFT Tead4 profile contribution')+
    theme_classic()+
    theme()


ggsave('figures/1_motif_summary/tead4_profile_contrib_only.pdf',contrib_tead4.plot, height = 6, width = 4)
```

In conclusion, you can see good binding to each motif as well as localized contribution validating the motif choices.

# Session Information

For the purposes of reproducibility, the R/Bioconductor session information is printed below:

```{r sessioninfo}
sessionInfo()
```














