---
title: "Tead4 anf Yap1 analysis"
author: "Khyati Dalal"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    depth: 3
    theme: sandstone
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
  word_document:
    toc: yes
header-includes:
- \usepackage{fancyhdr}
- \usepackage{lipsum}
- \pagestyle{fancy}
- \fancyfoot[CO,CE]{1028-02098-001-001}
- \fancyfoot[LE,RO]{\thepage}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options:
  chunk_output_type: console
---

# Introduction

The goal of this analysis is to use the actual binding, predicted binding, and contribution tracks of the `tsc` model across enhancers to validate and assess the TF behavior at these well-annotated sites.

# Computational Setup

```{r setup, warning=F, message=F}
#Standard packages
library(rtracklayer) ; library(GenomicRanges); library(magrittr) ; library(Biostrings);library(parallel)
library(ggplot2) ; library(reshape2); library(plyranges); library(Rsamtools);library(BSgenome.Mmusculus.UCSC.mm10)
library(dplyr); library(data.table); library(patchwork); library(readr);library(tidyverse);library(forcats); library(testit);library("TxDb.Mmusculus.UCSC.mm10.knownGene"); library(viridis);library(DECIPHER)

#KNITR Options
setwd("/n/projects/kd2200/publication/bpnet/analysis/")
options(knitr.figure_dir="figures/5_tead4_yap1_analysis", java.parameters = "- Xmx6g")

source("/n/projects/kd2200/publication/bpnet/analysis/scripts/granges_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/load_motif_instance_from_all_regions.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/caching.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_functions.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/multiplot.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/knitr_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/motif_functions.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/heatmap_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_functions_yue.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/1.motif_summary/tsc_variables.R")
```


# load bw file paths

```{r}

##chip-seq bws
seq.bw.list<-list(H3K27ac=list(
  rep1="/n/projects/kd2200/publication/data_preparation/bw/mm10/seq/combined/mtsc_h3k27ac_seq_combined_rpkm.bw",                   rep2="/n/projects/kd2200/publication/data_preparation/bw/mm10/seq/combined/mtsc_h3k27ac_seq_combined_rpkm.bw")) 

##nexus_bws
nexus.bw.list<-list(
  gata3=list(pos="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_gata3_nexus_filtered_combined_normalized_positive.bw", 
             neg="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_gata3_nexus_filtered_combined_normalized_negative.bw"),  
  cdx2=list(pos="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_cdx2_nexus_filtered_combined_normalized_positive.bw",
            neg="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_cdx2_nexus_filtered_combined_normalized_negative.bw"),
  tfap2c=list(pos="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_tfap2c_nexus_filtered_combined_normalized_positive.bw",
             neg="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_tfap2c_nexus_filtered_combined_normalized_negative.bw"),
  tead4=list(pos="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_tead4_nexus_filtered_combined_normalized_positive.bw",
            neg="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_tead4_nexus_filtered_combined_normalized_negative.bw"),
  yap1=list(pos="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_yap1_nexus_filtered_combined_normalized_positive.bw",
            neg="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_yap1_nexus_filtered_combined_normalized_negative.bw"),
  PolII=list(pos="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_polii_nexus_filtered_combined_normalized_positive.bw",
           neg="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_polii_nexus_filtered_combined_normalized_negative.bw"),
  ttseq=list(pos="/n/projects/kd2200/publication/data_preparation/bw/mm10/nascent_rna/combined/mtsc_ttseq_rna_combined_positive_rpkm.bw",
           neg="/n/projects/kd2200/publication/data_preparation/bw/mm10/nascent_rna/combined/mtsc_ttseq_rna_combined_negative_rpkm.bw"))
 

```

# peak correlation plots
```{r}
##Peaks
cdx2_peaks.gr <- plyranges::read_narrowpeaks("/n/projects/kd2200/publication/data_preparation/peaks/mm10/nexus/combined/mtsc_cdx2_nexus_combined_peaks.narrowPeak") %>% dplyr::mutate(., name = "Cdx2") 
tead4_peaks.gr <- plyranges::read_narrowpeaks("/n/projects/kd2200/publication/data_preparation/peaks/mm10/nexus/combined/mtsc_tead4_nexus_combined_peaks.narrowPeak") %>% dplyr::mutate(., name = "Tead4")
tfap2c_peaks.gr <- plyranges::read_narrowpeaks("/n/projects/kd2200/publication/data_preparation/peaks/mm10/nexus/combined/mtsc_tfap2c_nexus_combined_peaks.narrowPeak") %>% dplyr::mutate(., name = "Tfap2c") 
gata3_peaks.gr <- plyranges::read_narrowpeaks("/n/projects/kd2200/publication/data_preparation/peaks/mm10/nexus/combined/mtsc_gata3_nexus_combined_peaks.narrowPeak") %>% dplyr::mutate(., name = "Gata3")
yap1_peaks.gr <- plyranges::read_narrowpeaks("/n/projects/kd2200/publication/data_preparation/peaks/mm10/nexus/combined/mtsc_yap1_nexus_combined_peaks.narrowPeak") %>% dplyr::mutate(., name = "Yap1")

# remove ERVs
rm <- readRDS("/n/projects/kd2200/publication/bpnet/analysis/data/repeatmasker.mm10.gr.rds")
erv <- rm[grep("ERV", rm$repeat_class)]

cdx2_idr_peaks_no_erv.gr <- cdx2_peaks.gr[which(!overlapsAny(cdx2_peaks.gr,erv, ignore.strand=T))]
tead4_idr_peaks_no_erv.gr <- tead4_peaks.gr[which(!overlapsAny(tead4_peaks.gr,erv, ignore.strand=T))]
tfap2c_idr_peaks_no_erv.gr <- tfap2c_peaks.gr[which(!overlapsAny(tfap2c_peaks.gr,erv, ignore.strand=T))]
gata3_idr_peaks_no_erv.gr <- gata3_peaks.gr[which(!overlapsAny(gata3_peaks.gr,erv, ignore.strand=T))]
yap1_idr_peaks_no_erv.gr <- yap1_peaks.gr[which(!overlapsAny(yap1_peaks.gr,erv, ignore.strand=T))]


##motifs
model_path2= "/n/projects/kd2200/publication/bpnet/cwm_all_regions/"
cdx2.gr <- load_motif_instance_from_all_regions(model_path2,"cdx2","p0", filter_TE = TRUE) %>% dplyr::mutate(., tf = "Cdx2")
tead4.gr <- load_motif_instance_from_all_regions(model_path2,"tead4","p0", filter_TE = TRUE) %>% dplyr::mutate(., tf = "Tead4")
tfap2c.gr <- load_motif_instance_from_all_regions(model_path2,"tfap2c", "p0", filter_TE = TRUE) %>% dplyr::mutate(., tf = "Tfap2c")
gata3.gr <- load_motif_instance_from_all_regions(model_path2,"gata3","p4", filter_TE = TRUE) %>% dplyr::mutate(., tf = "Gata3")
tead4_double.gr <- load_motif_instance_from_all_regions(model_path2,"tead4","p1", filter_TE = TRUE) %>% dplyr::mutate(., tf = "Tead4_double") 
tead4_tead4double <- c(tead4.gr, tead4_double.gr)

# load promoter regions for mm10

mm10_promoters <-promoters(TxDb.Mmusculus.UCSC.mm10.knownGene,upstream=1000, downstream=500)

# remove promoter region mapped motifs
cdx2_no_promoter.gr <- cdx2.gr[which(!overlapsAny(cdx2.gr,mm10_promoters, ignore.strand=T))]
tead4_tead4double_no_promoter.gr <- tead4_tead4double[which(!overlapsAny(tead4_tead4double,mm10_promoters, ignore.strand=T))]
tfap2c_no_promoter.gr <- tfap2c.gr[which(!overlapsAny(tfap2c.gr,mm10_promoters, ignore.strand=T))]
gata3_no_promoter.gr <- gata3.gr[which(!overlapsAny(gata3.gr,mm10_promoters, ignore.strand=T))]

# find overlap between peaks and motifs for each TF
test <- findOverlaps(cdx2_no_promoter.gr,cdx2_idr_peaks_no_erv.gr)
cdx2_peaks_motif_ov <- cdx2_idr_peaks_no_erv.gr[test@to]
test <- findOverlaps(tfap2c_no_promoter.gr,tfap2c_idr_peaks_no_erv.gr)
tfap2c_peaks_motif_ov <- tfap2c_idr_peaks_no_erv.gr[test@to]
test <- findOverlaps(gata3_no_promoter.gr,gata3_idr_peaks_no_erv.gr)
gata3_peaks_motif_ov <- gata3_idr_peaks_no_erv.gr[test@to]
test <- findOverlaps(tead4_tead4double_no_promoter.gr,tead4_idr_peaks_no_erv.gr)
tead4_peaks_motif_ov <- tead4_idr_peaks_no_erv.gr[test@to]
test <- findOverlaps(tead4_tead4double_no_promoter.gr,yap1_idr_peaks_no_erv.gr)
yap1_peaks_motif_ov <- yap1_idr_peaks_no_erv.gr[test@to]

cdx2_peaks_motif_ov_no_promoter.gr <- cdx2_peaks_motif_ov[which(!overlapsAny(cdx2_peaks_motif_ov,mm10_promoters, ignore.strand=T))]
tead4_peaks_motif_ov_no_promoter.gr <- tead4_peaks_motif_ov[which(!overlapsAny(tead4_peaks_motif_ov,mm10_promoters, ignore.strand=T))]
tfap2c_peaks_motif_ov_no_promoter.gr <-tfap2c_peaks_motif_ov[which(!overlapsAny(tfap2c_peaks_motif_ov,mm10_promoters, ignore.strand=T))]
gata3_peaks_motif_ov_no_promoter.gr <- gata3_peaks_motif_ov[which(!overlapsAny(gata3_peaks_motif_ov,mm10_promoters, ignore.strand=T))]
yap1_peaks_motif_ov_no_promoter.gr <- yap1_peaks_motif_ov[which(!overlapsAny(yap1_peaks_motif_ov,mm10_promoters, ignore.strand=T))]


# calculate signal using regions sums for each TF 
cdx2_peaks_motif_ov_no_promoter.gr$signal <- nexus_regionSums(resize(cdx2_peaks_motif_ov_no_promoter.gr,500,"center"), nexus.bw.list$cdx2)
tfap2c_peaks_motif_ov_no_promoter.gr$signal <- nexus_regionSums(resize(tfap2c_peaks_motif_ov_no_promoter.gr,500,"center"), nexus.bw.list$tfap2c)
tead4_peaks_motif_ov_no_promoter.gr$signal <- nexus_regionSums(resize(tead4_peaks_motif_ov_no_promoter.gr,500,"center"), nexus.bw.list$tead4)
gata3_peaks_motif_ov_no_promoter.gr$signal <- nexus_regionSums(resize(gata3_peaks_motif_ov_no_promoter.gr,500,"center"), nexus.bw.list$gata3)
yap1_peaks_motif_ov_no_promoter.gr$signal <- nexus_regionSums(resize(yap1_peaks_motif_ov_no_promoter.gr,500,"center"), nexus.bw.list$yap1)

# select top 4000 regions for plotting to keep the number of regions consistent across TFs
cdx2_peaks_motif_ov_no_promoter.gr <- cdx2_peaks_motif_ov_no_promoter.gr[order(cdx2_peaks_motif_ov_no_promoter.gr$signal, decreasing = T)][1:4000]
tfap2c_peaks_motif_ov_no_promoter.gr <- tfap2c_peaks_motif_ov_no_promoter.gr[order(tfap2c_peaks_motif_ov_no_promoter.gr$signal, decreasing = T)][1:4000]
tead4_peaks_motif_ov_no_promoter.gr <- tead4_peaks_motif_ov_no_promoter.gr[order(tead4_peaks_motif_ov_no_promoter.gr$signal, decreasing = T)][1:4000]
yap1_peaks_motif_ov_no_promoter.gr <- yap1_peaks_motif_ov_no_promoter.gr[order(yap1_peaks_motif_ov_no_promoter.gr$signal, decreasing = T)][1:4000]

combine_bed.gr <- c(cdx2_peaks_motif_ov_no_promoter.gr,tead4_peaks_motif_ov_no_promoter.gr,tfap2c_peaks_motif_ov_no_promoter.gr,gata3_peaks_motif_ov_no_promoter.gr,yap1_peaks_motif_ov_no_promoter.gr) %>% resize(., 500, fix = "center")

#Calculate h3k27ac signal
seq_signals.df<-lapply(seq.bw.list, function(x){(regionMeans(regions = combine_bed.gr, x$rep1) + regionMeans(regions = combine_bed.gr, x$rep2))/2}) %>% as.data.frame

#Calculate TF, Pol II and nascent RNA signal
nexus_signals.df<-lapply(nexus.bw.list, function(x){
  return(abs(regionMeans(regions = combine_bed.gr, x$pos))+abs(regionMeans(regions = combine_bed.gr, x$neg)))
}) %>% as.data.frame


###Correlation Plot####
#########################
signals_peaks.df<-cbind(seq_signals.df,nexus_signals.df)

cormat <- cor(signals_peaks.df, method= "spearman")
q <- cormat %>% as.data.frame()

s_1 <- q %>% as.data.frame() %>% `[`(1,2:6)
s_2 <- q %>% as.data.frame() %>% `[`(7:8,2:6) 
s3 <- rbind(s_1,s_2)

row.names(s3) <- c("H3K27ac","PolII","TT-seq")
colnames(s3) <- c("Gata3","Cdx2","Tfap2c","Tead4","Yap1")

s3$id = row.names(s3) 
melted_cormat<- reshape2::melt(s3)

corr <- ggplot(melted_cormat, aes(x= id, y= variable, fill=value))+
  geom_raster()+scale_fill_viridis(option = "F",begin = 0.5,end= 1, direction = -1)+
  theme(text=element_text(size=14),axis.line = element_blank(),
        axis.ticks = element_blank(), axis.title = element_blank())+
  ggtitle("")+theme_classic()+ylab("TF binding at top 4k peak regions with motifs")+xlab("")+
  labs(fill="Spearman\ncorrelation")


corr <- corr + theme(legend.position="bottom")

ggsave("figures/5_tead4_yap1_analysis/peaks_correlation_with_markers.pdf",corr,height= 8, width=5)
ggsave("figures/5_tead4_yap1_analysis/peaks_correlation_with_markers.png",corr,height= 8, width=5)


```

# Tead4 and Yap1 peak correlation with other factors peaks
```{r, Tead4 & Yap1 corr}
#tfpeaks correlations
s_1 <- q %>% as.data.frame() %>% `[`(5:6,2:6)
row.names(s_1) <- c("Tead4","Yap1")
colnames(s_1) <- c("Gata3","Cdx2","Tfap2c","Tead4","Yap1")
s_1$id = row.names(s_1) 
melted_cormat<- reshape2::melt(s_1)
cor2 <- ggplot(melted_cormat, aes(x= id, y= variable, fill=value))+
    geom_raster()+
    scale_fill_viridis(option = "F",begin = 0.5,end= 1, direction = -1)+
    theme(text=element_text(size=14), axis.line = element_blank(),
          axis.ticks = element_blank(), axis.title = element_blank())+
    ggtitle("")+
    labs(fill="Spearman\ncorrelation")

ggsave("figures/5_tead4_yap1_analysis/tf_peaks_correlations.pdf",cor2, height = 6, width = 4)
ggsave("figures/5_tead4_yap1_analysis/tf_peaks_correlations.png",cor2, height = 6, width = 4)
```


# Tead4 and Yap1 heatmap at accesible Tead4 task returned BPNet mapped motifs 
```{r, heatmap}

m0_p0.gr <- load_motif_instance_from_all_regions(model_path2,"tead4","p0", filter_TE = TRUE) %>% dplyr::mutate(., tf = "m0_p0")
m0_p1.gr <- load_motif_instance_from_all_regions(model_path2,"tead4","p1", filter_TE = TRUE) %>% dplyr::mutate(., tf = "m0_p1")

motifs_from_tead4_task_sel <- c(m0_p0.gr,m0_p1.gr)#,m0_p2.gr,m0_p3.gr ,m0_p4.gr ,m0_p6.gr, m0_p7.gr, m0_p10.gr)

# heatmap 
motifs_from_tead4_task_sel$yap1_signal <- nexus_regionSums(resize(motifs_from_tead4_task_sel, 101, "center"), nexus.bw.list$yap1)
motifs_from_tead4_task_sel$tead4_signal <- nexus_regionSums(resize(motifs_from_tead4_task_sel, 101, "center"), nexus.bw.list$tead4)

# find overlap between  motifs important for Tead4 task and Tead4 peaks
test <- find_overlaps_within(motifs_from_tead4_task_sel, resize(tead4_peaks.gr, 1000, "center"))
test <- unique(test)

#select regions with minimum of tead4 signal i.e above median value
test$tead4_signal %>% summary()
test_medianabove <- test[test$tead4_signal>0.5]


## select and order by yap1 signal and take top and bottom 5000 regions 
test_medianabove$yap1_signal %>% summary()
test_medianabove2 <- test_medianabove[test_medianabove$yap1_signal>0.28]
test_medianabove2 <- test_medianabove2[order(test_medianabove2$yap1_signal, decreasing = T)]
yap1_high <- test_medianabove2[1:5000]
yap1_low <- test_medianabove2[12091:17090]
combine <- list(yap_hig=yap1_high, yap_low=yap1_low)

yap1_low_re <- resize(yap1_low,1,"start")
yap1_high_re <-  resize(yap1_high,1,"start")
combine2 <- list(yap_hig=yap1_high_re, yap_low=yap1_low_re)

 #plot
 p<-plot_multiple_gr_nexus_heatmap(combine,nexus.bw.list$PolII,upstream =1500, downstream = 1500, ncol = 2,text_size = 10,neg_color ="#961d35", pos_color = "#961d35",order="as.is",normalize_threshold = 0)
 tt<- plot_multiple_gr_nexus_heatmap(combine,nexus.bw.list$ttseq,upstream =1500, downstream = 1500, ncol = 2,text_size = 10,neg_color ="#3d7d33", pos_color = "#3d7d33",order="as.is",normalize_threshold = 0)
td<-plot_multiple_gr_nexus_heatmap(combine,nexus.bw.list$tead4,upstream =250, downstream = 250, ncol = 2,text_size = 10,neg_color ="#00b3b3", pos_color = "#00b3b3",order="as.is",normalize_threshold = 0)
yp<-plot_multiple_gr_nexus_heatmap(combine,nexus.bw.list$yap1,upstream =250, downstream = 250, ncol = 2,text_size = 10,neg_color ="#f37f20", pos_color = "#f37f20",order="as.is",normalize_threshold = 0)
h3<-plot_multiple_gr_standard_heatmap(combine2,seq.bw.list$H3K27ac$rep1,upstream =1500, downstream = 1500, ncol = 2,text_size = 10,low_color= "#fad0b4", high_color= "#a55d2d",order="as.is",normalize_threshold = 0)

g<- td + yp + h3 +p+ tt + patchwork::plot_layout(nrow = 1, widths = c(.5,.5, .5,.5,.5))
#(once the plot appears in the Plots section, save manually, that opens into another R window with the plot, downloaded that plot for plotting, as the resolution is better that way.)

#save
ggsave("figures/5_tead4_yap1_analysis/tead4_yap1_markers_heatmap.pdf",g, height = 8, width = 20)
ggsave("figures/5_tead4_yap1_analysis/tead4_yap1_markers_heatmap.png",g, height = 8, width = 20)

```

# Motif island analysis & Yap1 conntribution at single motif pairs
```{r, island analysis}

##CWM_all motif regions
model_path2= "/n/projects/kd2200/publication/bpnet/cwm_all_regions/"
cdx2.gr <- load_motif_instance_from_all_regions(model_path2,"cdx2","p0", filter_TE = TRUE) %>% dplyr::mutate(., tf = "Cdx2")
tead4.gr <- load_motif_instance_from_all_regions(model_path2,"tead4","p0", filter_TE = TRUE) %>% dplyr::mutate(., tf = "Tead4")
tfap2c.gr <- load_motif_instance_from_all_regions(model_path2,"tfap2c", "p0", filter_TE = TRUE) %>% dplyr::mutate(., tf = "Tfap2c")
gata3.gr <- load_motif_instance_from_all_regions(model_path2,"gata3","p4", filter_TE = TRUE) %>% dplyr::mutate(., tf = "Gata3")
tead4_double.gr <- load_motif_instance_from_all_regions(model_path2,"tead4","p1", filter_TE = TRUE) %>% dplyr::mutate(., tf = "Tead4_double") 

tead4_only.gr <- tead4.gr[which(!overlapsAny(tead4.gr, tead4_double.gr, ignore.strand=T))]
tead4_double_only.gr <- tead4_double.gr[which(!overlapsAny(tead4_double.gr, tead4.gr, ignore.strand=T))]


#combine motifs
all_tf.gr <- bind_rows(as.data.frame(cdx2.gr),as.data.frame(tead4_only.gr),as.data.frame(tfap2c.gr),as.data.frame(gata3.gr),as.data.frame(tead4_double_only.gr)) %>% makeGRangesFromDataFrame(., keep.extra.columns = T, starts.in.df.are.0based = F) 

# get promoter regions to remove it.
mm10_promoters <-promoters(TxDb.Mmusculus.UCSC.mm10.knownGene,upstream=1000, downstream=500)
all_tf_final_no_promoter.gr <- all_tf.gr[which(!overlapsAny(all_tf.gr,mm10_promoters, ignore.strand=T))]

# island generation
islands_new.gr<-all_tf_final_no_promoter.gr  %>% resize(500, 'center') %>% 
  GenomicRanges::reduce(ignore.strand = T) %>%
  GenomicRanges::sort(ignore.strand = T) %>%
  plyranges::mutate(island_id = 1:length(.))

instances_vs_islands.ov<-findOverlaps(all_tf_final_no_promoter.gr , islands_new.gr, ignore.strand = T)
all_tf_final_no_promoter.gr$island_id<-islands_new.gr$island_id[instances_vs_islands.ov@to]

#Show counts of # of motifs per island
all_tf_final_no_promoter.gr$island_id %>% table %>% table

island_names.df<-all_tf_final_no_promoter.gr %>% 
  GenomicRanges::sort(ignore.strand = T) %>% 
  as.data.frame %>%
  dplyr::group_by(island_id) %>%
  dplyr::summarize(island_content = paste(unique(sort(tf)), table(sort(tf)), sep = '-', collapse = '_'),
                   island_content_ordered = paste(tf, collapse = '_'),
                   island_content_unique = paste(unique(sort(tf)), collapse = '_'),
                   island_count = length(tf))


all_tf_final_no_promoter.gr <- all_tf_final_no_promoter.gr %>% as.data.frame %>%
  dplyr::left_join(., island_names.df, by = 'island_id') %>%
  makeGRangesFromDataFrame(keep.extra.columns = T, starts.in.df.are.0based = F)

island_names.df$island_content %>% table %>% sort(decreasing = T) %>% head(n=20)
island_names.df$island_content_ordered %>% table %>% sort(decreasing = T) %>% head(n=20)
island_names.df$island_content_unique %>% table %>% sort(decreasing = T) %>% head(n=20)


all_tf_final_no_promoter.gr$tead4_sig <- nexus_regionSums(resize(all_tf_final_no_promoter.gr , 100, "center"), nexus.bw.list$tead4) 
 all_tf_final_no_promoter.gr$tfap2c_sig <- nexus_regionSums(resize(all_tf_final_no_promoter.gr , 100, "center"), nexus.bw.list$tfap2c) 
 all_tf_final_no_promoter.gr$cdx2_sig <- nexus_regionSums(resize(all_tf_final_no_promoter.gr , 100, "center"), nexus.bw.list$cdx2) 
 all_tf_final_no_promoter.gr$gata3_sig <- nexus_regionSums(resize(all_tf_final_no_promoter.gr , 100, "center"), nexus.bw.list$gata3) 
 all_tf_final_no_promoter.gr$yap1_sig <- nexus_regionSums(resize(all_tf_final_no_promoter.gr , 100, "center"), nexus.bw.list$yap1) 
 all_tf_final_no_promoter.gr$h3k27ac_sig <- regionSums(resize(all_tf_final_no_promoter.gr , 1000, "center"), seq.bw.list$H3K27ac$rep2)
 all_tf_final_no_promoter.gr$tt_sig <- nexus_regionSums(resize(all_tf_final_no_promoter.gr , 1000, "center"), nexus.bw.list$ttseq)
 all_tf_final_no_promoter.gr$polii_sig <- nexus_regionSums(resize(all_tf_final_no_promoter.gr , 1000, "center"), nexus.bw.list$PolII)
all_tf_final_no_promoter.df <- as.data.frame(all_tf_final_no_promoter.gr)
write_tsv(all_tf_final_no_promoter.df,"tsv/all_tfregions_with_binding_activity.tsv")

#for making excel of binding and activity signal
islands_new.gr <- islands_new.gr %>% as.data.frame %>%
    dplyr::left_join(., island_names.df, by = 'island_id') %>%
    makeGRangesFromDataFrame(keep.extra.columns = T, starts.in.df.are.0based = F)

islands_new.gr$tead4_sig <- nexus_regionSums(resize(islands_new.gr , 500, "center"), nexus.bw.list$tead4) 
islands_new.gr$tfap2c_sig <- nexus_regionSums(resize(islands_new.gr , 500, "center"), nexus.bw.list$tfap2c) 
islands_new.gr$cdx2_sig <- nexus_regionSums(resize(islands_new.gr, 500, "center"), nexus.bw.list$cdx2) 
islands_new.gr$gata3_sig <- nexus_regionSums(resize(islands_new.gr, 500, "center"), nexus.bw.list$gata3) 
islands_new.gr$yap1_sig <- nexus_regionSums(resize(islands_new.gr, 100, "center"), nexus.bw.list$yap1) 
islands_new.gr$h3k27ac_sig <- regionSums(resize(islands_new.gr, 1000, "center"), seq.bw.list$H3K27ac$rep2)
islands_new.gr$tt_sig <- nexus_regionSums(resize(islands_new.gr, 1000, "center"), nexus.bw.list$ttseq)
islands_new.gr$polii_sig <- nexus_regionSums(resize(islands_new.gr, 1000, "center"), nexus.bw.list$PolII)
islands_new.df <- as.data.frame(islands_new.gr)
write_tsv(islands_new.df,"tsv/all_islands_with_binding_activity.tsv")
```

# Tead4-1_Tfap2c-1 motif pair heatmaps
```{r, Tead4-1_Tfap2c-1 heatmap}
tead4_tfap2c_1 <-subset(all_tf_final_no_promoter.gr,island_content=="Tead4-1_Tfap2c-1")
td.gr <-tead4_tfap2c_1 %>% subset(tf == "Tead4")
tf.gr <- tead4_tfap2c_1 %>% subset(tf== "Tfap2c")

t <- distanceToNearest(td.gr, tf.gr)
g <- t@elementMetadata
s <- g$distance<=160
s <- t[s]
td_tf <-tead4_tfap2c_1[s@from]
td_tf$td_tf_motif_dis <- s@elementMetadata
td_tf <- td_tf[order(td_tf$td_tf_motif_dis, decreasing = F)]
td_tf3 <- td_tf[1:1550,]

order_of_rows<-1:length(td_tf3)
mat <- standard_metapeak_matrix(td_tf3, tsc.contrib.profile.bws$yap1, upstream = 150, downstream = 150)
mat<-mat[order_of_rows, ]
df <- as_tibble(mat)
colnames(df) <- seq(150 - ncol(mat) + 1, 150)
df$rownames <- rownames(mat)
df$order <- seq(nrow(df))
df <- tidyr::gather(data = df, position, value, colnames(df)[1]:colnames(df)[ncol(df)-2])
df$position <- as.numeric(df$position)
df$value_pos <- scales::oob_squish_any(x = df$value, range = c(0, 0.1))
df$value_pos[df$value_pos == 0] <- NA
df$value_neg <- scales::oob_squish_any(x = df$value, range = c(-0.1, 0))
df$value_neg[df$value_neg == 0] <- NA

yp_flip <- df
yp_flip$position[yp_flip$position<0]<-(yp_flip$position + -11)
yp_flip$position <- abs(yp_flip$position)
new_c <-lapply(1:max(yp_flip$position), function(x) data.frame(region = x, position = -10:0, signal = 0)) %>% bind_rows()
new_df <- bind_rows(yp_flip, new_c)

p<-ggplot() +
  geom_tile(data=new_df, aes(x=position, y = order, fill = value_pos))+
  ggtitle("yap1 contrib") +
  {if (is.null("white")) scale_fill_gradient(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", limits = c(0, high.value), space = "Lab") } +
  {if (!is.null("white")) scale_fill_gradient2(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", midpoint = 0, limits = c(0, 0.1), space = "Lab") } +
  scale_x_continuous(expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0)) +
  theme(axis.line = element_line(colour = "black"),
        axis.text.y = element_text(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position="right",
        legend.title = element_text())+
  labs(x = "Position (bp)", y = "regions") + 
  labs(fill = "Contrib. score")

ggsave("figures/5_tead4_yap1_analysis/yap1_contrib_tead_tfap2c_motif_regions.pdf",p,height = 6, width = 4)  

```

# Tead4-1_Gata3-1 motif pair heatmaps
```{r, Tead4-1_Gata3-1 heatmap}
# tead4-gata3

tead4_gata3_1 <-subset(all_tf_final_no_promoter.gr,island_content=="Gata3-1_Tead4-1")
td.gr <- tead4_gata3_1 %>% subset(tf == "Tead4")
gat.gr <- tead4_gata3_1 %>% subset(tf== "Gata3")

t <- distanceToNearest(td.gr, gat.gr)
g <- t@elementMetadata
s <- g$distance<=160
s <- t[s]
td_gt <-tead4_gata3_1[s@from]
td_gt$td_gt_motif_dis <- s@elementMetadata
td_gt <- td_gt[order(td_gt$td_gt_motif_dis, decreasing = F)]

order_of_rows<-1:length(td_gt)
mat <- standard_metapeak_matrix(td_gt, tsc.contrib.profile.bws$yap1, upstream = 150, downstream = 150)
mat<-mat[order_of_rows, ]
df <- as_tibble(mat)
colnames(df) <- seq(150 - ncol(mat) + 1, 150)
df$rownames <- rownames(mat)
df$order <- seq(nrow(df))
df <- tidyr::gather(data = df, position, value, colnames(df)[1]:colnames(df)[ncol(df)-2])
df$position <- as.numeric(df$position)
df$value_pos <- scales::oob_squish_any(x = df$value, range = c(0, 0.1))
df$value_pos[df$value_pos == 0] <- NA
df$value_neg <- scales::oob_squish_any(x = df$value, range = c(-0.1, 0))
df$value_neg[df$value_neg == 0] <- NA

yp_flip <- df
yp_flip$position[yp_flip$position<0]<-(yp_flip$position + -15)
yp_flip$position <- abs(yp_flip$position)
new_c <-lapply(1:max(yp_flip$position), function(x) data.frame(region = x, position = -10:0, signal = 0)) %>% bind_rows()
new_df <- bind_rows(yp_flip, new_c)

yp_tg<-ggplot() +
  geom_tile(data=new_df, aes(x=position, y = order, fill = value_pos))+
  ggtitle("yap1 contrib") +
  {if (is.null("white")) scale_fill_gradient(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", limits = c(0, high.value), space = "Lab") } +
  {if (!is.null("white")) scale_fill_gradient2(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", midpoint = 0, limits = c(0, 0.1), space = "Lab") } +
  scale_x_continuous(expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0)) +
  theme(axis.line = element_line(colour = "black"),
        axis.text.y = element_text(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position="right",
        legend.title = element_text())+
  labs(x = "Position (bp)", y = "regions") + 
  labs(fill = "Contrib. score")

order_of_rows<-1:length(td_tf3)
mat <- standard_metapeak_matrix(td_tf3, tsc.contrib.profile.bws$tfap2c, upstream = 150, downstream = 150)
mat<-mat[order_of_rows, ]
df <- as_tibble(mat)
colnames(df) <- seq(150 - ncol(mat) + 1, 150)
df$rownames <- rownames(mat)
df$order <- seq(nrow(df))
df <- tidyr::gather(data = df, position, value, colnames(df)[1]:colnames(df)[ncol(df)-2])
df$position <- as.numeric(df$position)
df$value_pos <- scales::oob_squish_any(x = df$value, range = c(0, 0.25))
df$value_pos[df$value_pos == 0] <- NA
df$value_neg <- scales::oob_squish_any(x = df$value, range = c(-0.25, 0))
df$value_neg[df$value_neg == 0] <- NA

yp_flip <- df
yp_flip$position[yp_flip$position<0]<-(yp_flip$position + -11)
yp_flip$position <- abs(yp_flip$position)
new_c <-lapply(1:max(yp_flip$position), function(x) data.frame(region = x, position = -10:0, signal = 0)) %>% bind_rows()
new_df <- bind_rows(yp_flip, new_c)

q<-ggplot() +
  geom_tile(data=new_df, aes(x=position, y = order, fill = value_pos))+
  ggtitle("tfap2c contrib") +
  {if (is.null("white")) scale_fill_gradient(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", limits = c(0, 0.25), space = "Lab") } +
  {if (!is.null("white")) scale_fill_gradient2(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", midpoint = 0, limits = c(0, 0.25), space = "Lab") } +
  scale_x_continuous(expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0)) +
  theme(axis.line = element_line(colour = "black"),
        axis.text.y = element_text(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position="right",
        legend.title = element_text())+
  labs(x = "Position (bp)", y = "regions") + 
  labs(fill = "Contrib. score")



order_of_rows<-1:length(td_tf3)
mat <- standard_metapeak_matrix(td_tf3, tsc.contrib.profile.bws$tead4, upstream = 150, downstream = 150)
mat<-mat[order_of_rows, ]
df <- as_tibble(mat)
colnames(df) <- seq(150 - ncol(mat) + 1, 150)
df$rownames <- rownames(mat)
df$order <- seq(nrow(df))
df <- tidyr::gather(data = df, position, value, colnames(df)[1]:colnames(df)[ncol(df)-2])
df$position <- as.numeric(df$position)
df$value_pos <- scales::oob_squish_any(x = df$value, range = c(0, 0.25))
df$value_pos[df$value_pos == 0] <- NA
df$value_neg <- scales::oob_squish_any(x = df$value, range = c(-0.25, 0))
df$value_neg[df$value_neg == 0] <- NA

yp_flip <- df
yp_flip$position[yp_flip$position<0]<-(yp_flip$position + -11)
yp_flip$position <- abs(yp_flip$position)
new_c <-lapply(1:max(yp_flip$position), function(x) data.frame(region = x, position = -10:0, signal = 0)) %>% bind_rows()
new_df <- bind_rows(yp_flip, new_c)

w<-ggplot() +
  geom_tile(data=new_df, aes(x=position, y = order, fill = value_pos))+
  ggtitle("tead4 contrib") +
  {if (is.null("white")) scale_fill_gradient(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", limits = c(0, 0.25), space = "Lab") } +
  {if (!is.null("white")) scale_fill_gradient2(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", midpoint = 0, limits = c(0, 0.25), space = "Lab") } +
  scale_x_continuous(expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0)) +
  theme(axis.line = element_line(colour = "black"),
        axis.text.y = element_text(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position="right",
        legend.title = element_text())+
  labs(x = "Position (bp)", y = "regions") + 
  labs(fill = "Contrib. score")

g<-p + q + w+ patchwork::plot_layout(nrow = 1, widths = c(1, 1,1))


```

# Cdx2-1_Tead4-1 motif pair heatmaps
```{r,Cdx2-1_Tead4-1 heatmap}
# tead4-cdx2
tead4_cdx2_1 <-subset(all_tf_final_no_promoter.gr,island_content=="Cdx2-1_Tead4-1")
td.gr <-tead4_cdx2_1 %>% subset(tf == "Tead4")
cd.gr <- tead4_cdx2_1 %>% subset(tf== "Cdx2")

t <- distanceToNearest(td.gr, cd.gr)
g <- t@elementMetadata
s <- g$distance<=160
s <- t[s]
td_cd <-td.gr[s@from]
td_cd$td_tf_motif_dis <- s@elementMetadata
td_cd <- td_cd[order(td_cd$td_tf_motif_dis, decreasing = F)]


order_of_rows<-1:length(td_cd)
mat <- standard_metapeak_matrix(td_cd, tsc.contrib.profile.bws$yap1, upstream = 150, downstream = 150)
mat<-mat[order_of_rows, ]
df <- as_tibble(mat)
colnames(df) <- seq(150 - ncol(mat) + 1, 150)
df$rownames <- rownames(mat)
df$order <- seq(nrow(df))
df <- tidyr::gather(data = df, position, value, colnames(df)[1]:colnames(df)[ncol(df)-2])
df$position <- as.numeric(df$position)
df$value_pos <- scales::oob_squish_any(x = df$value, range = c(0, 0.1))
df$value_pos[df$value_pos == 0] <- NA
df$value_neg <- scales::oob_squish_any(x = df$value, range = c(-0.1, 0))
df$value_neg[df$value_neg == 0] <- NA

yp_flip <- df
yp_flip$position[yp_flip$position<0]<-(yp_flip$position + -11)
yp_flip$position <- abs(yp_flip$position)
new_c <-lapply(1:max(yp_flip$position), function(x) data.frame(region = x, position = -10:0, signal = 0)) %>% bind_rows()
new_df <- bind_rows(yp_flip, new_c)

p<-ggplot() +
  geom_tile(data=new_df, aes(x=position, y = order, fill = value_pos))+
  ggtitle("yap1 contrib") +
  {if (is.null("white")) scale_fill_gradient(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", limits = c(0, 0.1), space = "Lab") } +
  {if (!is.null("white")) scale_fill_gradient2(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", midpoint = 0, limits = c(0,0.1), space = "Lab") } +
  scale_x_continuous(expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0)) +
  theme(axis.line = element_line(colour = "black"),
        axis.text.y = element_text(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position="right",
        legend.title = element_text())+
  labs(x = "Position (bp)", y = "regions") + 
  labs(fill = "Contrib. score")

order_of_rows<-1:length(td_cd)
mat <- standard_metapeak_matrix(td_cd, tsc.contrib.profile.bws$cdx2, upstream = 150, downstream = 150)
mat<-mat[order_of_rows, ]
df <- as_tibble(mat)
colnames(df) <- seq(150 - ncol(mat) + 1, 150)
df$rownames <- rownames(mat)
df$order <- seq(nrow(df))
df <- tidyr::gather(data = df, position, value, colnames(df)[1]:colnames(df)[ncol(df)-2])
df$position <- as.numeric(df$position)
df$value_pos <- scales::oob_squish_any(x = df$value, range = c(0, 0.2))
df$value_pos[df$value_pos == 0] <- NA
df$value_neg <- scales::oob_squish_any(x = df$value, range = c(-0.2, 0))
df$value_neg[df$value_neg == 0] <- NA

yp_flip <- df
yp_flip$position[yp_flip$position<0]<-(yp_flip$position + -11)
yp_flip$position <- abs(yp_flip$position)
new_c <-lapply(1:max(yp_flip$position), function(x) data.frame(region = x, position = -10:0, signal = 0)) %>% bind_rows()
new_df <- bind_rows(yp_flip, new_c)

q<-ggplot() +
  geom_tile(data=new_df, aes(x=position, y = order, fill = value_pos))+
  ggtitle("cdx2 contrib") +
  {if (is.null("white")) scale_fill_gradient(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", limits = c(0, 0.2), space = "Lab") } +
  {if (!is.null("white")) scale_fill_gradient2(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", midpoint = 0, limits = c(0, 0.2), space = "Lab") } +
  scale_x_continuous(expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0)) +
  theme(axis.line = element_line(colour = "black"),
        axis.text.y = element_text(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position="right",
        legend.title = element_text())+
  labs(x = "Position (bp)", y = "regions") + 
  labs(fill = "Contrib. score")


order_of_rows<-1:length(td_cd)
mat <- standard_metapeak_matrix(td_cd, tsc.contrib.profile.bws$tead4, upstream = 150, downstream = 150)
mat<-mat[order_of_rows, ]
df <- as_tibble(mat)
colnames(df) <- seq(150 - ncol(mat) + 1, 150)
df$rownames <- rownames(mat)
df$order <- seq(nrow(df))
df <- tidyr::gather(data = df, position, value, colnames(df)[1]:colnames(df)[ncol(df)-2])
df$position <- as.numeric(df$position)
df$value_pos <- scales::oob_squish_any(x = df$value, range = c(0, 0.2))
df$value_pos[df$value_pos == 0] <- NA
df$value_neg <- scales::oob_squish_any(x = df$value, range = c(-0.2, 0))
df$value_neg[df$value_neg == 0] <- NA

yp_flip <- df
yp_flip$position[yp_flip$position<0]<-(yp_flip$position + -11)
yp_flip$position <- abs(yp_flip$position)
new_c <-lapply(1:max(yp_flip$position), function(x) data.frame(region = x, position = -10:0, signal = 0)) %>% bind_rows()
new_df <- bind_rows(yp_flip, new_c)

w<-ggplot() +
  geom_tile(data=new_df, aes(x=position, y = order, fill = value_pos))+
  ggtitle("tead4 contrib") +
  {if (is.null("white")) scale_fill_gradient(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", limits = c(0, 0.2), space = "Lab") } +
  {if (!is.null("white")) scale_fill_gradient2(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", midpoint = 0, limits = c(0, 0.2), space = "Lab") } +
  scale_x_continuous(expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0)) +
  theme(axis.line = element_line(colour = "black"),
        axis.text.y = element_text(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position="right",
        legend.title = element_text())+
  labs(x = "Position (bp)", y = "regions") + 
  labs(fill = "Contrib. score")


g<-p + q + w+ patchwork::plot_layout(nrow = 1, widths = c(1, 1,1))

```

# Gata3-1_Tead4-1 motif pair heatmaps
```{r,Gata3-1_Tead4-1 heatmap}
# tead4-gata3
tead4_gata3_1 <-subset(all_tf_final_no_promoter.gr,island_content=="Gata3-1_Tead4-1")
td.gr <- tead4_gata3_1 %>% subset(tf == "Tead4")
gt.gr <-  tead4_gata3_1 %>% subset(tf== "Gata3")

t <- distanceToNearest(td.gr, gt.gr)
g <- t@elementMetadata
s <- g$distance<=160
s <- t[s]
td_gt <- tead4_gata3_1[s@from]
td_gt$td_tf_motif_dis <- s@elementMetadata
td_gt <- td_gt[order(td_gt$td_tf_motif_dis, decreasing = F)]


order_of_rows<-1:length(td_gt)
mat <- standard_metapeak_matrix(td_gt, tsc.contrib.profile.bws$yap1, upstream = 150, downstream = 150)
mat<-mat[order_of_rows, ]
df <- as_tibble(mat)
colnames(df) <- seq(150 - ncol(mat) + 1, 150)
df$rownames <- rownames(mat)
df$order <- seq(nrow(df))
df <- tidyr::gather(data = df, position, value, colnames(df)[1]:colnames(df)[ncol(df)-2])
df$position <- as.numeric(df$position)
df$value_pos <- scales::oob_squish_any(x = df$value, range = c(0, 0.1))
df$value_pos[df$value_pos == 0] <- NA
df$value_neg <- scales::oob_squish_any(x = df$value, range = c(-0.1, 0))
df$value_neg[df$value_neg == 0] <- NA

yp_flip <- df
yp_flip$position[yp_flip$position<0]<-(yp_flip$position + -16)
yp_flip$position <- abs(yp_flip$position)
new_c <-lapply(1:max(yp_flip$position), function(x) data.frame(region = x, position = -10:0, signal = 0)) %>% bind_rows()
new_df <- bind_rows(yp_flip, new_c)

p<-ggplot() +
  geom_tile(data=new_df, aes(x=position, y = order, fill = value_pos))+
  ggtitle("yap1 contrib") +
  {if (is.null("white")) scale_fill_gradient(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", limits = c(0, 0.1), space = "Lab") } +
  {if (!is.null("white")) scale_fill_gradient2(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", midpoint = 0, limits = c(0,0.1), space = "Lab") } +
  scale_x_continuous(expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0)) +
  theme(axis.line = element_line(colour = "black"),
        axis.text.y = element_text(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position="right",
        legend.title = element_text())+
  labs(x = "Position (bp)", y = "regions") + 
  labs(fill = "Contrib. score")

order_of_rows<-1:length(td_gt)
mat <- standard_metapeak_matrix(td_gt, tsc.contrib.profile.bws$gata3, upstream = 150, downstream = 150)
mat<-mat[order_of_rows, ]
df <- as_tibble(mat)
colnames(df) <- seq(150 - ncol(mat) + 1, 150)
df$rownames <- rownames(mat)
df$order <- seq(nrow(df))
df <- tidyr::gather(data = df, position, value, colnames(df)[1]:colnames(df)[ncol(df)-2])
df$position <- as.numeric(df$position)
df$value_pos <- scales::oob_squish_any(x = df$value, range = c(0, 0.25))
df$value_pos[df$value_pos == 0] <- NA
df$value_neg <- scales::oob_squish_any(x = df$value, range = c(-0.25, 0))
df$value_neg[df$value_neg == 0] <- NA

yp_flip <- df
yp_flip$position[yp_flip$position<0]<-(yp_flip$position + -16)
yp_flip$position <- abs(yp_flip$position)
new_c <-lapply(1:max(yp_flip$position), function(x) data.frame(region = x, position = -10:0, signal = 0)) %>% bind_rows()
new_df <- bind_rows(yp_flip, new_c)

q<-ggplot() +
  geom_tile(data=new_df, aes(x=position, y = order, fill = value_pos))+
  ggtitle("gata3 contrib") +
  {if (is.null("white")) scale_fill_gradient(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", limits = c(0, 0.25), space = "Lab") } +
  {if (!is.null("white")) scale_fill_gradient2(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", midpoint = 0, limits = c(0, 0.25), space = "Lab") } +
  scale_x_continuous(expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0)) +
  theme(axis.line = element_line(colour = "black"),
        axis.text.y = element_text(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position="right",
        legend.title = element_text())+
  labs(x = "Position (bp)", y = "regions") + 
  labs(fill = "Contrib. score")


order_of_rows<-1:length(td_gt)
mat <- standard_metapeak_matrix(td_gt, tsc.contrib.profile.bws$tead4, upstream = 150, downstream = 150)
mat<-mat[order_of_rows, ]
df <- as_tibble(mat)
colnames(df) <- seq(150 - ncol(mat) + 1, 150)
df$rownames <- rownames(mat)
df$order <- seq(nrow(df))
df <- tidyr::gather(data = df, position, value, colnames(df)[1]:colnames(df)[ncol(df)-2])
df$position <- as.numeric(df$position)
df$value_pos <- scales::oob_squish_any(x = df$value, range = c(0, 0.25))
df$value_pos[df$value_pos == 0] <- NA
df$value_neg <- scales::oob_squish_any(x = df$value, range = c(-0.25, 0))
df$value_neg[df$value_neg == 0] <- NA

yp_flip <- df
yp_flip$position[yp_flip$position<0]<-(yp_flip$position + -16)
yp_flip$position <- abs(yp_flip$position)
new_c <-lapply(1:max(yp_flip$position), function(x) data.frame(region = x, position = -10:0, signal = 0)) %>% bind_rows()
new_df <- bind_rows(yp_flip, new_c)

w<-ggplot() +
  geom_tile(data=new_df, aes(x=position, y = order, fill = value_pos))+
  ggtitle("tead4 contrib") +
  {if (is.null("white")) scale_fill_gradient(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", limits = c(0, 0.25), space = "Lab") } +
  {if (!is.null("white")) scale_fill_gradient2(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", midpoint = 0, limits = c(0, 0.25), space = "Lab") } +
  scale_x_continuous(expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0)) +
  theme(axis.line = element_line(colour = "black"),
        axis.text.y = element_text(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position="right",
        legend.title = element_text())+
  labs(x = "Position (bp)", y = "regions") + 
  labs(fill = "Contrib. score")


g<-p + q + w+ patchwork::plot_layout(nrow = 1, widths = c(1, 1,1))


```

# Tead4_double-1_Tfap2c-1 motif pair heatmaps
```{r,Tead4_double-1_Tfap2c-1 heatmap}
tead4double_tfap2c_1 <-subset(all_tf_final_no_promoter.gr,island_content=="Tead4_double-1_Tfap2c-1")
tdbl.gr <-tead4double_tfap2c_1 %>% subset(tf == "Tead4_double")
tf.gr <- tead4double_tfap2c_1 %>% subset(tf== "Tfap2c")

t <- distanceToNearest(tdbl.gr, tf.gr)
g <- t@elementMetadata
s <- g$distance<=160
s <- t[s]
tdbl_tf <- tdbl.gr[s@from]
tdbl_tf$td_tf_motif_dis <- s@elementMetadata
tdbl_tf <- tdbl_tf[order(tdbl_tf$td_tf_motif_dis, decreasing = F)]


order_of_rows<-1:length(tdbl_tf)
mat <- standard_metapeak_matrix(tdbl_tf, tsc.contrib.profile.bws$yap1, upstream = 150, downstream = 150)
mat <- mat[apply(mat[,-1], 1, function(x) !all(x==0)),]
mat<-mat[order_of_rows, ]
df <- as_tibble(mat)
colnames(df) <- seq(150 - ncol(mat) + 1, 150)
df$rownames <- rownames(mat)
df$order <- seq(nrow(df))
df <- tidyr::gather(data = df, position, value, colnames(df)[1]:colnames(df)[ncol(df)-2])
df$position <- as.numeric(df$position)
df$value_pos <- scales::oob_squish_any(x = df$value, range = c(0, 0.1))
df$value_pos[df$value_pos == 0] <- NA
df$value_neg <- scales::oob_squish_any(x = df$value, range = c(-0.1, 0))
df$value_neg[df$value_neg == 0] <- NA

yp_flip <- df
yp_flip$position[yp_flip$position<0]<-(yp_flip$position + -17)
yp_flip$position <- abs(yp_flip$position)
new_c <-lapply(1:max(yp_flip$position), function(x) data.frame(region = x, position = -10:0, signal = 0)) %>% bind_rows()
new_df <- bind_rows(yp_flip, new_c)

p<-ggplot() +
  geom_tile(data=new_df, aes(x=position, y = order, fill = value_pos))+
  ggtitle("yap1 contrib") +
  {if (is.null("white")) scale_fill_gradient(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", limits = c(0, 0.1), space = "Lab") } +
  {if (!is.null("white")) scale_fill_gradient2(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", midpoint = 0, limits = c(0,0.1), space = "Lab") } +
  scale_x_continuous(expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0)) +
  theme(axis.line = element_line(colour = "black"),
        axis.text.y = element_text(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position="right",
        legend.title = element_text())+
  labs(x = "Position (bp)", y = "regions") + 
  labs(fill = "Contrib. score")

order_of_rows<-1:length(tdbl_tf)
mat <- standard_metapeak_matrix(tdbl_tf, tsc.contrib.profile.bws$tfap2c, upstream = 150, downstream = 150)
mat <- mat[apply(mat[,-1], 1, function(x) !all(x==0)),]
mat<-mat[order_of_rows, ]
df <- as_tibble(mat)
colnames(df) <- seq(150 - ncol(mat) + 1, 150)
df$rownames <- rownames(mat)
df$order <- seq(nrow(df))
df <- tidyr::gather(data = df, position, value, colnames(df)[1]:colnames(df)[ncol(df)-2])
df$position <- as.numeric(df$position)
df$value_pos <- scales::oob_squish_any(x = df$value, range = c(0, 0.25))
df$value_pos[df$value_pos == 0] <- NA
df$value_neg <- scales::oob_squish_any(x = df$value, range = c(-0.25, 0))
df$value_neg[df$value_neg == 0] <- NA

yp_flip <- df
yp_flip$position[yp_flip$position<0]<-(yp_flip$position + -17)
yp_flip$position <- abs(yp_flip$position)
new_c <-lapply(1:max(yp_flip$position), function(x) data.frame(region = x, position = -10:0, signal = 0)) %>% bind_rows()
new_df <- bind_rows(yp_flip, new_c)

q<-ggplot() +
  geom_tile(data=new_df, aes(x=position, y = order, fill = value_pos))+
  ggtitle("tfap2c contrib") +
  {if (is.null("white")) scale_fill_gradient(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", limits = c(0, 0.25), space = "Lab") } +
  {if (!is.null("white")) scale_fill_gradient2(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", midpoint = 0, limits = c(0, 0.25), space = "Lab") } +
  scale_x_continuous(expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0)) +
  theme(axis.line = element_line(colour = "black"),
        axis.text.y = element_text(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position="right",
        legend.title = element_text())+
  labs(x = "Position (bp)", y = "regions") + 
  labs(fill = "Contrib. score")


order_of_rows<-1:length(tdbl_tf)
mat <- standard_metapeak_matrix(tdbl_tf, tsc.contrib.profile.bws$tead4, upstream = 150, downstream = 150)
mat <- mat[apply(mat[,-1], 1, function(x) !all(x==0)),]
mat<-mat[order_of_rows, ]
df <- as_tibble(mat)
colnames(df) <- seq(150 - ncol(mat) + 1, 150)
df$rownames <- rownames(mat)
df$order <- seq(nrow(df))
df <- tidyr::gather(data = df, position, value, colnames(df)[1]:colnames(df)[ncol(df)-2])
df$position <- as.numeric(df$position)
df$value_pos <- scales::oob_squish_any(x = df$value, range = c(0, 0.25))
df$value_pos[df$value_pos == 0] <- NA
df$value_neg <- scales::oob_squish_any(x = df$value, range = c(-0.25, 0))
df$value_neg[df$value_neg == 0] <- NA

yp_flip <- df
yp_flip$position[yp_flip$position<0]<-(yp_flip$position + -17)
yp_flip$position <- abs(yp_flip$position)
new_c <-lapply(1:max(yp_flip$position), function(x) data.frame(region = x, position = -10:0, signal = 0)) %>% bind_rows()
new_df <- bind_rows(yp_flip, new_c)

w<-ggplot() +
  geom_tile(data=new_df, aes(x=position, y = order, fill = value_pos))+
  ggtitle("tead4 double contrib") +
  {if (is.null("white")) scale_fill_gradient(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", limits = c(0, 0.25), space = "Lab") } +
  {if (!is.null("white")) scale_fill_gradient2(high = "#67000d", low = "blue", mid = "white", guide = guide_legend(reverse = FALSE), na.value = "transparent", midpoint = 0, limits = c(0, 0.25), space = "Lab") } +
  scale_x_continuous(expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0)) +
  theme(axis.line = element_line(colour = "black"),
        axis.text.y = element_text(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        legend.position="right",
        legend.title = element_text())+
  labs(x = "Position (bp)", y = "regions") + 
  labs(fill = "Contrib. score")


g<-p + q + w+ patchwork::plot_layout(nrow = 1, widths = c(1, 1,1))
```


# GO analysis with Tead single-Tfap2c motif-pair 
```{r, go analysis}
##go term
library("TxDb.Mmusculus.UCSC.mm10.knownGene")
library(org.Mm.eg.db)
library(clusterProfiler)
library(bumphunter)
genes <- annotateTranscripts(TxDb.Mmusculus.UCSC.mm10.knownGene)
tab_tdtf<- matchGenes(td_tf ,genes)
go <- enrichGO(gene          = tab_tdtf$Geneid,
               OrgDb         = org.Mm.eg.db,
               keyType       = 'ENTREZID',
               ont           = "BP",
               pAdjustMethod = "BH",
               pvalueCutoff  = 0.01,
               qvalueCutoff  = 0.01)
go <- setReadable(go, OrgDb = org.Mm.eg.db)
#clusterProfiler::dotplot(go, showCategory=10)
go_df <- as.data.frame(go)

## selecting the category of interests that are less than or equal to 0.05 of p.adjust and qvalues 
selc_cat <- dplyr::filter(go_df, Description=='cell-cell signaling by wnt'| Description=='pattern specification process'|Description=='regulation of epithelial cell proliferation'|Description=='cell junction assembly'|Description== 'regulation of DNA-binding transcription factor activity'|Description=='positive regulation of cell development'|Description=='actin filament organization'|Description=='cell fate commitment'|Description=='extracellular matrix organization' |Description=='regulation of GTPase activity')

# plotting dotplot
dot<-  ggplot(selc_cat, aes(x = GeneRatio, y = fct_reorder(Description, GeneRatio))) + 
    geom_point(aes(size = Count, color = p.adjust)) +
    theme_bw(base_size = 14) +
    scale_colour_gradient(low="red") +
    ylab(NULL) +
    ggtitle("GO pathway enrichment")
 
ggsave("figures/5_tead4_yap1_analysis/tt_1_go_term.pdf",dot,height = 6, width = 4)  
```

#Yap1 insilico distance prediction analysis
```{r, yap1 insilico pred heatmap}

insilico_yap <- read_tsv("tsv/4b_yaptask_summary_cttygjectg_bothdouble_insilico_distance_analysis.tsv")  

motif_order <- c('tead4_double','tead4','junfos','tfap2c','ctcf','elf5','gata3','cdx2','gata3_double')     
insilico_yap$motif_pair <- factor(insilico_yap$side_motif, levels = rev(motif_order))
insilico_yap$name <- "tsc"
  
yap<- ggplot(insilico_yap, aes(x=name,y=motif_pair, fill = log2_fc_max_cooperativity))+
  geom_tile(color = 'black')+
  xlab("")+scale_y_discrete(name = 'Motifs')+scale_fill_gradientn(colours = rev(c('#A3985A', '#C2BA91','#E0DDC8', 'white')))+
  labs(fill='Average log2\npredicted binding\nenhancement')+
  theme_classic()
  
ggsave("figures/5_tead4_yap1_analysis/yap1_pred_insilico_near_motifs.pdf",yap,height = 6, width = 4)  
```

# Session Information

For the purposes of reproducibility, the R/Bioconductor session information is printed below:

```{r sessioninfo}
sessionInfo()
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
