{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "This analysis to quantify the cooperative binding of Tead4 on Tead double motifs genome-wide using trined TSC model. We  with measure the fold increase in binding to the intact double motif as compared to the sum of the binding to each motif alone.\n",
    "\n",
    "# Computational setup\n",
    "\n",
    "## Environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n",
      "2024-01-19 15:17:49,317 [INFO] Note: detected 80 virtual cores but NumExpr set to maximum of 64, check \"NUMEXPR_MAX_THREADS\" environment variable.\n",
      "2024-01-19 15:17:49,320 [INFO] Note: NumExpr detected 80 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "2024-01-19 15:17:49,321 [INFO] NumExpr defaulting to 8 threads.\n",
      "2024-01-19 15:17:54,925 [INFO] Failed to extract font properties from /usr/share/fonts/google-noto-emoji/NotoColorEmoji.ttf: In FT2Font: Can not load face.  Unknown file format.\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from tensorflow.python.util import deprecation\n",
    "deprecation._PRINT_DEPRECATION_WARNINGS = False\n",
    "\n",
    "#Packages\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import plotnine\n",
    "#!pip install logomaker\n",
    "import seaborn as sns\n",
    "from pybedtools import BedTool\n",
    "from tqdm import tqdm\n",
    "from bpnet.utils import create_tf_session\n",
    "from bpnet.BPNet import BPNetSeqModel\n",
    "from bpnet.simulate import motif_coords\n",
    "from bpnet.preproc import resize_interval\n",
    "from concise.preprocessing import encodeDNA\n",
    "from bpnet.extractors import extract_seq\n",
    "from plotnine import *\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# Settings\n",
    "os.chdir(f'/n/projects/kd2200/publication/bpnet/analysis/')\n",
    "pd.set_option('display.max_columns', 100)\n",
    "create_tf_session('0',1)\n",
    "\n",
    "# Custom functions\n",
    "sys.path.insert(0, f'/n/projects/kd2200/publication/bpnet/analysis/scripts/bpnet/scripts')\n",
    "from perturb_functions import insert_motif, one_hot_decode_sequence, one_hot_encode_sequences, one_hot_encode_sequence\n",
    "from motif_functions import remove_palindromic_motif_duplicates\n",
    "from data_format_functions import tidy_bpnet_predictions_nexus,tidy_bpnet_contributions\n",
    "sys.path.insert(0, f'/n/projects/kd2200/publication/bpnet/analysis/scripts/py')\n",
    "from motifs import extract_seqs_from_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%matplotlib inline\n",
    "# Pre-existing variables\n",
    "fasta_file = f'/n/projects/kd2200/publication/bpnet/fasta/mm10.fa'\n",
    "model_dir = f'/n/projects/kd2200/publication/bpnet/model/dataspec.yaml_default_fold_5/'\n",
    "\n",
    "# Independent variables\n",
    "#tead4_motifs_path = f'tsv/tead4_double_motifs_no_ov_noerv_no_promoter_cwm_fold5_test.tsv.gz'\n",
    "tead4_motifs_path = f'/n/projects/kd2200/publication/bpnet/cwm_all_regions/tead4/instances_by_pattern/m0_p1_motif-instances-all-regions.tsv.gz'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load BPNet model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TF-MoDISco is using the TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "model = BPNetSeqModel.from_mdir(model_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Import Tead4 double motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>example_chrom</th>\n",
       "      <th>pattern_start_abs</th>\n",
       "      <th>pattern_end_abs</th>\n",
       "      <th>pattern</th>\n",
       "      <th>contrib_weighted_p</th>\n",
       "      <th>strand</th>\n",
       "      <th>match_weighted_p</th>\n",
       "      <th>example_idx</th>\n",
       "      <th>pattern_start</th>\n",
       "      <th>pattern_end</th>\n",
       "      <th>pattern_center</th>\n",
       "      <th>pattern_len</th>\n",
       "      <th>match_weighted</th>\n",
       "      <th>match_weighted_cat</th>\n",
       "      <th>match_max</th>\n",
       "      <th>match_max_task</th>\n",
       "      <th>contrib_weighted</th>\n",
       "      <th>contrib_weighted_cat</th>\n",
       "      <th>contrib_max</th>\n",
       "      <th>contrib_max_task</th>\n",
       "      <th>seq_match</th>\n",
       "      <th>seq_match_p</th>\n",
       "      <th>seq_match_cat</th>\n",
       "      <th>match/tead4</th>\n",
       "      <th>contrib/tead4</th>\n",
       "      <th>pattern_short</th>\n",
       "      <th>example_start</th>\n",
       "      <th>example_end</th>\n",
       "      <th>example_strand</th>\n",
       "      <th>example_interval_from_task</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30420800</td>\n",
       "      <td>30420817</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.430212</td>\n",
       "      <td>+</td>\n",
       "      <td>0.470039</td>\n",
       "      <td>8</td>\n",
       "      <td>654</td>\n",
       "      <td>671</td>\n",
       "      <td>663</td>\n",
       "      <td>17</td>\n",
       "      <td>0.496458</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.496458</td>\n",
       "      <td>tead4</td>\n",
       "      <td>3.093251</td>\n",
       "      <td>medium</td>\n",
       "      <td>3.093251</td>\n",
       "      <td>tead4</td>\n",
       "      <td>10.025424</td>\n",
       "      <td>0.383925</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.496458</td>\n",
       "      <td>3.093251</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>30420146</td>\n",
       "      <td>30421146</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30420802</td>\n",
       "      <td>30420819</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.402942</td>\n",
       "      <td>-</td>\n",
       "      <td>0.731970</td>\n",
       "      <td>8</td>\n",
       "      <td>656</td>\n",
       "      <td>673</td>\n",
       "      <td>664</td>\n",
       "      <td>17</td>\n",
       "      <td>0.544908</td>\n",
       "      <td>high</td>\n",
       "      <td>0.544908</td>\n",
       "      <td>tead4</td>\n",
       "      <td>2.988721</td>\n",
       "      <td>medium</td>\n",
       "      <td>2.988721</td>\n",
       "      <td>tead4</td>\n",
       "      <td>13.275682</td>\n",
       "      <td>0.898457</td>\n",
       "      <td>high</td>\n",
       "      <td>0.544908</td>\n",
       "      <td>2.988721</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>30420146</td>\n",
       "      <td>30421146</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182250577</td>\n",
       "      <td>182250594</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.067815</td>\n",
       "      <td>-</td>\n",
       "      <td>0.328669</td>\n",
       "      <td>16</td>\n",
       "      <td>293</td>\n",
       "      <td>310</td>\n",
       "      <td>301</td>\n",
       "      <td>17</td>\n",
       "      <td>0.468899</td>\n",
       "      <td>low</td>\n",
       "      <td>0.468899</td>\n",
       "      <td>tead4</td>\n",
       "      <td>1.222768</td>\n",
       "      <td>low</td>\n",
       "      <td>1.222768</td>\n",
       "      <td>tead4</td>\n",
       "      <td>9.284872</td>\n",
       "      <td>0.270542</td>\n",
       "      <td>low</td>\n",
       "      <td>0.468899</td>\n",
       "      <td>1.222768</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>182250284</td>\n",
       "      <td>182251284</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>180419635</td>\n",
       "      <td>180419652</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.020093</td>\n",
       "      <td>-</td>\n",
       "      <td>0.947973</td>\n",
       "      <td>17</td>\n",
       "      <td>305</td>\n",
       "      <td>322</td>\n",
       "      <td>313</td>\n",
       "      <td>17</td>\n",
       "      <td>0.602751</td>\n",
       "      <td>high</td>\n",
       "      <td>0.602751</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.964003</td>\n",
       "      <td>low</td>\n",
       "      <td>0.964003</td>\n",
       "      <td>tead4</td>\n",
       "      <td>10.964270</td>\n",
       "      <td>0.543595</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.602751</td>\n",
       "      <td>0.964003</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>180419330</td>\n",
       "      <td>180420330</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>180419786</td>\n",
       "      <td>180419803</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.359168</td>\n",
       "      <td>-</td>\n",
       "      <td>0.348762</td>\n",
       "      <td>17</td>\n",
       "      <td>456</td>\n",
       "      <td>473</td>\n",
       "      <td>464</td>\n",
       "      <td>17</td>\n",
       "      <td>0.472677</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.472677</td>\n",
       "      <td>tead4</td>\n",
       "      <td>2.774343</td>\n",
       "      <td>medium</td>\n",
       "      <td>2.774343</td>\n",
       "      <td>tead4</td>\n",
       "      <td>8.084553</td>\n",
       "      <td>0.123430</td>\n",
       "      <td>low</td>\n",
       "      <td>0.472677</td>\n",
       "      <td>2.774343</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>180419330</td>\n",
       "      <td>180420330</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18365</th>\n",
       "      <td>chrX</td>\n",
       "      <td>151233079</td>\n",
       "      <td>151233096</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.689630</td>\n",
       "      <td>+</td>\n",
       "      <td>0.771439</td>\n",
       "      <td>187740</td>\n",
       "      <td>620</td>\n",
       "      <td>637</td>\n",
       "      <td>629</td>\n",
       "      <td>17</td>\n",
       "      <td>0.551809</td>\n",
       "      <td>high</td>\n",
       "      <td>0.551809</td>\n",
       "      <td>tead4</td>\n",
       "      <td>4.271722</td>\n",
       "      <td>high</td>\n",
       "      <td>4.271722</td>\n",
       "      <td>tead4</td>\n",
       "      <td>11.921370</td>\n",
       "      <td>0.707930</td>\n",
       "      <td>high</td>\n",
       "      <td>0.551809</td>\n",
       "      <td>4.271722</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>151232459</td>\n",
       "      <td>151233459</td>\n",
       "      <td>.</td>\n",
       "      <td>gata3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18366</th>\n",
       "      <td>chrX</td>\n",
       "      <td>36805292</td>\n",
       "      <td>36805309</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.519555</td>\n",
       "      <td>-</td>\n",
       "      <td>0.937567</td>\n",
       "      <td>187742</td>\n",
       "      <td>349</td>\n",
       "      <td>366</td>\n",
       "      <td>357</td>\n",
       "      <td>17</td>\n",
       "      <td>0.598039</td>\n",
       "      <td>high</td>\n",
       "      <td>0.598039</td>\n",
       "      <td>tead4</td>\n",
       "      <td>3.522194</td>\n",
       "      <td>medium</td>\n",
       "      <td>3.522194</td>\n",
       "      <td>tead4</td>\n",
       "      <td>12.923700</td>\n",
       "      <td>0.846430</td>\n",
       "      <td>high</td>\n",
       "      <td>0.598039</td>\n",
       "      <td>3.522194</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>36804943</td>\n",
       "      <td>36805943</td>\n",
       "      <td>.</td>\n",
       "      <td>gata3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18367</th>\n",
       "      <td>chrX</td>\n",
       "      <td>64238848</td>\n",
       "      <td>64238865</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.399713</td>\n",
       "      <td>+</td>\n",
       "      <td>0.367061</td>\n",
       "      <td>187753</td>\n",
       "      <td>442</td>\n",
       "      <td>459</td>\n",
       "      <td>451</td>\n",
       "      <td>17</td>\n",
       "      <td>0.476800</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.476800</td>\n",
       "      <td>tead4</td>\n",
       "      <td>2.974250</td>\n",
       "      <td>medium</td>\n",
       "      <td>2.974250</td>\n",
       "      <td>tead4</td>\n",
       "      <td>9.598459</td>\n",
       "      <td>0.318263</td>\n",
       "      <td>low</td>\n",
       "      <td>0.476800</td>\n",
       "      <td>2.974250</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>64238406</td>\n",
       "      <td>64239406</td>\n",
       "      <td>.</td>\n",
       "      <td>gata3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18368</th>\n",
       "      <td>chrX</td>\n",
       "      <td>151148657</td>\n",
       "      <td>151148674</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.757086</td>\n",
       "      <td>+</td>\n",
       "      <td>0.326516</td>\n",
       "      <td>187767</td>\n",
       "      <td>132</td>\n",
       "      <td>149</td>\n",
       "      <td>141</td>\n",
       "      <td>17</td>\n",
       "      <td>0.468207</td>\n",
       "      <td>low</td>\n",
       "      <td>0.468207</td>\n",
       "      <td>tead4</td>\n",
       "      <td>4.556071</td>\n",
       "      <td>high</td>\n",
       "      <td>4.556071</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.986142</td>\n",
       "      <td>0.001076</td>\n",
       "      <td>low</td>\n",
       "      <td>0.468207</td>\n",
       "      <td>4.556071</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>151148525</td>\n",
       "      <td>151149525</td>\n",
       "      <td>.</td>\n",
       "      <td>gata3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18369</th>\n",
       "      <td>chrX</td>\n",
       "      <td>151148659</td>\n",
       "      <td>151148676</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.840689</td>\n",
       "      <td>-</td>\n",
       "      <td>0.793326</td>\n",
       "      <td>187767</td>\n",
       "      <td>134</td>\n",
       "      <td>151</td>\n",
       "      <td>142</td>\n",
       "      <td>17</td>\n",
       "      <td>0.556961</td>\n",
       "      <td>high</td>\n",
       "      <td>0.556961</td>\n",
       "      <td>tead4</td>\n",
       "      <td>4.993919</td>\n",
       "      <td>high</td>\n",
       "      <td>4.993919</td>\n",
       "      <td>tead4</td>\n",
       "      <td>12.246571</td>\n",
       "      <td>0.763186</td>\n",
       "      <td>high</td>\n",
       "      <td>0.556961</td>\n",
       "      <td>4.993919</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>151148525</td>\n",
       "      <td>151149525</td>\n",
       "      <td>.</td>\n",
       "      <td>gata3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18370 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      example_chrom  pattern_start_abs  pattern_end_abs  \\\n",
       "0              chr1           30420800         30420817   \n",
       "1              chr1           30420802         30420819   \n",
       "2              chr1          182250577        182250594   \n",
       "3              chr1          180419635        180419652   \n",
       "4              chr1          180419786        180419803   \n",
       "...             ...                ...              ...   \n",
       "18365          chrX          151233079        151233096   \n",
       "18366          chrX           36805292         36805309   \n",
       "18367          chrX           64238848         64238865   \n",
       "18368          chrX          151148657        151148674   \n",
       "18369          chrX          151148659        151148676   \n",
       "\n",
       "                       pattern  contrib_weighted_p strand  match_weighted_p  \\\n",
       "0      metacluster_0/pattern_1            0.430212      +          0.470039   \n",
       "1      metacluster_0/pattern_1            0.402942      -          0.731970   \n",
       "2      metacluster_0/pattern_1            0.067815      -          0.328669   \n",
       "3      metacluster_0/pattern_1            0.020093      -          0.947973   \n",
       "4      metacluster_0/pattern_1            0.359168      -          0.348762   \n",
       "...                        ...                 ...    ...               ...   \n",
       "18365  metacluster_0/pattern_1            0.689630      +          0.771439   \n",
       "18366  metacluster_0/pattern_1            0.519555      -          0.937567   \n",
       "18367  metacluster_0/pattern_1            0.399713      +          0.367061   \n",
       "18368  metacluster_0/pattern_1            0.757086      +          0.326516   \n",
       "18369  metacluster_0/pattern_1            0.840689      -          0.793326   \n",
       "\n",
       "       example_idx  pattern_start  pattern_end  pattern_center  pattern_len  \\\n",
       "0                8            654          671             663           17   \n",
       "1                8            656          673             664           17   \n",
       "2               16            293          310             301           17   \n",
       "3               17            305          322             313           17   \n",
       "4               17            456          473             464           17   \n",
       "...            ...            ...          ...             ...          ...   \n",
       "18365       187740            620          637             629           17   \n",
       "18366       187742            349          366             357           17   \n",
       "18367       187753            442          459             451           17   \n",
       "18368       187767            132          149             141           17   \n",
       "18369       187767            134          151             142           17   \n",
       "\n",
       "       match_weighted match_weighted_cat  match_max match_max_task  \\\n",
       "0            0.496458             medium   0.496458          tead4   \n",
       "1            0.544908               high   0.544908          tead4   \n",
       "2            0.468899                low   0.468899          tead4   \n",
       "3            0.602751               high   0.602751          tead4   \n",
       "4            0.472677             medium   0.472677          tead4   \n",
       "...               ...                ...        ...            ...   \n",
       "18365        0.551809               high   0.551809          tead4   \n",
       "18366        0.598039               high   0.598039          tead4   \n",
       "18367        0.476800             medium   0.476800          tead4   \n",
       "18368        0.468207                low   0.468207          tead4   \n",
       "18369        0.556961               high   0.556961          tead4   \n",
       "\n",
       "       contrib_weighted contrib_weighted_cat  contrib_max contrib_max_task  \\\n",
       "0              3.093251               medium     3.093251            tead4   \n",
       "1              2.988721               medium     2.988721            tead4   \n",
       "2              1.222768                  low     1.222768            tead4   \n",
       "3              0.964003                  low     0.964003            tead4   \n",
       "4              2.774343               medium     2.774343            tead4   \n",
       "...                 ...                  ...          ...              ...   \n",
       "18365          4.271722                 high     4.271722            tead4   \n",
       "18366          3.522194               medium     3.522194            tead4   \n",
       "18367          2.974250               medium     2.974250            tead4   \n",
       "18368          4.556071                 high     4.556071            tead4   \n",
       "18369          4.993919                 high     4.993919            tead4   \n",
       "\n",
       "       seq_match  seq_match_p seq_match_cat  match/tead4  contrib/tead4  \\\n",
       "0      10.025424     0.383925        medium     0.496458       3.093251   \n",
       "1      13.275682     0.898457          high     0.544908       2.988721   \n",
       "2       9.284872     0.270542           low     0.468899       1.222768   \n",
       "3      10.964270     0.543595        medium     0.602751       0.964003   \n",
       "4       8.084553     0.123430           low     0.472677       2.774343   \n",
       "...          ...          ...           ...          ...            ...   \n",
       "18365  11.921370     0.707930          high     0.551809       4.271722   \n",
       "18366  12.923700     0.846430          high     0.598039       3.522194   \n",
       "18367   9.598459     0.318263           low     0.476800       2.974250   \n",
       "18368   0.986142     0.001076           low     0.468207       4.556071   \n",
       "18369  12.246571     0.763186          high     0.556961       4.993919   \n",
       "\n",
       "      pattern_short  example_start  example_end example_strand  \\\n",
       "0             m0_p1       30420146     30421146              .   \n",
       "1             m0_p1       30420146     30421146              .   \n",
       "2             m0_p1      182250284    182251284              .   \n",
       "3             m0_p1      180419330    180420330              .   \n",
       "4             m0_p1      180419330    180420330              .   \n",
       "...             ...            ...          ...            ...   \n",
       "18365         m0_p1      151232459    151233459              .   \n",
       "18366         m0_p1       36804943     36805943              .   \n",
       "18367         m0_p1       64238406     64239406              .   \n",
       "18368         m0_p1      151148525    151149525              .   \n",
       "18369         m0_p1      151148525    151149525              .   \n",
       "\n",
       "      example_interval_from_task  \n",
       "0                           cdx2  \n",
       "1                           cdx2  \n",
       "2                           cdx2  \n",
       "3                           cdx2  \n",
       "4                           cdx2  \n",
       "...                          ...  \n",
       "18365                      gata3  \n",
       "18366                      gata3  \n",
       "18367                      gata3  \n",
       "18368                      gata3  \n",
       "18369                      gata3  \n",
       "\n",
       "[18370 rows x 30 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motifs_df = pd.read_csv(tead4_motifs_path, sep = '\\t')\n",
    "motifs_df.shape\n",
    "motifs_df   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get repeat masker annotation and remove ERVs from dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18370"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make grnages using pyrange\n",
    "import pyranges as pr\n",
    "\n",
    "motifs_df['row_idx'] = np.arange(len(motifs_df))\n",
    "motifs_df['Chromosome'] = motifs_df.example_chrom\n",
    "motifs_df['Start'] = motifs_df.pattern_start_abs\n",
    "motifs_df['End'] = motifs_df.pattern_end_abs\n",
    "motifs_df['Strand'] = motifs_df.strand\n",
    "\n",
    "cwm_scan_regions_pr = pr.PyRanges(motifs_df)\n",
    "len(cwm_scan_regions_pr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#functions\n",
    "def read_repeat_masker(file_path):\n",
    "    dfrm = pd.read_table(file_path, delim_whitespace=True, header=[1])\n",
    "\n",
    "    dfrm.columns = [x.replace(\"\\n\", \"_\") for x in dfrm.columns]\n",
    "\n",
    "    dfrm['name'] = dfrm['repeat'] + \"//\" + dfrm['class/family']\n",
    "\n",
    "    dfrm = dfrm[['ins.', 'sequence', 'begin', 'name']]\n",
    "    dfrm.columns = ['chrom', 'start', 'end', 'name']\n",
    "    return dfrm\n",
    "\n",
    "\n",
    "def intersect_repeat_masker(pattern_name, seqlets: BedTool, repeat_masker: BedTool, f=1.0):\n",
    "    \"\"\"Intersect the seqlets bed file with \n",
    "    \"\"\"\n",
    "    try:\n",
    "        dfint = seqlets.intersect(repeat_masker, wa=True, wb=True, f=f).to_dataframe()\n",
    "    except Exception:\n",
    "        return None\n",
    "    t = dfint.blockCount.str.split(\"//\", expand=True)\n",
    "    dfint['pattern_name'] = pattern_name\n",
    "    dfint['repeat_name'] = t[0]\n",
    "    dfint['repeat_family'] = t[1]\n",
    "    dfint['n_pattern'] = seqlets.to_dataframe()[['chrom', 'start', 'end']].drop_duplicates().shape[0]\n",
    "    dfint['interval'] = dfint['chrom'] + \":\" + dfint['start'].astype(str) + \"-\" + dfint['end'].astype(str)\n",
    "    return dfint[['chrom', 'start', 'end', 'interval', 'pattern_name', 'n_pattern', 'repeat_name', 'repeat_family']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False    15749\n",
      "True      2621\n",
      "Name: row_idx, dtype: int64\n",
      "Excluding: 2621 / 18370 rows\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False    15749\n",
       "Name: is_erv, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load repeat masker file and remove ervs\n",
    "rm_filepath = f'/n/projects/kd2200/publication/bpnet/analysis/data/repeatmasker.mm10.fa.out.gz'\n",
    "dfrm = read_repeat_masker(rm_filepath)\n",
    "dfrm_erv = dfrm[dfrm.name.str.contains(\"ERV\")]\n",
    "dfrm_erv['Start'] = dfrm_erv['start']\n",
    "dfrm_erv['End'] = dfrm_erv['end']\n",
    "dfrm_erv['Chromosome'] = dfrm_erv['chrom']\n",
    "dfrm_erv = pr.PyRanges(dfrm_erv)\n",
    "\n",
    "# exclude erv's\n",
    "exclude = cwm_scan_regions_pr.overlap(dfrm_erv).df.row_idx.unique()\n",
    "exclude_rows = motifs_df.row_idx.isin(exclude)\n",
    "print(exclude_rows.value_counts())\n",
    "print(f\"Excluding: {len(exclude)} / {len(motifs_df)} rows\")\n",
    "motifs_df['is_erv'] = exclude_rows #.astype(str)\n",
    "\n",
    "#check motif counts without ERVs\n",
    "dfi = motifs_df[(motifs_df['is_erv'].astype(str)=='False')]\n",
    "dfi.is_erv.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>example_chrom</th>\n",
       "      <th>pattern_start_abs</th>\n",
       "      <th>pattern_end_abs</th>\n",
       "      <th>pattern</th>\n",
       "      <th>contrib_weighted_p</th>\n",
       "      <th>strand</th>\n",
       "      <th>match_weighted_p</th>\n",
       "      <th>example_idx</th>\n",
       "      <th>pattern_start</th>\n",
       "      <th>pattern_end</th>\n",
       "      <th>pattern_center</th>\n",
       "      <th>pattern_len</th>\n",
       "      <th>match_weighted</th>\n",
       "      <th>match_weighted_cat</th>\n",
       "      <th>match_max</th>\n",
       "      <th>match_max_task</th>\n",
       "      <th>contrib_weighted</th>\n",
       "      <th>contrib_weighted_cat</th>\n",
       "      <th>contrib_max</th>\n",
       "      <th>contrib_max_task</th>\n",
       "      <th>seq_match</th>\n",
       "      <th>seq_match_p</th>\n",
       "      <th>seq_match_cat</th>\n",
       "      <th>match/tead4</th>\n",
       "      <th>contrib/tead4</th>\n",
       "      <th>pattern_short</th>\n",
       "      <th>example_start</th>\n",
       "      <th>example_end</th>\n",
       "      <th>example_strand</th>\n",
       "      <th>example_interval_from_task</th>\n",
       "      <th>row_idx</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>is_erv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30420800</td>\n",
       "      <td>30420817</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.430212</td>\n",
       "      <td>+</td>\n",
       "      <td>0.470039</td>\n",
       "      <td>8</td>\n",
       "      <td>654</td>\n",
       "      <td>671</td>\n",
       "      <td>663</td>\n",
       "      <td>17</td>\n",
       "      <td>0.496458</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.496458</td>\n",
       "      <td>tead4</td>\n",
       "      <td>3.093251</td>\n",
       "      <td>medium</td>\n",
       "      <td>3.093251</td>\n",
       "      <td>tead4</td>\n",
       "      <td>10.025424</td>\n",
       "      <td>0.383925</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.496458</td>\n",
       "      <td>3.093251</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>30420146</td>\n",
       "      <td>30421146</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30420800</td>\n",
       "      <td>30420818</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>185336727</td>\n",
       "      <td>185336744</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.479010</td>\n",
       "      <td>+</td>\n",
       "      <td>0.380337</td>\n",
       "      <td>23</td>\n",
       "      <td>525</td>\n",
       "      <td>542</td>\n",
       "      <td>534</td>\n",
       "      <td>17</td>\n",
       "      <td>0.480125</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.480125</td>\n",
       "      <td>tead4</td>\n",
       "      <td>3.350234</td>\n",
       "      <td>medium</td>\n",
       "      <td>3.350234</td>\n",
       "      <td>tead4</td>\n",
       "      <td>7.793720</td>\n",
       "      <td>0.100108</td>\n",
       "      <td>low</td>\n",
       "      <td>0.480125</td>\n",
       "      <td>3.350234</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>185336202</td>\n",
       "      <td>185337202</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1</td>\n",
       "      <td>185336727</td>\n",
       "      <td>185336745</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>170438850</td>\n",
       "      <td>170438867</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.827772</td>\n",
       "      <td>+</td>\n",
       "      <td>0.964478</td>\n",
       "      <td>55</td>\n",
       "      <td>495</td>\n",
       "      <td>512</td>\n",
       "      <td>504</td>\n",
       "      <td>17</td>\n",
       "      <td>0.613330</td>\n",
       "      <td>high</td>\n",
       "      <td>0.613330</td>\n",
       "      <td>tead4</td>\n",
       "      <td>4.918233</td>\n",
       "      <td>high</td>\n",
       "      <td>4.918233</td>\n",
       "      <td>tead4</td>\n",
       "      <td>12.900725</td>\n",
       "      <td>0.843201</td>\n",
       "      <td>high</td>\n",
       "      <td>0.613330</td>\n",
       "      <td>4.918233</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>170438355</td>\n",
       "      <td>170439355</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>8</td>\n",
       "      <td>chr1</td>\n",
       "      <td>170438850</td>\n",
       "      <td>170438868</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>51906459</td>\n",
       "      <td>51906476</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.027628</td>\n",
       "      <td>+</td>\n",
       "      <td>0.509508</td>\n",
       "      <td>114</td>\n",
       "      <td>693</td>\n",
       "      <td>710</td>\n",
       "      <td>702</td>\n",
       "      <td>17</td>\n",
       "      <td>0.503989</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.503989</td>\n",
       "      <td>tead4</td>\n",
       "      <td>1.008890</td>\n",
       "      <td>low</td>\n",
       "      <td>1.008890</td>\n",
       "      <td>tead4</td>\n",
       "      <td>8.346114</td>\n",
       "      <td>0.151417</td>\n",
       "      <td>low</td>\n",
       "      <td>0.503989</td>\n",
       "      <td>1.008890</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>51905766</td>\n",
       "      <td>51906766</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>14</td>\n",
       "      <td>chr1</td>\n",
       "      <td>51906459</td>\n",
       "      <td>51906477</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>152121957</td>\n",
       "      <td>152121974</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.023681</td>\n",
       "      <td>+</td>\n",
       "      <td>0.208109</td>\n",
       "      <td>119</td>\n",
       "      <td>43</td>\n",
       "      <td>60</td>\n",
       "      <td>52</td>\n",
       "      <td>17</td>\n",
       "      <td>0.436122</td>\n",
       "      <td>low</td>\n",
       "      <td>0.436122</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.985851</td>\n",
       "      <td>low</td>\n",
       "      <td>0.985851</td>\n",
       "      <td>tead4</td>\n",
       "      <td>3.619336</td>\n",
       "      <td>0.004306</td>\n",
       "      <td>low</td>\n",
       "      <td>0.436122</td>\n",
       "      <td>0.985851</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>152121914</td>\n",
       "      <td>152122914</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>15</td>\n",
       "      <td>chr1</td>\n",
       "      <td>152121957</td>\n",
       "      <td>152121975</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15744</th>\n",
       "      <td>chrX</td>\n",
       "      <td>152071148</td>\n",
       "      <td>152071165</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.075350</td>\n",
       "      <td>-</td>\n",
       "      <td>0.365626</td>\n",
       "      <td>187707</td>\n",
       "      <td>179</td>\n",
       "      <td>196</td>\n",
       "      <td>187</td>\n",
       "      <td>17</td>\n",
       "      <td>0.476687</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.476687</td>\n",
       "      <td>tead4</td>\n",
       "      <td>1.253692</td>\n",
       "      <td>low</td>\n",
       "      <td>1.253692</td>\n",
       "      <td>tead4</td>\n",
       "      <td>7.667224</td>\n",
       "      <td>0.095802</td>\n",
       "      <td>low</td>\n",
       "      <td>0.476687</td>\n",
       "      <td>1.253692</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>152070969</td>\n",
       "      <td>152071969</td>\n",
       "      <td>.</td>\n",
       "      <td>gata3</td>\n",
       "      <td>18362</td>\n",
       "      <td>chrX</td>\n",
       "      <td>152071147</td>\n",
       "      <td>152071165</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15745</th>\n",
       "      <td>chrX</td>\n",
       "      <td>12191672</td>\n",
       "      <td>12191689</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.242914</td>\n",
       "      <td>-</td>\n",
       "      <td>0.253678</td>\n",
       "      <td>187714</td>\n",
       "      <td>354</td>\n",
       "      <td>371</td>\n",
       "      <td>362</td>\n",
       "      <td>17</td>\n",
       "      <td>0.448724</td>\n",
       "      <td>low</td>\n",
       "      <td>0.448724</td>\n",
       "      <td>tead4</td>\n",
       "      <td>2.188614</td>\n",
       "      <td>low</td>\n",
       "      <td>2.188614</td>\n",
       "      <td>tead4</td>\n",
       "      <td>11.372528</td>\n",
       "      <td>0.614281</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.448724</td>\n",
       "      <td>2.188614</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>12191318</td>\n",
       "      <td>12192318</td>\n",
       "      <td>.</td>\n",
       "      <td>gata3</td>\n",
       "      <td>18363</td>\n",
       "      <td>chrX</td>\n",
       "      <td>12191671</td>\n",
       "      <td>12191689</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15746</th>\n",
       "      <td>chrX</td>\n",
       "      <td>36805292</td>\n",
       "      <td>36805309</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.519555</td>\n",
       "      <td>-</td>\n",
       "      <td>0.937567</td>\n",
       "      <td>187742</td>\n",
       "      <td>349</td>\n",
       "      <td>366</td>\n",
       "      <td>357</td>\n",
       "      <td>17</td>\n",
       "      <td>0.598039</td>\n",
       "      <td>high</td>\n",
       "      <td>0.598039</td>\n",
       "      <td>tead4</td>\n",
       "      <td>3.522194</td>\n",
       "      <td>medium</td>\n",
       "      <td>3.522194</td>\n",
       "      <td>tead4</td>\n",
       "      <td>12.923700</td>\n",
       "      <td>0.846430</td>\n",
       "      <td>high</td>\n",
       "      <td>0.598039</td>\n",
       "      <td>3.522194</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>36804943</td>\n",
       "      <td>36805943</td>\n",
       "      <td>.</td>\n",
       "      <td>gata3</td>\n",
       "      <td>18366</td>\n",
       "      <td>chrX</td>\n",
       "      <td>36805291</td>\n",
       "      <td>36805309</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15747</th>\n",
       "      <td>chrX</td>\n",
       "      <td>151148659</td>\n",
       "      <td>151148676</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.840689</td>\n",
       "      <td>-</td>\n",
       "      <td>0.793326</td>\n",
       "      <td>187767</td>\n",
       "      <td>134</td>\n",
       "      <td>151</td>\n",
       "      <td>142</td>\n",
       "      <td>17</td>\n",
       "      <td>0.556961</td>\n",
       "      <td>high</td>\n",
       "      <td>0.556961</td>\n",
       "      <td>tead4</td>\n",
       "      <td>4.993919</td>\n",
       "      <td>high</td>\n",
       "      <td>4.993919</td>\n",
       "      <td>tead4</td>\n",
       "      <td>12.246571</td>\n",
       "      <td>0.763186</td>\n",
       "      <td>high</td>\n",
       "      <td>0.556961</td>\n",
       "      <td>4.993919</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>151148525</td>\n",
       "      <td>151149525</td>\n",
       "      <td>.</td>\n",
       "      <td>gata3</td>\n",
       "      <td>18369</td>\n",
       "      <td>chrX</td>\n",
       "      <td>151148658</td>\n",
       "      <td>151148676</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15748</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90802665</td>\n",
       "      <td>90802682</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.090779</td>\n",
       "      <td>-</td>\n",
       "      <td>0.241837</td>\n",
       "      <td>107484</td>\n",
       "      <td>220</td>\n",
       "      <td>237</td>\n",
       "      <td>228</td>\n",
       "      <td>17</td>\n",
       "      <td>0.446307</td>\n",
       "      <td>low</td>\n",
       "      <td>0.446307</td>\n",
       "      <td>tead4</td>\n",
       "      <td>1.343004</td>\n",
       "      <td>low</td>\n",
       "      <td>1.343004</td>\n",
       "      <td>tead4</td>\n",
       "      <td>8.319051</td>\n",
       "      <td>0.143524</td>\n",
       "      <td>low</td>\n",
       "      <td>0.446307</td>\n",
       "      <td>1.343004</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>90802445</td>\n",
       "      <td>90803445</td>\n",
       "      <td>.</td>\n",
       "      <td>tfap2c</td>\n",
       "      <td>7661</td>\n",
       "      <td>chrY</td>\n",
       "      <td>90802664</td>\n",
       "      <td>90802682</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15749 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      example_chrom  pattern_start_abs  pattern_end_abs  \\\n",
       "0              chr1           30420800         30420817   \n",
       "1              chr1          185336727        185336744   \n",
       "2              chr1          170438850        170438867   \n",
       "3              chr1           51906459         51906476   \n",
       "4              chr1          152121957        152121974   \n",
       "...             ...                ...              ...   \n",
       "15744          chrX          152071148        152071165   \n",
       "15745          chrX           12191672         12191689   \n",
       "15746          chrX           36805292         36805309   \n",
       "15747          chrX          151148659        151148676   \n",
       "15748          chrY           90802665         90802682   \n",
       "\n",
       "                       pattern  contrib_weighted_p strand  match_weighted_p  \\\n",
       "0      metacluster_0/pattern_1            0.430212      +          0.470039   \n",
       "1      metacluster_0/pattern_1            0.479010      +          0.380337   \n",
       "2      metacluster_0/pattern_1            0.827772      +          0.964478   \n",
       "3      metacluster_0/pattern_1            0.027628      +          0.509508   \n",
       "4      metacluster_0/pattern_1            0.023681      +          0.208109   \n",
       "...                        ...                 ...    ...               ...   \n",
       "15744  metacluster_0/pattern_1            0.075350      -          0.365626   \n",
       "15745  metacluster_0/pattern_1            0.242914      -          0.253678   \n",
       "15746  metacluster_0/pattern_1            0.519555      -          0.937567   \n",
       "15747  metacluster_0/pattern_1            0.840689      -          0.793326   \n",
       "15748  metacluster_0/pattern_1            0.090779      -          0.241837   \n",
       "\n",
       "       example_idx  pattern_start  pattern_end  pattern_center  pattern_len  \\\n",
       "0                8            654          671             663           17   \n",
       "1               23            525          542             534           17   \n",
       "2               55            495          512             504           17   \n",
       "3              114            693          710             702           17   \n",
       "4              119             43           60              52           17   \n",
       "...            ...            ...          ...             ...          ...   \n",
       "15744       187707            179          196             187           17   \n",
       "15745       187714            354          371             362           17   \n",
       "15746       187742            349          366             357           17   \n",
       "15747       187767            134          151             142           17   \n",
       "15748       107484            220          237             228           17   \n",
       "\n",
       "       match_weighted match_weighted_cat  match_max match_max_task  \\\n",
       "0            0.496458             medium   0.496458          tead4   \n",
       "1            0.480125             medium   0.480125          tead4   \n",
       "2            0.613330               high   0.613330          tead4   \n",
       "3            0.503989             medium   0.503989          tead4   \n",
       "4            0.436122                low   0.436122          tead4   \n",
       "...               ...                ...        ...            ...   \n",
       "15744        0.476687             medium   0.476687          tead4   \n",
       "15745        0.448724                low   0.448724          tead4   \n",
       "15746        0.598039               high   0.598039          tead4   \n",
       "15747        0.556961               high   0.556961          tead4   \n",
       "15748        0.446307                low   0.446307          tead4   \n",
       "\n",
       "       contrib_weighted contrib_weighted_cat  contrib_max contrib_max_task  \\\n",
       "0              3.093251               medium     3.093251            tead4   \n",
       "1              3.350234               medium     3.350234            tead4   \n",
       "2              4.918233                 high     4.918233            tead4   \n",
       "3              1.008890                  low     1.008890            tead4   \n",
       "4              0.985851                  low     0.985851            tead4   \n",
       "...                 ...                  ...          ...              ...   \n",
       "15744          1.253692                  low     1.253692            tead4   \n",
       "15745          2.188614                  low     2.188614            tead4   \n",
       "15746          3.522194               medium     3.522194            tead4   \n",
       "15747          4.993919                 high     4.993919            tead4   \n",
       "15748          1.343004                  low     1.343004            tead4   \n",
       "\n",
       "       seq_match  seq_match_p seq_match_cat  match/tead4  contrib/tead4  \\\n",
       "0      10.025424     0.383925        medium     0.496458       3.093251   \n",
       "1       7.793720     0.100108           low     0.480125       3.350234   \n",
       "2      12.900725     0.843201          high     0.613330       4.918233   \n",
       "3       8.346114     0.151417           low     0.503989       1.008890   \n",
       "4       3.619336     0.004306           low     0.436122       0.985851   \n",
       "...          ...          ...           ...          ...            ...   \n",
       "15744   7.667224     0.095802           low     0.476687       1.253692   \n",
       "15745  11.372528     0.614281        medium     0.448724       2.188614   \n",
       "15746  12.923700     0.846430          high     0.598039       3.522194   \n",
       "15747  12.246571     0.763186          high     0.556961       4.993919   \n",
       "15748   8.319051     0.143524           low     0.446307       1.343004   \n",
       "\n",
       "      pattern_short  example_start  example_end example_strand  \\\n",
       "0             m0_p1       30420146     30421146              .   \n",
       "1             m0_p1      185336202    185337202              .   \n",
       "2             m0_p1      170438355    170439355              .   \n",
       "3             m0_p1       51905766     51906766              .   \n",
       "4             m0_p1      152121914    152122914              .   \n",
       "...             ...            ...          ...            ...   \n",
       "15744         m0_p1      152070969    152071969              .   \n",
       "15745         m0_p1       12191318     12192318              .   \n",
       "15746         m0_p1       36804943     36805943              .   \n",
       "15747         m0_p1      151148525    151149525              .   \n",
       "15748         m0_p1       90802445     90803445              .   \n",
       "\n",
       "      example_interval_from_task  row_idx Chromosome      Start        End  \\\n",
       "0                           cdx2        0       chr1   30420800   30420818   \n",
       "1                           cdx2        6       chr1  185336727  185336745   \n",
       "2                           cdx2        8       chr1  170438850  170438868   \n",
       "3                           cdx2       14       chr1   51906459   51906477   \n",
       "4                           cdx2       15       chr1  152121957  152121975   \n",
       "...                          ...      ...        ...        ...        ...   \n",
       "15744                      gata3    18362       chrX  152071147  152071165   \n",
       "15745                      gata3    18363       chrX   12191671   12191689   \n",
       "15746                      gata3    18366       chrX   36805291   36805309   \n",
       "15747                      gata3    18369       chrX  151148658  151148676   \n",
       "15748                     tfap2c     7661       chrY   90802664   90802682   \n",
       "\n",
       "      Strand  is_erv  \n",
       "0          +   False  \n",
       "1          +   False  \n",
       "2          +   False  \n",
       "3          +   False  \n",
       "4          +   False  \n",
       "...      ...     ...  \n",
       "15744      -   False  \n",
       "15745      -   False  \n",
       "15746      -   False  \n",
       "15747      -   False  \n",
       "15748      -   False  \n",
       "\n",
       "[15749 rows x 36 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# above returened dataframe without erv, I will make pyrnage to resize the motif to 18 bp from start and make df again for downstream analysis\n",
    "motif_pr = pr.PyRanges(dfi)\n",
    "motif_pr = motif_pr.extend({'3':1})\n",
    "x =motif_pr.df\n",
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>example_chrom</th>\n",
       "      <th>pattern_start_abs</th>\n",
       "      <th>pattern_end_abs</th>\n",
       "      <th>pattern</th>\n",
       "      <th>contrib_weighted_p</th>\n",
       "      <th>strand</th>\n",
       "      <th>match_weighted_p</th>\n",
       "      <th>example_idx</th>\n",
       "      <th>pattern_start</th>\n",
       "      <th>pattern_end</th>\n",
       "      <th>pattern_center</th>\n",
       "      <th>pattern_len</th>\n",
       "      <th>match_weighted</th>\n",
       "      <th>match_weighted_cat</th>\n",
       "      <th>match_max</th>\n",
       "      <th>match_max_task</th>\n",
       "      <th>contrib_weighted</th>\n",
       "      <th>contrib_weighted_cat</th>\n",
       "      <th>contrib_max</th>\n",
       "      <th>contrib_max_task</th>\n",
       "      <th>seq_match</th>\n",
       "      <th>seq_match_p</th>\n",
       "      <th>seq_match_cat</th>\n",
       "      <th>match/tead4</th>\n",
       "      <th>contrib/tead4</th>\n",
       "      <th>pattern_short</th>\n",
       "      <th>example_start</th>\n",
       "      <th>example_end</th>\n",
       "      <th>example_strand</th>\n",
       "      <th>example_interval_from_task</th>\n",
       "      <th>row_idx</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>is_erv</th>\n",
       "      <th>fwd_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30420800</td>\n",
       "      <td>30420817</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.430212</td>\n",
       "      <td>+</td>\n",
       "      <td>0.470039</td>\n",
       "      <td>8</td>\n",
       "      <td>654</td>\n",
       "      <td>671</td>\n",
       "      <td>663</td>\n",
       "      <td>17</td>\n",
       "      <td>0.496458</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.496458</td>\n",
       "      <td>tead4</td>\n",
       "      <td>3.093251</td>\n",
       "      <td>medium</td>\n",
       "      <td>3.093251</td>\n",
       "      <td>tead4</td>\n",
       "      <td>10.025424</td>\n",
       "      <td>0.383925</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.496458</td>\n",
       "      <td>3.093251</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>30420146</td>\n",
       "      <td>30421146</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30420800</td>\n",
       "      <td>30420818</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>TAAAATGCCTGGCATTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>185336727</td>\n",
       "      <td>185336744</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.479010</td>\n",
       "      <td>+</td>\n",
       "      <td>0.380337</td>\n",
       "      <td>23</td>\n",
       "      <td>525</td>\n",
       "      <td>542</td>\n",
       "      <td>534</td>\n",
       "      <td>17</td>\n",
       "      <td>0.480125</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.480125</td>\n",
       "      <td>tead4</td>\n",
       "      <td>3.350234</td>\n",
       "      <td>medium</td>\n",
       "      <td>3.350234</td>\n",
       "      <td>tead4</td>\n",
       "      <td>7.793720</td>\n",
       "      <td>0.100108</td>\n",
       "      <td>low</td>\n",
       "      <td>0.480125</td>\n",
       "      <td>3.350234</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>185336202</td>\n",
       "      <td>185337202</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1</td>\n",
       "      <td>185336727</td>\n",
       "      <td>185336745</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>AGGCATTCACTGAATGTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>170438850</td>\n",
       "      <td>170438867</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.827772</td>\n",
       "      <td>+</td>\n",
       "      <td>0.964478</td>\n",
       "      <td>55</td>\n",
       "      <td>495</td>\n",
       "      <td>512</td>\n",
       "      <td>504</td>\n",
       "      <td>17</td>\n",
       "      <td>0.613330</td>\n",
       "      <td>high</td>\n",
       "      <td>0.613330</td>\n",
       "      <td>tead4</td>\n",
       "      <td>4.918233</td>\n",
       "      <td>high</td>\n",
       "      <td>4.918233</td>\n",
       "      <td>tead4</td>\n",
       "      <td>12.900725</td>\n",
       "      <td>0.843201</td>\n",
       "      <td>high</td>\n",
       "      <td>0.613330</td>\n",
       "      <td>4.918233</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>170438355</td>\n",
       "      <td>170439355</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>8</td>\n",
       "      <td>chr1</td>\n",
       "      <td>170438850</td>\n",
       "      <td>170438868</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>TGGAATGCCAGGAATTGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>51906459</td>\n",
       "      <td>51906476</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.027628</td>\n",
       "      <td>+</td>\n",
       "      <td>0.509508</td>\n",
       "      <td>114</td>\n",
       "      <td>693</td>\n",
       "      <td>710</td>\n",
       "      <td>702</td>\n",
       "      <td>17</td>\n",
       "      <td>0.503989</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.503989</td>\n",
       "      <td>tead4</td>\n",
       "      <td>1.008890</td>\n",
       "      <td>low</td>\n",
       "      <td>1.008890</td>\n",
       "      <td>tead4</td>\n",
       "      <td>8.346114</td>\n",
       "      <td>0.151417</td>\n",
       "      <td>low</td>\n",
       "      <td>0.503989</td>\n",
       "      <td>1.008890</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>51905766</td>\n",
       "      <td>51906766</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>14</td>\n",
       "      <td>chr1</td>\n",
       "      <td>51906459</td>\n",
       "      <td>51906477</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>GAGAATGCTATGAATCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>152121957</td>\n",
       "      <td>152121974</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.023681</td>\n",
       "      <td>+</td>\n",
       "      <td>0.208109</td>\n",
       "      <td>119</td>\n",
       "      <td>43</td>\n",
       "      <td>60</td>\n",
       "      <td>52</td>\n",
       "      <td>17</td>\n",
       "      <td>0.436122</td>\n",
       "      <td>low</td>\n",
       "      <td>0.436122</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.985851</td>\n",
       "      <td>low</td>\n",
       "      <td>0.985851</td>\n",
       "      <td>tead4</td>\n",
       "      <td>3.619336</td>\n",
       "      <td>0.004306</td>\n",
       "      <td>low</td>\n",
       "      <td>0.436122</td>\n",
       "      <td>0.985851</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>152121914</td>\n",
       "      <td>152122914</td>\n",
       "      <td>.</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>15</td>\n",
       "      <td>chr1</td>\n",
       "      <td>152121957</td>\n",
       "      <td>152121975</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>GTAAATACCTGGAATTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15744</th>\n",
       "      <td>chrX</td>\n",
       "      <td>152071148</td>\n",
       "      <td>152071165</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.075350</td>\n",
       "      <td>-</td>\n",
       "      <td>0.365626</td>\n",
       "      <td>187707</td>\n",
       "      <td>179</td>\n",
       "      <td>196</td>\n",
       "      <td>187</td>\n",
       "      <td>17</td>\n",
       "      <td>0.476687</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.476687</td>\n",
       "      <td>tead4</td>\n",
       "      <td>1.253692</td>\n",
       "      <td>low</td>\n",
       "      <td>1.253692</td>\n",
       "      <td>tead4</td>\n",
       "      <td>7.667224</td>\n",
       "      <td>0.095802</td>\n",
       "      <td>low</td>\n",
       "      <td>0.476687</td>\n",
       "      <td>1.253692</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>152070969</td>\n",
       "      <td>152071969</td>\n",
       "      <td>.</td>\n",
       "      <td>gata3</td>\n",
       "      <td>18362</td>\n",
       "      <td>chrX</td>\n",
       "      <td>152071147</td>\n",
       "      <td>152071165</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>TGGATTTCAGGGATTGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15745</th>\n",
       "      <td>chrX</td>\n",
       "      <td>12191672</td>\n",
       "      <td>12191689</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.242914</td>\n",
       "      <td>-</td>\n",
       "      <td>0.253678</td>\n",
       "      <td>187714</td>\n",
       "      <td>354</td>\n",
       "      <td>371</td>\n",
       "      <td>362</td>\n",
       "      <td>17</td>\n",
       "      <td>0.448724</td>\n",
       "      <td>low</td>\n",
       "      <td>0.448724</td>\n",
       "      <td>tead4</td>\n",
       "      <td>2.188614</td>\n",
       "      <td>low</td>\n",
       "      <td>2.188614</td>\n",
       "      <td>tead4</td>\n",
       "      <td>11.372528</td>\n",
       "      <td>0.614281</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.448724</td>\n",
       "      <td>2.188614</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>12191318</td>\n",
       "      <td>12192318</td>\n",
       "      <td>.</td>\n",
       "      <td>gata3</td>\n",
       "      <td>18363</td>\n",
       "      <td>chrX</td>\n",
       "      <td>12191671</td>\n",
       "      <td>12191689</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>TTGATTCCTGGAATACCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15746</th>\n",
       "      <td>chrX</td>\n",
       "      <td>36805292</td>\n",
       "      <td>36805309</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.519555</td>\n",
       "      <td>-</td>\n",
       "      <td>0.937567</td>\n",
       "      <td>187742</td>\n",
       "      <td>349</td>\n",
       "      <td>366</td>\n",
       "      <td>357</td>\n",
       "      <td>17</td>\n",
       "      <td>0.598039</td>\n",
       "      <td>high</td>\n",
       "      <td>0.598039</td>\n",
       "      <td>tead4</td>\n",
       "      <td>3.522194</td>\n",
       "      <td>medium</td>\n",
       "      <td>3.522194</td>\n",
       "      <td>tead4</td>\n",
       "      <td>12.923700</td>\n",
       "      <td>0.846430</td>\n",
       "      <td>high</td>\n",
       "      <td>0.598039</td>\n",
       "      <td>3.522194</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>36804943</td>\n",
       "      <td>36805943</td>\n",
       "      <td>.</td>\n",
       "      <td>gata3</td>\n",
       "      <td>18366</td>\n",
       "      <td>chrX</td>\n",
       "      <td>36805291</td>\n",
       "      <td>36805309</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>TACATACCAGGGATTCCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15747</th>\n",
       "      <td>chrX</td>\n",
       "      <td>151148659</td>\n",
       "      <td>151148676</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.840689</td>\n",
       "      <td>-</td>\n",
       "      <td>0.793326</td>\n",
       "      <td>187767</td>\n",
       "      <td>134</td>\n",
       "      <td>151</td>\n",
       "      <td>142</td>\n",
       "      <td>17</td>\n",
       "      <td>0.556961</td>\n",
       "      <td>high</td>\n",
       "      <td>0.556961</td>\n",
       "      <td>tead4</td>\n",
       "      <td>4.993919</td>\n",
       "      <td>high</td>\n",
       "      <td>4.993919</td>\n",
       "      <td>tead4</td>\n",
       "      <td>12.246571</td>\n",
       "      <td>0.763186</td>\n",
       "      <td>high</td>\n",
       "      <td>0.556961</td>\n",
       "      <td>4.993919</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>151148525</td>\n",
       "      <td>151149525</td>\n",
       "      <td>.</td>\n",
       "      <td>gata3</td>\n",
       "      <td>18369</td>\n",
       "      <td>chrX</td>\n",
       "      <td>151148658</td>\n",
       "      <td>151148676</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>CAAATACCTGTAATTCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15748</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90802665</td>\n",
       "      <td>90802682</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>0.090779</td>\n",
       "      <td>-</td>\n",
       "      <td>0.241837</td>\n",
       "      <td>107484</td>\n",
       "      <td>220</td>\n",
       "      <td>237</td>\n",
       "      <td>228</td>\n",
       "      <td>17</td>\n",
       "      <td>0.446307</td>\n",
       "      <td>low</td>\n",
       "      <td>0.446307</td>\n",
       "      <td>tead4</td>\n",
       "      <td>1.343004</td>\n",
       "      <td>low</td>\n",
       "      <td>1.343004</td>\n",
       "      <td>tead4</td>\n",
       "      <td>8.319051</td>\n",
       "      <td>0.143524</td>\n",
       "      <td>low</td>\n",
       "      <td>0.446307</td>\n",
       "      <td>1.343004</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>90802445</td>\n",
       "      <td>90803445</td>\n",
       "      <td>.</td>\n",
       "      <td>tfap2c</td>\n",
       "      <td>7661</td>\n",
       "      <td>chrY</td>\n",
       "      <td>90802664</td>\n",
       "      <td>90802682</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>TTAATTCCTTGAATAGTT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15749 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      example_chrom  pattern_start_abs  pattern_end_abs  \\\n",
       "0              chr1           30420800         30420817   \n",
       "1              chr1          185336727        185336744   \n",
       "2              chr1          170438850        170438867   \n",
       "3              chr1           51906459         51906476   \n",
       "4              chr1          152121957        152121974   \n",
       "...             ...                ...              ...   \n",
       "15744          chrX          152071148        152071165   \n",
       "15745          chrX           12191672         12191689   \n",
       "15746          chrX           36805292         36805309   \n",
       "15747          chrX          151148659        151148676   \n",
       "15748          chrY           90802665         90802682   \n",
       "\n",
       "                       pattern  contrib_weighted_p strand  match_weighted_p  \\\n",
       "0      metacluster_0/pattern_1            0.430212      +          0.470039   \n",
       "1      metacluster_0/pattern_1            0.479010      +          0.380337   \n",
       "2      metacluster_0/pattern_1            0.827772      +          0.964478   \n",
       "3      metacluster_0/pattern_1            0.027628      +          0.509508   \n",
       "4      metacluster_0/pattern_1            0.023681      +          0.208109   \n",
       "...                        ...                 ...    ...               ...   \n",
       "15744  metacluster_0/pattern_1            0.075350      -          0.365626   \n",
       "15745  metacluster_0/pattern_1            0.242914      -          0.253678   \n",
       "15746  metacluster_0/pattern_1            0.519555      -          0.937567   \n",
       "15747  metacluster_0/pattern_1            0.840689      -          0.793326   \n",
       "15748  metacluster_0/pattern_1            0.090779      -          0.241837   \n",
       "\n",
       "       example_idx  pattern_start  pattern_end  pattern_center  pattern_len  \\\n",
       "0                8            654          671             663           17   \n",
       "1               23            525          542             534           17   \n",
       "2               55            495          512             504           17   \n",
       "3              114            693          710             702           17   \n",
       "4              119             43           60              52           17   \n",
       "...            ...            ...          ...             ...          ...   \n",
       "15744       187707            179          196             187           17   \n",
       "15745       187714            354          371             362           17   \n",
       "15746       187742            349          366             357           17   \n",
       "15747       187767            134          151             142           17   \n",
       "15748       107484            220          237             228           17   \n",
       "\n",
       "       match_weighted match_weighted_cat  match_max match_max_task  \\\n",
       "0            0.496458             medium   0.496458          tead4   \n",
       "1            0.480125             medium   0.480125          tead4   \n",
       "2            0.613330               high   0.613330          tead4   \n",
       "3            0.503989             medium   0.503989          tead4   \n",
       "4            0.436122                low   0.436122          tead4   \n",
       "...               ...                ...        ...            ...   \n",
       "15744        0.476687             medium   0.476687          tead4   \n",
       "15745        0.448724                low   0.448724          tead4   \n",
       "15746        0.598039               high   0.598039          tead4   \n",
       "15747        0.556961               high   0.556961          tead4   \n",
       "15748        0.446307                low   0.446307          tead4   \n",
       "\n",
       "       contrib_weighted contrib_weighted_cat  contrib_max contrib_max_task  \\\n",
       "0              3.093251               medium     3.093251            tead4   \n",
       "1              3.350234               medium     3.350234            tead4   \n",
       "2              4.918233                 high     4.918233            tead4   \n",
       "3              1.008890                  low     1.008890            tead4   \n",
       "4              0.985851                  low     0.985851            tead4   \n",
       "...                 ...                  ...          ...              ...   \n",
       "15744          1.253692                  low     1.253692            tead4   \n",
       "15745          2.188614                  low     2.188614            tead4   \n",
       "15746          3.522194               medium     3.522194            tead4   \n",
       "15747          4.993919                 high     4.993919            tead4   \n",
       "15748          1.343004                  low     1.343004            tead4   \n",
       "\n",
       "       seq_match  seq_match_p seq_match_cat  match/tead4  contrib/tead4  \\\n",
       "0      10.025424     0.383925        medium     0.496458       3.093251   \n",
       "1       7.793720     0.100108           low     0.480125       3.350234   \n",
       "2      12.900725     0.843201          high     0.613330       4.918233   \n",
       "3       8.346114     0.151417           low     0.503989       1.008890   \n",
       "4       3.619336     0.004306           low     0.436122       0.985851   \n",
       "...          ...          ...           ...          ...            ...   \n",
       "15744   7.667224     0.095802           low     0.476687       1.253692   \n",
       "15745  11.372528     0.614281        medium     0.448724       2.188614   \n",
       "15746  12.923700     0.846430          high     0.598039       3.522194   \n",
       "15747  12.246571     0.763186          high     0.556961       4.993919   \n",
       "15748   8.319051     0.143524           low     0.446307       1.343004   \n",
       "\n",
       "      pattern_short  example_start  example_end example_strand  \\\n",
       "0             m0_p1       30420146     30421146              .   \n",
       "1             m0_p1      185336202    185337202              .   \n",
       "2             m0_p1      170438355    170439355              .   \n",
       "3             m0_p1       51905766     51906766              .   \n",
       "4             m0_p1      152121914    152122914              .   \n",
       "...             ...            ...          ...            ...   \n",
       "15744         m0_p1      152070969    152071969              .   \n",
       "15745         m0_p1       12191318     12192318              .   \n",
       "15746         m0_p1       36804943     36805943              .   \n",
       "15747         m0_p1      151148525    151149525              .   \n",
       "15748         m0_p1       90802445     90803445              .   \n",
       "\n",
       "      example_interval_from_task  row_idx Chromosome      Start        End  \\\n",
       "0                           cdx2        0       chr1   30420800   30420818   \n",
       "1                           cdx2        6       chr1  185336727  185336745   \n",
       "2                           cdx2        8       chr1  170438850  170438868   \n",
       "3                           cdx2       14       chr1   51906459   51906477   \n",
       "4                           cdx2       15       chr1  152121957  152121975   \n",
       "...                          ...      ...        ...        ...        ...   \n",
       "15744                      gata3    18362       chrX  152071147  152071165   \n",
       "15745                      gata3    18363       chrX   12191671   12191689   \n",
       "15746                      gata3    18366       chrX   36805291   36805309   \n",
       "15747                      gata3    18369       chrX  151148658  151148676   \n",
       "15748                     tfap2c     7661       chrY   90802664   90802682   \n",
       "\n",
       "      Strand  is_erv             fwd_seq  \n",
       "0          +   False  TAAAATGCCTGGCATTCT  \n",
       "1          +   False  AGGCATTCACTGAATGTA  \n",
       "2          +   False  TGGAATGCCAGGAATTGC  \n",
       "3          +   False  GAGAATGCTATGAATCCA  \n",
       "4          +   False  GTAAATACCTGGAATTAT  \n",
       "...      ...     ...                 ...  \n",
       "15744      -   False  TGGATTTCAGGGATTGCA  \n",
       "15745      -   False  TTGATTCCTGGAATACCA  \n",
       "15746      -   False  TACATACCAGGGATTCCT  \n",
       "15747      -   False  CAAATACCTGTAATTCCA  \n",
       "15748      -   False  TTAATTCCTTGAATAGTT  \n",
       "\n",
       "[15749 rows x 37 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Get sequences on the forward strand\n",
    "x['fwd_seq'] = extract_seqs_from_df(coords_df = x, fasta_path = fasta_file, chrom_column = 'Chromosome', start_column = 'Start', end_column = 'End')\n",
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "15749it [04:22, 59.99it/s]\n"
     ]
    }
   ],
   "source": [
    "# Reorient the negative strand sequences\n",
    "motifs_oriented_df = pd.DataFrame()\n",
    "for i,row in tqdm(x.iterrows()):\n",
    "    if row.strand=='-':\n",
    "        fwd_seq = Seq(row.fwd_seq)\n",
    "        rev_seq = str(fwd_seq.reverse_complement())\n",
    "        row['seq'] = rev_seq\n",
    "    else:\n",
    "        row['seq'] = row.fwd_seq\n",
    "    motifs_oriented_df = motifs_oriented_df.append(row)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract first half site and second half site groups."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "first_half_site = (0,9) \n",
    "second_half_site = (9,19) \n",
    "\n",
    "motifs_oriented_df['first_half_site_seq'] = [s[first_half_site[0]:first_half_site[1]] for s in motifs_oriented_df.seq.values]\n",
    "motifs_oriented_df['second_half_site_seq'] = [s[second_half_site[0]:second_half_site[1]] for s in motifs_oriented_df.seq.values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>End</th>\n",
       "      <th>Start</th>\n",
       "      <th>Strand</th>\n",
       "      <th>contrib/tead4</th>\n",
       "      <th>contrib_max</th>\n",
       "      <th>contrib_max_task</th>\n",
       "      <th>contrib_weighted</th>\n",
       "      <th>contrib_weighted_cat</th>\n",
       "      <th>contrib_weighted_p</th>\n",
       "      <th>example_chrom</th>\n",
       "      <th>example_end</th>\n",
       "      <th>example_idx</th>\n",
       "      <th>example_interval_from_task</th>\n",
       "      <th>example_start</th>\n",
       "      <th>example_strand</th>\n",
       "      <th>fwd_seq</th>\n",
       "      <th>is_erv</th>\n",
       "      <th>match/tead4</th>\n",
       "      <th>match_max</th>\n",
       "      <th>match_max_task</th>\n",
       "      <th>match_weighted</th>\n",
       "      <th>match_weighted_cat</th>\n",
       "      <th>match_weighted_p</th>\n",
       "      <th>pattern</th>\n",
       "      <th>pattern_center</th>\n",
       "      <th>pattern_end</th>\n",
       "      <th>pattern_end_abs</th>\n",
       "      <th>pattern_len</th>\n",
       "      <th>pattern_short</th>\n",
       "      <th>pattern_start</th>\n",
       "      <th>pattern_start_abs</th>\n",
       "      <th>row_idx</th>\n",
       "      <th>seq</th>\n",
       "      <th>seq_match</th>\n",
       "      <th>seq_match_cat</th>\n",
       "      <th>seq_match_p</th>\n",
       "      <th>strand</th>\n",
       "      <th>first_half_site_seq</th>\n",
       "      <th>second_half_site_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30420818.0</td>\n",
       "      <td>30420800.0</td>\n",
       "      <td>+</td>\n",
       "      <td>3.093251</td>\n",
       "      <td>3.093251</td>\n",
       "      <td>tead4</td>\n",
       "      <td>3.093251</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.430212</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30421146.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>30420146.0</td>\n",
       "      <td>.</td>\n",
       "      <td>TAAAATGCCTGGCATTCT</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.496458</td>\n",
       "      <td>0.496458</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.496458</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.470039</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>663.0</td>\n",
       "      <td>671.0</td>\n",
       "      <td>30420817.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>654.0</td>\n",
       "      <td>30420800.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TAAAATGCCTGGCATTCT</td>\n",
       "      <td>10.025424</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.383925</td>\n",
       "      <td>+</td>\n",
       "      <td>TAAAATGCC</td>\n",
       "      <td>TGGCATTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>185336745.0</td>\n",
       "      <td>185336727.0</td>\n",
       "      <td>+</td>\n",
       "      <td>3.350234</td>\n",
       "      <td>3.350234</td>\n",
       "      <td>tead4</td>\n",
       "      <td>3.350234</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.479010</td>\n",
       "      <td>chr1</td>\n",
       "      <td>185337202.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>185336202.0</td>\n",
       "      <td>.</td>\n",
       "      <td>AGGCATTCACTGAATGTA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.480125</td>\n",
       "      <td>0.480125</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.480125</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.380337</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>534.0</td>\n",
       "      <td>542.0</td>\n",
       "      <td>185336744.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>525.0</td>\n",
       "      <td>185336727.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>AGGCATTCACTGAATGTA</td>\n",
       "      <td>7.793720</td>\n",
       "      <td>low</td>\n",
       "      <td>0.100108</td>\n",
       "      <td>+</td>\n",
       "      <td>AGGCATTCA</td>\n",
       "      <td>CTGAATGTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>170438868.0</td>\n",
       "      <td>170438850.0</td>\n",
       "      <td>+</td>\n",
       "      <td>4.918233</td>\n",
       "      <td>4.918233</td>\n",
       "      <td>tead4</td>\n",
       "      <td>4.918233</td>\n",
       "      <td>high</td>\n",
       "      <td>0.827772</td>\n",
       "      <td>chr1</td>\n",
       "      <td>170439355.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>170438355.0</td>\n",
       "      <td>.</td>\n",
       "      <td>TGGAATGCCAGGAATTGC</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.613330</td>\n",
       "      <td>0.613330</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.613330</td>\n",
       "      <td>high</td>\n",
       "      <td>0.964478</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>504.0</td>\n",
       "      <td>512.0</td>\n",
       "      <td>170438867.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>495.0</td>\n",
       "      <td>170438850.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>TGGAATGCCAGGAATTGC</td>\n",
       "      <td>12.900725</td>\n",
       "      <td>high</td>\n",
       "      <td>0.843201</td>\n",
       "      <td>+</td>\n",
       "      <td>TGGAATGCC</td>\n",
       "      <td>AGGAATTGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>51906477.0</td>\n",
       "      <td>51906459.0</td>\n",
       "      <td>+</td>\n",
       "      <td>1.008890</td>\n",
       "      <td>1.008890</td>\n",
       "      <td>tead4</td>\n",
       "      <td>1.008890</td>\n",
       "      <td>low</td>\n",
       "      <td>0.027628</td>\n",
       "      <td>chr1</td>\n",
       "      <td>51906766.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>51905766.0</td>\n",
       "      <td>.</td>\n",
       "      <td>GAGAATGCTATGAATCCA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.503989</td>\n",
       "      <td>0.503989</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.503989</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.509508</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>702.0</td>\n",
       "      <td>710.0</td>\n",
       "      <td>51906476.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>693.0</td>\n",
       "      <td>51906459.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>GAGAATGCTATGAATCCA</td>\n",
       "      <td>8.346114</td>\n",
       "      <td>low</td>\n",
       "      <td>0.151417</td>\n",
       "      <td>+</td>\n",
       "      <td>GAGAATGCT</td>\n",
       "      <td>ATGAATCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>152121975.0</td>\n",
       "      <td>152121957.0</td>\n",
       "      <td>+</td>\n",
       "      <td>0.985851</td>\n",
       "      <td>0.985851</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.985851</td>\n",
       "      <td>low</td>\n",
       "      <td>0.023681</td>\n",
       "      <td>chr1</td>\n",
       "      <td>152122914.0</td>\n",
       "      <td>119.0</td>\n",
       "      <td>cdx2</td>\n",
       "      <td>152121914.0</td>\n",
       "      <td>.</td>\n",
       "      <td>GTAAATACCTGGAATTAT</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.436122</td>\n",
       "      <td>0.436122</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.436122</td>\n",
       "      <td>low</td>\n",
       "      <td>0.208109</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>52.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>152121974.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>43.0</td>\n",
       "      <td>152121957.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>GTAAATACCTGGAATTAT</td>\n",
       "      <td>3.619336</td>\n",
       "      <td>low</td>\n",
       "      <td>0.004306</td>\n",
       "      <td>+</td>\n",
       "      <td>GTAAATACC</td>\n",
       "      <td>TGGAATTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15744</th>\n",
       "      <td>chrX</td>\n",
       "      <td>152071165.0</td>\n",
       "      <td>152071147.0</td>\n",
       "      <td>-</td>\n",
       "      <td>1.253692</td>\n",
       "      <td>1.253692</td>\n",
       "      <td>tead4</td>\n",
       "      <td>1.253692</td>\n",
       "      <td>low</td>\n",
       "      <td>0.075350</td>\n",
       "      <td>chrX</td>\n",
       "      <td>152071969.0</td>\n",
       "      <td>187707.0</td>\n",
       "      <td>gata3</td>\n",
       "      <td>152070969.0</td>\n",
       "      <td>.</td>\n",
       "      <td>TGGATTTCAGGGATTGCA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.476687</td>\n",
       "      <td>0.476687</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.476687</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.365626</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>187.0</td>\n",
       "      <td>196.0</td>\n",
       "      <td>152071165.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>179.0</td>\n",
       "      <td>152071148.0</td>\n",
       "      <td>18362.0</td>\n",
       "      <td>TGCAATCCCTGAAATCCA</td>\n",
       "      <td>7.667224</td>\n",
       "      <td>low</td>\n",
       "      <td>0.095802</td>\n",
       "      <td>-</td>\n",
       "      <td>TGCAATCCC</td>\n",
       "      <td>TGAAATCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15745</th>\n",
       "      <td>chrX</td>\n",
       "      <td>12191689.0</td>\n",
       "      <td>12191671.0</td>\n",
       "      <td>-</td>\n",
       "      <td>2.188614</td>\n",
       "      <td>2.188614</td>\n",
       "      <td>tead4</td>\n",
       "      <td>2.188614</td>\n",
       "      <td>low</td>\n",
       "      <td>0.242914</td>\n",
       "      <td>chrX</td>\n",
       "      <td>12192318.0</td>\n",
       "      <td>187714.0</td>\n",
       "      <td>gata3</td>\n",
       "      <td>12191318.0</td>\n",
       "      <td>.</td>\n",
       "      <td>TTGATTCCTGGAATACCA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.448724</td>\n",
       "      <td>0.448724</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.448724</td>\n",
       "      <td>low</td>\n",
       "      <td>0.253678</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>362.0</td>\n",
       "      <td>371.0</td>\n",
       "      <td>12191689.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>354.0</td>\n",
       "      <td>12191672.0</td>\n",
       "      <td>18363.0</td>\n",
       "      <td>TGGTATTCCAGGAATCAA</td>\n",
       "      <td>11.372528</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.614281</td>\n",
       "      <td>-</td>\n",
       "      <td>TGGTATTCC</td>\n",
       "      <td>AGGAATCAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15746</th>\n",
       "      <td>chrX</td>\n",
       "      <td>36805309.0</td>\n",
       "      <td>36805291.0</td>\n",
       "      <td>-</td>\n",
       "      <td>3.522194</td>\n",
       "      <td>3.522194</td>\n",
       "      <td>tead4</td>\n",
       "      <td>3.522194</td>\n",
       "      <td>medium</td>\n",
       "      <td>0.519555</td>\n",
       "      <td>chrX</td>\n",
       "      <td>36805943.0</td>\n",
       "      <td>187742.0</td>\n",
       "      <td>gata3</td>\n",
       "      <td>36804943.0</td>\n",
       "      <td>.</td>\n",
       "      <td>TACATACCAGGGATTCCT</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.598039</td>\n",
       "      <td>0.598039</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.598039</td>\n",
       "      <td>high</td>\n",
       "      <td>0.937567</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>357.0</td>\n",
       "      <td>366.0</td>\n",
       "      <td>36805309.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>349.0</td>\n",
       "      <td>36805292.0</td>\n",
       "      <td>18366.0</td>\n",
       "      <td>AGGAATCCCTGGTATGTA</td>\n",
       "      <td>12.923700</td>\n",
       "      <td>high</td>\n",
       "      <td>0.846430</td>\n",
       "      <td>-</td>\n",
       "      <td>AGGAATCCC</td>\n",
       "      <td>TGGTATGTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15747</th>\n",
       "      <td>chrX</td>\n",
       "      <td>151148676.0</td>\n",
       "      <td>151148658.0</td>\n",
       "      <td>-</td>\n",
       "      <td>4.993919</td>\n",
       "      <td>4.993919</td>\n",
       "      <td>tead4</td>\n",
       "      <td>4.993919</td>\n",
       "      <td>high</td>\n",
       "      <td>0.840689</td>\n",
       "      <td>chrX</td>\n",
       "      <td>151149525.0</td>\n",
       "      <td>187767.0</td>\n",
       "      <td>gata3</td>\n",
       "      <td>151148525.0</td>\n",
       "      <td>.</td>\n",
       "      <td>CAAATACCTGTAATTCCA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.556961</td>\n",
       "      <td>0.556961</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.556961</td>\n",
       "      <td>high</td>\n",
       "      <td>0.793326</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>142.0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>151148676.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>134.0</td>\n",
       "      <td>151148659.0</td>\n",
       "      <td>18369.0</td>\n",
       "      <td>TGGAATTACAGGTATTTG</td>\n",
       "      <td>12.246571</td>\n",
       "      <td>high</td>\n",
       "      <td>0.763186</td>\n",
       "      <td>-</td>\n",
       "      <td>TGGAATTAC</td>\n",
       "      <td>AGGTATTTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15748</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90802682.0</td>\n",
       "      <td>90802664.0</td>\n",
       "      <td>-</td>\n",
       "      <td>1.343004</td>\n",
       "      <td>1.343004</td>\n",
       "      <td>tead4</td>\n",
       "      <td>1.343004</td>\n",
       "      <td>low</td>\n",
       "      <td>0.090779</td>\n",
       "      <td>chrY</td>\n",
       "      <td>90803445.0</td>\n",
       "      <td>107484.0</td>\n",
       "      <td>tfap2c</td>\n",
       "      <td>90802445.0</td>\n",
       "      <td>.</td>\n",
       "      <td>TTAATTCCTTGAATAGTT</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.446307</td>\n",
       "      <td>0.446307</td>\n",
       "      <td>tead4</td>\n",
       "      <td>0.446307</td>\n",
       "      <td>low</td>\n",
       "      <td>0.241837</td>\n",
       "      <td>metacluster_0/pattern_1</td>\n",
       "      <td>228.0</td>\n",
       "      <td>237.0</td>\n",
       "      <td>90802682.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>m0_p1</td>\n",
       "      <td>220.0</td>\n",
       "      <td>90802665.0</td>\n",
       "      <td>7661.0</td>\n",
       "      <td>AACTATTCAAGGAATTAA</td>\n",
       "      <td>8.319051</td>\n",
       "      <td>low</td>\n",
       "      <td>0.143524</td>\n",
       "      <td>-</td>\n",
       "      <td>AACTATTCA</td>\n",
       "      <td>AGGAATTAA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15749 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome          End        Start Strand  contrib/tead4  contrib_max  \\\n",
       "0           chr1   30420818.0   30420800.0      +       3.093251     3.093251   \n",
       "1           chr1  185336745.0  185336727.0      +       3.350234     3.350234   \n",
       "2           chr1  170438868.0  170438850.0      +       4.918233     4.918233   \n",
       "3           chr1   51906477.0   51906459.0      +       1.008890     1.008890   \n",
       "4           chr1  152121975.0  152121957.0      +       0.985851     0.985851   \n",
       "...          ...          ...          ...    ...            ...          ...   \n",
       "15744       chrX  152071165.0  152071147.0      -       1.253692     1.253692   \n",
       "15745       chrX   12191689.0   12191671.0      -       2.188614     2.188614   \n",
       "15746       chrX   36805309.0   36805291.0      -       3.522194     3.522194   \n",
       "15747       chrX  151148676.0  151148658.0      -       4.993919     4.993919   \n",
       "15748       chrY   90802682.0   90802664.0      -       1.343004     1.343004   \n",
       "\n",
       "      contrib_max_task  contrib_weighted contrib_weighted_cat  \\\n",
       "0                tead4          3.093251               medium   \n",
       "1                tead4          3.350234               medium   \n",
       "2                tead4          4.918233                 high   \n",
       "3                tead4          1.008890                  low   \n",
       "4                tead4          0.985851                  low   \n",
       "...                ...               ...                  ...   \n",
       "15744            tead4          1.253692                  low   \n",
       "15745            tead4          2.188614                  low   \n",
       "15746            tead4          3.522194               medium   \n",
       "15747            tead4          4.993919                 high   \n",
       "15748            tead4          1.343004                  low   \n",
       "\n",
       "       contrib_weighted_p example_chrom  example_end  example_idx  \\\n",
       "0                0.430212          chr1   30421146.0          8.0   \n",
       "1                0.479010          chr1  185337202.0         23.0   \n",
       "2                0.827772          chr1  170439355.0         55.0   \n",
       "3                0.027628          chr1   51906766.0        114.0   \n",
       "4                0.023681          chr1  152122914.0        119.0   \n",
       "...                   ...           ...          ...          ...   \n",
       "15744            0.075350          chrX  152071969.0     187707.0   \n",
       "15745            0.242914          chrX   12192318.0     187714.0   \n",
       "15746            0.519555          chrX   36805943.0     187742.0   \n",
       "15747            0.840689          chrX  151149525.0     187767.0   \n",
       "15748            0.090779          chrY   90803445.0     107484.0   \n",
       "\n",
       "      example_interval_from_task  example_start example_strand  \\\n",
       "0                           cdx2     30420146.0              .   \n",
       "1                           cdx2    185336202.0              .   \n",
       "2                           cdx2    170438355.0              .   \n",
       "3                           cdx2     51905766.0              .   \n",
       "4                           cdx2    152121914.0              .   \n",
       "...                          ...            ...            ...   \n",
       "15744                      gata3    152070969.0              .   \n",
       "15745                      gata3     12191318.0              .   \n",
       "15746                      gata3     36804943.0              .   \n",
       "15747                      gata3    151148525.0              .   \n",
       "15748                     tfap2c     90802445.0              .   \n",
       "\n",
       "                  fwd_seq  is_erv  match/tead4  match_max match_max_task  \\\n",
       "0      TAAAATGCCTGGCATTCT     0.0     0.496458   0.496458          tead4   \n",
       "1      AGGCATTCACTGAATGTA     0.0     0.480125   0.480125          tead4   \n",
       "2      TGGAATGCCAGGAATTGC     0.0     0.613330   0.613330          tead4   \n",
       "3      GAGAATGCTATGAATCCA     0.0     0.503989   0.503989          tead4   \n",
       "4      GTAAATACCTGGAATTAT     0.0     0.436122   0.436122          tead4   \n",
       "...                   ...     ...          ...        ...            ...   \n",
       "15744  TGGATTTCAGGGATTGCA     0.0     0.476687   0.476687          tead4   \n",
       "15745  TTGATTCCTGGAATACCA     0.0     0.448724   0.448724          tead4   \n",
       "15746  TACATACCAGGGATTCCT     0.0     0.598039   0.598039          tead4   \n",
       "15747  CAAATACCTGTAATTCCA     0.0     0.556961   0.556961          tead4   \n",
       "15748  TTAATTCCTTGAATAGTT     0.0     0.446307   0.446307          tead4   \n",
       "\n",
       "       match_weighted match_weighted_cat  match_weighted_p  \\\n",
       "0            0.496458             medium          0.470039   \n",
       "1            0.480125             medium          0.380337   \n",
       "2            0.613330               high          0.964478   \n",
       "3            0.503989             medium          0.509508   \n",
       "4            0.436122                low          0.208109   \n",
       "...               ...                ...               ...   \n",
       "15744        0.476687             medium          0.365626   \n",
       "15745        0.448724                low          0.253678   \n",
       "15746        0.598039               high          0.937567   \n",
       "15747        0.556961               high          0.793326   \n",
       "15748        0.446307                low          0.241837   \n",
       "\n",
       "                       pattern  pattern_center  pattern_end  pattern_end_abs  \\\n",
       "0      metacluster_0/pattern_1           663.0        671.0       30420817.0   \n",
       "1      metacluster_0/pattern_1           534.0        542.0      185336744.0   \n",
       "2      metacluster_0/pattern_1           504.0        512.0      170438867.0   \n",
       "3      metacluster_0/pattern_1           702.0        710.0       51906476.0   \n",
       "4      metacluster_0/pattern_1            52.0         60.0      152121974.0   \n",
       "...                        ...             ...          ...              ...   \n",
       "15744  metacluster_0/pattern_1           187.0        196.0      152071165.0   \n",
       "15745  metacluster_0/pattern_1           362.0        371.0       12191689.0   \n",
       "15746  metacluster_0/pattern_1           357.0        366.0       36805309.0   \n",
       "15747  metacluster_0/pattern_1           142.0        151.0      151148676.0   \n",
       "15748  metacluster_0/pattern_1           228.0        237.0       90802682.0   \n",
       "\n",
       "       pattern_len pattern_short  pattern_start  pattern_start_abs  row_idx  \\\n",
       "0             17.0         m0_p1          654.0         30420800.0      0.0   \n",
       "1             17.0         m0_p1          525.0        185336727.0      6.0   \n",
       "2             17.0         m0_p1          495.0        170438850.0      8.0   \n",
       "3             17.0         m0_p1          693.0         51906459.0     14.0   \n",
       "4             17.0         m0_p1           43.0        152121957.0     15.0   \n",
       "...            ...           ...            ...                ...      ...   \n",
       "15744         17.0         m0_p1          179.0        152071148.0  18362.0   \n",
       "15745         17.0         m0_p1          354.0         12191672.0  18363.0   \n",
       "15746         17.0         m0_p1          349.0         36805292.0  18366.0   \n",
       "15747         17.0         m0_p1          134.0        151148659.0  18369.0   \n",
       "15748         17.0         m0_p1          220.0         90802665.0   7661.0   \n",
       "\n",
       "                      seq  seq_match seq_match_cat  seq_match_p strand  \\\n",
       "0      TAAAATGCCTGGCATTCT  10.025424        medium     0.383925      +   \n",
       "1      AGGCATTCACTGAATGTA   7.793720           low     0.100108      +   \n",
       "2      TGGAATGCCAGGAATTGC  12.900725          high     0.843201      +   \n",
       "3      GAGAATGCTATGAATCCA   8.346114           low     0.151417      +   \n",
       "4      GTAAATACCTGGAATTAT   3.619336           low     0.004306      +   \n",
       "...                   ...        ...           ...          ...    ...   \n",
       "15744  TGCAATCCCTGAAATCCA   7.667224           low     0.095802      -   \n",
       "15745  TGGTATTCCAGGAATCAA  11.372528        medium     0.614281      -   \n",
       "15746  AGGAATCCCTGGTATGTA  12.923700          high     0.846430      -   \n",
       "15747  TGGAATTACAGGTATTTG  12.246571          high     0.763186      -   \n",
       "15748  AACTATTCAAGGAATTAA   8.319051           low     0.143524      -   \n",
       "\n",
       "      first_half_site_seq second_half_site_seq  \n",
       "0               TAAAATGCC            TGGCATTCT  \n",
       "1               AGGCATTCA            CTGAATGTA  \n",
       "2               TGGAATGCC            AGGAATTGC  \n",
       "3               GAGAATGCT            ATGAATCCA  \n",
       "4               GTAAATACC            TGGAATTAT  \n",
       "...                   ...                  ...  \n",
       "15744           TGCAATCCC            TGAAATCCA  \n",
       "15745           TGGTATTCC            AGGAATCAA  \n",
       "15746           AGGAATCCC            TGGTATGTA  \n",
       "15747           TGGAATTAC            AGGTATTTG  \n",
       "15748           AACTATTCA            AGGAATTAA  \n",
       "\n",
       "[15749 rows x 40 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Save results.\n",
    "motifs_oriented_df.to_csv('tsv/cwm_tead4double_noervs_nopromoter_withseq.tsv.gz', sep = '\\t', index = False)\n",
    "\n",
    "motifs_oriented_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Group results based on half site summaries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "summarized_tead4_sites_df = motifs_oriented_df.groupby(['first_half_site_seq', 'second_half_site_seq']).size().reset_index()\n",
    "summarized_tead4_sites_df.columns = ['first_half_site_seq', 'second_half_site_seq', 'frequency']\n",
    "summarized_tead4_sites_df = summarized_tead4_sites_df.sort_values('frequency', ascending = False)\n",
    "summarized_tead4_sites_df = summarized_tead4_sites_df.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>first_half_site_seq</th>\n",
       "      <th>second_half_site_seq</th>\n",
       "      <th>frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7398</td>\n",
       "      <td>TGGAATTAA</td>\n",
       "      <td>AGGCATGTG</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7393</td>\n",
       "      <td>TGGAATTAA</td>\n",
       "      <td>AGGCATGCA</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7457</td>\n",
       "      <td>TGGAATTAC</td>\n",
       "      <td>AGGCATGTG</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7453</td>\n",
       "      <td>TGGAATTAC</td>\n",
       "      <td>AGGCATGCA</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8194</td>\n",
       "      <td>TGGTATGTG</td>\n",
       "      <td>TGGTATGTG</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4995</th>\n",
       "      <td>8432</td>\n",
       "      <td>TTGAATTCC</td>\n",
       "      <td>ATAAATTCC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4996</th>\n",
       "      <td>48</td>\n",
       "      <td>AAAAATGCT</td>\n",
       "      <td>AAGAATGTA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4997</th>\n",
       "      <td>8438</td>\n",
       "      <td>TTGAATTCC</td>\n",
       "      <td>TGGAATTCC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4998</th>\n",
       "      <td>49</td>\n",
       "      <td>AAAAATGCT</td>\n",
       "      <td>GGACATGTC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4999</th>\n",
       "      <td>8436</td>\n",
       "      <td>TTGAATTCC</td>\n",
       "      <td>TGACATTCT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index first_half_site_seq second_half_site_seq  frequency\n",
       "0      7398           TGGAATTAA            AGGCATGTG         46\n",
       "1      7393           TGGAATTAA            AGGCATGCA         29\n",
       "2      7457           TGGAATTAC            AGGCATGTG         28\n",
       "3      7453           TGGAATTAC            AGGCATGCA         23\n",
       "4      8194           TGGTATGTG            TGGTATGTG         17\n",
       "...     ...                 ...                  ...        ...\n",
       "4995   8432           TTGAATTCC            ATAAATTCC          1\n",
       "4996     48           AAAAATGCT            AAGAATGTA          1\n",
       "4997   8438           TTGAATTCC            TGGAATTCC          1\n",
       "4998     49           AAAAATGCT            GGACATGTC          1\n",
       "4999   8436           TTGAATTCC            TGACATTCT          1\n",
       "\n",
       "[5000 rows x 4 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summarized_tead4_sites_df1=summarized_tead4_sites_df[0:5000]\n",
    "\n",
    "summarized_tead4_sites_df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>first_half_site_seq</th>\n",
       "      <th>second_half_site_seq</th>\n",
       "      <th>frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7398</td>\n",
       "      <td>TGGAATTAA</td>\n",
       "      <td>AGGCATGTG</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7393</td>\n",
       "      <td>TGGAATTAA</td>\n",
       "      <td>AGGCATGCA</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7457</td>\n",
       "      <td>TGGAATTAC</td>\n",
       "      <td>AGGCATGTG</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7453</td>\n",
       "      <td>TGGAATTAC</td>\n",
       "      <td>AGGCATGCA</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8194</td>\n",
       "      <td>TGGTATGTG</td>\n",
       "      <td>TGGTATGTG</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8526</th>\n",
       "      <td>1960</td>\n",
       "      <td>AGGAATCCT</td>\n",
       "      <td>GGGTATGGA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8527</th>\n",
       "      <td>1961</td>\n",
       "      <td>AGGAATCCT</td>\n",
       "      <td>GGGTATTTG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8528</th>\n",
       "      <td>1964</td>\n",
       "      <td>AGGAATCCT</td>\n",
       "      <td>TGGCATTAG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8529</th>\n",
       "      <td>4961</td>\n",
       "      <td>GGAAATTCA</td>\n",
       "      <td>CGGAATTTC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8530</th>\n",
       "      <td>4265</td>\n",
       "      <td>CTGCATTCC</td>\n",
       "      <td>CAGAATGTC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8531 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index first_half_site_seq second_half_site_seq  frequency\n",
       "0      7398           TGGAATTAA            AGGCATGTG         46\n",
       "1      7393           TGGAATTAA            AGGCATGCA         29\n",
       "2      7457           TGGAATTAC            AGGCATGTG         28\n",
       "3      7453           TGGAATTAC            AGGCATGCA         23\n",
       "4      8194           TGGTATGTG            TGGTATGTG         17\n",
       "...     ...                 ...                  ...        ...\n",
       "8526   1960           AGGAATCCT            GGGTATGGA          1\n",
       "8527   1961           AGGAATCCT            GGGTATTTG          1\n",
       "8528   1964           AGGAATCCT            TGGCATTAG          1\n",
       "8529   4961           GGAAATTCA            CGGAATTTC          1\n",
       "8530   4265           CTGCATTCC            CAGAATGTC          1\n",
       "\n",
       "[8531 rows x 4 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summarized_tead4_sites_df.shape\n",
    "summarized_tead4_sites_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get in silico contributions for each combination of half sites. \n",
    "\n",
    "### A. For each combination...\n",
    "    1. Generate random sequences.\n",
    "    2. Inject each half site at appropriate distances.\n",
    "    3. Get predict_all across all trials\n",
    "    4. Average across appropriate areas for contribution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## functions for generating random sequences & encoding and decoding in 1he format\n",
    "\n",
    "# seq in form of ACGT letters\n",
    "def generate_random_seq(seqlen, weights = [.25, .25, .25, .25]):\n",
    "    \"\"\"\n",
    "    Purpose: Generate a random DNA sequence of a specified length.\n",
    "    \"\"\"\n",
    "    import random\n",
    "    seq = random.choices(['A','C','G','T'], weights = weights, k=seqlen)\n",
    "    return(''.join(seq))\n",
    "\n",
    "# 1he seq\n",
    "def one_hot_encode_sequences(sequences):\n",
    "    \"\"\"\n",
    "    Purpose: Given an array of sequences, one-hot-encode into a [region x position x 4] array.\n",
    "    \"\"\"\n",
    "    return(np.stack([one_hot_encode_sequence(s) for s in sequences]))\n",
    "\n",
    "\n",
    "def one_hot_encode_sequence(sequence):\n",
    "    \"\"\"\n",
    "    Kudos to Charles: /n/projects/cm2363/bpnet-nucleosomes/work/localimportance/allLocalImportances.py\n",
    "    Purpose: Given a SINGLE sequence string, one-hot encode the data.\n",
    "        + default control_profiles and control_logcounts is to be zeroed out\n",
    "        + naively detects whether the sequence is one-hot-encoded.\n",
    "    \"\"\"\n",
    "    onehot_mapping = {\n",
    "    'A': [1,0,0,0],\n",
    "    'C': [0,1,0,0],\n",
    "    'G': [0,0,1,0],\n",
    "    'T': [0,0,0,1],\n",
    "    'a': [1,0,0,0],\n",
    "    'c': [0,1,0,0],\n",
    "    'g': [0,0,1,0],\n",
    "    't': [0,0,0,1],\n",
    "    'N': [0,0,0,0]\n",
    "    }\n",
    "    return np.array([onehot_mapping[x] for x in sequence])\n",
    "\n",
    "\n",
    "def one_hot_decode_sequence(array):\n",
    "    \"\"\"\n",
    "    Purpose: Given an array [position x 4], decode sequence to a string.\n",
    "    \"\"\"\n",
    "    onehot_decoder = {\n",
    "    0: 'A',\n",
    "    1: 'C',\n",
    "    2: 'G',\n",
    "    3: 'T'\n",
    "    }\n",
    "\n",
    "    idxs = np.where(array)[1]\n",
    "    return (''.join([onehot_decoder[i] for i in idxs]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate 64 random seq\n",
    "import random\n",
    "random.seed(10)\n",
    "trials = 64\n",
    "random_seqs = [generate_random_seq(seqlen = 1000) for i in range(trials)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(64, 1000, 4)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make it 1he\n",
    "random_seqs_1he =one_hot_encode_sequences(random_seqs)\n",
    "flanks = (random_seqs_1he.shape[1]-1000)//2\n",
    "random_seqs_1he = random_seqs_1he[:, flanks:(1000+flanks)]\n",
    "random_seqs_1he.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Collect all sequences with a pd.df that keeps indexes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "8531it [3:12:45,  1.36s/it]\n"
     ]
    }
   ],
   "source": [
    "combination_index_df = pd.DataFrame()\n",
    "combination_w_trials_list = []\n",
    "\n",
    "for i,row in tqdm(summarized_tead4_sites_df.iterrows()):\n",
    "        \n",
    "    #define half sites\n",
    "    first_half_site = row.first_half_site_seq\n",
    "    second_half_site = row.second_half_site_seq\n",
    "    \n",
    "    #Create sequences\n",
    "    for j,s in enumerate(random_seqs_1he):\n",
    "        seq = one_hot_decode_sequence(s)\n",
    "        seq_w_first = insert_motif(seq = seq, motif = first_half_site, position = 500)\n",
    "        seq_w_second = insert_motif(seq = seq_w_first, motif = second_half_site, position = 509)\n",
    "        combination_w_trials_list.append(seq_w_second)\n",
    "        row['trial_index'] = j\n",
    "        row['combo_index'] = i\n",
    "        combination_index_df = combination_index_df.append(row)\n",
    "\n",
    "#Create row index that combines trial index and ocmbo index and matches index of contrib predictions\n",
    "combination_index_df['row_index'] = (combination_index_df.combo_index * 64 + combination_index_df.trial_index).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "545984"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(combination_w_trials_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#to save the dataframe\n",
    "#combination_index_df.to_pickle('tsv/tsc_tead4double_combination_index_df.pkl')    \n",
    "#combination_index_df = pd.read_pickle('tsv/tsc_tead4double_combination_index_df.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### In a single prediction step, get all contribution scores (`list(3K) -> ['contrib_score']['tead4/profile']`), save as .pkl file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n",
      "DeepExplain: running \"deeplift\" explanation method (5)\n",
      "Model with multiple inputs:  True\n"
     ]
    }
   ],
   "source": [
    "%%script false --no-raise-error\n",
    "import pickle\n",
    "\n",
    "#One hot encode sequences\n",
    "combination_w_trixals_seq_1he = one_hot_encode_sequences(combination_w_trials_list)\n",
    "\n",
    "#Predict all sequences\n",
    "combinations_w_trials_pred_dict = model.predict_all(combination_w_trixals_seq_1he, contrib_method='deeplift')\n",
    "with open('tsv/all_combination_of_halfsites_w_trials.pkl', 'wb') as f:\n",
    "    pickle.dump(combinations_w_trials_pred_dict, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import pkl file and isolate and average values. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "with open('tsv/all_combination_of_halfsites_w_trials.pkl', 'rb') as f:\n",
    "    combinations_w_trials_pred_dict = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  6%|▌         | 31460/545984 [01:05<18:18, 468.28it/s]"
     ]
    }
   ],
   "source": [
    "summarized_tead4_sites_w_contrib_df = pd.DataFrame()\n",
    "\n",
    "for i in tqdm(range(len(combinations_w_trials_pred_dict))):\n",
    "     #Collect contribution scores under proper coordinates for all trials\n",
    "    first_half_site_contrib = (combinations_w_trials_pred_dict[i]['contrib_score']['tead4/profile'][496:505] * combinations_w_trials_pred_dict[i]['seq'][496:505]).sum()\n",
    "    second_half_site_contrib = (combinations_w_trials_pred_dict[i]['contrib_score']['tead4/profile'][505:514] * combinations_w_trials_pred_dict[i]['seq'][505:514]).sum()\n",
    "    df = pd.DataFrame()\n",
    "    df['first_half_site_contrib'] = [first_half_site_contrib]\n",
    "    df['second_half_site_contrib'] = [second_half_site_contrib]\n",
    "    df['row_index'] = i\n",
    "\n",
    "    summarized_tead4_sites_w_contrib_df = summarized_tead4_sites_w_contrib_df.append(df) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reset index and average the contribution over trial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "contrib_over_combo_over_trials_df = combination_index_df.reset_index().join(summarized_tead4_sites_w_contrib_df.reset_index(), on = ['row_index'], how = 'left', rsuffix = 'test')\n",
    "contrib_over_combo_df = contrib_over_combo_over_trials_df.groupby(['combo_index', 'first_half_site_seq', 'second_half_site_seq']).agg({'first_half_site_contrib' : 'mean', 'second_half_site_contrib' : 'mean'}).reset_index()\n",
    "contrib_over_combo_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cut the contrib values in each half into quantiles and save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "first_half_site_cutoffs = np.quantile(contrib_over_combo_df.first_half_site_contrib.values, [0, .25, .5, .75, 1])\n",
    "second_half_site_cutoffs = np.quantile(contrib_over_combo_df.second_half_site_contrib.values, [0, .25, .5, .75, 1])\n",
    "\n",
    "contrib_over_combo_df['first_half_site_q'] = pd.cut(contrib_over_combo_df.first_half_site_contrib, bins=first_half_site_cutoffs, labels=['q1','q2','q3','q4'], include_lowest = True)\n",
    "contrib_over_combo_df['second_half_site_q'] = pd.cut(contrib_over_combo_df.second_half_site_contrib, bins=second_half_site_cutoffs, labels=['q1','q2','q3','q4'], include_lowest = True)\n",
    "\n",
    "#save\n",
    "contrib_over_combo_df.to_csv('tsv/tead4_double_motif_contrib_over_combo_completehalfs_quantiles.tsv.gz', sep = '\\t', index = False)\n",
    "\n",
    "#check freq\n",
    "df = contrib_over_combo_df.value_counts(['first_half_site_q', 'second_half_site_q']).reset_index()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get insilico predictions keeping the combo index from above df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. get random seq decoded as list\n",
    "2. get first_half_site & second half sites from contrib_over_combo_df, keeping the combo index.\n",
    "3. inject each tead4 combo and measure total counts with and without entire motif (background).\n",
    "4. append to df above and save the df for plotting in R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "random_seqs_1he.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## get random seq decoded as list\n",
    "seqs = []\n",
    "for j,s in enumerate(random_seqs_1he):\n",
    "    seqs.append(one_hot_decode_sequence(s))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def collect_summarized_counts(contrib_over_combo_df, random_seqs, tasks_of_interest, second_motif_injection_site, motif_window, measurement_window):\n",
    "\n",
    "    \"\"\"\n",
    "    Purpose: Given WT, dA, dB, and dAdB states of a WT genomic tead4 halfs -> randomize other halfs, how does the double motif enrichment change?\n",
    "    Inputs:\n",
    "    Outputs:\n",
    "    \"\"\"\n",
    "    \n",
    "    from tqdm import tqdm\n",
    "    from concise.preprocessing import encodeDNA\n",
    "    def generate_seq(random_seq, central_motif, side_motif=None, side_distances=[], seqlen=1000):\n",
    "        from bpnet.simulate import insert_motif\n",
    "        injected_seq = insert_motif(random_seq, central_motif, seqlen // 2)\n",
    "        for d in side_distances:\n",
    "            injected_seq = insert_motif(injected_seq, side_motif, d)\n",
    "        return injected_seq\n",
    "    \n",
    "    summary_for_all_combos_df = pd.DataFrame()\n",
    "    for i,row in tqdm(contrib_over_combo_df.iterrows()):\n",
    "        \n",
    "        #define half sites\n",
    "        first_half_site = row.first_half_site_seq\n",
    "        second_half_site = row.second_half_site_seq\n",
    "        combo_index = row.combo_index\n",
    "\n",
    "        #Generate sequences with variant states\n",
    "        WT_seqs = [generate_seq(random_seq = i, central_motif = first_half_site, side_motif = second_half_site, \n",
    "                                side_distances = [second_motif_injection_site]) for i in random_seqs]\n",
    "        dAdB_seqs = [generate_seq(random_seq = i, central_motif = '', side_motif = '', \n",
    "                                  side_distances = []) for i in random_seqs]\n",
    "        dA_seqs = [generate_seq(random_seq = i, central_motif = '', side_motif = second_half_site, \n",
    "                                side_distances = [second_motif_injection_site]) for i in random_seqs]\n",
    "        dB_seqs = [generate_seq(random_seq = i, central_motif = first_half_site, side_motif = '', \n",
    "                                side_distances = [second_motif_injection_site]) for i in random_seqs]\n",
    "\n",
    "        #Collect into dictionary\n",
    "        seq_dict = {'WT': WT_seqs, 'dAdB': dAdB_seqs, 'dA': dA_seqs, 'dB': dB_seqs}\n",
    "\n",
    "\n",
    "        #Predict each variant state({inj_state -> {task -> [64 x 1000 x 2]}))\n",
    "        preds_dict = {k: model.predict(encodeDNA(v)) for k,v in seq_dict.items()}\n",
    "        \n",
    "        # Average across trials and measurement window({inj_state -> {task -> #}})\n",
    "        preds_single_dict = {inj_state: {task: np.mean(np.sum(v1[:, measurement_window[0]:measurement_window[1], :], axis = (1,2))) \n",
    "                     for task,v1 in v.items()} \n",
    "         for inj_state,v in preds_dict.items()}\n",
    "\n",
    "        #Summarize counts together pd.df based on predictions\n",
    "        summary_df = pd.DataFrame()\n",
    "        for inj_state, v in preds_single_dict.items():\n",
    "            df = pd.DataFrame.from_dict(v, orient = 'index').transpose()\n",
    "            df['inj_state'] = inj_state\n",
    "            df = df.melt(id_vars = ['inj_state'], var_name = 'task', value_name = 'counts')\n",
    "            summary_df = summary_df.append(df)\n",
    "        summary_df['first_half_site'] = first_half_site\n",
    "        summary_df['second_half_site'] = second_half_site        \n",
    "        summary_df['combo_index'] = combo_index\n",
    "        summary_for_all_combos_df = summary_for_all_combos_df.append(summary_df)\n",
    "    return(summary_for_all_combos_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Collect summarized counts and profiles across trials for each halfs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_df = collect_summarized_counts(contrib_over_combo_df=contrib_over_combo_df,random_seqs = seqs, \n",
    "                                   tasks_of_interest = ['tead4'], #','yap1'\n",
    "                                   second_motif_injection_site = 509,\n",
    "                                   motif_window = (496, 514), measurement_window = (475, 535))\n",
    "#counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save .tsv.gz for each object\n",
    "counts_df.to_csv(f'csv/all_counts_summary_predictions_genomic_regions_9mer.csv.gz', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bpnet-gpu",
   "language": "python",
   "name": "bpnet-gpu"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
