{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "The purpose of this .ipynb is to compare affinity of the single Tead4 and double Tead4 motif by injecting motif sequences in randomly generated sequences.\n",
    "\n",
    "# Computational setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n",
      "2024-08-21 09:20:24,561 [INFO] Note: detected 80 virtual cores but NumExpr set to maximum of 64, check \"NUMEXPR_MAX_THREADS\" environment variable.\n",
      "2024-08-21 09:20:24,563 [INFO] Note: NumExpr detected 80 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "2024-08-21 09:20:24,565 [INFO] NumExpr defaulting to 8 threads.\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from tensorflow.python.util import deprecation\n",
    "deprecation._PRINT_DEPRECATION_WARNINGS = False\n",
    "\n",
    "#Packages\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pybedtools import BedTool\n",
    "from tqdm import tqdm\n",
    "\n",
    "# Settings\n",
    "os.chdir(f'/n/projects/kd2200/publication/rebuttal/analysis/')\n",
    "pd.set_option('display.max_columns', 100)\n",
    "\n",
    "# Custom commands\n",
    "sys.path.insert(0, f'/n/projects/kd2200/publication/bpnet/analysis/scripts/bpnet/scripts/')\n",
    "from data_format_functions import myround, myfloor, myceiling\n",
    "\n",
    "# function to return key for any value \n",
    "def get_key(val, my_dict): \n",
    "    for key, value in my_dict.items(): \n",
    "        if val == value: \n",
    "            return key \n",
    "    return \"key doesn't exist\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Pre-existing variables\n",
    "fasta_file = f'/n/projects/kd2200/publication/bpnet/fasta/mm10.fa'\n",
    "model_dir = f'/n/projects/kd2200/publication/bpnet/model/dataspec.yaml_default_fold_5/'\n",
    "modisco_dir = f'/n/projects/kd2200/publication/bpnet/modisco/'\n",
    "cwm_scan_all_dir = f'/n/projects/kd2200/publication/bpnet/modisco/profile/dataspec.yaml_default_fold_5/'\n",
    "pwm_scan_task_dir = f'/n/projects/kd2200/publication/bpnet/analysis/tsv/'\n",
    "# Determined variables\n",
    "motifs_of_interest_dict = {#'Tead4': 'Tead4/m0_p0'}\n",
    "                           'Tead4-Tead4': 'Tead4/m0_p1'}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "import Bio\n",
    "from Bio.Seq import Seq\n",
    "from bpnet.extractors import extract_seq\n",
    "\n",
    "dfi_dict = {}\n",
    "for t in motifs_of_interest_dict.keys():\n",
    "\n",
    "    #Import the task\n",
    "    task, pattern_short = motifs_of_interest_dict[t].split('/')\n",
    "    #dfi = pd.read_csv(f'/n/projects/kd2200/publication/rebuttal/analysis/tsv/cwm_contrib_sel.tsv.gz',sep = '\\t')\n",
    "    #dfi = pd.read_csv(f'/n/projects/kd2200/publication/rebuttal/analysis/tsv/pwm_freq_sel.tsv.gz',sep = '\\t')\n",
    "    #dfi = pd.read_csv(f'/n/projects/kd2200/publication/rebuttal/analysis/tsv/cwm_freq_sel.tsv.gz',sep = '\\t')\n",
    "    dfi = pd.read_csv(f'/n/projects/kd2200/publication/rebuttal/analysis/tsv/marg_cwm_freq_all.tsv.gz',sep = '\\t')\n",
    "    #dfi = pd.read_csv(f'/n/projects/kd2200/publication/rebuttal/analysis/tsv/cwm_contrib_sel_context.tsv.gz',sep = '\\t')\n",
    "    \n",
    "    #dfi = pd.read_csv(f'/n/projects/kd2200/publication/rebuttal/analysis/tsv/td_cwm_contrib_sel.tsv.gz',sep = '\\t')\n",
    "    #dfi = pd.read_csv(f'/n/projects/kd2200/publication/rebuttal/analysis/tsv/td_pwm_freq_sel.tsv.gz',sep = '\\t')\n",
    "    #dfi = pd.read_csv(f'/n/projects/kd2200/publication/rebuttal/analysis/tsv/td_cwm_freq_sel.tsv.gz',sep = '\\t')\n",
    "    #dfi = pd.read_csv(f'/n/projects/kd2200/publication/rebuttal/analysis/tsv/td_marg_cwm_freq_sel.tsv.gz',sep = '\\t')\n",
    "    #dfi = pd.read_csv(f'/n/projects/kd2200/publication/rebuttal/analysis/tsv/td_cwm_contrib_sel_context.tsv.gz',sep = '\\t')\n",
    "\n",
    "    dfi['row_idx']=range(dfi.shape[0])\n",
    "\n",
    "    #Get coordinates in Interval format\n",
    "    coords_interv = [pybedtools.create_interval_from_list([row.seqnames, int(row.start), int(row.end), \n",
    "                                                       str(dfi.row_idx), '.', row.strand]) for i,row in dfi.iterrows()] \n",
    "\n",
    "    #Extract sequences\n",
    "    seqs = []\n",
    "    for i in coords_interv:\n",
    "        seq = Seq(extract_seq(i, variant = None, fasta_file = fasta_file, one_hot = False).upper())\n",
    "        if i.strand=='+': #reverse complement based on which strand the motif maps to\n",
    "            seq = seq.reverse_complement()\n",
    "        seqs.append(str(seq))\n",
    "\n",
    "    dfi['sequence'] = seqs\n",
    "    dfi_dict[t] = dfi    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>motif_id</th>\n",
       "      <th>motif_alt_id</th>\n",
       "      <th>score</th>\n",
       "      <th>p.value</th>\n",
       "      <th>q.value</th>\n",
       "      <th>matched_sequence</th>\n",
       "      <th>peak_region_id</th>\n",
       "      <th>motif</th>\n",
       "      <th>seq</th>\n",
       "      <th>pwm_match_score</th>\n",
       "      <th>pwm_match_score_q</th>\n",
       "      <th>row_idx</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>78049760</td>\n",
       "      <td>78049776</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.69767</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9568</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>ACATATTCCAGATACCT</td>\n",
       "      <td>9.529146</td>\n",
       "      <td>0.005714</td>\n",
       "      <td>0</td>\n",
       "      <td>GGTATCTGGAATATGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>46569428</td>\n",
       "      <td>46569444</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.51163</td>\n",
       "      <td>0.000097</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19672</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>GAATATCTGAGATATCT</td>\n",
       "      <td>9.139072</td>\n",
       "      <td>0.003214</td>\n",
       "      <td>1</td>\n",
       "      <td>GATATCTCAGATATTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>79679687</td>\n",
       "      <td>79679703</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.60465</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2377</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>ACATATTCAGAATATCT</td>\n",
       "      <td>8.784098</td>\n",
       "      <td>0.001786</td>\n",
       "      <td>2</td>\n",
       "      <td>GATATTCTGAATATGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>20456848</td>\n",
       "      <td>20456864</td>\n",
       "      <td>17</td>\n",
       "      <td>+</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.06980</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>54624</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>AGATACCCCACATATCA</td>\n",
       "      <td>8.653703</td>\n",
       "      <td>0.001429</td>\n",
       "      <td>3</td>\n",
       "      <td>TGATATGTGGGGTATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr4</td>\n",
       "      <td>18803828</td>\n",
       "      <td>18803844</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.09300</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18938</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>ACATATTTAAAATATCT</td>\n",
       "      <td>8.589107</td>\n",
       "      <td>0.001071</td>\n",
       "      <td>4</td>\n",
       "      <td>GATATTTTAAATATGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18831</th>\n",
       "      <td>chr19</td>\n",
       "      <td>26709228</td>\n",
       "      <td>26709244</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.74420</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>86240</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>GCATTCCAAGCATGACA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>18831</td>\n",
       "      <td>GTCATGCTTGGAATGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18832</th>\n",
       "      <td>chr19</td>\n",
       "      <td>37341504</td>\n",
       "      <td>37341520</td>\n",
       "      <td>17</td>\n",
       "      <td>+</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.25580</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>86586</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>GCATGCCTTTCATCCCA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>18832</td>\n",
       "      <td>TGGGATGAAAGGCATG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18833</th>\n",
       "      <td>chrX</td>\n",
       "      <td>38601047</td>\n",
       "      <td>38601063</td>\n",
       "      <td>17</td>\n",
       "      <td>+</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.72090</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>88149</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>CCATTCCTAGTATTCTA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>18833</td>\n",
       "      <td>TAGAATACTAGGAATG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18834</th>\n",
       "      <td>chrX</td>\n",
       "      <td>42069606</td>\n",
       "      <td>42069622</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.06980</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>88173</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>GCATGGCTAGCATGGTA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>18834</td>\n",
       "      <td>ACCATGCTAGCCATGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18835</th>\n",
       "      <td>chrX</td>\n",
       "      <td>139635503</td>\n",
       "      <td>139635519</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.46510</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>89400</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>GCATTGCTATTATTCCA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>18835</td>\n",
       "      <td>GGAATAATAGCAATGC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18836 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqnames      start        end  width strand      motif_id  \\\n",
       "0         chr2   78049760   78049776     17      -  tead4_double   \n",
       "1         chr4   46569428   46569444     17      -  tead4_double   \n",
       "2         chr1   79679687   79679703     17      -  tead4_double   \n",
       "3        chr11   20456848   20456864     17      +  tead4_double   \n",
       "4         chr4   18803828   18803844     17      -  tead4_double   \n",
       "...        ...        ...        ...    ...    ...           ...   \n",
       "18831    chr19   26709228   26709244     17      -  tead4_double   \n",
       "18832    chr19   37341504   37341520     17      +  tead4_double   \n",
       "18833     chrX   38601047   38601063     17      +  tead4_double   \n",
       "18834     chrX   42069606   42069622     17      -  tead4_double   \n",
       "18835     chrX  139635503  139635519     17      -  tead4_double   \n",
       "\n",
       "       motif_alt_id     score   p.value  q.value  matched_sequence  \\\n",
       "0               NaN   9.69767  0.000089      NaN               NaN   \n",
       "1               NaN   9.51163  0.000097      NaN               NaN   \n",
       "2               NaN   9.60465  0.000093      NaN               NaN   \n",
       "3               NaN  10.06980  0.000075      NaN               NaN   \n",
       "4               NaN  10.09300  0.000074      NaN               NaN   \n",
       "...             ...       ...       ...      ...               ...   \n",
       "18831           NaN  13.74420  0.000010      NaN               NaN   \n",
       "18832           NaN  11.25580  0.000042      NaN               NaN   \n",
       "18833           NaN  11.72090  0.000033      NaN               NaN   \n",
       "18834           NaN  11.06980  0.000047      NaN               NaN   \n",
       "18835           NaN  10.46510  0.000063      NaN               NaN   \n",
       "\n",
       "       peak_region_id         motif                seq  pwm_match_score  \\\n",
       "0                9568  tead4_double  ACATATTCCAGATACCT         9.529146   \n",
       "1               19672  tead4_double  GAATATCTGAGATATCT         9.139072   \n",
       "2                2377  tead4_double  ACATATTCAGAATATCT         8.784098   \n",
       "3               54624  tead4_double  AGATACCCCACATATCA         8.653703   \n",
       "4               18938  tead4_double  ACATATTTAAAATATCT         8.589107   \n",
       "...               ...           ...                ...              ...   \n",
       "18831           86240  tead4_double  GCATTCCAAGCATGACA         0.000000   \n",
       "18832           86586  tead4_double  GCATGCCTTTCATCCCA         0.000000   \n",
       "18833           88149  tead4_double  CCATTCCTAGTATTCTA         0.000000   \n",
       "18834           88173  tead4_double  GCATGGCTAGCATGGTA         0.000000   \n",
       "18835           89400  tead4_double  GCATTGCTATTATTCCA         0.000000   \n",
       "\n",
       "       pwm_match_score_q  row_idx          sequence  \n",
       "0               0.005714        0  GGTATCTGGAATATGT  \n",
       "1               0.003214        1  GATATCTCAGATATTC  \n",
       "2               0.001786        2  GATATTCTGAATATGT  \n",
       "3               0.001429        3  TGATATGTGGGGTATC  \n",
       "4               0.001071        4  GATATTTTAAATATGT  \n",
       "...                  ...      ...               ...  \n",
       "18831           0.000000    18831  GTCATGCTTGGAATGC  \n",
       "18832           0.000000    18832  TGGGATGAAAGGCATG  \n",
       "18833           0.000000    18833  TAGAATACTAGGAATG  \n",
       "18834           0.000000    18834  ACCATGCTAGCCATGC  \n",
       "18835           0.000000    18835  GGAATAATAGCAATGC  \n",
       "\n",
       "[18836 rows x 18 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# to change the seq column to be considered\n",
    "dfi.rename(columns = {'sequence':'seqe'}, inplace = True)\n",
    "dfi.rename(columns = {'seq':'sequence'}, inplace = True)\n",
    "#dfi.rename(columns = {'sequence':'seqe'}, inplace = True)\n",
    "#dfi.rename(columns = {'dna2':'sequence'}, inplace = True)\n",
    "#dfi.rename(columns = {'sequence':'seqe'}, inplace = True)\n",
    "#dfi.rename(columns = {'dna3':'sequence'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>motif_id</th>\n",
       "      <th>motif_alt_id</th>\n",
       "      <th>score</th>\n",
       "      <th>p.value</th>\n",
       "      <th>q.value</th>\n",
       "      <th>matched_sequence</th>\n",
       "      <th>peak_region_id</th>\n",
       "      <th>motif</th>\n",
       "      <th>sequence</th>\n",
       "      <th>pwm_match_score</th>\n",
       "      <th>pwm_match_score_q</th>\n",
       "      <th>row_idx</th>\n",
       "      <th>seqe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>78049760</td>\n",
       "      <td>78049776</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.69767</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9568</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>ACATATTCCAGATACCT</td>\n",
       "      <td>9.529146</td>\n",
       "      <td>0.005714</td>\n",
       "      <td>0</td>\n",
       "      <td>GGTATCTGGAATATGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>46569428</td>\n",
       "      <td>46569444</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.51163</td>\n",
       "      <td>0.000097</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19672</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>GAATATCTGAGATATCT</td>\n",
       "      <td>9.139072</td>\n",
       "      <td>0.003214</td>\n",
       "      <td>1</td>\n",
       "      <td>GATATCTCAGATATTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>79679687</td>\n",
       "      <td>79679703</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.60465</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2377</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>ACATATTCAGAATATCT</td>\n",
       "      <td>8.784098</td>\n",
       "      <td>0.001786</td>\n",
       "      <td>2</td>\n",
       "      <td>GATATTCTGAATATGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>20456848</td>\n",
       "      <td>20456864</td>\n",
       "      <td>17</td>\n",
       "      <td>+</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.06980</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>54624</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>AGATACCCCACATATCA</td>\n",
       "      <td>8.653703</td>\n",
       "      <td>0.001429</td>\n",
       "      <td>3</td>\n",
       "      <td>TGATATGTGGGGTATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr4</td>\n",
       "      <td>18803828</td>\n",
       "      <td>18803844</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.09300</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18938</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>ACATATTTAAAATATCT</td>\n",
       "      <td>8.589107</td>\n",
       "      <td>0.001071</td>\n",
       "      <td>4</td>\n",
       "      <td>GATATTTTAAATATGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18831</th>\n",
       "      <td>chr19</td>\n",
       "      <td>26709228</td>\n",
       "      <td>26709244</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.74420</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>86240</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>GCATTCCAAGCATGACA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>18831</td>\n",
       "      <td>GTCATGCTTGGAATGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18832</th>\n",
       "      <td>chr19</td>\n",
       "      <td>37341504</td>\n",
       "      <td>37341520</td>\n",
       "      <td>17</td>\n",
       "      <td>+</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.25580</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>86586</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>GCATGCCTTTCATCCCA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>18832</td>\n",
       "      <td>TGGGATGAAAGGCATG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18833</th>\n",
       "      <td>chrX</td>\n",
       "      <td>38601047</td>\n",
       "      <td>38601063</td>\n",
       "      <td>17</td>\n",
       "      <td>+</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.72090</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>88149</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>CCATTCCTAGTATTCTA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>18833</td>\n",
       "      <td>TAGAATACTAGGAATG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18834</th>\n",
       "      <td>chrX</td>\n",
       "      <td>42069606</td>\n",
       "      <td>42069622</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.06980</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>88173</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>GCATGGCTAGCATGGTA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>18834</td>\n",
       "      <td>ACCATGCTAGCCATGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18835</th>\n",
       "      <td>chrX</td>\n",
       "      <td>139635503</td>\n",
       "      <td>139635519</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.46510</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>89400</td>\n",
       "      <td>tead4_double</td>\n",
       "      <td>GCATTGCTATTATTCCA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>18835</td>\n",
       "      <td>GGAATAATAGCAATGC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18836 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqnames      start        end  width strand      motif_id  \\\n",
       "0         chr2   78049760   78049776     17      -  tead4_double   \n",
       "1         chr4   46569428   46569444     17      -  tead4_double   \n",
       "2         chr1   79679687   79679703     17      -  tead4_double   \n",
       "3        chr11   20456848   20456864     17      +  tead4_double   \n",
       "4         chr4   18803828   18803844     17      -  tead4_double   \n",
       "...        ...        ...        ...    ...    ...           ...   \n",
       "18831    chr19   26709228   26709244     17      -  tead4_double   \n",
       "18832    chr19   37341504   37341520     17      +  tead4_double   \n",
       "18833     chrX   38601047   38601063     17      +  tead4_double   \n",
       "18834     chrX   42069606   42069622     17      -  tead4_double   \n",
       "18835     chrX  139635503  139635519     17      -  tead4_double   \n",
       "\n",
       "       motif_alt_id     score   p.value  q.value  matched_sequence  \\\n",
       "0               NaN   9.69767  0.000089      NaN               NaN   \n",
       "1               NaN   9.51163  0.000097      NaN               NaN   \n",
       "2               NaN   9.60465  0.000093      NaN               NaN   \n",
       "3               NaN  10.06980  0.000075      NaN               NaN   \n",
       "4               NaN  10.09300  0.000074      NaN               NaN   \n",
       "...             ...       ...       ...      ...               ...   \n",
       "18831           NaN  13.74420  0.000010      NaN               NaN   \n",
       "18832           NaN  11.25580  0.000042      NaN               NaN   \n",
       "18833           NaN  11.72090  0.000033      NaN               NaN   \n",
       "18834           NaN  11.06980  0.000047      NaN               NaN   \n",
       "18835           NaN  10.46510  0.000063      NaN               NaN   \n",
       "\n",
       "       peak_region_id         motif           sequence  pwm_match_score  \\\n",
       "0                9568  tead4_double  ACATATTCCAGATACCT         9.529146   \n",
       "1               19672  tead4_double  GAATATCTGAGATATCT         9.139072   \n",
       "2                2377  tead4_double  ACATATTCAGAATATCT         8.784098   \n",
       "3               54624  tead4_double  AGATACCCCACATATCA         8.653703   \n",
       "4               18938  tead4_double  ACATATTTAAAATATCT         8.589107   \n",
       "...               ...           ...                ...              ...   \n",
       "18831           86240  tead4_double  GCATTCCAAGCATGACA         0.000000   \n",
       "18832           86586  tead4_double  GCATGCCTTTCATCCCA         0.000000   \n",
       "18833           88149  tead4_double  CCATTCCTAGTATTCTA         0.000000   \n",
       "18834           88173  tead4_double  GCATGGCTAGCATGGTA         0.000000   \n",
       "18835           89400  tead4_double  GCATTGCTATTATTCCA         0.000000   \n",
       "\n",
       "       pwm_match_score_q  row_idx              seqe  \n",
       "0               0.005714        0  GGTATCTGGAATATGT  \n",
       "1               0.003214        1  GATATCTCAGATATTC  \n",
       "2               0.001786        2  GATATTCTGAATATGT  \n",
       "3               0.001429        3  TGATATGTGGGGTATC  \n",
       "4               0.001071        4  GATATTTTAAATATGT  \n",
       "...                  ...      ...               ...  \n",
       "18831           0.000000    18831  GTCATGCTTGGAATGC  \n",
       "18832           0.000000    18832  TGGGATGAAAGGCATG  \n",
       "18833           0.000000    18833  TAGAATACTAGGAATG  \n",
       "18834           0.000000    18834  ACCATGCTAGCCATGC  \n",
       "18835           0.000000    18835  GGAATAATAGCAATGC  \n",
       "\n",
       "[18836 rows x 18 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# save as csv\n",
    "#dfi.to_csv(f'new_tead4_double_noerv_withseq.gz', sep = '\\t', index = False)\n",
    "#dfi.to_csv(f'new_tead4_noerv_withseq.gz', sep = '\\t', index = False)\n",
    "dfi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assess the distribution of motif variant frequency.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_list = [v.sequence.value_counts() for v in dfi_dict.values()]\n",
    "seq_df = pd.DataFrame()\n",
    "for i,l in enumerate(seq_list):\n",
    "    df = pd.DataFrame(l).reset_index()\n",
    "    df.columns = ['sequence','frequency']\n",
    "    df['motif'] = list(motifs_of_interest_dict.keys())[i]\n",
    "    seq_df = seq_df.append(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "seq_list = [v.sequence for v in dfi_dict.values()]  # Get unique sequences without counting frequency\n",
    "seq_df = pd.DataFrame()\n",
    "\n",
    "for i, seqs in enumerate(seq_list):\n",
    "    df = pd.DataFrame(seqs, columns=['sequence'])  # Create a DataFrame with sequences\n",
    "    df['motif'] = list(motifs_of_interest_dict.keys())[i]  # Add the motif name\n",
    "    seq_df = seq_df.append(df, ignore_index=True)  # Append to the main DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>frequency</th>\n",
       "      <th>motif</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCATGCCTTTAATCCCA</td>\n",
       "      <td>692</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACATGCCTTTAATCCCA</td>\n",
       "      <td>372</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GAATTTATGAAATTCCT</td>\n",
       "      <td>65</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCATGCCTTTAATTCCA</td>\n",
       "      <td>37</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCATACCTTTAATCCCA</td>\n",
       "      <td>34</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16607</th>\n",
       "      <td>CAATGCTTAGCATTCCT</td>\n",
       "      <td>1</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16608</th>\n",
       "      <td>GCATTCTAGCCATGGTG</td>\n",
       "      <td>1</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16609</th>\n",
       "      <td>ACATTTTATGCATTCAG</td>\n",
       "      <td>1</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16610</th>\n",
       "      <td>GCATACCAAAAATTTTT</td>\n",
       "      <td>1</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16611</th>\n",
       "      <td>TGATACTCAGAATTCCA</td>\n",
       "      <td>1</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16612 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                sequence  frequency        motif\n",
       "0      GCATGCCTTTAATCCCA        692  Tead4-Tead4\n",
       "1      ACATGCCTTTAATCCCA        372  Tead4-Tead4\n",
       "2      GAATTTATGAAATTCCT         65  Tead4-Tead4\n",
       "3      GCATGCCTTTAATTCCA         37  Tead4-Tead4\n",
       "4      GCATACCTTTAATCCCA         34  Tead4-Tead4\n",
       "...                  ...        ...          ...\n",
       "16607  CAATGCTTAGCATTCCT          1  Tead4-Tead4\n",
       "16608  GCATTCTAGCCATGGTG          1  Tead4-Tead4\n",
       "16609  ACATTTTATGCATTCAG          1  Tead4-Tead4\n",
       "16610  GCATACCAAAAATTTTT          1  Tead4-Tead4\n",
       "16611  TGATACTCAGAATTCCA          1  Tead4-Tead4\n",
       "\n",
       "[16612 rows x 3 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"15\" halign=\"left\">frequency</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>0%</th>\n",
       "      <th>10%</th>\n",
       "      <th>20%</th>\n",
       "      <th>30%</th>\n",
       "      <th>40%</th>\n",
       "      <th>50%</th>\n",
       "      <th>60%</th>\n",
       "      <th>70%</th>\n",
       "      <th>80%</th>\n",
       "      <th>90%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>motif</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Tead4-Tead4</th>\n",
       "      <td>16612.0</td>\n",
       "      <td>1.133879</td>\n",
       "      <td>6.152538</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>692.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            frequency                                                         \\\n",
       "                count      mean       std  min   0%  10%  20%  30%  40%  50%   \n",
       "motif                                                                          \n",
       "Tead4-Tead4   16612.0  1.133879  6.152538  1.0  1.0  1.0  1.0  1.0  1.0  1.0   \n",
       "\n",
       "                                        \n",
       "             60%  70%  80%  90%    max  \n",
       "motif                                   \n",
       "Tead4-Tead4  1.0  1.0  1.0  1.0  692.0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_summary_df = seq_df.groupby('motif').describe(percentiles = np.arange(0.0, 1.0, 0.1))\n",
    "seq_summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>frequency</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>motif</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Tead4-Tead4</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             frequency\n",
       "motif                 \n",
       "Tead4-Tead4        1.0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = seq_df.groupby('motif').quantile(q=0.5)\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tead_single_df = seq_df[seq_df['motif']==\"Tead4-Tead4\"]\n",
    "tead_single_df\n",
    "tead_single_df_arr = tead_single_df['frequency']\n",
    "np.percentile(tead_single_df_arr, 50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter meaningful sequences by requiring that they occur at least 10 times or are in the top 90th percentile of the frequency counts for Tead4. Because Tead4-Tead4 has lot of variability so, we will use 10 times and top 90th percentile together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tead4-Tead4    16612\n",
       "Name: motif, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_filt_df = pd.DataFrame()\n",
    "for m in seq_df.motif.unique(): \n",
    "    df = seq_df[seq_df.motif==m]\n",
    "    #df = df[(df.frequency>seq_summary_df.loc[m, ('frequency', '90%')]) & (df.frequency >= 5)] #6 for pwms_tdbl \n",
    "    df['row'] = range(df.shape[0])\n",
    "    seq_filt_df = seq_filt_df.append(df)\n",
    "seq_filt_df.motif.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>frequency</th>\n",
       "      <th>motif</th>\n",
       "      <th>row</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCATGCCTTTAATCCCA</td>\n",
       "      <td>692</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACATGCCTTTAATCCCA</td>\n",
       "      <td>372</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GAATTTATGAAATTCCT</td>\n",
       "      <td>65</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCATGCCTTTAATTCCA</td>\n",
       "      <td>37</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCATACCTTTAATCCCA</td>\n",
       "      <td>34</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16607</th>\n",
       "      <td>CAATGCTTAGCATTCCT</td>\n",
       "      <td>1</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>16607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16608</th>\n",
       "      <td>GCATTCTAGCCATGGTG</td>\n",
       "      <td>1</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>16608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16609</th>\n",
       "      <td>ACATTTTATGCATTCAG</td>\n",
       "      <td>1</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>16609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16610</th>\n",
       "      <td>GCATACCAAAAATTTTT</td>\n",
       "      <td>1</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>16610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16611</th>\n",
       "      <td>TGATACTCAGAATTCCA</td>\n",
       "      <td>1</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>16611</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16612 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                sequence  frequency        motif    row\n",
       "0      GCATGCCTTTAATCCCA        692  Tead4-Tead4      0\n",
       "1      ACATGCCTTTAATCCCA        372  Tead4-Tead4      1\n",
       "2      GAATTTATGAAATTCCT         65  Tead4-Tead4      2\n",
       "3      GCATGCCTTTAATTCCA         37  Tead4-Tead4      3\n",
       "4      GCATACCTTTAATCCCA         34  Tead4-Tead4      4\n",
       "...                  ...        ...          ...    ...\n",
       "16607  CAATGCTTAGCATTCCT          1  Tead4-Tead4  16607\n",
       "16608  GCATTCTAGCCATGGTG          1  Tead4-Tead4  16608\n",
       "16609  ACATTTTATGCATTCAG          1  Tead4-Tead4  16609\n",
       "16610  GCATACCAAAAATTTTT          1  Tead4-Tead4  16610\n",
       "16611  TGATACTCAGAATTCCA          1  Tead4-Tead4  16611\n",
       "\n",
       "[16612 rows x 4 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_filt_df#.frequency.sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assess BPNet predictions for variant sequences within each motif\n",
    "\n",
    "Now, with the variety of sequences that can be selected, assess whether BPNet predicts that some sequences result in higher binding than others. To do this, predict the average signal 256 times. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate simulated sequences, predict, and summarize for each motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TF-MoDISco is using the TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "from bpnet.BPNet import BPNetSeqModel\n",
    "bpn = BPNetSeqModel.from_mdir(model_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16612/16612 [10:28<00:00, 26.43it/s]\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "from bpnet.simulate import generate_sim\n",
    "lower_bounds = 0\n",
    "upper_bounds = 1000\n",
    "repeats = 28\n",
    "random.seed(10)\n",
    "\n",
    "#For each sequence variant, run BPNet simulate\n",
    "preds_summary = pd.DataFrame()\n",
    "for i,row in tqdm(seq_filt_df.iterrows(), total=seq_filt_df.shape[0]):\n",
    "    motif_variant = row.sequence\n",
    "    motif = row.motif\n",
    "    \n",
    "    pred = bpn.sim_pred(central_motif=motif_variant, side_motif=None, repeat = repeats, side_distances=[])\n",
    "    pred_summary =  pd.DataFrame.from_dict({k: [np.sum(v[lower_bounds:upper_bounds,:])] for k,v in pred.items()})\n",
    "    pred_summary.columns = [k.replace('profile/','') for k in pred.keys()]\n",
    "    pred_summary['sequence'] = motif_variant\n",
    "    pred_summary['motif'] = motif\n",
    "    pred_summary['row'] = row.row\n",
    "    #pred_summary['frequency'] = row.frequency\n",
    "    preds_summary = preds_summary.append(pred_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cdx2</th>\n",
       "      <th>tfap2c</th>\n",
       "      <th>tead4</th>\n",
       "      <th>yap1</th>\n",
       "      <th>gata3</th>\n",
       "      <th>sequence</th>\n",
       "      <th>motif</th>\n",
       "      <th>row</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>47.674011</td>\n",
       "      <td>15.999769</td>\n",
       "      <td>84.487434</td>\n",
       "      <td>26.559311</td>\n",
       "      <td>15.988951</td>\n",
       "      <td>ACATTCTTTGCATTCCA</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>5285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>52.462959</td>\n",
       "      <td>20.853617</td>\n",
       "      <td>82.714333</td>\n",
       "      <td>27.221485</td>\n",
       "      <td>16.383532</td>\n",
       "      <td>ACATTCTTGGAATTCTT</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>7222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>45.735836</td>\n",
       "      <td>21.094540</td>\n",
       "      <td>76.925552</td>\n",
       "      <td>25.716116</td>\n",
       "      <td>14.091481</td>\n",
       "      <td>AAATTCCTGGAATTCCA</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>14218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>49.480087</td>\n",
       "      <td>21.416353</td>\n",
       "      <td>76.559479</td>\n",
       "      <td>25.545521</td>\n",
       "      <td>15.831800</td>\n",
       "      <td>ACATTCCACGCATTCCA</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>12476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>43.089569</td>\n",
       "      <td>16.997982</td>\n",
       "      <td>76.042038</td>\n",
       "      <td>24.626364</td>\n",
       "      <td>14.666025</td>\n",
       "      <td>ACATTCCTGACATTCCA</td>\n",
       "      <td>Tead4-Tead4</td>\n",
       "      <td>15562</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        cdx2     tfap2c      tead4       yap1      gata3           sequence  \\\n",
       "0  47.674011  15.999769  84.487434  26.559311  15.988951  ACATTCTTTGCATTCCA   \n",
       "0  52.462959  20.853617  82.714333  27.221485  16.383532  ACATTCTTGGAATTCTT   \n",
       "0  45.735836  21.094540  76.925552  25.716116  14.091481  AAATTCCTGGAATTCCA   \n",
       "0  49.480087  21.416353  76.559479  25.545521  15.831800  ACATTCCACGCATTCCA   \n",
       "0  43.089569  16.997982  76.042038  24.626364  14.666025  ACATTCCTGACATTCCA   \n",
       "\n",
       "         motif    row  \n",
       "0  Tead4-Tead4   5285  \n",
       "0  Tead4-Tead4   7222  \n",
       "0  Tead4-Tead4  14218  \n",
       "0  Tead4-Tead4  12476  \n",
       "0  Tead4-Tead4  15562  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preds_summary = preds_summary.sort_values(['tead4','yap1'], ascending = False) \n",
    "\n",
    "\n",
    "#preds_summary.to_csv(f'new_tead4_double_noervs_preds_summary.csv', index = False)\n",
    "#preds_summary.to_csv(f'new_tead4_double_noervs_preds_summary_allseq.csv', index = False)\n",
    "#preds_summary.to_csv(f'new_tead4_double_noervs_preds_summary_allseq_v2.csv', index = False)\n",
    "#preds_summary.to_csv(f'new_tead4_double_cwms_preds_summary_allseq_genomic_1000_bp_v2.csv', index = False)\n",
    "\n",
    "#preds_summary.to_csv(f'new_tead4_noervs_preds_summary.csv', index = False)\n",
    "#preds_summary.to_csv(f'new_tead4_noervs_preds_summary_allseq.csv', index = False)\n",
    "#preds_summary.to_csv(f'new_tead4_noervs_preds_summary_allseq_v2.csv', index = False)\n",
    "#preds_summary.to_csv(f'new_tead4_cwms_preds_summary_allseq_genomic_1000bp.csv', index = False)\n",
    "\n",
    "\n",
    "#preds_summary.to_csv(f'cwm_contrib_sel.csv', index = False)\n",
    "#preds_summary.to_csv(f'pwm_freq_sel.csv', index = False)\n",
    "#preds_summary.to_csv(f'cwm_freq_sel.csv', index = False)\n",
    "#preds_summary.to_csv(f'cwm_contrib_sel_context.csv', index = False)\n",
    "preds_summary.to_csv(f'marg_cwm_freq_all.csv', index = False)\n",
    "\n",
    "\n",
    "#preds_summary.to_csv(f'td_cwm_contrib_sel.csv', index = False)\n",
    "#preds_summary.to_csv(f'td_pwm_freq_sel.csv', index = False)\n",
    "#preds_summary.to_csv(f'td_cwm_freq_sel.csv', index = False)\n",
    "#preds_summary.to_csv(f'td_cwm_contrib_sel_context.csv', index = False)\n",
    "#preds_summary.to_csv(f'td_marg_cwm_freq_sel.csv', index = False)\n",
    "preds_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bpnet-gpu",
   "language": "python",
   "name": "bpnet-gpu"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
