---
title: "Replicate correlation"
author: "Khyati Dalal"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    depth: 3
    theme: sandstone
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
  word_document:
    toc: yes
header-includes:
- \usepackage{fancyhdr}
- \usepackage{lipsum}
- \pagestyle{fancy}
- \fancyfoot[CO,CE]{1028-02098-001-001}
- \fancyfoot[LE,RO]{\thepage}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options:
  chunk_output_type: console
---

# 1. Introduction

The goal of this analysis is to plot replicate correlations for all genomics data sets used in this study. This markdown includs ChIP-nexus, ATAC-seq, ChIP-seq. 

# 2. Computational Setup

```{r, warning=F, message=F}
#Standard packages
library(rtracklayer) ; library(GenomicRanges); library(magrittr) ; library(Biostrings)
library(ggplot2) ; library(reshape2); library(plyranges); library(Rsamtools); library(ggpubr)
library(dplyr); library(data.table); library(patchwork); library(readr); library(testit)

#KNITR Options
setwd("/n/projects/kd2200/publication/bpnet/analysis/")
options(knitr.figure_dir="figures/0_replicate_correlation/", java.parameters = "- Xmx6g")

#Lab sources
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/granges_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/knitr_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/caching.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_functions.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/multiplot.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/motif_functions.R")
source("scripts/1.motif_summary/tsc_variables.R")

#Specific packages
library(readxl)
library(ggseqlogo)
library(BSgenome.Mmusculus.UCSC.mm10)
source("scripts/1.motif_summary/tsc_variables.R")

```

# 3. Load relevant data

```{r coverage}

# ChIP-nexus

cdx2_bw.list <- list(cdx2_rep6 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_cdx2_nexus_6_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_cdx2_nexus_6_negative.bw"),
                     cdx2_rep7 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_cdx2_nexus_7_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_cdx2_nexus_7_negative.bw"),
                     cdx2_rep5 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_cdx2_nexus_5_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_cdx2_nexus_5_negative.bw"))

tead4_bw.list <- list(tead4_rep1 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_tead4_nexus_1_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_tead4_nexus_1_negative.bw"),
                     tead4_rep2 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_tead4_nexus_2_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_tead4_nexus_2_negative.bw"),
                     tead4_rep3 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_tead4_nexus_3_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_tead4_nexus_3_negative.bw"))

tfap2c_bw.list <- list(tfap2c_rep6 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_tfap2c_nexus_6_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_tfap2c_nexus_6_negative.bw"),
                     tfap2c_rep7 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_tfap2c_nexus_7_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_tfap2c_nexus_7_negative.bw"),
                     tfap2c_rep5 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_tfap2c_nexus_5_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_tfap2c_nexus_5_negative.bw"))

gata3_bw.list <- list(gata3_rep1 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_gata3_nexus_1_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_gata3_nexus_1_negative.bw"),
                     gata3_rep2 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_gata3_nexus_2_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_gata3_nexus_2_negative.bw"),
                     gata3_rep3 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_gata3_nexus_4_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_gata3_nexus_4_negative.bw"))

yap1_bw.list <- list(yap1_rep1 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_yap1_nexus_1_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_yap1_nexus_1_negative.bw"),
                     yap1_rep2 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_yap1_nexus_2_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_yap1_nexus_2_negative.bw"),
                     yap1_rep3 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_yap1_nexus_4_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_yap1_nexus_4_negative.bw"))

# ATAC
atac_bw.list <- list(rep1 ="/n/projects/kd2200/publication/data_preparation/bw/mm10/atac/individual/mtsc_native_atac_1.bw", rep2="/n/projects/kd2200/publication/data_preparation/bw/mm10/atac/individual/mtsc_native_atac_2.bw", rep3="/n/projects/kd2200/publication/data_preparation/bw/mm10/atac/individual/mtsc_native_atac_3.bw")

# TT-seq

ttseq_bw.list <- list(ttseq_rep1 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nascent_rna/individual/mtsc_ttseq_rna_1_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nascent_rna/individual/mtsc_ttseq_rna_1_negative.bw"),
                     ttseq_rep2 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nascent_rna/individual/mtsc_ttseq_rna_2_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nascent_rna/individual/mtsc_ttseq_rna_2_negative.bw"),
                      ttseq_rep3 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nascent_rna/individual/mtsc_ttseq_rna_3_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nascent_rna/individual/mtsc_ttseq_rna_3_negative.bw"))


# H3K27ac
h3k27ac_bw.list <- list(rep1 ="/n/projects/kd2200/publication/data_preparation/bw/mm10/seq/individual/mtsc_h3k27ac_seq_1.bw", rep2="/n/projects/kd2200/publication/data_preparation/bw/mm10/seq/individual/mtsc_h3k27ac_seq_2.bw", rep3="/n/projects/kd2200/publication/data_preparation/bw/mm10/seq/individual/mtsc_h3k27ac_seq_3.bw")

# Pol II
polii_bw.list <- list(polii_rep1 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_polii_nexus_1_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_polii_nexus_1_negative.bw"),
                     polii_rep2 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_polii_nexus_3_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_polii_nexus_3_negative.bw"),
                      polii_rep3 = list(pos = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_polii_nexus_4_positive.bw",neg = "/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/individual/mtsc_polii_nexus_4_negative.bw"))

# ChIP-nexus peaks

cdx2_peaks <- import("/n/projects/kd2200/publication/data_preparation/peaks/mm10/nexus/combined/mtsc_cdx2_nexus_combined_peaks.narrowPeak")  %>% keepStandardChromosomes(pruning.mode="coarse")
tead4_peaks <- import("/n/projects/kd2200/publication/data_preparation/peaks/mm10/nexus/combined/mtsc_tead4_nexus_combined_peaks.narrowPeak")  %>% keepStandardChromosomes(pruning.mode="coarse")
tfap2c_peaks <- import("/n/projects/kd2200/publication/data_preparation/peaks/mm10/nexus/combined/mtsc_tfap2c_nexus_combined_peaks.narrowPeak")  %>% keepStandardChromosomes(pruning.mode="coarse")
gata3_peaks <- import("/n/projects/kd2200/publication/data_preparation/peaks/mm10/nexus/combined/mtsc_gata3_nexus_combined_peaks.narrowPeak")  %>% keepStandardChromosomes(pruning.mode="coarse")
yap1_peaks <- import("/n/projects/kd2200/publication/data_preparation/peaks/mm10/nexus/combined/mtsc_yap1_nexus_combined_peaks.narrowPeak")  %>% keepStandardChromosomes(pruning.mode="coarse")

#atac_peaks
atac_peaks <- import("/n/projects/kd2200/publication/data_preparation/peaks/mm10/atac/combined/mtsc_native_atac_combined_peaks.narrowPeak")%>% keepStandardChromosomes(pruning.mode="coarse")

#polii_peaks
polii_peaks <- import("/n/projects/kd2200/publication/data_preparation/peaks/mm10/nexus/combined/mtsc_polii_nexus_combined_peaks.narrowPeak")%>% keepStandardChromosomes(pruning.mode="coarse")

#h3k27ac_peaks
h3k27ac_peaks <- import("/n/projects/kd2200/publication/data_preparation/peaks/mm10/seq/combined/mtsc_h3k27ac_seq_combined_peaks.narrowPeak")%>% keepStandardChromosomes(pruning.mode="coarse")

# ttseq-seq: use TSSs
#Pre-existing variables
library(GenomicFeatures)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
tss.gr <- transcripts(TxDb.Mmusculus.UCSC.mm10.knownGene) %>% resize(., 1, 'start') %>% keepStandardChromosomes(pruning.mode="coarse")

#calcuate signal
cdx2_peaks$rep6signal <- abs(regionSums(resize(cdx2_peaks, 200, 'center'), cdx2_bw.list$cdx2_rep6$pos)) + abs(regionSums(resize(cdx2_peaks, 200, 'center'), cdx2_bw.list$cdx2_rep6$neg))
cdx2_peaks$rep7signal <- abs(regionSums(resize(cdx2_peaks, 200, 'center'), cdx2_bw.list$cdx2_rep7$pos)) + abs(regionSums(resize(cdx2_peaks, 200, 'center'), cdx2_bw.list$cdx2_rep7$neg))
cdx2_peaks$rep5signal <- abs(regionSums(resize(cdx2_peaks, 200, 'center'), cdx2_bw.list$cdx2_rep5$pos)) + abs(regionSums(resize(cdx2_peaks, 200, 'center'), cdx2_bw.list$cdx2_rep5$neg))

tfap2c_peaks$rep6signal <- abs(regionSums(resize(tfap2c_peaks, 200, 'center'), tfap2c_bw.list$tfap2c_rep6$pos)) + abs(regionSums(resize(tfap2c_peaks, 200, 'center'), tfap2c_bw.list$tfap2c_rep6$neg))
tfap2c_peaks$rep7signal <- abs(regionSums(resize(tfap2c_peaks, 200, 'center'), tfap2c_bw.list$tfap2c_rep7$pos)) + abs(regionSums(resize(tfap2c_peaks, 200, 'center'), tfap2c_bw.list$tfap2c_rep7$neg))
tfap2c_peaks$rep5signal <- abs(regionSums(resize(tfap2c_peaks, 200, 'center'), tfap2c_bw.list$tfap2c_rep5$pos)) + abs(regionSums(resize(tfap2c_peaks, 200, 'center'), tfap2c_bw.list$tfap2c_rep5$neg))


tead4_peaks$rep1signal <- abs(regionSums(resize(tead4_peaks, 200, 'center'), tead4_bw.list$tead4_rep1$pos)) + abs(regionSums(resize(tead4_peaks, 200, 'center'), tead4_bw.list$tead4_rep1$neg))
tead4_peaks$rep2signal <- abs(regionSums(resize(tead4_peaks, 200, 'center'), tead4_bw.list$tead4_rep2$pos)) + abs(regionSums(resize(tead4_peaks, 200, 'center'), tead4_bw.list$tead4_rep2$neg))
tead4_peaks$rep3signal <- abs(regionSums(resize(tead4_peaks, 200, 'center'), tead4_bw.list$tead4_rep3$pos)) + abs(regionSums(resize(tead4_peaks, 200, 'center'), tead4_bw.list$tead4_rep3$neg))

gata3_peaks$rep1signal <- abs(regionSums(resize(gata3_peaks, 200, 'center'), gata3_bw.list$gata3_rep1$pos)) + abs(regionSums(resize(gata3_peaks, 200, 'center'), gata3_bw.list$gata3_rep1$neg))
gata3_peaks$rep2signal <- abs(regionSums(resize(gata3_peaks, 200, 'center'), gata3_bw.list$gata3_rep2$pos)) + abs(regionSums(resize(gata3_peaks, 200, 'center'), gata3_bw.list$gata3_rep2$neg))
gata3_peaks$rep3signal <- abs(regionSums(resize(gata3_peaks, 200, 'center'), gata3_bw.list$gata3_rep3$pos)) + abs(regionSums(resize(gata3_peaks, 200, 'center'), gata3_bw.list$gata3_rep3$neg))

yap1_peaks$rep1signal <- abs(regionSums(resize(yap1_peaks, 200, 'center'), yap1_bw.list$yap1_rep1$pos)) + abs(regionSums(resize(yap1_peaks, 200, 'center'), yap1_bw.list$yap1_rep1$neg))
yap1_peaks$rep2signal <- abs(regionSums(resize(yap1_peaks, 200, 'center'), yap1_bw.list$yap1_rep2$pos)) + abs(regionSums(resize(yap1_peaks, 200, 'center'), yap1_bw.list$yap1_rep2$neg))
yap1_peaks$rep3signal <- abs(regionSums(resize(yap1_peaks, 200, 'center'), yap1_bw.list$yap1_rep3$pos)) + abs(regionSums(resize(yap1_peaks, 200, 'center'), yap1_bw.list$yap1_rep3$neg))

#atac
atac_peaks$rep1signal <- regionSums(resize(atac_peaks, 600, 'center'), atac_bw.list$rep1)
atac_peaks$rep2signal <- regionSums(resize(atac_peaks, 600, 'center'), atac_bw.list$rep2)
atac_peaks$rep3signal <- regionSums(resize(atac_peaks, 600, 'center'), atac_bw.list$rep3)

#h3k27ac
h3k27ac_peaks$rep1signal <- regionSums(resize(h3k27ac_peaks, 1000, 'center'),h3k27ac_bw.list$rep1)
h3k27ac_peaks$rep2signal <- regionSums(resize(h3k27ac_peaks, 1000, 'center'),h3k27ac_bw.list$rep2)
h3k27ac_peaks$rep3signal <- regionSums(resize(h3k27ac_peaks, 1000, 'center'), h3k27ac_bw.list$rep3)

#polii

polii_peaks$rep1signal <- abs(regionSums(resize(polii_peaks, 200, 'center'), polii_bw.list$polii_rep1$pos)) + abs(regionSums(resize(polii_peaks, 200, 'center'), polii_bw.list$polii_rep1$neg))
polii_peaks$rep2signal <- abs(regionSums(resize(polii_peaks, 200, 'center'), polii_bw.list$polii_rep2$pos)) + abs(regionSums(resize(polii_peaks, 200, 'center'), polii_bw.list$polii_rep2$neg))
polii_peaks$rep3signal <- abs(regionSums(resize(polii_peaks, 200, 'center'), polii_bw.list$polii_rep3$pos)) + abs(regionSums(resize(polii_peaks, 200, 'center'), polii_bw.list$polii_rep3$neg))

#ttseq

tss.gr$rep1signal <- abs(regionSums(resize(tss.gr, 500, 'center'), ttseq_bw.list$ttseq_rep1$pos)) + abs(regionSums(resize(tss.gr, 500, 'center'), ttseq_bw.list$ttseq_rep1$neg))
tss.gr$rep2signal <- abs(regionSums(resize(tss.gr, 500, 'center'), ttseq_bw.list$ttseq_rep2$pos)) + abs(regionSums(resize(tss.gr, 500, 'center'), ttseq_bw.list$ttseq_rep2$neg))
tss.gr$rep3signal <- abs(regionSums(resize(tss.gr, 500, 'center'), ttseq_bw.list$ttseq_rep3$pos)) + abs(regionSums(resize(tss.gr,500, 'center'),ttseq_bw.list$ttseq_rep3$neg))

#correlate
#nexus
cor(as.data.frame(log(cdx2_peaks$rep6signal+1)), as.data.frame(log(cdx2_peaks$rep7signal+1)), method = "pearson")
cor(as.data.frame(log(tfap2c_peaks$rep6signal+1)), as.data.frame(log(tfap2c_peaks$rep7signal+1)), method = "pearson")
cor(as.data.frame(log(tead4_peaks$rep1signal+1)), as.data.frame(log(tead4_peaks$rep2signal+1)), method = "pearson")
cor(as.data.frame(log(gata3_peaks$rep1signal+1)), as.data.frame(log(gata3_peaks$rep2signal+1)), method = "pearson")
cor(as.data.frame(log(yap1_peaks$rep1signal+1)), as.data.frame(log(yap1_peaks$rep2signal+1)), method = "pearson")


cor(as.data.frame(log(cdx2_peaks$rep5signal+1)), as.data.frame(log(cdx2_peaks$rep6signal+1)), method = "pearson")
cor(as.data.frame(log(tfap2c_peaks$rep5signal+1)), as.data.frame(log(tfap2c_peaks$rep6signal+1)), method = "pearson")
cor(as.data.frame(log(tead4_peaks$rep2signal+1)), as.data.frame(log(tead4_peaks$rep3signal+1)), method = "pearson")
cor(as.data.frame(log(gata3_peaks$rep1signal+1)), as.data.frame(log(gata3_peaks$rep3signal+1)), method = "pearson")
cor(as.data.frame(log(yap1_peaks$rep1signal+1)), as.data.frame(log(yap1_peaks$rep3signal+1)), method = "pearson")

cor(as.data.frame(log(cdx2_peaks$rep5signal+1)), as.data.frame(log(cdx2_peaks$rep7signal+1)), method = "pearson")
cor(as.data.frame(log(tfap2c_peaks$rep5signal+1)), as.data.frame(log(tfap2c_peaks$rep7signal+1)), method = "pearson")
cor(as.data.frame(log(tead4_peaks$rep1signal+1)), as.data.frame(log(tead4_peaks$rep3signal+1)), method = "pearson")
cor(as.data.frame(log(gata3_peaks$rep2signal+1)), as.data.frame(log(gata3_peaks$rep3signal+1)), method = "pearson")
cor(as.data.frame(log(yap1_peaks$rep2signal+1)), as.data.frame(log(yap1_peaks$rep3signal+1)), method = "pearson")

#polii
cor(as.data.frame(log(polii_peaks$rep1signal+1)), as.data.frame(log(polii_peaks$rep2signal+1)), method = "pearson")
cor(as.data.frame(log(polii_peaks$rep2signal+1)), as.data.frame(log(polii_peaks$rep3signal+1)), method = "pearson")
cor(as.data.frame(log(polii_peaks$rep1signal+1)), as.data.frame(log(polii_peaks$rep3signal+1)), method = "pearson")

#h3k27ac
cor(as.data.frame(log(h3k27ac_peaks$rep1signal+1)), as.data.frame(log(h3k27ac_peaks$rep2signal+1)), method = "pearson")
cor(as.data.frame(log(h3k27ac_peaks$rep2signal+1)), as.data.frame(log(h3k27ac_peaks$rep3signal+1)), method = "pearson")
cor(as.data.frame(log(h3k27ac_peaks$rep1signal+1)), as.data.frame(log(h3k27ac_peaks$rep3signal+1)), method = "pearson")
#atac
cor(as.data.frame(log(atac_peaks$rep1signal+1)), as.data.frame(log(atac_peaks$rep2signal+1)), method = "pearson")
cor(as.data.frame(log(atac_peaks$rep2signal+1)), as.data.frame(log(atac_peaks$rep3signal+1)), method = "pearson")
cor(as.data.frame(log(atac_peaks$rep1signal+1)), as.data.frame(log(atac_peaks$rep3signal+1)), method = "pearson")

#ttseq

cor(as.data.frame(log(tss.gr$rep1signal+1)), as.data.frame(log(tss.gr$rep2signal+1)), method = "pearson")
cor(as.data.frame(log(tss.gr$rep1signal+1)), as.data.frame(log(tss.gr$rep3signal+1)), method = "pearson")
cor(as.data.frame(log(tss.gr$rep2signal+1)), as.data.frame(log(tss.gr$rep3signal+1)), method = "pearson")

#calcuate signal

library(readxl)

rep_corr_r.df <- readxl::read_xlsx(path="/n/projects/kd2200/publication/bpnet/analysis/csv/rep_corr_df.xlsx") %>% 
    as.data.frame()

rep_corr_summary.plot <- ggplot(rep_corr_r.df, aes(x=comparison)) +
  geom_tile(aes(y=state, fill=r)) +
  geom_text(aes(y=state, label=r)) +
  facet_wrap(exp_type ~ condition, scales = "free_y") +
  scale_fill_viridis(limits = c(0,1), breaks = c(0, .2, 0.4, 0.6, 0.8, 1), begin = 0, end = .92) +
  theme_classic()

rep_corr_summary.plot

ggsave("rep_corr_summary.pdf", plot = rep_corr_summary.plot, path = "figures/0_replicate_correlation/", width = 15, height = 8, units = "cm")
ggsave("rep_corr_summary.png", plot = rep_corr_summary.plot, path = "figures/0_replicate_correlation/", width = 15, height = 8, units = "cm")

```
## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

# Session Information

For the purposes of reproducibility, the R/Bioconductor session information is printed below:

```{r sessioninfo}
sessionInfo()
```
