---
title: "13_plot_enhancers"
author: "Khyati Dalal"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    depth: 3
    theme: sandstone
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
  word_document:
    toc: yes
header-includes:
- \usepackage{fancyhdr}
- \usepackage{lipsum}
- \pagestyle{fancy}
- \fancyfoot[CO,CE]{1028-02098-001-001}
- \fancyfoot[LE,RO]{\thepage}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options:
  chunk_output_type: console
---

# Introduction

Plot the region of CRISPR regions with activity markers and TFs nexus

# Computational Setup

```{r, warning=F, message=F}
#Standard packages
library(rtracklayer) ; library(GenomicRanges); library(magrittr) ; library(Biostrings)
library(ggplot2) ; library(reshape2); library(plyranges); library(Rsamtools);library(parallel)
library(dplyr); library(data.table); library(patchwork); library(readr); library(testit);library(BSgenome.Mmusculus.UCSC.mm10)

#KNITR Options
setwd("/n/projects/kd2200/publication/bpnet/analysis/")
options(knitr.figure_dir="figures/13_plot_enhancers/", java.parameters = "- Xmx6g")

#Lab sources
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/granges_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/caching.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_functions.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/multiplot.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/knitr_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/motif_functions.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/1.motif_summary/tsc_variables.R")
```

##Define data needed for plotting
enhancer.gr: a GRanges of a single enhancer region
```{r}
enhancers.gr<-c(
  GRanges('chr12', IRanges(start = 102260924, end = 102263463), strand = '*', name = 'Rin3'),
  GRanges('chr15', IRanges(start = 102015446, end = 102017257), strand = '*', name = 'Krt8_18'),
  GRanges('chr17', IRanges(start = 6827105, end = 6828904), strand = '*', name = 'Ezr'),
  GRanges('chr2', IRanges(start = 172759453, end = 172761334), strand = '*', name = 'tty2'),#bmp7
  GRanges('chr9', IRanges(start = 102725580, end = 102727379), strand = '*', name = 'Amotl2'),
  GRanges('chr7', IRanges(start = 65429623, end = 65431460), strand = '*', name = 'Tjp1'),
  GRanges('chr10', IRanges(start = 17578782, end = 17580451), strand = '*', name = 'Cited2')
  )
```

motifs.gr: a GRanges of motifs that are expected to occur across the enhancer. Can be more comprehensive of a set–function will subset.
```{r}
motifs.gr<-readr::read_tsv('tsv/curated_motif_regoion_tsc_model_motifs_fold5_1based_noerv.tsv.gz') %>%
  makeGRangesFromDataFrame(keep.extra.columns = T, starts.in.df.are.0based = T) %>%
  GenomicRanges::sort(ignore.strand = T)


#Specific variables
#Specific variables
library(RColorBrewer)
n_motifs<-motifs.gr$pattern_name %>% unique %>% length
colors <- colorRampPalette(brewer.pal(n_motifs, "Accent"))(n_motifs)
names(colors)<-motifs.gr$pattern_name %>% unique
motif.colors.list<-as.list(colors)
```


bw.list: a list of sample .bw filepaths desired to plot across this coordinate region. Can be single-stranded, double-stranded or a mix of both.
```{r}

bw.list<-list(`ATAC` = "../../data_preparation/bw/mm10/atac/combined/mtsc_native_atac_combined_rpkm.bw",
              `Tfap2c_nexus` = list(pos="../../data_preparation/bw/mm10/nexus/combined/mtsc_tfap2c_nexus_combined_normalized_positive.bw",
                                    neg="../../data_preparation/bw/mm10/nexus/combined/mtsc_tfap2c_nexus_combined_normalized_negative.bw"),
              `Tead4_nexus` = list(pos="../../data_preparation/bw/mm10/nexus/combined/mtsc_tead4_nexus_combined_normalized_positive.bw",
                                   neg="../../data_preparation/bw/mm10/nexus/combined/mtsc_tead4_nexus_combined_normalized_negative.bw"),
              `Yap1_nexus` = list(pos="../../data_preparation/bw/mm10/nexus/combined/mtsc_yap1_nexus_combined_normalized_positive.bw",
                                  neg="../../data_preparation/bw/mm10/nexus/combined/mtsc_yap1_nexus_combined_normalized_negative.bw"),
              `Cdx2_nexus` = list(pos="../../data_preparation/bw/mm10/nexus/combined/mtsc_cdx2_nexus_combined_normalized_positive.bw",
                                  neg="../../data_preparation/bw/mm10/nexus/combined/mtsc_cdx2_nexus_combined_normalized_negative.bw"),
              `Gata3_nexus` = list(pos="../../data_preparation/bw/mm10/nexus/combined/mtsc_gata3_nexus_combined_normalized_positive.bw",
                                   neg="../../data_preparation/bw/mm10/nexus/combined/mtsc_gata3_nexus_combined_normalized_negative.bw"),                    `H3K27ac` = "../../data_preparation/bw/mm10/seq/combined/mtsc_h3k27ac_seq_combined_rpkm.bw",
              `PolII_nexus` = list(pos='../../data_preparation/bw/mm10/nexus/combined/mtsc_polii_nexus_combined_normalized_positive.bw',
                                   neg='../../data_preparation/bw/mm10/nexus/combined/mtsc_polii_nexus_combined_normalized_negative.bw'),
              `Nascent_RNA_seq`=list(pos='../../data_preparation/bw/mm10/nascent_rna/combined/mtsc_ttseq_rna_combined_positive_rpkm.bw',
                                     neg='../../data_preparation/bw/mm10/nascent_rna/combined/mtsc_ttseq_rna_combined_negative_rpkm.bw'),
              `Yap1_contrib` = "../pred_contrib_bws/yap1.contrib.profile.bw")

```

The rest are general parameters.

##Function to simultaneously plot enhancer coordinates with coverage
Inputs:

enhancer.gr: a GRanges of a single enhancer region
motifs.gr: a GRanges of motifs that are expected to occur across the enhancer. Can be more comprehensive of a set–function will subset.
bw.list: a list of sample .bw filepaths desired to plot across this coordinate region. Can be single-stranded, double-stranded or a mix of both.
motif.colors.list: a list of motifs corresponding to motifs.gr that provide a hex value for intended color
title: title of the figure
region_width: how wide the figure should plot across, regardless of enhancer coordinates
cores: number of cores to extract coverage

Outputs:
ggplot object of coverage and enhancer coordinates

```{r}
plot_coverage_and_enhancer<-function(enhancer.gr, motifs.gr, bw.list, motif.colors.list, title, region_width = 2800, 
                                     cores = 6, scales = 'fixed', rounding_margin = 50, label_margin = 200){
  
  #Assertion requirements to make the function run correctly.
  testit::assert("Input enhancer GRanges needs to contain only 1 region.", length(enhancer.gr)==1)
  testit::assert("Motifs need to be a GRanges object", is(motifs.gr, "GRanges"))
  testit::assert("Enhancers need to be a GRanges object", is(enhancer.gr, "GRanges"))
  testit::assert("Region width should be larger than enhancer width.", width(enhancer.gr)<=region_width)
  testit::assert("Bigwig object needs to be a list of samples. If samples are ChIP-nexus/double-stranded samples, they need to be in a nested list.", 
                 is(bw.list, "list") & all(!(names(bw.list) %in% c("pos", "neg"))))
  testit::assert("Motif object needs a column titled pattern_name, motif_name, or name to distinguish motifs apart.", 
                 any(grepl("pattern_name|name|motif_name", colnames(motifs.gr@elementMetadata))))
  
  #Find overlaps between motif set and enhancer
  motifs_in_enhancer.gr<-subsetByOverlaps(motifs.gr, resize(enhancer.gr, region_width, "center"), ignore.strand = T)
  if("pattern_name" %in% colnames(motifs_in_enhancer.gr@elementMetadata)){
    motifs_in_enhancer.gr$motif_name<-motifs_in_enhancer.gr$pattern_name
  }
  if("name" %in% colnames(motifs_in_enhancer.gr@elementMetadata)){
    motifs_in_enhancer.gr$motif_name<-motifs_in_enhancer.gr$name
  }
  
  #Extract colors for enhancer
  fills_in_order<-lapply(motifs_in_enhancer.gr$motif_name, function(x) motif.colors.list[[x]]) %>% unlist %>% unique %>% as.character
  motifs_in_order<-motifs_in_enhancer.gr$motif_name %>% unique %>% as.character
  motifs_in_enhancer.df<-motifs_in_enhancer.gr %>% as.data.frame
  motifs_in_enhancer.df$motif_name<-factor(motifs_in_enhancer.df$motif_name, levels = motifs_in_order)
  
  #Convert enhancer coordinates to df
  enhancer.df<-enhancer.gr %>% as.data.frame
  
  #Get half of the enhancer
  enh_half<-floor(width(enhancer.gr)/2)
  lower_bounds<-plyr::round_any(enhancer.df$start - (region_width/2 - enh_half), rounding_margin, f = floor)
  upper_bounds<-plyr::round_any(enhancer.df$end + (region_width/2 - enh_half), rounding_margin, f = ceiling)
  
  enhancer.plot<-ggplot()+
    geom_hline(yintercept = .5, color = "gray 85")+
    geom_rect(data = enhancer.df, aes(xmin = start, xmax = end, ymin = 0, ymax = 1), fill = "gray 85")+
    geom_text(data = enhancer.df, aes(x = ((start - end)/2 + start), y = .5, label = paste0(seqnames, ":", start, "-", end)), 
              color = "black")
  if(nrow(motifs_in_enhancer.df)>0){
    enhancer.plot<-enhancer.plot + geom_rect(data = motifs_in_enhancer.df, 
                                             aes(xmin = start, xmax = end, ymin = 1.1, ymax = 2, fill = motif_name), color = 'black', size = .1)
  }
  enhancer.plot<-enhancer.plot+
    scale_fill_manual(values = fills_in_order)+
    scale_x_continuous(breaks = seq(plyr::round_any(enhancer.df$start - (region_width/2 - enh_half), rounding_margin, f = floor), 
                                    plyr::round_any(enhancer.df$end + (region_width/2 - enh_half), rounding_margin, f = ceiling), label_margin), 
                       limits = c(lower_bounds,upper_bounds))+
    scale_y_continuous(breaks = c(.5, 1.75), labels = c("Enhancer", "Motif(s)"), position = "right")+
    theme_classic()+
    theme(axis.title = element_blank(), axis.line = element_blank(), axis.ticks.y = element_blank(),
          legend.position = "none")
  
  # Get coverage across bigwig list
  coverage.df<-mclapply(names(bw.list), function(x){
    if(length(bw.list[[x]])==1){
      df<-standard_metapeak(smooth = 4,resize(enhancer.gr, 1, "center"), bw.list[[x]], upstream = floor(region_width/2) + rounding_margin, downstream = floor(region_width/2)+200) %>% 
        dplyr::rename(window_distance = tss_distance) %>% 
        dplyr::mutate(position = window_distance + start(resize(enhancer.gr, 1, "center")), sample_name = x)
    }
    if(length(bw.list[[x]])==2){
      df<-exo_metapeak(smooth = 4,resize(enhancer.gr, 1, "center"), bw.list[[x]], upstream = region_width/2 + rounding_margin, downstream = region_width/2+rounding_margin) %>%
        dplyr::rename(window_distance = tss_distance) %>% 
        dplyr::mutate(position = window_distance + start(resize(enhancer.gr, 1, "center")), sample_name = x)
    }
    return(df)
  }, mc.cores = cores) %>% rbindlist(fill = TRUE) %>%
    dplyr::filter(position >= lower_bounds, position <= upper_bounds) %>%
    dplyr::mutate(sample_name = factor(sample_name, levels = names(bw.list)))
  
  #Check for strand if there are no 2-channeled samples.
  if(!("strand" %in% colnames(coverage.df))){
    coverage.df$strand<-NA
  }
  
  #Plot the enhancer coverage.
  coverage.plot<-ggplot(coverage.df)
  if(nrow(motifs_in_enhancer.df)>0){
    coverage.plot<-coverage.plot + geom_rect(data = motifs_in_enhancer.df, aes(xmin = start, xmax = end, ymin = -Inf, ymax = Inf, fill = motif_name), alpha = .2)
  }
  coverage.plot<-coverage.plot +
    geom_area(aes(x = position, y = reads, group = strand), fill = "darkblue")+
    geom_text(aes(x = Inf, y = Inf, label = sample_name), hjust = -.01)+
    scale_fill_manual(values = fills_in_order)+
    scale_x_continuous(breaks = seq(plyr::round_any(enhancer.df$start - (region_width/2 - enh_half), rounding_margin, f = floor), 
                                    plyr::round_any(enhancer.df$end + (region_width/2 - enh_half), rounding_margin, f = ceiling), label_margin), 
                       limits = c(lower_bounds,upper_bounds))+
    facet_grid(sample_name ~ ., scales = scales)+
    ggtitle(title)+
    theme_classic()+
    theme(axis.title = element_blank(), axis.line.x = element_blank(), strip.background = element_blank())  
  
  g<-coverage.plot + enhancer.plot + plot_layout(ncol = 1, heights = c(5, 1))
  return(g)
}
```

##Plot enhancers

```{r}
mclapply(1:length(enhancers.gr), function(x){
  enhancer.gr = enhancers.gr[x]
  g<-plot_coverage_and_enhancer(enhancer.gr = enhancer.gr, motifs.gr = motifs.gr, bw.list = bw.list, 
                                motif.colors.list = motif.colors.list, title = enhancer.gr$pattern_name, 
                                region_width = width(enhancer.gr), cores = 6, scales = 'free')
  ggsave(paste0("figures/13_plot_enhancers/", enhancer.gr$name,".pdf"), g, height = 6, width = 14)
  ggsave(paste0("figures/13_plot_enhancers/", enhancer.gr$name,".png"), g, height = 6, width = 14)
  return(g)
}, mc.cores = 6)

```

# Session Information

For the purposes of reproducibility, the R/Bioconductor session information is printed below:

```{r sessioninfo}
sessionInfo()
```


## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. 

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
