---
title: "Tead4 Tfap2c distance analysis"
author: "Khyati Dalal"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    depth: 3
    theme: sandstone
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
  word_document:
    toc: yes
header-includes:
- \usepackage{fancyhdr}
- \usepackage{lipsum}
- \pagestyle{fancy}
- \fancyfoot[CO,CE]{1028-02098-001-001}
- \fancyfoot[LE,RO]{\thepage}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options:
  chunk_output_type: console
---
Tead4_Tfap2c distance analysis and selecting regions for luciferase and CRISPR

```{r setup, warning=F, message=F}

#Standard packages
library(rtracklayer) ; library(GenomicRanges); library(magrittr) ; library(Biostrings);library(parallel)
library(ggplot2) ; library(reshape2); library(plyranges); library(Rsamtools);library(BSgenome.Mmusculus.UCSC.mm10)
library(dplyr); library(data.table); library(patchwork); library(readr);library(tidyverse);library(forcats); library(testit);library("TxDb.Mmusculus.UCSC.mm10.knownGene"); library(viridis);library(DECIPHER)

#KNITR Options
setwd("/n/projects/kd2200/publication/bpnet/analysis/")
options(knitr.figure_dir="figures", width=16, height=20)

source("/n/projects/kd2200/publication/bpnet/analysis/scripts/granges_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/load_motif_instance_from_all_regions.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/caching.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_functions.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/multiplot.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/knitr_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/motif_functions.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/heatmap_common.r")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/metapeak_functions_yue.R")
source("/n/projects/kd2200/publication/bpnet/analysis/scripts/1.motif_summary/tsc_variables.R")


#Specific packages
library(stringr)
library(grid)
library(gridExtra)

```

# Aims

I want to generate excel with tead4-tfap2c motif distance relation with new tsc model. This can be further used to narrow down the region for luciferase assay.

#load bigwig
```{r}

##chip-seq bws
seq.bigwig.list<-list(PolII=list
                      (pos="/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/combined/mtsc_polii_nexus_combined_normalized_positive.bw",
                        neg="/n/projects/kd2200/publication/data_preparation/bw/mm10/nexus/combined/mtsc_polii_nexus_combined_normalized_negative.bw"),
H3K27ac=list(rep1="/n/projects/kd2200/publication/data_preparation/bw/mm10/seq/combined/mtsc_h3k27ac_seq_combined_rpkm.bw",        rep2="/n/projects/kd2200/publication/data_preparation/bw/mm10/seq/combined/mtsc_h3k27ac_seq_combined_rpkm.bw"),
atac=list(rep1="/n/projects/kd2200/publication/data_preparation/bw/mm10/atac/combined/mtsc_native_atac_combined_rpkm.bw",
          rep2="/n/projects/kd2200/publication/data_preparation/bw/mm10/atac/combined/mtsc_native_atac_combined_rpkm.bw"))


##nexus_bws
nexus.bigwig.list<-
  list(gata3=list(pos="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_gata3_nexus_filtered_combined_normalized_positive.bw", 
             neg="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_gata3_nexus_filtered_combined_normalized_negative.bw"),  
  cdx2=list(pos="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_cdx2_nexus_filtered_combined_normalized_positive.bw",
            neg="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_cdx2_nexus_filtered_combined_normalized_negative.bw"),
  tfap2c=list(pos="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_tfap2c_nexus_filtered_combined_normalized_positive.bw",
             neg="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_tfap2c_nexus_filtered_combined_normalized_negative.bw"),
  tead4=list(pos="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_tead4_nexus_filtered_combined_normalized_positive.bw",
            neg="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_tead4_nexus_filtered_combined_normalized_negative.bw"),
  yap1=list(pos="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_yap1_nexus_filtered_combined_normalized_positive.bw",
            neg="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_yap1_nexus_filtered_combined_normalized_negative.bw"),
  PolII=list(pos="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_polii_nexus_filtered_combined_normalized_positive.bw",
           neg="/n/projects/kd2200/publication/bpnet_prep/bw/normalized/mtsc_polii_nexus_filtered_combined_normalized_negative.bw"),
  ttseq=list(pos="/n/projects/kd2200/publication/data_preparation/bw/mm10/nascent_rna/combined/mtsc_ttseq_rna_combined_positive_rpkm.bw",
           neg="/n/projects/kd2200/publication/data_preparation/bw/mm10/nascent_rna/combined/mtsc_ttseq_rna_combined_negative_rpkm.bw"))
 
```

## Island analysis
```{r}
##CWM_all motif regions
model_path= "/n/projects/kd2200/publication/bpnet/cwm_all_regions/"
cdx2.gr <- load_motif_instance_from_all_regions(model_path,"cdx2","p0") %>% dplyr::mutate(., tf = "Cdx2")
tead4.gr <- load_motif_instance_from_all_regions(model_path,"tead4","p0") %>% dplyr::mutate(., tf = "Tead4")
tfap2c.gr <- load_motif_instance_from_all_regions(model_path,"tfap2c", "p0") %>% dplyr::mutate(., tf = "Tfap2c")
gata3.gr <- load_motif_instance_from_all_regions(model_path,"gata3","p4") %>% dplyr::mutate(., tf = "Gata3")
yap1.gr <- load_motif_instance_from_all_regions(model_path,"yap1","p2") %>% dplyr::mutate(., tf = "Yap1_tead4_double_motif")


mm10_promoters <-promoters(TxDb.Mmusculus.UCSC.mm10.knownGene,upstream=1000, downstream=200)
cdx2_no_promoter.gr <- cdx2.gr[which(!overlapsAny(cdx2.gr,mm10_promoters, ignore.strand=T))]
tead4_no_promoter.gr <- tead4.gr[which(!overlapsAny(tead4.gr,mm10_promoters, ignore.strand=T))]
tfap2c_no_promoter.gr <- tfap2c.gr[which(!overlapsAny(tfap2c.gr,mm10_promoters, ignore.strand=T))]
gata3_no_promoter.gr <- gata3.gr[which(!overlapsAny(gata3.gr,mm10_promoters, ignore.strand=T))]
yap1_no_promoter.gr <- yap1.gr[which(!overlapsAny(yap1.gr,mm10_promoters, ignore.strand=T))]


all_tf.gr <- bind_rows(as.data.frame(cdx2_no_promoter.gr),as.data.frame(tead4_no_promoter.gr),as.data.frame(tfap2c_no_promoter.gr),as.data.frame(gata3_no_promoter.gr)) %>% makeGRangesFromDataFrame(., keep.extra.columns = T, starts.in.df.are.0based = F)
all_tf.gr$tf_center <- resize(all_tf.gr, 1, "center") %>% start
width(all_tf.gr) %>% table() 

islands.gr<- resize(all_tf.gr,500, "center") %>% GenomicRanges::reduce(.) 
# Filter islands that motifs are greater or equal to 500bp
islands.gr <- islands.gr[width(islands.gr)>=500,]
#Assign island id and island center
islands.gr$island_id<-1:length(islands.gr)
islands.gr$island_center <- resize(islands.gr, 1, "center") %>% start


islands.gr$h3k27ac_sig <- regionSums(resize(islands.gr, 500, "center"), seq.bigwig.list$H3K27ac$rep1)
islands.gr$atac <- regionSums(resize(islands.gr, 200, "center"), seq.bigwig.list$atac$rep1)

islands.gr$tead4_sig <- abs(regionSums(resize(islands.gr, 51, "center"), nexus.bigwig.list$tead4$pos) + abs(regionSums(regions = resize(islands.gr,51,"center"), nexus.bigwig.list$tead4$neg)))
islands.gr$tfap2c_sig <- abs(regionSums(resize(islands.gr, 51, "center"), nexus.bigwig.list$tfap2c$pos) + abs(regionSums(regions = resize(islands.gr,51,"center"), nexus.bigwig.list$tfap2c$neg)))
islands.gr$cdx2_sig <- abs(regionSums(resize(islands.gr, 51, "center"), nexus.bigwig.list$cdx2$pos) + abs(regionSums(regions = resize(islands.gr,51,"center"), nexus.bigwig.list$cdx2$neg)))
islands.gr$yap1_sig <- abs(regionSums(resize(islands.gr,100, "center"), nexus.bigwig.list$yap1$pos) + abs(regionSums(regions = resize(islands.gr,100,"center"), nexus.bigwig.list$yap1$neg)))
islands.gr$gata3_sig <- abs(regionSums(resize(islands.gr,51, "center"), nexus.bigwig.list$gata3$pos) + abs(regionSums(regions = resize(islands.gr,51,"center"), nexus.bigwig.list$gata3$neg)))
islands.gr$polII_sig <- abs(regionSums(resize(islands.gr, 100, "center"), nexus.bigwig.list$PolII$pos) + abs(regionSums(regions = resize(islands.gr,100,"center"), nexus.bigwig.list$PolII$neg)))

ov<-findOverlaps(all_tf.gr, islands.gr, ignore.strand = T)
all_tf_ov.gr<-all_tf.gr[ov@from]
all_tf_ov.gr$island_id<-islands.gr[ov@to]$island_id
all_tf_ov.gr$island_center<-islands.gr[ov@to]$island_center
all_tf_ov.gr$island_id %>% table %>% table

#Find all motif combinations across each island
islands_vs_motifs.df<-mclapply(unique(all_tf.gr$tf), function(x){
  df<-data.frame(island_id = as.factor(1:length(islands.gr)))
  all_tf_ov_subset.gr<-subset(all_tf_ov.gr, tf == x)
  instance_to_island.df <-all_tf_ov_subset.gr$island_id %>% table %>% as.data.frame()
  instance_to_island.df$Freq <- paste(x, instance_to_island.df$Freq, sep = "_")
  colnames(instance_to_island.df) <- c("island_id", "tf")
  df <- left_join(df,instance_to_island.df)
  return(df)
}, mc.cores = 11) %>% as.data.frame() %>% dplyr::select(tf,tf.1,tf.2,tf.3)
colnames(islands_vs_motifs.df)<-unique(all_tf.gr$tf)
rownames(islands_vs_motifs.df)<-1:length(islands.gr)

#Extract island combinations and summarize
island_combos.vec<-apply(islands_vs_motifs.df, 1, function(x){
  vec<-x[!is.na(x)]
  str<-paste0(vec, collapse = "/")
  return(str)
})
islands_vs_motifs.df$motif_pair <- island_combos.vec
islands_vs_motifs.df$island_id <- rownames(islands_vs_motifs.df)


islands_vs_motifs.df$polII_sig <- islands.gr$polII_sig
islands_vs_motifs.df$h3k27ac_sig <- islands.gr$h3k27ac_sig 
islands_vs_motifs.df$atac <- islands.gr$atac

islands_vs_motifs.df$tead4_sig <- islands.gr$tead4_sig
islands_vs_motifs.df$tfap2c_sig <- islands.gr$tfap2c_sig 
islands_vs_motifs.df$cdx2_sig <- islands.gr$cdx2_sig 
islands_vs_motifs.df$yap1_sig <- islands.gr$yap1_sig
islands_vs_motifs.df$gata3_sig <- islands.gr$gata3_sig 

island_combos.freq.df <- islands_vs_motifs.df %>% group_by(motif_pair) %>% summarise(Freq = length(unique(island_id))) # The frequency here is the number of islands containing this motif combination
island_combos.freq.df <-island_combos.freq.df[order(island_combos.freq.df$Freq, decreasing = T),]
island_combos.freq.df 
write.table(island_combos.freq.df, "cttg_factor_tsc_model_island_combo.freq_taskcwm_regions.xls",col.names = T, sep="\t")

#Tidy the all_tf_ov.gr $ islands.gr
all_tf_ov.df <- as.data.frame(all_tf_ov.gr)
all_tf_ov.df$island_id <- as.character(all_tf_ov.df$island_id)
all_tf_ov.gr <- left_join(all_tf_ov.df,islands_vs_motifs.df %>% dplyr::select(motif_pair,island_id), by = "island_id") %>% makeGRangesFromDataFrame(keep.extra.columns = T)
islands.gr$motif_pair <- islands_vs_motifs.df$motif_pair

#motif distribution
tt.gr <- all_tf_ov.gr %>% dplyr::filter(motif_pair == "Tead4_1/Tfap2c_1")
 tt.gr$island_vs_motif_dis <- tt.gr$island_center-tt.gr$tf_center
 
 ggplot(as.data.frame(tt.gr ))+
     geom_tile(aes(y = as.factor(island_id), x =island_vs_motif_dis, fill = tf, color = tf),size = 1)+
     theme(axis.title.y=element_blank(),
           axis.text.y=element_blank(),
           axis.ticks.y=element_blank())
 
 ##Motif distribution wrt to motif center
##Tead4-Tfap2c
tt.gr <- all_tf_ov.gr %>% dplyr::filter(motif_pair == "Tead4_1/Tfap2c_1")
td.gr <- all_tf_ov.gr %>% dplyr::filter(motif_pair == "Tead4_1/Tfap2c_1"  & tf == "Tead4") 
tf.gr <- all_tf_ov.gr %>% dplyr::filter(motif_pair == "Tead4_1/Tfap2c_1" & tf== "Tfap2c") 
tt.gr$td_tf_motif_dis <-  td.gr$tf_center-tf.gr$tf_center
tt.gr$td_tf_motif_dis <- abs(tt.gr$td_tf_motif_dis)
c.df <- as.data.frame(tt.gr)
write.table(c.df, "TT_motif_distance_allcwm_regions.xls",col.names = T, sep="\t")

tead4_tfap2c_dis.gr <- makeGRangesFromDataFrame(c.df, keep.extra.columns = TRUE, starts.in.df.are.0based = FALSE)
saveRDS(tead4_tfap2c_dis.gr, "rds/tead4_tfap2c_distance_from_island_analysis.rds")

# add signal
tead4_tfap2c_dis.gr$tead4_sig <- abs(regionSums(resize(tead4_tfap2c_dis.gr, 51, "center"), nexus.bigwig.list$tead4$pos) + abs(regionSums(regions = resize(tead4_tfap2c_dis.gr,51,"center"), nexus.bigwig.list$tead4$neg)))
tead4_tfap2c_dis.gr$tfap2c_sig <- abs(regionSums(resize(tead4_tfap2c_dis.gr, 51, "center"), nexus.bigwig.list$tfap2c$pos) + abs(regionSums(regions = resize(tead4_tfap2c_dis.gr,51,"center"), nexus.bigwig.list$tfap2c$neg)))
tead4_tfap2c_dis.gr$yap1_sig <- abs(regionSums(resize(tead4_tfap2c_dis.gr, 101, "center"), nexus.bigwig.list$yap1$pos) + abs(regionSums(regions = resize(tead4_tfap2c_dis.gr,101,"center"), nexus.bigwig.list$yap1$neg)))

tead4_tfap2c_dis.df <- as.data.frame(tead4_tfap2c_dis.gr) ## 1_based as as.dataframe doesnt convert it to 0-based
write.table(tead4_tfap2c_dis.df, "TT_motif_distance_allcwm_regions_withsignal.xls",col.names = T, sep="\t")
```

##nearest gene analysis
```{r, nearest gene analysis}
## nearest gene analysis
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
tss.gr<-genes(txdb) %>% resize(1, fix = "start")
nearest.idx_ov<-nearest(tead4_tfap2c_dis.gr, tss.gr, ignore.strand = T)
tead4_tfap2c_dis.gr$nearest_gene_id<-tss.gr[nearest.idx_ov]$gene_id
#Get gene name
library(biomaRt)
library(plyranges)
#connect to db with the correct dataset to pull from
mm10.ensembl <- useMart("ensembl",dataset="mmusculus_gene_ensembl")
tead4_tfap2c_dis_gene_info.df<-getBM(attributes = c("external_gene_name", "ensembl_gene_id","entrezgene_id","description"),filters = "entrezgene_id",
                          values = tead4_tfap2c_dis.gr$nearest_gene_id,
                          mart = mm10.ensembl)

peaks_with_genes_tead4_tfap2c_dis.gr.df<-merge(x = as.data.frame(tead4_tfap2c_dis.gr, stringsAsFactors = F), y = tead4_tfap2c_dis_gene_info.df, by.x = "nearest_gene_id", by.y = "entrezgene_id")

write.table(peaks_with_genes_tead4_tfap2c_dis.gr.df, file="tead4_tfap2c_motif_distance_and_nearesergene_biomart.xls", quote=FALSE, row.names=FALSE, col.names=TRUE, sep="\t")



```

##enhancer regions
```{r, tead4_tfap2c_mapped_regions}

genes_tead4_tfap2c_dis.gr <- makeGRangesFromDataFrame(peaks_with_genes_tead4_tfap2c_dis.gr.df, keep.extra.columns = T, seqnames.field = "seqnames", start.field = "start", end="end",starts.in.df.are.0based = FALSE)

#filter by motif distances
#Create independent granges of enhancer regions.
enhancers.gr<-genes_tead4_tfap2c_dis.gr %>% 
  plyranges::filter(td_tf_motif_dis<=200) %>%
  GenomicRanges::resize(400, 'center') %>%
  GenomicRanges::reduce(ignore.strand = T) %>%
  plyranges::mutate(enhancer_id = 0:(length(.)-1),
                    name = enhancer_id) %>%
  resize(1000, 'center') 
rtracklayer::export(enhancers.gr, 'bed/enhancers_genes_tead4_tfap2c_dis_0based.bed', format = 'BED')


##tfap2c motif
genes_tfap2c_dis.gr<- genes_tead4_tfap2c_dis.gr %>% 
  plyranges::filter(tf=="Tfap2c") %>%
  unique(.) %>%
  subsetByOverlaps(., enhancers.gr, ignore.strand = T) %>%
  plyranges::mutate(name = enhancers.gr[findOverlaps(., enhancers.gr, ignore.strand = T)@to]$enhancer_id)
rtracklayer::export(genes_tfap2c_dis.gr, 'bed/genes_tfap2c_dis_0based.bed', format = 'BED')
testit::assert('There is more than 1 tfap2c in an enhancer region.', (genes_tfap2c_dis.gr$name %>% unique %>% length)==length(genes_tfap2c_dis.gr))

##tead4 motif
genes_tead4_dis.gr<- genes_tead4_tfap2c_dis.gr %>% 
  plyranges::filter(tf=="Tead4") %>%
  unique(.) %>%
  subsetByOverlaps(., enhancers.gr, ignore.strand = T) %>%
  plyranges::mutate(name = enhancers.gr[findOverlaps(., enhancers.gr, ignore.strand = T)@to]$enhancer_id)
rtracklayer::export(genes_tead4_dis.gr, 'bed/genes_tead4_dis_0based.bed', format = 'BED')
testit::assert('There is more than 1 tead4 in an enhancer region.', (genes_tead4_dis.gr$name %>% unique %>% length)==length(genes_tead4_dis.gr))


#checking the motif start [between 0 and 1 based]
genes_tfap2c_dis.gr %>% plyranges::arrange(name) %>% head(n=1)
width(genes_tfap2c_dis.gr)

```

## plots
We, mostly Melanie has generated predictions on 154 Tfap2c-Tead4 (Tfap2c_1/Tead4_1) island motif regions, where we anchored one motif and moved another within 200 bp window. For that we first disrupt moved motifs original high contributing position/nucleotides to least contributing nucleotides and then injected moved motif at different distances in multiples of 15 bp

```{r, plots}
motif_specs.list<-list(
  tfap2c = list(color = '#391b40', tf = "Tfap2c"),
  tead4 = list(color = '#914236', tf = "Tead4"),
  cdx2 = list(color = '#d77d49', tf = "Cdx2"),
  gata3 = list(color = '#00565f', tf = "Gata3"))

motifs_of_interest<-c('tfap2c','tead4','cdx2', 'gata3')

```

##Dst
```{r}

profile_summary.df<-readr::read_tsv("/n/projects/kd2200/publication/bpnet/analysis/test/two_mutation/tsv/individual_profile_by_enhancer_id_dis_10_multiples/0_profiles_df.tsv.gz") %>% as.data.frame(stringsAsFactors = F)
coordinates_summary.df<-readr::read_tsv("/n/projects/kd2200/publication/bpnet/analysis/test/two_mutation/tsv/individual_pred_summary_by_enhancer_id/0_pred_summary_df.tsv.gz") %>% as.data.frame(stringsAsFactors = F)

## 42_distance ## Dst
eid <- 0
distance<- 90 #20
anchor<-'tead4'

#Get coordinates of motifs for plotting in enhancer profiles
enhancer_coords.df<-coordinates_summary.df %>%
    dplyr::filter(enhancer_id == eid, distance_between_moved_center_and_anchor_center==distance, anchor_motif==anchor) %>%
    dplyr::select(enhancer_id, distance_between_moved_center_and_anchor_center, anchor_motif, moved_motif,
                  anchor_motif_window_start_0based, anchor_motif_window_end_0based, 
                  moved_motif_window_orig_start_0based, moved_motif_window_orig_end_0based) %>%
    dplyr::rename(anchor_start = anchor_motif_window_start_0based, anchor_end = anchor_motif_window_end_0based,
                  mut_start = moved_motif_window_orig_start_0based, mut_end = moved_motif_window_orig_end_0based)  %>%
    dplyr::mutate(moved_start = anchor_start + distance_between_moved_center_and_anchor_center,
                  moved_end = anchor_end + distance_between_moved_center_and_anchor_center) %>%
    dplyr::mutate(bounded_motif_min = min(anchor_start, anchor_end, mut_start, mut_end, moved_start, moved_end),
                  bounded_motif_max = max(anchor_start, anchor_end, mut_start, mut_end,moved_start, moved_end))

enhancer_profile.df<-profile_summary.df %>%
    dplyr::filter(enhancer_id == eid, anchor_motif==anchor, type %in% c('wt','mut','mut_inj/90')) 

g <- ggplot()+
    annotate("rect", xmin = enhancer_coords.df$anchor_start, xmax = enhancer_coords.df$anchor_end,
             ymin = -Inf, ymax = Inf, alpha = .2)+
    annotate("text", x = enhancer_coords.df$anchor_start, label = enhancer_coords.df$anchor_start, y = Inf, vjust = 1.5)+
    annotate("rect", xmin = enhancer_coords.df$mut_start, xmax = enhancer_coords.df$mut_end,
             ymin = -Inf, ymax = Inf, alpha = .2)+
    annotate("text", x = enhancer_coords.df$mut_start, label = enhancer_coords.df$mut_start, y = Inf, vjust = 1.5)+
   annotate("rect", xmin = enhancer_coords.df$moved_start, xmax = enhancer_coords.df$moved_end, colour = "lightblue",
             ymin = -Inf, ymax = Inf, alpha = .1)+
    annotate("text", x = enhancer_coords.df$moved_start, label = enhancer_coords.df$moved_start, y = Inf, vjust = 1.5)+
    geom_area(data = enhancer_profile.df %>% dplyr::filter(type == 'wt'), mapping = aes(position, signal, group = strand, fill = task), alpha = .7)+
    geom_line(data = enhancer_profile.df %>% dplyr::filter(type == 'mut_inj/90'), mapping = aes(position, signal, group = strand, color = task))+
    facet_grid(task ~ ., scales = "free")+
    scale_x_continuous(limits = c(400,700),
                       name = "Position (bp)")+
    scale_y_continuous(name = "BPNet predictions")+
    scale_fill_manual(values = c('#6e828e', '#7f778c', '#b47382','#788e6e', '#968b71', '#96718f'), name = "WT")+
    scale_color_manual(values = c('#314d5e', '#493d5c','#94384d','#3e8f1a', '#a87c14','#8c1f78'), name = "Mut")+
    theme_classic()+ggtitle("can_0_WT & mut_moved_tfap2c_0")
g

#Save plots

ggsave("figures/distance_analysis/can_0_42_dis_to_90_dis_td_anchor.png",g,height = 12, width = 24)
ggsave("figures/distance_analysis/can_0_42_dis_to_90_dis_td_anchor.pdf",g,height =12, width = 24)

ggsave("figures/distance_analysis/can_0_42_dis_to_20_dis_td_anchor.png",g,height = 12, width = 24)
ggsave("figures/distance_analysis/can_0_42_dis_to_20_dis_td_anchor.pdf",g,height =12, width = 24)

# o <- filter(coordinates_summary.df, anchor_motif=="tead4")
# o_90 <- filter(o, distance_between_moved_center_and_anchor_center==90)
# write.table(o_90, file="/n/projects/kd2200/publication/bpnet/analysis/test/can_0_dis_90_td_anchor.xls", quote=FALSE, row.names=FALSE, col.names=TRUE, sep="\t")

```

##Adcy7
```{r, enhancer number 114}
profile_summary.df<-readr::read_tsv("/n/projects/kd2200/publication/bpnet/analysis/test/two_mutation/tsv/individual_profile_by_enhancer_id_dis_15_multiples/114_profiles_df.tsv.gz") %>% as.data.frame(stringsAsFactors = F)
coordinates_summary.df<-readr::read_tsv("/n/projects/kd2200/publication/bpnet/analysis/test/two_mutation/tsv/individual_pred_summary_by_enhancer_id/114_pred_summary_df.tsv.gz") %>% as.data.frame(stringsAsFactors = F)

## 118_distance ## Adcy7
eid <- 114
distance<- -75 #-30
anchor<-'tfap2c'

#Get coordinates of motifs for plotting in enhancer profiles
enhancer_coords.df<-coordinates_summary.df %>%
    dplyr::filter(enhancer_id == eid, distance_between_moved_center_and_anchor_center==distance, anchor_motif==anchor) %>%
    dplyr::select(enhancer_id, distance_between_moved_center_and_anchor_center, anchor_motif, moved_motif,
                  anchor_motif_window_start_0based, anchor_motif_window_end_0based, 
                  moved_motif_window_orig_start_0based, moved_motif_window_orig_end_0based) %>%
    dplyr::rename(anchor_start = anchor_motif_window_start_0based, anchor_end = anchor_motif_window_end_0based,
                  mut_start = moved_motif_window_orig_start_0based, mut_end = moved_motif_window_orig_end_0based)  %>%
    dplyr::mutate(moved_start = anchor_start + distance_between_moved_center_and_anchor_center,
                  moved_end = anchor_end + distance_between_moved_center_and_anchor_center) %>%
    dplyr::mutate(bounded_motif_min = min(anchor_start, anchor_end, mut_start, mut_end, moved_start, moved_end),
                  bounded_motif_max = max(anchor_start, anchor_end, mut_start, mut_end,moved_start, moved_end))

enhancer_profile.df<-profile_summary.df %>%
    dplyr::filter(enhancer_id == eid, anchor_motif==anchor, type %in% c('wt','mut','mut_inj/-75')) 

g <- ggplot()+
    annotate("rect", xmin = enhancer_coords.df$anchor_start, xmax = enhancer_coords.df$anchor_end,
             ymin = -Inf, ymax = Inf, alpha = .2)+
    annotate("text", x = enhancer_coords.df$anchor_start, label = enhancer_coords.df$anchor_start, y = Inf, vjust = 1.5)+
    annotate("rect", xmin = enhancer_coords.df$mut_start, xmax = enhancer_coords.df$mut_end,
             ymin = -Inf, ymax = Inf, alpha = .2)+
    annotate("text", x = enhancer_coords.df$mut_start, label = enhancer_coords.df$mut_start, y = Inf, vjust = 1.5)+
   annotate("rect", xmin = enhancer_coords.df$moved_start, xmax = enhancer_coords.df$moved_end, colour = "lightblue",
             ymin = -Inf, ymax = Inf, alpha = .1)+
    annotate("text", x = enhancer_coords.df$moved_start, label = enhancer_coords.df$moved_start, y = Inf, vjust = 1.5)+
    geom_area(data = enhancer_profile.df %>% dplyr::filter(type == 'wt'), mapping = aes(position, signal, group = strand, fill = task), alpha = .7)+
    geom_line(data = enhancer_profile.df %>% dplyr::filter(type == 'mut_inj/-75'), mapping = aes(position, signal, group = strand, color = task))+
    facet_grid(task ~ ., scales = "free")+
    scale_x_continuous(limits = c(400,650),
                       name = "Position (bp)")+
    scale_y_continuous(name = "BPNet predictions")+
    scale_fill_manual(values = c('#6e828e', '#7f778c', '#b47382','#788e6e', '#968b71', '#96718f'), name = "WT")+
    scale_color_manual(values = c('#314d5e', '#493d5c','#94384d','#3e8f1a', '#a87c14','#8c1f78'), name = "Mut")+
    theme_classic()+ggtitle("can_114_WT & mut_moved_tead4_114")
g

#Save plots
ggsave("figures/distance_analysis/can_114_118_dis_to_-75_dis_tf_anchor.png",g,height = 12, width = 24)
ggsave("figures/distance_analysis/can_114_118_dis_to_-75_dis_tf_anchor.pdf",g,height =12, width = 24)

# y <- filter(coordinates_summary.df, anchor_motif=="tfap2c")
# y_75 <- filter(y, distance_between_moved_center_and_anchor_center==-75)
# write.table(y_75, file="/n/projects/kd2200/publication/bpnet/analysis/test/can_114_dis_-75_tf_anchor.xls", quote=FALSE, row.names=FALSE, col.names=TRUE, sep="\t")
```


##Rin3 
```{r, enhancer number 154}
profile_summary.df<-readr::read_tsv("/n/projects/kd2200/publication/bpnet/analysis/test/two_mutation/tsv/individual_profile_by_enhancer_id_dis_15_multiples/154_profiles_df.tsv.gz") %>% as.data.frame(stringsAsFactors = F)
coordinates_summary.df<-readr::read_tsv("/n/projects/kd2200/publication/bpnet/analysis/test/two_mutation/tsv/individual_pred_summary_by_enhancer_id/154_pred_summary_df.tsv.gz") %>% as.data.frame(stringsAsFactors = F)

## rin3
eid <- 154
distance<- 60 
anchor<-'tfap2c'

#Get coordinates of motifs for plotting in enhancer profiles
enhancer_coords.df<-coordinates_summary.df %>%
    dplyr::filter(enhancer_id == eid, distance_between_moved_center_and_anchor_center==distance, anchor_motif==anchor) %>%
    dplyr::select(enhancer_id, distance_between_moved_center_and_anchor_center, anchor_motif, moved_motif,
                  anchor_motif_window_start_0based, anchor_motif_window_end_0based, 
                  moved_motif_window_orig_start_0based, moved_motif_window_orig_end_0based) %>%
    dplyr::rename(anchor_start = anchor_motif_window_start_0based, anchor_end = anchor_motif_window_end_0based,
                  mut_start = moved_motif_window_orig_start_0based, mut_end = moved_motif_window_orig_end_0based)  %>%
    dplyr::mutate(moved_start = anchor_start + distance_between_moved_center_and_anchor_center,
                  moved_end = anchor_end + distance_between_moved_center_and_anchor_center) %>%
    dplyr::mutate(bounded_motif_min = min(anchor_start, anchor_end, mut_start, mut_end, moved_start, moved_end),
                  bounded_motif_max = max(anchor_start, anchor_end, mut_start, mut_end,moved_start, moved_end))

enhancer_profile.df<-profile_summary.df %>%
    dplyr::filter(enhancer_id == eid, anchor_motif==anchor, type %in% c('wt','mut','mut_inj/60')) 

g <- ggplot()+
    annotate("rect", xmin = enhancer_coords.df$anchor_start, xmax = enhancer_coords.df$anchor_end,
             ymin = -Inf, ymax = Inf, alpha = .2)+
    annotate("text", x = enhancer_coords.df$anchor_start, label = enhancer_coords.df$anchor_start, y = Inf, vjust = 1.5)+
    annotate("rect", xmin = enhancer_coords.df$mut_start, xmax = enhancer_coords.df$mut_end,
             ymin = -Inf, ymax = Inf, alpha = .2)+
    annotate("text", x = enhancer_coords.df$mut_start, label = enhancer_coords.df$mut_start, y = Inf, vjust = 1.5)+
    annotate("rect", xmin = enhancer_coords.df$moved_start, xmax = enhancer_coords.df$moved_end, colour = "lightblue",
             ymin = -Inf, ymax = Inf, alpha = .1)+
    annotate("text", x = enhancer_coords.df$moved_start, label = enhancer_coords.df$moved_start, y = Inf, vjust = 1.5)+
    geom_area(data = enhancer_profile.df %>% dplyr::filter(type == 'wt'), mapping = aes(position, signal, group = strand, fill = task), alpha = .7)+
    geom_line(data = enhancer_profile.df %>% dplyr::filter(type == 'mut_inj/60'), mapping = aes(position, signal, group = strand, color = task))+
    facet_grid(task ~ ., scales = "free")+
    scale_x_continuous(limits = c(250,550),
                       name = "Position (bp)")+
    scale_y_continuous(name = "BPNet predictions")+
    scale_fill_manual(values = c('#6e828e', '#7f778c', '#b47382','#788e6e', '#968b71', '#96718f'), name = "WT")+
    scale_color_manual(values = c('#314d5e', '#493d5c','#94384d','#3e8f1a', '#a87c14','#8c1f78'), name = "Mut")+
    theme_classic()+ggtitle("can_154_WT & mut_moved_tead4")
g

#Save plots
ggsave("figures/distance_analysis/can_154_15_dis_to_55_dis_tf_anchor.png",g,height = 12, width = 24)
ggsave("figures/distance_analysis/can_154_15_dis_to_55_dis_tf_anchor.pdf",g,height =12, width = 24)

#y <- filter(coordinates_summary.df, anchor_motif=="tfap2c")
# y_60 <- filter(y, distance_between_moved_center_and_anchor_center==60)
# write.table(y_60, file="/n/projects/kd2200/publication/bpnet/analysis/test/can_114_dis_60_tf_anchor.xls", quote=FALSE, row.names=FALSE, col.names=TRUE, sep="\t")
```

# SessionInfo

For the purposes of reproducibility, the R/Bioconductor session information is printed below:

```{r sessioninfo}
sessionInfo()
```


